{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook to analyzse the efficiency of minimap mapping against a mock community"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting points from Tavish  \n",
    "reference_dataframe at '/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv'  \n",
    "custom_database at '/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta'  \n",
    "taxonomy_file at '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### workflow\n",
    "\n",
    "* Two databases\n",
    "* subsample 15000 reads per each mock community species. Save those out.\n",
    "* map reads against both databases with minimap safe out data in paf format.\n",
    "* get best hit per species (see what this means while looking at the data).\n",
    "* add the full taxonomy to each best match using the taxonomy file.\n",
    "* summarize data at different taxonomic ranks for each species.\n",
    "* pull this all together somehow (summary across all the samples? focus on species of interest e.g. deleted from analyis?)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### requirment\n",
    "\n",
    "* Bbmap (conda install bbmap https://anaconda.org/bioconda/bbmap)  \n",
    "* minimap2 (conda install minimap2 https://anaconda.org/bioconda/minimap2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fixes on the command line\n",
    "\n",
    "* fixed the taxonomy file to fit with the quiime format\n",
    "\n",
    "cat /media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file.csv | sed 's/,/\\t/' > /media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v2.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fix the taxonomy_file_v2.csv more to reflect Qiime style\n",
    "\n",
    "* This requires to make the species name genus_species and not _species.... hope this makes sense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "old_taxonomy_file_fn = '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v2.csv'\n",
    "new_taxonomy_file_fn = '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v3.csv'\n",
    "with open(new_taxonomy_file_fn, 'w') as out_fh:\n",
    "    with open(old_taxonomy_file_fn, 'r') as in_fh:\n",
    "        for line in in_fh:\n",
    "            line = line.rstrip()\n",
    "            #print(line)\n",
    "            first_half = line.split('s__')[0]\n",
    "            second_half = line.split('s__')[1]\n",
    "            pattern = re.compile(r'g__\\w+;')\n",
    "            genus = re.findall(pattern, first_half)[0].replace('g__','').replace(';','')\n",
    "            new_line = F\"{first_half}s__{genus}_{second_half}\"\n",
    "            print(new_line,file=out_fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!head -2 {old_taxonomy_file_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!head -2 {new_taxonomy_file_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Initial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv')\n",
    "max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta')\n",
    "taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_qiime.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#threads to use\n",
    "threads = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = os.path.abspath('../../analysis/Mapping_mock_gsref')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### list of species in the max database\n",
    "max_species = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ###Removed from second test databes\n",
    "species_delete = [\n",
    "# 'Candida_orthopsilosis',\n",
    "#                  'Candida_metapsilosis',\n",
    "#                  'Aspergillus_niger'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###species to be searched against both databases\n",
    "# mock_community = ['Penicillium_chrysogenum',\n",
    "#  'Aspergillus_flavus',\n",
    "#  'Aspergillus_niger',\n",
    "#  'Pichia_kudriavzevii',\n",
    "#  'Pichia_membranifaciens',\n",
    "#  'Candida_albicans',\n",
    "#  'Candida_parapsilosis',\n",
    "#  'Candida_orthopsilosis',\n",
    "#  'Candida_metapsilosis']\n",
    "\n",
    "mock_community = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fixed_old_names = ['Kluyveromyces_lactis',\n",
    "#                    'Candida_zeylanoides',\n",
    "#                    'Cladophialophora_sp.',\n",
    "#                    'Diaporthe_sp.',\n",
    "#                    'CCL060',\n",
    "#                    'CCL068',\n",
    "#                    'CCL067',\n",
    "#                    'Aspergillus_sp.',\n",
    "#                    'Entoleuca_sp.',\n",
    "#                    'Tapesia_yallundae_CCL029',\n",
    "#                    'Tapesia_yallundae_CCL031',\n",
    "#                    'Cryptococcus_neoformans']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fixed_new_names = ['candida_unidentified',\n",
    "#                    'debaryomyces_unidentified',\n",
    "#                    'cladophialophora_unidentified',\n",
    "#                    'diaporthe_unidentified',\n",
    "#                    'asteroma_ccl060',\n",
    "#                    'asteroma_ccl068',\n",
    "#                    'diaporthe_ccl067',\n",
    "#                    'aspergillus_unidentified',\n",
    "#                    'entoleuca_unidentified',\n",
    "#                    'oculimacula_yallundae-ccl029',\n",
    "#                    'oculimacula_yallundae-ccl031',\n",
    "#                    'kluyveromyces_unidentified']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old_to_new_names = dict(zip(fixed_old_names, fixed_new_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old_to_new_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fix databases and names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = pd.read_csv(reference_dataframe_fn)\n",
    "ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df.name_species.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_fn = os.path.join(OUT_DIR, 'gsref.db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_list = []\n",
    "old_db_list = []\n",
    "for seq in SeqIO.parse(max_custom_database_fn, 'fasta'):\n",
    "    old_db_list.append(seq.id)\n",
    "    if seq.id.lower() in ref_df.name_species.tolist():\n",
    "        #print(seq.id)\n",
    "        seq.id = seq.name = seq.description = seq.id.lower()\n",
    "        new_db_list.append(seq)\n",
    "    else:\n",
    "        print(seq.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(new_db_list) == len(old_db_list):\n",
    "    SeqIO.write(new_db_list, new_db_fn, 'fasta')\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_db_fn = os.path.join(OUT_DIR, 'gsref.subdb.fasta')\n",
    "sub_db_list = []\n",
    "for seq in new_db_list:\n",
    "    if seq.id not in [x.lower() for x in species_delete]:\n",
    "        sub_db_list.append(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(sub_db_list) + len(species_delete) == len(new_db_list):\n",
    "    SeqIO.write(sub_db_list, sub_db_fn, 'fasta' )\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[x.id for x in sub_db_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = [x.lower() for x in mock_community]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subsample reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 2000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MC_READ_DIR = os.path.join(OUT_DIR, 'MC_READS')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map with minimap against both databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results(fn, species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the query species: {species}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')\n",
    "    hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/custom_results/%s.json' % species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###this is running the reads against the full database\n",
    "for species, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###this is running against a database that have ['Candida_orthopsilosis', 'Candida_metapsilosis', 'Aspergillus_niger'] deleted\n",
    "for species, hit_fn in sub_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pull in mapping results and analyse them at all available levels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### idea\n",
    "\n",
    "* pull in query taxid as a dictionary\n",
    "* assign taxid for each tname species from minimap2\n",
    "* generate a summary dictionary that checks concordance at each taxonmic rank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['mquality'].count().tolist(), sub_df.groupby('tname')['mquality'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "    return hit_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getquery_taxfileid(refdf_fn, species):\n",
    "    \"\"\"\n",
    "    Takes the reference dataframe filename and the species name.\n",
    "    Returns the taxfileid, which is the date/flowcellid (column 0 value) of the ref_df.\n",
    "    \"\"\"\n",
    "    ref_df = pd.read_csv(refdf_fn)\n",
    "    ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']\n",
    "    return ref_df[ref_df.name_species == species].iloc[:,0].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxid_dict(taxid_fn, taxfileid):\n",
    "    \"\"\"\n",
    "    Takes a taxonomy assignment file filename in the Qiime format and a taxonomic identifier.\n",
    "    Returns the a dictionary with the taxonomic assignment at each rank.\n",
    "    \"\"\"\n",
    "    tax_dict = {}\n",
    "    with open(taxid_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(taxfileid):\n",
    "                taxrankids = line.rstrip().split('\\t')[1].split(';')\n",
    "                for taxrank in taxrankids:\n",
    "                    tax_dict[taxrank.split('__')[0]] = taxrank.split('__')[1]\n",
    "    return tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_taxranks_results(mapping_df, tax_fn, ref_df_fn = False):\n",
    "    \"\"\"\n",
    "    This function assigns the taxonomic ranks for each hit in the mapping results dataframe.\n",
    "    It takes a mapping_df, taxonomy assignment file, and if required a reference dataframe filename.\n",
    "    Returns the mapping dataframe with assignment. \n",
    "    \"\"\"\n",
    "    for tname in mapping_df.index:\n",
    "        if ref_df_fn:\n",
    "            tmp_taxfileid = getquery_taxfileid(ref_df_fn, tname)\n",
    "        else:\n",
    "            tmp_taxfileid = tname\n",
    "        tmp_tax_dict = get_taxid_dict(tax_fn, tmp_taxfileid)\n",
    "        for key, value in tmp_tax_dict.items():\n",
    "            mapping_df.loc[tname, key] = value\n",
    "    return mapping_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarieses the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = {}\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'candida_albicans'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'aspergillus_niger'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test run on the qiime2 Database\n",
    "\n",
    "##### Prep on the command line\n",
    "\n",
    "cp sh_refs_qiime_ver8_dynamic_02.02.2019.fasta /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/.  \n",
    "cp sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results_v2(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    #sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['mquality'].count().tolist(), sub_df.groupby('tname')['mquality'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.path.abspath(os.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run on test species 'penicillium_chrysogenum'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### These are wired results that might be linked to\n",
    "\n",
    "* database issues as you can see\n",
    "\n",
    "sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt: k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_results[mapping_results['s'] == 'Penicillium_chrysogenum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]\n",
    "query_tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.path.abspath(os.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run on test species 'penicillium_chrysogenum'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### These are wired results that might be linked to\n",
    "\n",
    "* database issues as you can see\n",
    "\n",
    "sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt: k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_results[mapping_results['s'] == 'Penicillium_chrysogenum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]\n",
    "query_tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Might want to double check how your families are matched in your taxonomic input file for your known test species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_results[mapping_results['s'] == 'Candida_albicans']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# query_tax_dict['f'] = 'Saccharomycetales_fam_Incertae_sedis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### It appears that the filtering of results by mapping quality works well for the long ITS database but not for the qiime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.tname.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.tname.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.index = full_results_df.tname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Asign taxonomic ranks to the full_results_df\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df[full_results_df['g'] == 'Candida']['qseqid'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df[full_results_df['g'] == 'Candida']['qseqid'].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at the results unfiltered doesn't really work very well either. Might need to look into different filtering of the alignments or the qiime2 database might be just not really useful for the noise reads. Simulated reads with higher accuracy should get better here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.groupby('g').count()['k'].sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on other species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['cortinarius_globuliformis']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on all species using v3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from collections import OrderedDict\n",
    "\n",
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarises the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = OrderedDict()\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        tmps_df = pd.DataFrame(data=None)\n",
    "        if tax_rank == 's':\n",
    "            for index, row in mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]].iterrows():\n",
    "                if row['s'] == 'unidentified' and row['g'] != query_tax_dict['g']:\n",
    "                    mapping_df.drop(index, axis=0, inplace=True)\n",
    "                else:\n",
    "                    continue\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        else:\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict\n",
    "\n",
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df\n",
    "    \n",
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "test_species_list = []\n",
    "for entry in ref_df.name_species.tolist():\n",
    "#     if entry[-7:] != '-ccl031' and entry[-7:] != '-ccl029':\n",
    "#         test_species_list.append(entry)\n",
    "#     else:\n",
    "#         test_species_list.append(entry[:-7])\n",
    "#         print(entry[:-7])\n",
    "    test_species_list.append(entry)\n",
    "    \n",
    "for test_species in test_species_list:\n",
    "    \n",
    "    print(test_species)\n",
    "    \n",
    "    #subsample tests species\n",
    "    fn_subsampling = {}\n",
    "    test_species = [test_species]\n",
    "    for x in test_species:\n",
    "        fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "        fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "    sub_reads_fn = {}\n",
    "    n_reads = 20000\n",
    "    for key, value in fn_subsampling.items():\n",
    "        species = key\n",
    "        in_fn = value\n",
    "        out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "        subsamplereads(in_fn, out_fn, n_reads)\n",
    "        sub_reads_fn[species] = out_fn\n",
    "        \n",
    "    ###Map the reads\n",
    "    db_fn = qiime_db_fn\n",
    "    sub_db_mapping_fn = {}\n",
    "    for species, fasta_fn in sub_reads_fn.items():\n",
    "        db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "        out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "        sub_db_mapping_fn[species] = out_fn\n",
    "        minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "        \n",
    "    ###Test out the summary results statistic for a single mapping result\n",
    "    species = test_species[0]\n",
    "    mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "    mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "    taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "    \n",
    "    query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "    print(type(query_tax_dict))\n",
    "    print(query_tax_dict)\n",
    "    \n",
    "#     sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "        \n",
    "#     print(json.dumps(sensitivity_dict, indent=1))\n",
    "#     print('\\n')\n",
    "#     with open('/media/MassStorage/tmp/TE/honours/analysis/Mapping/qiime_results/%s.json' % species, 'w+') as fp:\n",
    "#         json.dump(sensitivity_dict, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test run on the qiime2 Database 99\n",
    "\n",
    "##### Prep on the command line\n",
    "\n",
    "cp sh_refs_qiime_ver8_99_02.02.2019.fasta /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/. \n",
    "cp sh_taxonomy_qiime_ver8_99_02.02.2019.txt /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_99_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_99_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "print(mapping_dir)\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###fix the query_tax_dict with the data found in the qiime database\n",
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use with Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'candida_albicans'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply to wheat species for qiime and consensus respectively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>species1</th>\n",
       "      <th>species2</th>\n",
       "      <th># raw reads</th>\n",
       "      <th># reads after homology filtering</th>\n",
       "      <th># reads after length filtering</th>\n",
       "      <th># for use</th>\n",
       "      <th>path to raw reads</th>\n",
       "      <th>path to homology filtering</th>\n",
       "      <th>path to length filtering</th>\n",
       "      <th>path for use</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>Puccinia_striiformis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>115052</td>\n",
       "      <td>23477</td>\n",
       "      <td>21590</td>\n",
       "      <td>21590.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode01/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode01/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode01/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode01/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>barcode02</td>\n",
       "      <td>Zymoseptoria_tritici</td>\n",
       "      <td>NaN</td>\n",
       "      <td>151005</td>\n",
       "      <td>20416</td>\n",
       "      <td>18100</td>\n",
       "      <td>18100.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode02/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode02/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode02/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode02/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>barcode03</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>92927</td>\n",
       "      <td>23616</td>\n",
       "      <td>21919</td>\n",
       "      <td>21919.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode03/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode03/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode03/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode03/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>barcode04</td>\n",
       "      <td>Puccinia_striiformis</td>\n",
       "      <td>Zymoseptoria_tritici</td>\n",
       "      <td>115407</td>\n",
       "      <td>22716</td>\n",
       "      <td>20568</td>\n",
       "      <td>20568.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode04/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode04/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode04/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode04/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>Pyrenophora_tritici-repentis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>134094</td>\n",
       "      <td>24785</td>\n",
       "      <td>22433</td>\n",
       "      <td>22433.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode05/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode05/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode05/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode05/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150739</td>\n",
       "      <td>25462</td>\n",
       "      <td>23769</td>\n",
       "      <td>23769.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode06/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode06/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode06/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode06/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109201</td>\n",
       "      <td>24688</td>\n",
       "      <td>22931</td>\n",
       "      <td>22931.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode07/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode07/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode07/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode07/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>barcode08</td>\n",
       "      <td>Puccinia_striiformis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>114215</td>\n",
       "      <td>21725</td>\n",
       "      <td>20105</td>\n",
       "      <td>20105.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode08/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode08/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode08/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode08/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>barcode09</td>\n",
       "      <td>Puccinia_striiformis</td>\n",
       "      <td>Pyrenophora_tritici-repentis</td>\n",
       "      <td>106976</td>\n",
       "      <td>22012</td>\n",
       "      <td>20451</td>\n",
       "      <td>20451.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode09/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode09/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode09/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode09/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>barcode10</td>\n",
       "      <td>Puccinia_striiformis</td>\n",
       "      <td>Zymoseptoria_tritici</td>\n",
       "      <td>108583</td>\n",
       "      <td>21514</td>\n",
       "      <td>19810</td>\n",
       "      <td>19810.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode10/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode10/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode10/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode10/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>barcode11</td>\n",
       "      <td>Zymoseptoria_tritici</td>\n",
       "      <td>Puccinia_striiformis</td>\n",
       "      <td>118345</td>\n",
       "      <td>22462</td>\n",
       "      <td>20832</td>\n",
       "      <td>20832.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode11/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode11/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode11/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode11/lengt...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                      species1                      species2  \\\n",
       "0   barcode01          Puccinia_striiformis                           NaN   \n",
       "1   barcode02          Zymoseptoria_tritici                           NaN   \n",
       "2   barcode03                       unknown                           NaN   \n",
       "3   barcode04          Puccinia_striiformis          Zymoseptoria_tritici   \n",
       "4   barcode05  Pyrenophora_tritici-repentis                           NaN   \n",
       "5   barcode06                       unknown                           NaN   \n",
       "6   barcode07                       unknown                           NaN   \n",
       "7   barcode08          Puccinia_striiformis                           NaN   \n",
       "8   barcode09          Puccinia_striiformis  Pyrenophora_tritici-repentis   \n",
       "9   barcode10          Puccinia_striiformis          Zymoseptoria_tritici   \n",
       "10  barcode11          Zymoseptoria_tritici          Puccinia_striiformis   \n",
       "\n",
       "    # raw reads  # reads after homology filtering  \\\n",
       "0        115052                             23477   \n",
       "1        151005                             20416   \n",
       "2         92927                             23616   \n",
       "3        115407                             22716   \n",
       "4        134094                             24785   \n",
       "5        150739                             25462   \n",
       "6        109201                             24688   \n",
       "7        114215                             21725   \n",
       "8        106976                             22012   \n",
       "9        108583                             21514   \n",
       "10       118345                             22462   \n",
       "\n",
       "    # reads after length filtering  # for use  \\\n",
       "0                            21590    21590.0   \n",
       "1                            18100    18100.0   \n",
       "2                            21919    21919.0   \n",
       "3                            20568    20568.0   \n",
       "4                            22433    22433.0   \n",
       "5                            23769    23769.0   \n",
       "6                            22931    22931.0   \n",
       "7                            20105    20105.0   \n",
       "8                            20451    20451.0   \n",
       "9                            19810    19810.0   \n",
       "10                           20832    20832.0   \n",
       "\n",
       "                                    path to raw reads  \\\n",
       "0   analysis/Concatenated/wheat/barcode01/merged.f...   \n",
       "1   analysis/Concatenated/wheat/barcode02/merged.f...   \n",
       "2   analysis/Concatenated/wheat/barcode03/merged.f...   \n",
       "3   analysis/Concatenated/wheat/barcode04/merged.f...   \n",
       "4   analysis/Concatenated/wheat/barcode05/merged.f...   \n",
       "5   analysis/Concatenated/wheat/barcode06/merged.f...   \n",
       "6   analysis/Concatenated/wheat/barcode07/merged.f...   \n",
       "7   analysis/Concatenated/wheat/barcode08/merged.f...   \n",
       "8   analysis/Concatenated/wheat/barcode09/merged.f...   \n",
       "9   analysis/Concatenated/wheat/barcode10/merged.f...   \n",
       "10  analysis/Concatenated/wheat/barcode11/merged.f...   \n",
       "\n",
       "                           path to homology filtering  \\\n",
       "0   analysis/Python_Processing/wheat/barcode01/com...   \n",
       "1   analysis/Python_Processing/wheat/barcode02/com...   \n",
       "2   analysis/Python_Processing/wheat/barcode03/com...   \n",
       "3   analysis/Python_Processing/wheat/barcode04/com...   \n",
       "4   analysis/Python_Processing/wheat/barcode05/com...   \n",
       "5   analysis/Python_Processing/wheat/barcode06/com...   \n",
       "6   analysis/Python_Processing/wheat/barcode07/com...   \n",
       "7   analysis/Python_Processing/wheat/barcode08/com...   \n",
       "8   analysis/Python_Processing/wheat/barcode09/com...   \n",
       "9   analysis/Python_Processing/wheat/barcode10/com...   \n",
       "10  analysis/Python_Processing/wheat/barcode11/com...   \n",
       "\n",
       "                             path to length filtering  \\\n",
       "0   analysis/Length_Filtered/wheat/barcode01/lengt...   \n",
       "1   analysis/Length_Filtered/wheat/barcode02/lengt...   \n",
       "2   analysis/Length_Filtered/wheat/barcode03/lengt...   \n",
       "3   analysis/Length_Filtered/wheat/barcode04/lengt...   \n",
       "4   analysis/Length_Filtered/wheat/barcode05/lengt...   \n",
       "5   analysis/Length_Filtered/wheat/barcode06/lengt...   \n",
       "6   analysis/Length_Filtered/wheat/barcode07/lengt...   \n",
       "7   analysis/Length_Filtered/wheat/barcode08/lengt...   \n",
       "8   analysis/Length_Filtered/wheat/barcode09/lengt...   \n",
       "9   analysis/Length_Filtered/wheat/barcode10/lengt...   \n",
       "10  analysis/Length_Filtered/wheat/barcode11/lengt...   \n",
       "\n",
       "                                         path for use  \n",
       "0   analysis/Length_Filtered/wheat/barcode01/lengt...  \n",
       "1   analysis/Length_Filtered/wheat/barcode02/lengt...  \n",
       "2   analysis/Length_Filtered/wheat/barcode03/lengt...  \n",
       "3   analysis/Length_Filtered/wheat/barcode04/lengt...  \n",
       "4   analysis/Length_Filtered/wheat/barcode05/lengt...  \n",
       "5   analysis/Length_Filtered/wheat/barcode06/lengt...  \n",
       "6   analysis/Length_Filtered/wheat/barcode07/lengt...  \n",
       "7   analysis/Length_Filtered/wheat/barcode08/lengt...  \n",
       "8   analysis/Length_Filtered/wheat/barcode09/lengt...  \n",
       "9   analysis/Length_Filtered/wheat/barcode10/lengt...  \n",
       "10  analysis/Length_Filtered/wheat/barcode11/lengt...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "\n",
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')\n",
    "subsampling_dir = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/subsample_reads')\n",
    "mapping_dir = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/qiime_results')\n",
    "\n",
    "wheat_reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/wheat_reference_dataframe.csv')\n",
    "wheat_max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/wheat_database_labelled.fasta')\n",
    "# taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_qiime.csv')\n",
    "\n",
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "\n",
    "wheat_ref_df = pd.read_csv(wheat_reference_dataframe_fn)\n",
    "wheat_ref_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode01\n",
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    514                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             364                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              278            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             206                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              159          Knoxdaviesia_cecropiae\n",
      "SH1558908.08FU_KF800655_reps              144                    unidentified\n",
      "SH1563343.08FU_FJ430715_reps              123                    unidentified\n",
      "SH1560162.08FU_AY225488_reps              119               Lignincola_laevis\n",
      "SH1680040.08FU_KC992945_refs_singleton    110  Cystoagaricus_hirtosquamulosus\n",
      "SH1563342.08FU_UDB028691_reps              88                    unidentified\n",
      "SH1566136.08FU_KC992951_refs               75        Lacrymaria_subcinnamomea\n",
      "SH1573885.08FU_MF594688_reps               26       Deltopyxis_triangulispora\n",
      "SH1651389.08FU_KY782284_reps               18                    unidentified\n",
      "SH1598232.08FU_HM057122_refs                7            Puccinia_striiformis\n",
      "SH1516349.08FU_KY242684_reps                5                    unidentified\n",
      "SH2127282.08FU_AY762947_refs                1          Alternaria_oregonensis\n",
      "SH1920677.08FU_KT389479_refs                1                Ascochyta_rabiei\n",
      "SH2232092.08FU_GU237898_refs                1             Ascochyta_herbicola\n",
      "SH1505864.08FU_JQ693660_refs                1       Alternaria_metachromatica\n",
      "SH1650816.08FU_FM178254_refs                1              Papiliotrema_frias\n",
      "SH1598253.08FU_AY956558_reps                1            Puccinia_striiformis\n",
      "SH1573994.08FU_KP170660_reps                1  Sphaerellopsis_macroconidialis\n",
      "SH1505885.08FU_HQ873338_reps                1              Alternaria_molesta\n",
      "SH1572792.08FU_EF679363_refs                1         Mycosphaerella_tassiana\n",
      "SH1549852.08FU_UDB024177_reps               1             Toninia_physaroides\n",
      "SH1528219.08FU_JF495242_reps                1                    unidentified\n",
      "SH2232193.08FU_KT389475_refs                1                Ascochyta_phacae\n",
      "{'s': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"s\": 0.0035587188612099642\n",
      "}\n",
      "barcode02\n",
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    572                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             228                    unidentified\n",
      "SH1558908.08FU_KF800655_reps              186                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              139            Bannoa_ogasawarensis\n",
      "SH1523077.08FU_MH019936_reps              109                    unidentified\n",
      "SH1654757.08FU_UDB014954_reps              95                    unidentified\n",
      "SH1680040.08FU_KC992945_refs_singleton     91  Cystoagaricus_hirtosquamulosus\n",
      "SH1577464.08FU_JF700867_refs               82             Zymoseptoria_brevis\n",
      "SH1563344.08FU_AM267268_reps               72          Knoxdaviesia_cecropiae\n",
      "SH1563342.08FU_UDB028691_reps              72                    unidentified\n",
      "SH1563343.08FU_FJ430715_reps               69                    unidentified\n",
      "SH1577481.08FU_AF181701_reps               51          Zymoseptoria_halophila\n",
      "SH1541142.08FU_KR015769_reps               50                    unidentified\n",
      "SH1566136.08FU_KC992951_refs               45        Lacrymaria_subcinnamomea\n",
      "SH1560162.08FU_AY225488_reps               39               Lignincola_laevis\n",
      "SH1523966.08FU_HE820879_reps               36                    unidentified\n",
      "SH2127280.08FU_AF347034_refs               23           Alternaria_infectoria\n",
      "SH1573885.08FU_MF594688_reps               23       Deltopyxis_triangulispora\n",
      "SH1577471.08FU_KC005781_refs               21           Zymoseptoria_verkleyi\n",
      "SH1613737.08FU_GQ303275_refs               17                    unidentified\n",
      "SH1505864.08FU_JQ693660_refs               14       Alternaria_metachromatica\n",
      "SH1541263.08FU_KF436104_reps               11                    unidentified\n",
      "SH2127282.08FU_AY762947_refs               11          Alternaria_oregonensis\n",
      "SH1554120.08FU_KU747561_reps               11                Xylaria_cubensis\n",
      "SH1541118.08FU_KR016816_reps                9                    unidentified\n",
      "SH1651389.08FU_KY782284_reps                8                    unidentified\n",
      "SH1649730.08FU_HE820818_reps                7                    unidentified\n",
      "SH1554929.08FU_JF773605_reps                7                    unidentified\n",
      "SH1546322.08FU_JF433037_refs                7         Fusarium_pseudensiforme\n",
      "SH1515312.08FU_KJ713215_reps                5                    unidentified\n",
      "SH1516349.08FU_KY242684_reps                5                    unidentified\n",
      "SH1853051.08FU_KF924559_refs_singleton      4         Conioscypha_minutispora\n",
      "SH1507266.08FU_LC177377_reps                4                    unidentified\n",
      "SH1645391.08FU_FJ415474_refs                4         Elasticomyces_elasticus\n",
      "SH1505584.08FU_MF767833_reps                3                    unidentified\n",
      "SH1546461.08FU_JF433034_refs                3            Fusarium_rectiphorum\n",
      "SH1651458.08FU_EF536346_refs                3                    unidentified\n",
      "SH1750412.08FU_KF251253_refs                3    Sphaerulina_pseudovirgaureae\n",
      "SH1687540.08FU_AB042642_refs                3             Leveillula_elaeagni\n",
      "SH1508982.08FU_AY740160_reps                2             Sporisorium_lepturi\n",
      "SH1549852.08FU_UDB024177_reps               2             Toninia_physaroides\n",
      "SH1612303.08FU_EU715613_reps                2                    unidentified\n",
      "SH1516182.08FU_FJ430752_reps                2                    unidentified\n",
      "SH1548073.08FU_FJ839626_reps                2                    unidentified\n",
      "SH1718969.08FU_AF173308_refs                2           Dissoconium_aciculare\n",
      "SH1552276.08FU_EU707897_refs                2             Dissoconium_proteae\n",
      "SH1524268.08FU_FJ820763_reps                2                    unidentified\n",
      "SH1930245.08FU_EF394855_refs                2           Dissoconium_eucalypti\n",
      "SH1548076.08FU_EU514291_refs                2                 Zasmidium_musae\n",
      "SH1711498.08FU_JF433036_refs_singleton      2          Fusarium_kurunegalense\n",
      "SH1919165.08FU_JF433041_refs                1              Fusarium_kelerajum\n",
      "SH1664303.08FU_KC584202_refs_singleton      1         Alternaria_juxtiseptata\n",
      "SH1664302.08FU_KC584224_refs_singleton      1         Alternaria_vaccariicola\n",
      "SH1664237.08FU_JN383490_refs                1               Alternaria_eureka\n",
      "SH1820842.08FU_AF444354_refs                1          Filobasidium_chernovii\n",
      "SH1652815.08FU_MF509780_reps                1             Peziza_retrocurvata\n",
      "SH1858974.08FU_JQ693639_refs                1                Alternaria_rosae\n",
      "SH1849720.08FU_JF433035_refs_singleton      1          Fusarium_haematococcum\n",
      "SH1577480.08FU_KC677913_refs                1            Zasmidium_dalbergiae\n",
      "SH1650852.08FU_HE820847_reps                1                    unidentified\n",
      "SH1645457.08FU_EF394833_reps                1     Xenomycosphaerella_elongata\n",
      "SH1505885.08FU_HQ873338_reps                1              Alternaria_molesta\n",
      "SH1505898.08FU_DQ420888_reps                1                    unidentified\n",
      "SH1526446.08FU_KC584194_reps                1          Alternaria_dianthicola\n",
      "SH1528207.08FU_AF444469_refs                1         Vishniacozyma_victoriae\n",
      "SH1546365.08FU_JF433045_refs                1              Fusarium_mahasenii\n",
      "SH1570265.08FU_DQ303009_reps                1                    unidentified\n",
      "SH1572792.08FU_EF679363_refs                1         Mycosphaerella_tassiana\n",
      "SH1575716.08FU_EF452459_reps                1               Pyrenophora_bromi\n",
      "SH1577467.08FU_JF770452_refs                1       Mycosphaerella_etlingerae\n",
      "SH1577485.08FU_EU853477_reps                1                    unidentified\n",
      "SH1577495.08FU_KX287497_reps                1               Ramularia_rubella\n",
      "SH1577503.08FU_KT328901_reps                1                    unidentified\n",
      "SH1577538.08FU_KX287556_reps                1                    unidentified\n",
      "SH1598232.08FU_HM057122_refs                1            Puccinia_striiformis\n",
      "SH1619615.08FU_FJ501296_reps                1                    unidentified\n",
      "SH1630812.08FU_KX444437_reps                1                    unidentified\n",
      "SH1631055.08FU_EU561491_reps                1                    unidentified\n",
      "SH1645444.08FU_AB277789_reps                1      Cercospora_cryptomeriicola\n",
      "SH1645454.08FU_AY646479_reps                1         Mycosphaerella_musicola\n",
      "SH2127294.08FU_FJ196308_refs                1              Alternaria_molesta\n",
      "{'s': 'Zymoseptoria_tritici'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "barcode03\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    513                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             386                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              319            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             198                    unidentified\n",
      "SH1558908.08FU_KF800655_reps              156                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              156          Knoxdaviesia_cecropiae\n",
      "SH1563343.08FU_FJ430715_reps              137                    unidentified\n",
      "SH1680040.08FU_KC992945_refs_singleton    132  Cystoagaricus_hirtosquamulosus\n",
      "SH1566136.08FU_KC992951_refs               92        Lacrymaria_subcinnamomea\n",
      "SH1560162.08FU_AY225488_reps               88               Lignincola_laevis\n",
      "SH1563342.08FU_UDB028691_reps              75                    unidentified\n",
      "SH1573885.08FU_MF594688_reps               29       Deltopyxis_triangulispora\n",
      "SH1651389.08FU_KY782284_reps               26                    unidentified\n",
      "SH1516349.08FU_KY242684_reps                6                    unidentified\n",
      "SH1508982.08FU_AY740160_reps                1             Sporisorium_lepturi\n",
      "{'s': 'unknown'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "barcode04\n",
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    518                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             365                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              300            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             181                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              153          Knoxdaviesia_cecropiae\n",
      "SH1558908.08FU_KF800655_reps              146                    unidentified\n",
      "SH1563343.08FU_FJ430715_reps              120                    unidentified\n",
      "SH1680040.08FU_KC992945_refs_singleton    110  Cystoagaricus_hirtosquamulosus\n",
      "SH1563342.08FU_UDB028691_reps              84                    unidentified\n",
      "SH1560162.08FU_AY225488_reps               83               Lignincola_laevis\n",
      "SH1566136.08FU_KC992951_refs               73        Lacrymaria_subcinnamomea\n",
      "SH1573885.08FU_MF594688_reps               39       Deltopyxis_triangulispora\n",
      "SH1651389.08FU_KY782284_reps               18                    unidentified\n",
      "SH1523077.08FU_MH019936_reps               12                    unidentified\n",
      "SH1577464.08FU_JF700867_refs                9             Zymoseptoria_brevis\n",
      "SH1523966.08FU_HE820879_reps                8                    unidentified\n",
      "SH1554120.08FU_KU747561_reps                5                Xylaria_cubensis\n",
      "SH1577481.08FU_AF181701_reps                4          Zymoseptoria_halophila\n",
      "SH1541263.08FU_KF436104_reps                4                    unidentified\n",
      "SH1516349.08FU_KY242684_reps                4                    unidentified\n",
      "SH2127280.08FU_AF347034_refs                4           Alternaria_infectoria\n",
      "SH1541142.08FU_KR015769_reps                3                    unidentified\n",
      "SH1541118.08FU_KR016816_reps                3                    unidentified\n",
      "SH1528207.08FU_AF444469_refs                3         Vishniacozyma_victoriae\n",
      "SH1609774.08FU_AF444379_refs                3             Dioszegia_hungarica\n",
      "SH1598232.08FU_HM057122_refs                3            Puccinia_striiformis\n",
      "SH1853051.08FU_KF924559_refs_singleton      2         Conioscypha_minutispora\n",
      "SH1649730.08FU_HE820818_reps                2                    unidentified\n",
      "SH2525086.08FU_DQ222455_refs                2          Filobasidium_stepposum\n",
      "SH1577471.08FU_KC005781_refs                2           Zymoseptoria_verkleyi\n",
      "SH1508982.08FU_AY740160_reps                2             Sporisorium_lepturi\n",
      "SH1548073.08FU_FJ839626_reps                2                    unidentified\n",
      "SH1598238.08FU_GU598104_reps                1            Puccinia_striiformis\n",
      "SH2320230.08FU_EF679384_refs                1       Cladosporium_ramotenellum\n",
      "SH2320203.08FU_HM148081_refs                1        Cladosporium_delicatulum\n",
      "SH2179746.08FU_KC584242_refs                1               Alternaria_aspera\n",
      "SH1549852.08FU_UDB024177_reps               1             Toninia_physaroides\n",
      "SH1688430.08FU_KC584249_refs                1        Alternaria_subcucurbitae\n",
      "SH1552276.08FU_EU707897_refs                1             Dissoconium_proteae\n",
      "SH1554929.08FU_JF773605_reps                1                    unidentified\n",
      "SH1651458.08FU_EF536346_refs                1                    unidentified\n",
      "SH1570378.08FU_MF767825_reps                1                Candida_albicans\n",
      "SH1650089.08FU_AF444329_refs                1      Cystofilobasidium_macerans\n",
      "SH1572792.08FU_EF679363_refs                1         Mycosphaerella_tassiana\n",
      "SH1576431.08FU_JQ747672_reps                1                    unidentified\n",
      "SH1645391.08FU_FJ415474_refs                1         Elasticomyces_elasticus\n",
      "SH1613737.08FU_GQ303275_refs                1                    unidentified\n",
      "SH1612303.08FU_EU715613_reps                1                    unidentified\n",
      "SH1610809.08FU_HG936981_reps                1      Holtermanniella_takashimae\n",
      "SH1505875.08FU_EU747141_reps                1                    unidentified\n",
      "SH1598253.08FU_AY956558_reps                1            Puccinia_striiformis\n",
      "SH1598251.08FU_KM391669_reps                1            Puccinia_brachypodii\n",
      "SH1505584.08FU_MF767833_reps                1                    unidentified\n",
      "{'s': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"s\": 0.0021853146853146855\n",
      "}\n",
      "{'s': 'Zymoseptoria_tritici'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "barcode05\n",
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    374                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             240                    unidentified\n",
      "SH1575705.08FU_AM887506_reps              230    Pyrenophora_tritici-repentis\n",
      "SH1538821.08FU_AB035719_reps              174            Bannoa_ogasawarensis\n",
      "SH1523966.08FU_HE820879_reps              137                    unidentified\n",
      "SH1654757.08FU_UDB014954_reps             132                    unidentified\n",
      "SH1558908.08FU_KF800655_reps              120                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              117          Knoxdaviesia_cecropiae\n",
      "SH1563343.08FU_FJ430715_reps               91                    unidentified\n",
      "SH1563342.08FU_UDB028691_reps              89                    unidentified\n",
      "SH1680040.08FU_KC992945_refs_singleton     74  Cystoagaricus_hirtosquamulosus\n",
      "SH1560162.08FU_AY225488_reps               68               Lignincola_laevis\n",
      "SH1541142.08FU_KR015769_reps               64                    unidentified\n",
      "SH1566136.08FU_KC992951_refs               57        Lacrymaria_subcinnamomea\n",
      "SH1523077.08FU_MH019936_reps               25                    unidentified\n",
      "SH1549852.08FU_UDB024177_reps              22             Toninia_physaroides\n",
      "SH1573885.08FU_MF594688_reps               19       Deltopyxis_triangulispora\n",
      "SH1853051.08FU_KF924559_refs_singleton     15         Conioscypha_minutispora\n",
      "SH1651458.08FU_EF536346_refs               15                    unidentified\n",
      "SH1508982.08FU_AY740160_reps               14             Sporisorium_lepturi\n",
      "SH1515312.08FU_KJ713215_reps               12                    unidentified\n",
      "SH1651389.08FU_KY782284_reps               11                    unidentified\n",
      "SH1649730.08FU_HE820818_reps                8                    unidentified\n",
      "SH1541118.08FU_KR016816_reps                8                    unidentified\n",
      "SH1575708.08FU_KP097730_reps                6                    unidentified\n",
      "SH1575712.08FU_MF491854_reps                5                    unidentified\n",
      "SH2127280.08FU_AF347034_refs                4           Alternaria_infectoria\n",
      "SH1575717.08FU_JN943658_reps                4                Drechslera_phlei\n",
      "SH1575716.08FU_EF452459_reps                4               Pyrenophora_bromi\n",
      "SH1575714.08FU_AY004801_reps                4                 Drechslera_poae\n",
      "SH2127282.08FU_AY762947_refs                3          Alternaria_oregonensis\n",
      "SH1505584.08FU_MF767833_reps                3                    unidentified\n",
      "SH1505864.08FU_JQ693660_refs                3       Alternaria_metachromatica\n",
      "SH1541263.08FU_KF436104_reps                2                    unidentified\n",
      "SH1572792.08FU_EF679363_refs                2         Mycosphaerella_tassiana\n",
      "SH1516182.08FU_FJ430752_reps                2                    unidentified\n",
      "SH1554929.08FU_JF773605_reps                2                    unidentified\n",
      "SH1554120.08FU_KU747561_reps                2                Xylaria_cubensis\n",
      "SH1546322.08FU_JF433037_refs                2         Fusarium_pseudensiforme\n",
      "SH1526441.08FU_FJ357315_refs                1           Alternaria_planifunda\n",
      "SH2179828.08FU_JF780937_refs                1              Alternaria_simsimi\n",
      "SH1505887.08FU_KU935649_reps                1          Alternaria_botryospora\n",
      "SH1507266.08FU_LC177377_reps                1                    unidentified\n",
      "SH1820842.08FU_AF444354_refs                1          Filobasidium_chernovii\n",
      "SH1524234.08FU_KX452348_reps                1                    unidentified\n",
      "SH1711498.08FU_JF433036_refs_singleton      1          Fusarium_kurunegalense\n",
      "SH1688438.08FU_KF156010_refs                1     Paradendryphiella_arenariae\n",
      "SH1687540.08FU_AB042642_refs                1             Leveillula_elaeagni\n",
      "SH1575706.08FU_JN712471_reps                1                    unidentified\n",
      "SH1664313.08FU_AY278839_refs_singleton      1              Alternaria_caricis\n",
      "SH1526444.08FU_KY883868_reps                1                Stemphylium_loti\n",
      "SH1575711.08FU_MH399405_reps                1                    unidentified\n",
      "SH1614538.08FU_KX515525_reps                1                    unidentified\n",
      "SH1613737.08FU_GQ303275_refs                1                    unidentified\n",
      "SH1609774.08FU_AF444379_refs                1             Dioszegia_hungarica\n",
      "SH1575719.08FU_AY004784_reps                1                    unidentified\n",
      "SH1566490.08FU_KX928959_reps                1              Stemphylium_solani\n",
      "SH1573369.08FU_KX953702_reps                1                    unidentified\n",
      "SH2320230.08FU_EF679384_refs                1       Cladosporium_ramotenellum\n",
      "{'s': 'Pyrenophora_tritici-repentis'}\n",
      "{\n",
      " \"s\": 0.10531135531135531\n",
      "}\n",
      "barcode06\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    510                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             393                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              295            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             208                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              158          Knoxdaviesia_cecropiae\n",
      "SH1558908.08FU_KF800655_reps              142                    unidentified\n",
      "SH1563343.08FU_FJ430715_reps              126                    unidentified\n",
      "SH1680040.08FU_KC992945_refs_singleton    114  Cystoagaricus_hirtosquamulosus\n",
      "SH1560162.08FU_AY225488_reps              103               Lignincola_laevis\n",
      "SH1563342.08FU_UDB028691_reps              91                    unidentified\n",
      "SH1566136.08FU_KC992951_refs               83        Lacrymaria_subcinnamomea\n",
      "SH1573885.08FU_MF594688_reps               29       Deltopyxis_triangulispora\n",
      "SH1651389.08FU_KY782284_reps               20                    unidentified\n",
      "SH1516349.08FU_KY242684_reps               10                    unidentified\n",
      "{'s': 'unknown'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "barcode07\n",
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    508                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             386                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              306            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             200                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              158          Knoxdaviesia_cecropiae\n",
      "SH1558908.08FU_KF800655_reps              153                    unidentified\n",
      "SH1563343.08FU_FJ430715_reps              138                    unidentified\n",
      "SH1680040.08FU_KC992945_refs_singleton    132  Cystoagaricus_hirtosquamulosus\n",
      "SH1560162.08FU_AY225488_reps              105               Lignincola_laevis\n",
      "SH1566136.08FU_KC992951_refs               88        Lacrymaria_subcinnamomea\n",
      "SH1563342.08FU_UDB028691_reps              64                    unidentified\n",
      "SH1573885.08FU_MF594688_reps               27       Deltopyxis_triangulispora\n",
      "SH1651389.08FU_KY782284_reps               14                    unidentified\n",
      "SH1516349.08FU_KY242684_reps               10                    unidentified\n",
      "SH1508982.08FU_AY740160_reps                1             Sporisorium_lepturi\n",
      "{'s': 'unknown'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "barcode08\n",
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    525                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             367                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              296            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             195                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              157          Knoxdaviesia_cecropiae\n",
      "SH1563343.08FU_FJ430715_reps              136                    unidentified\n",
      "SH1558908.08FU_KF800655_reps              133                    unidentified\n",
      "SH1560162.08FU_AY225488_reps              110               Lignincola_laevis\n",
      "SH1680040.08FU_KC992945_refs_singleton    109  Cystoagaricus_hirtosquamulosus\n",
      "SH1563342.08FU_UDB028691_reps              93                    unidentified\n",
      "SH1566136.08FU_KC992951_refs               69        Lacrymaria_subcinnamomea\n",
      "SH1573885.08FU_MF594688_reps               42       Deltopyxis_triangulispora\n",
      "SH1651389.08FU_KY782284_reps               20                    unidentified\n",
      "SH1516349.08FU_KY242684_reps                9                    unidentified\n",
      "SH1598232.08FU_HM057122_refs                4            Puccinia_striiformis\n",
      "SH1651458.08FU_EF536346_refs                2                    unidentified\n",
      "SH1598253.08FU_AY956558_reps                1            Puccinia_striiformis\n",
      "SH2076848.08FU_AF444349_refs                1           Filobasidium_oeirense\n",
      "{'s': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"s\": 0.0022036139268400176\n",
      "}\n",
      "barcode09\n",
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    526                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             363                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              324            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             203                    unidentified\n",
      "SH1563343.08FU_FJ430715_reps              137                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              134          Knoxdaviesia_cecropiae\n",
      "SH1558908.08FU_KF800655_reps              122                    unidentified\n",
      "SH1560162.08FU_AY225488_reps              116               Lignincola_laevis\n",
      "SH1680040.08FU_KC992945_refs_singleton    112  Cystoagaricus_hirtosquamulosus\n",
      "SH1563342.08FU_UDB028691_reps              84                    unidentified\n",
      "SH1566136.08FU_KC992951_refs               72        Lacrymaria_subcinnamomea\n",
      "SH1573885.08FU_MF594688_reps               32       Deltopyxis_triangulispora\n",
      "SH1598232.08FU_HM057122_refs               15            Puccinia_striiformis\n",
      "SH1651389.08FU_KY782284_reps               13                    unidentified\n",
      "SH1516349.08FU_KY242684_reps                8                    unidentified\n",
      "SH1508982.08FU_AY740160_reps                1             Sporisorium_lepturi\n",
      "SH1651458.08FU_EF536346_refs                1                    unidentified\n",
      "SH1598238.08FU_GU598104_reps                1            Puccinia_striiformis\n",
      "SH1572792.08FU_EF679363_refs                1         Mycosphaerella_tassiana\n",
      "SH2320385.08FU_EF679383_refs_singleton      1        Cladosporium_pseudiridis\n",
      "{'s': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"s\": 0.00706090026478376\n",
      "}\n",
      "{'s': 'Pyrenophora_tritici-repentis'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "barcode10\n",
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    563                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             346                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              278            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             170                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              158          Knoxdaviesia_cecropiae\n",
      "SH1558908.08FU_KF800655_reps              135                    unidentified\n",
      "SH1563343.08FU_FJ430715_reps              132                    unidentified\n",
      "SH1680040.08FU_KC992945_refs_singleton    111  Cystoagaricus_hirtosquamulosus\n",
      "SH1563342.08FU_UDB028691_reps             101                    unidentified\n",
      "SH1560162.08FU_AY225488_reps               84               Lignincola_laevis\n",
      "SH1566136.08FU_KC992951_refs               80        Lacrymaria_subcinnamomea\n",
      "SH1573885.08FU_MF594688_reps               25       Deltopyxis_triangulispora\n",
      "SH1651389.08FU_KY782284_reps               21                    unidentified\n",
      "SH1516349.08FU_KY242684_reps               10                    unidentified\n",
      "SH2320203.08FU_HM148081_refs                3        Cladosporium_delicatulum\n",
      "SH1572792.08FU_EF679363_refs                3         Mycosphaerella_tassiana\n",
      "SH2320230.08FU_EF679384_refs                2       Cladosporium_ramotenellum\n",
      "SH1650089.08FU_AF444329_refs                2      Cystofilobasidium_macerans\n",
      "SH1650091.08FU_KU163965_reps                2      Cystofilobasidium_macerans\n",
      "SH1820842.08FU_AF444354_refs                1          Filobasidium_chernovii\n",
      "SH1858974.08FU_JQ693639_refs                1                Alternaria_rosae\n",
      "SH1515312.08FU_KJ713215_reps                1                    unidentified\n",
      "SH1549852.08FU_UDB024177_reps               1             Toninia_physaroides\n",
      "SH1541142.08FU_KR015769_reps                1                    unidentified\n",
      "SH1528208.08FU_KF036588_refs                1        Vishniacozyma_carnescens\n",
      "SH1523966.08FU_HE820879_reps                1                    unidentified\n",
      "SH1523077.08FU_MH019936_reps                1                    unidentified\n",
      "SH2525086.08FU_DQ222455_refs                1          Filobasidium_stepposum\n",
      "{'s': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "{'s': 'Zymoseptoria_tritici'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "barcode11\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                        count                               s\n",
      "tname                                                                        \n",
      "SH1732842.08FU_FJ430779_refs_singleton    506                  Acidea_extrema\n",
      "SH1649638.08FU_UDB016769_refs             381                    unidentified\n",
      "SH1538821.08FU_AB035719_reps              274            Bannoa_ogasawarensis\n",
      "SH1654757.08FU_UDB014954_reps             189                    unidentified\n",
      "SH1563344.08FU_AM267268_reps              140          Knoxdaviesia_cecropiae\n",
      "SH1563343.08FU_FJ430715_reps              138                    unidentified\n",
      "SH1558908.08FU_KF800655_reps              136                    unidentified\n",
      "SH1680040.08FU_KC992945_refs_singleton    133  Cystoagaricus_hirtosquamulosus\n",
      "SH1563342.08FU_UDB028691_reps              96                    unidentified\n",
      "SH1560162.08FU_AY225488_reps               94               Lignincola_laevis\n",
      "SH1566136.08FU_KC992951_refs               87        Lacrymaria_subcinnamomea\n",
      "SH1573885.08FU_MF594688_reps               36       Deltopyxis_triangulispora\n",
      "SH1651389.08FU_KY782284_reps               17                    unidentified\n",
      "SH1572792.08FU_EF679363_refs               17         Mycosphaerella_tassiana\n",
      "SH2320230.08FU_EF679384_refs                7       Cladosporium_ramotenellum\n",
      "SH1516349.08FU_KY242684_reps                4                    unidentified\n",
      "SH1598232.08FU_HM057122_refs                3            Puccinia_striiformis\n",
      "SH2320203.08FU_HM148081_refs                3        Cladosporium_delicatulum\n",
      "SH2320385.08FU_EF679383_refs_singleton      3        Cladosporium_pseudiridis\n",
      "SH1523077.08FU_MH019936_reps                3                    unidentified\n",
      "SH1820842.08FU_AF444354_refs                2          Filobasidium_chernovii\n",
      "SH2127280.08FU_AF347034_refs                2           Alternaria_infectoria\n",
      "SH1650852.08FU_HE820847_reps                2                    unidentified\n",
      "SH2525086.08FU_DQ222455_refs                2          Filobasidium_stepposum\n",
      "SH1609774.08FU_AF444379_refs                2             Dioszegia_hungarica\n",
      "SH1528207.08FU_AF444469_refs                2         Vishniacozyma_victoriae\n",
      "SH1572842.08FU_KU821452_reps                2        Cladosporium_tenuissimum\n",
      "SH1650091.08FU_KU163965_reps                1      Cystofilobasidium_macerans\n",
      "SH1547054.08FU_AB520459_reps                1                    unidentified\n",
      "SH1541142.08FU_KR015769_reps                1                    unidentified\n",
      "SH2320220.08FU_HM148092_refs                1      Cladosporium_flabelliforme\n",
      "SH1541263.08FU_KF436104_reps                1                    unidentified\n",
      "SH2179882.08FU_KC584233_refs                1            Alternaria_hyacinthi\n",
      "SH1744747.08FU_JN906979_refs_singleton      1        Cladosporium_arthropodii\n",
      "SH1554120.08FU_KU747561_reps                1                Xylaria_cubensis\n",
      "SH1610809.08FU_HG936981_reps                1      Holtermanniella_takashimae\n",
      "SH1559775.08FU_KM041828_reps                1                    unidentified\n",
      "SH1687540.08FU_AB042642_refs                1             Leveillula_elaeagni\n",
      "SH1505864.08FU_JQ693660_refs                1       Alternaria_metachromatica\n",
      "SH1651458.08FU_EF536346_refs                1                    unidentified\n",
      "SH1505584.08FU_MF767833_reps                1                    unidentified\n",
      "{'s': 'Zymoseptoria_tritici'}\n",
      "{\n",
      " \"s\": 0.0\n",
      "}\n",
      "{'s': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"s\": 0.00130718954248366\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from collections import OrderedDict\n",
    "pd.set_option('display.max_rows', None)\n",
    "\n",
    "def assign_taxranks_results(mapping_df, tax_fn, ref_df_fn = False):\n",
    "    \"\"\"\n",
    "    This function assigns the taxonomic ranks for each hit in the mapping results dataframe.\n",
    "    It takes a mapping_df, taxonomy assignment file, and if required a reference dataframe filename.\n",
    "    Returns the mapping dataframe with assignment. \n",
    "    \"\"\"\n",
    "    for tname in mapping_df.index:\n",
    "        if ref_df_fn:\n",
    "            tmp_taxfileid = getquery_taxfileid(ref_df_fn, tname)\n",
    "        else:\n",
    "            tmp_taxfileid = tname\n",
    "        tmp_tax_dict = get_taxid_dict(tax_fn, tmp_taxfileid)\n",
    "        for key, value in tmp_tax_dict.items():\n",
    "            mapping_df.loc[tname, key] = value\n",
    "    return mapping_df\n",
    "\n",
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarises the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = OrderedDict()\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['s']:\n",
    "        tmps_df = pd.DataFrame(data=None)\n",
    "        hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict\n",
    "\n",
    "def get_taxid_dict(taxid_fn, taxfileid):\n",
    "    \"\"\"\n",
    "    Takes a taxonomy assignment file filename in the Qiime format and a taxonomic identifier.\n",
    "    Returns the a dictionary with the taxonomic assignment at each rank.\n",
    "    \"\"\"\n",
    "    tax_dict = {}\n",
    "    with open(taxid_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(taxfileid):\n",
    "                taxrankids = line.rstrip().split('\\t')[1].split(';')\n",
    "                for taxrank in taxrankids:\n",
    "                    tax_dict[taxrank.split('__')[0]] = taxrank.split('__')[1]\n",
    "    return tax_dict\n",
    "\n",
    "def mapping_results(fn, species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the query species: {species}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')\n",
    "    hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/custom_results/%s.json' % species)\n",
    "\n",
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df\n",
    "    \n",
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "barcode_list = []\n",
    "for entry in wheat_ref_df['Unnamed: 0']:\n",
    "#     if entry[-7:] != '-ccl031' and entry[-7:] != '-ccl029':\n",
    "#         test_species_list.append(entry)\n",
    "#     else:\n",
    "#         test_species_list.append(entry[:-7])\n",
    "#         print(entry[:-7])\n",
    "    barcode_list.append(entry)\n",
    "    \n",
    "counter = 0\n",
    "for barcode in barcode_list:\n",
    "    \n",
    "    print(barcode)\n",
    "    \n",
    "    #subsample tests species\n",
    "    fn_subsampling = {}\n",
    "    barcode = [barcode]\n",
    "    for x in barcode:\n",
    "        fn_subsampling[x] = wheat_ref_df[wheat_ref_df['Unnamed: 0'] == x]['path for use'].tolist()[0]\n",
    "        fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "    sub_reads_fn = {}\n",
    "    n_reads = 2000\n",
    "    for key, value in fn_subsampling.items():\n",
    "        barcode = key\n",
    "        in_fn = value\n",
    "        out_fn = os.path.join(subsampling_dir, F'{barcode}.{n_reads}.fasta')\n",
    "        subsamplereads(in_fn, out_fn, n_reads)\n",
    "        sub_reads_fn[barcode] = out_fn\n",
    "        \n",
    "    ###Map the reads\n",
    "    db_fn = qiime_db_fn\n",
    "    sub_db_mapping_fn = {}\n",
    "    for barcode, fasta_fn in sub_reads_fn.items():\n",
    "        db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "        out_fn = os.path.join(mapping_dir, F\"{db_name}.{barcode}.minimap2.paf\")\n",
    "        sub_db_mapping_fn[barcode] = out_fn\n",
    "        minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "        \n",
    "    ###Test out the summary results statistic for a single mapping result\n",
    "    mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[barcode])\n",
    "    mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "    print(mapping_results.loc[:,['count', 's']])\n",
    "#     print(\"\\n\\n\\n\")\n",
    "    \n",
    "    specs = []\n",
    "    if wheat_ref_df[wheat_ref_df['Unnamed: 0'] == barcode]['species2'].any():\n",
    "        specs = [wheat_ref_df[wheat_ref_df['Unnamed: 0'] == barcode].loc[counter,'species1'],wheat_ref_df[wheat_ref_df['Unnamed: 0'] == barcode].loc[counter,'species2']]\n",
    "    else:\n",
    "        specs = [wheat_ref_df[wheat_ref_df['Unnamed: 0'] == barcode].loc[counter,'species1']]\n",
    "\n",
    "    counter += 1\n",
    "    \n",
    "    for element in specs:\n",
    "        query_tax_dict = {'s': element}\n",
    "        print(query_tax_dict)\n",
    "        sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "        \n",
    "        print(json.dumps(sensitivity_dict, indent=1))\n",
    "#     print('\\n')\n",
    "#     with open('/media/MassStorage/tmp/TE/honours/analysis/Mapping/qiime_results/%s.json' % species, 'w+') as fp:\n",
    "#         json.dump(sensitivity_dict, fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### list of species in the max database\n",
    "max_species = ['Puccinia_striiformis',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = ['barcode01','barcode02','barcode03','barcode04',\n",
    "                 'barcode05','barcode06','barcode07','barcode08',\n",
    "                 'barcode09','barcode10','barcode11']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 2000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = os.path.abspath('../../analysis/Mapping/wheat')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)\n",
    "MC_READ_DIR = os.path.join(OUT_DIR, 'MC_READS')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)\n",
    "sub_db_fn = os.path.join('../../analysis/Mapping/gsref.subdb.fasta')\n",
    "new_db_fn = os.path.join('../../analysis/Mapping/gsref.db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'species1', 'species2', '# raw reads',\n",
       "       '# reads after homology filtering', '# reads after length filtering',\n",
       "       '# for use', 'path to raw reads', 'path to homology filtering',\n",
       "       'path to length filtering', 'path for use'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wheat_ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode01\n",
      "barcode02\n",
      "barcode03\n",
      "barcode04\n",
      "barcode05\n",
      "barcode06\n",
      "barcode07\n",
      "barcode08\n",
      "barcode09\n",
      "barcode10\n",
      "barcode11\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'barcode01': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta',\n",
       " 'barcode02': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta',\n",
       " 'barcode03': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode03/length_restricted_reads.fasta',\n",
       " 'barcode04': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode04/length_restricted_reads.fasta',\n",
       " 'barcode05': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta',\n",
       " 'barcode06': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode06/length_restricted_reads.fasta',\n",
       " 'barcode07': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode07/length_restricted_reads.fasta',\n",
       " 'barcode08': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode08/length_restricted_reads.fasta',\n",
       " 'barcode09': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode09/length_restricted_reads.fasta',\n",
       " 'barcode10': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode10/length_restricted_reads.fasta',\n",
       " 'barcode11': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode11/length_restricted_reads.fasta'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    print(x)\n",
    "    fn_subsampling[x] = wheat_ref_df[wheat_ref_df['Unnamed: 0'] == x]['path for use'].tolist()[0]\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode01\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.2000.fasta\n",
      "\n",
      "barcode02\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.2000.fasta\n",
      "\n",
      "barcode03\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode03/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode03/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode03.2000.fasta\n",
      "\n",
      "barcode04\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode04/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode04/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode04.2000.fasta\n",
      "\n",
      "barcode05\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.2000.fasta\n",
      "\n",
      "barcode06\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode06/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode06/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode06.2000.fasta\n",
      "\n",
      "barcode07\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode07/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode07/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode07.2000.fasta\n",
      "\n",
      "barcode08\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode08/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode08/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode08.2000.fasta\n",
      "\n",
      "barcode09\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode09/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode09/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode09.2000.fasta\n",
      "\n",
      "barcode10\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode10/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode10/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode10.2000.fasta\n",
      "\n",
      "barcode11\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode11/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode11/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode11.2000.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    print(key)\n",
    "    print(value)\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map with minimap against both databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../analysis/Mapping/gsref.subdb.fasta\n",
      "../../analysis/Mapping/gsref.db.fasta\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'../../analysis/Mapping/gsref.subdb.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb',\n",
       " '../../analysis/Mapping/gsref.db.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    print(x)\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode01.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode02.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode03.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode03.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode04.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode04.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode05.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode06.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode06.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode07.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode07.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode08.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode08.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode09.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode09.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode10.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode10.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode11.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode11.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode01.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode02.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode03.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode03.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode04.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode04.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode05.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode06.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode06.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode07.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode07.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode08.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode08.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode09.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode09.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode10.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode10.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode11.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode11.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results(fn, species,expected_species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the sample type expected: {expected_species[species]}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')\n",
    "    hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/custom_results/%s.json' % species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "expected_species = {'barcode01': \"Puccinia striiformis-tritici\",\n",
    "                   'barcode02': 'Zymoseptoria tritici',\n",
    "                   'barcode03': 'Healthy wheat sample',\n",
    "                   'barcode04': 'Puccinia striiformis-tritici AND Zymoseptoria tritici',\n",
    "                   'barcode05': 'Pyrenophora tritici-repentis',\n",
    "                   'barcode06': 'Healthy resistant wheat sample',\n",
    "                   'barcode07': 'Healthy susceptible wheat sample',\n",
    "                   'barcode08': 'Puccinia striiformis-tritici',\n",
    "                   'barcode09': 'Puccinia striiformis-tritici AND Pyrenophora tritici-repentis',\n",
    "                   'barcode10': 'Puccinia striiformis-tritici AND Zymoseptoria tritici',\n",
    "                   'barcode11': 'Zymoseptoria tritici AND Puccinia striiformis-tritici'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.641251\n",
      "blastobotrys_proliferans         0.160313\n",
      "saccharomyces_cerevisiae         0.073314\n",
      "cortinarius_globuliformis        0.040567\n",
      "rhodotorula_mucilaginosa         0.008309\n",
      "aspergillus_flavus               0.007820\n",
      "meyerozyma_guillermondii         0.006354\n",
      "puccinia_striiformis-tritici     0.005376\n",
      "aspergillus_unidentified         0.003910\n",
      "candida_albicans                 0.003910\n",
      "candida_unidentified             0.003910\n",
      "kluyveromyces_unidentified       0.003910\n",
      "zymoseptoria_tritici             0.003910\n",
      "wickerhamomyces_anomalus         0.003910\n",
      "entoleuca_unidentified           0.003421\n",
      "candida_metapsilosis             0.003421\n",
      "quambalaria_cyanescens           0.003421\n",
      "debaryomyces_unidentified        0.002933\n",
      "pyrenophora_tritici-repentis     0.002933\n",
      "dothiorella_vidmadera            0.002444\n",
      "oculimacula_yallundae-ccl031     0.002444\n",
      "kluyveromyces_marxianus          0.001955\n",
      "penicillium_chrysogenum          0.001466\n",
      "zygoascus_hellenicus             0.000978\n",
      "pichia_kudriavzevii              0.000978\n",
      "oculimacula_yallundae-ccl029     0.000978\n",
      "cladophialophora_unidentified    0.000978\n",
      "tuber_brumale                    0.000978\n",
      "asteroma_ccl068                  0.000489\n",
      "aspergillus_niger                0.000489\n",
      "yarrowia_lipolytica              0.000489\n",
      "clavispora_lusitaniae            0.000489\n",
      "yamadazyma_scolyti               0.000489\n",
      "candida_orthopsilosis            0.000489\n",
      "candida_parapsilosis             0.000489\n",
      "fusarium_oxysporum               0.000489\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Zymoseptoria tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zymoseptoria_tritici             0.517073\n",
      "debaryomyces_unidentified        0.122927\n",
      "cryptococcus_zero                0.108780\n",
      "cortinarius_globuliformis        0.060976\n",
      "pyrenophora_tritici-repentis     0.042439\n",
      "cladophialophora_unidentified    0.029756\n",
      "blastobotrys_proliferans         0.025366\n",
      "aspergillus_unidentified         0.024878\n",
      "quambalaria_cyanescens           0.007805\n",
      "aspergillus_flavus               0.006341\n",
      "saccharomyces_cerevisiae         0.006341\n",
      "candida_unidentified             0.005854\n",
      "meyerozyma_guillermondii         0.004390\n",
      "kluyveromyces_unidentified       0.004390\n",
      "rhodotorula_mucilaginosa         0.004390\n",
      "candida_albicans                 0.003902\n",
      "dothiorella_vidmadera            0.003902\n",
      "candida_orthopsilosis            0.002439\n",
      "penicillium_chrysogenum          0.002439\n",
      "candida_metapsilosis             0.002439\n",
      "candida_parapsilosis             0.001951\n",
      "entoleuca_unidentified           0.001951\n",
      "aspergillus_niger                0.001463\n",
      "wickerhamomyces_anomalus         0.001463\n",
      "zygoascus_hellenicus             0.000976\n",
      "kluyveromyces_marxianus          0.000976\n",
      "tuber_brumale                    0.000976\n",
      "yamadazyma_mexicana              0.000488\n",
      "pichia_membranifaciens           0.000488\n",
      "clavispora_lusitaniae            0.000488\n",
      "diaporthe_unidentified           0.000488\n",
      "oculimacula_yallundae-ccl029     0.000488\n",
      "oculimacula_yallundae-ccl031     0.000488\n",
      "scedosporium_boydii              0.000488\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Healthy wheat sample\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.342029\n",
      "candida_albicans                 0.270531\n",
      "cortinarius_globuliformis        0.135749\n",
      "blastobotrys_proliferans         0.086473\n",
      "saccharomyces_cerevisiae         0.013527\n",
      "galactomyces_geotrichum          0.013043\n",
      "aspergillus_flavus               0.013043\n",
      "meyerozyma_guillermondii         0.012560\n",
      "candida_unidentified             0.011111\n",
      "entoleuca_unidentified           0.011111\n",
      "rhodotorula_mucilaginosa         0.010145\n",
      "quambalaria_cyanescens           0.007729\n",
      "pyrenophora_tritici-repentis     0.006763\n",
      "dothiorella_vidmadera            0.005797\n",
      "diaporthe_ccl067                 0.005797\n",
      "candida_metapsilosis             0.005314\n",
      "zymoseptoria_tritici             0.005314\n",
      "wickerhamomyces_anomalus         0.003865\n",
      "zygoascus_hellenicus             0.003865\n",
      "candida_orthopsilosis            0.003865\n",
      "kluyveromyces_unidentified       0.003382\n",
      "candida_parapsilosis             0.003382\n",
      "aspergillus_unidentified         0.002899\n",
      "aspergillus_niger                0.002415\n",
      "yarrowia_lipolytica              0.002415\n",
      "clavispora_lusitaniae            0.002415\n",
      "oculimacula_yallundae-ccl029     0.002415\n",
      "kluyveromyces_marxianus          0.001932\n",
      "debaryomyces_unidentified        0.001932\n",
      "oculimacula_yallundae-ccl031     0.001932\n",
      "cladophialophora_unidentified    0.001449\n",
      "pichia_membranifaciens           0.000966\n",
      "penicillium_chrysogenum          0.000966\n",
      "puccinia_striiformis-tritici     0.000966\n",
      "yamadazyma_scolyti               0.000966\n",
      "tuber_brumale                    0.000966\n",
      "yamadazyma_mexicana              0.000483\n",
      "fusarium_oxysporum               0.000483\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici AND Zymoseptoria tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.326035\n",
      "cortinarius_globuliformis        0.200857\n",
      "candida_metapsilosis             0.087577\n",
      "zymoseptoria_tritici             0.060447\n",
      "blastobotrys_proliferans         0.050452\n",
      "kodamaea_ohmeri                  0.040933\n",
      "quambalaria_cyanescens           0.020466\n",
      "saccharomyces_cerevisiae         0.018563\n",
      "pyrenophora_tritici-repentis     0.016659\n",
      "rhodotorula_mucilaginosa         0.016183\n",
      "wickerhamomyces_anomalus         0.015707\n",
      "candida_albicans                 0.014755\n",
      "meyerozyma_guillermondii         0.010471\n",
      "diaporthe_unidentified           0.010471\n",
      "aspergillus_flavus               0.009519\n",
      "candida_parapsilosis             0.009043\n",
      "entoleuca_unidentified           0.009043\n",
      "candida_unidentified             0.008091\n",
      "cladophialophora_unidentified    0.007615\n",
      "yamadazyma_scolyti               0.006663\n",
      "aspergillus_unidentified         0.006663\n",
      "dothiorella_vidmadera            0.006663\n",
      "debaryomyces_unidentified        0.006188\n",
      "puccinia_striiformis-tritici     0.006188\n",
      "kluyveromyces_unidentified       0.006188\n",
      "kluyveromyces_marxianus          0.004760\n",
      "oculimacula_yallundae-ccl029     0.003808\n",
      "candida_orthopsilosis            0.003332\n",
      "aspergillus_niger                0.002380\n",
      "zygoascus_hellenicus             0.002380\n",
      "penicillium_chrysogenum          0.001904\n",
      "yarrowia_lipolytica              0.001428\n",
      "galactomyces_geotrichum          0.001428\n",
      "oculimacula_yallundae-ccl031     0.001428\n",
      "fusarium_oxysporum               0.001428\n",
      "yamadazyma_mexicana              0.000952\n",
      "clavispora_lusitaniae            0.000952\n",
      "asteroma_ccl068                  0.000952\n",
      "tuber_brumale                    0.000952\n",
      "pichia_kudriavzevii              0.000476\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pyrenophora tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pyrenophora_tritici-repentis     0.365292\n",
      "cryptococcus_zero                0.191755\n",
      "meyerozyma_guillermondii         0.112656\n",
      "cortinarius_globuliformis        0.080537\n",
      "candida_orthopsilosis            0.044104\n",
      "blastobotrys_proliferans         0.039310\n",
      "oculimacula_yallundae-ccl031     0.030201\n",
      "candida_unidentified             0.013423\n",
      "tuber_brumale                    0.012464\n",
      "quambalaria_cyanescens           0.011505\n",
      "saccharomyces_cerevisiae         0.011505\n",
      "candida_albicans                 0.009108\n",
      "aspergillus_flavus               0.007191\n",
      "dothiorella_vidmadera            0.006711\n",
      "rhodotorula_mucilaginosa         0.006232\n",
      "kluyveromyces_unidentified       0.006232\n",
      "wickerhamomyces_anomalus         0.005753\n",
      "entoleuca_unidentified           0.004794\n",
      "candida_metapsilosis             0.004794\n",
      "zymoseptoria_tritici             0.004794\n",
      "yamadazyma_scolyti               0.003835\n",
      "aspergillus_unidentified         0.003835\n",
      "kluyveromyces_marxianus          0.003356\n",
      "cladophialophora_unidentified    0.003356\n",
      "debaryomyces_unidentified        0.002876\n",
      "oculimacula_yallundae-ccl029     0.002876\n",
      "aspergillus_niger                0.002876\n",
      "candida_parapsilosis             0.001918\n",
      "galactomyces_geotrichum          0.001918\n",
      "yamadazyma_mexicana              0.001438\n",
      "pichia_kudriavzevii              0.000959\n",
      "zygoascus_hellenicus             0.000479\n",
      "clavispora_lusitaniae            0.000479\n",
      "penicillium_chrysogenum          0.000479\n",
      "yarrowia_lipolytica              0.000479\n",
      "asteroma_ccl060                  0.000479\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Healthy resistant wheat sample\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cortinarius_globuliformis        0.550556\n",
      "cryptococcus_zero                0.172714\n",
      "wickerhamomyces_anomalus         0.106434\n",
      "candida_parapsilosis             0.046928\n",
      "blastobotrys_proliferans         0.023706\n",
      "saccharomyces_cerevisiae         0.013546\n",
      "aspergillus_flavus               0.007741\n",
      "quambalaria_cyanescens           0.007257\n",
      "oculimacula_yallundae-ccl029     0.006773\n",
      "meyerozyma_guillermondii         0.005806\n",
      "debaryomyces_unidentified        0.005806\n",
      "candida_unidentified             0.004838\n",
      "dothiorella_vidmadera            0.004838\n",
      "entoleuca_unidentified           0.004354\n",
      "rhodotorula_mucilaginosa         0.004354\n",
      "candida_albicans                 0.003870\n",
      "zymoseptoria_tritici             0.003870\n",
      "kluyveromyces_unidentified       0.003387\n",
      "aspergillus_unidentified         0.003387\n",
      "cladophialophora_unidentified    0.002903\n",
      "kluyveromyces_marxianus          0.002419\n",
      "yamadazyma_scolyti               0.002419\n",
      "candida_metapsilosis             0.002419\n",
      "penicillium_chrysogenum          0.001935\n",
      "aspergillus_niger                0.001451\n",
      "yarrowia_lipolytica              0.000968\n",
      "zygoascus_hellenicus             0.000968\n",
      "galactomyces_geotrichum          0.000484\n",
      "yamadazyma_mexicana              0.000484\n",
      "clavispora_lusitaniae            0.000484\n",
      "tuber_brumale                    0.000484\n",
      "oculimacula_yallundae-ccl031     0.000484\n",
      "kodamaea_ohmeri                  0.000484\n",
      "diaporthe_ccl067                 0.000484\n",
      "diaporthe_unidentified           0.000484\n",
      "fusarium_oxysporum               0.000484\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Healthy susceptible wheat sample\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.334871\n",
      "cortinarius_globuliformis        0.141144\n",
      "dothiorella_vidmadera            0.087638\n",
      "blastobotrys_proliferans         0.074723\n",
      "yamadazyma_mexicana              0.063192\n",
      "meyerozyma_guillermondii         0.032288\n",
      "saccharomyces_cerevisiae         0.031827\n",
      "candida_unidentified             0.029059\n",
      "aspergillus_niger                0.027214\n",
      "rhodotorula_mucilaginosa         0.023985\n",
      "quambalaria_cyanescens           0.021218\n",
      "candida_albicans                 0.016605\n",
      "aspergillus_flavus               0.013376\n",
      "candida_orthopsilosis            0.011070\n",
      "kluyveromyces_unidentified       0.010609\n",
      "wickerhamomyces_anomalus         0.009225\n",
      "candida_metapsilosis             0.008303\n",
      "entoleuca_unidentified           0.008303\n",
      "kluyveromyces_marxianus          0.007380\n",
      "zygoascus_hellenicus             0.005996\n",
      "debaryomyces_unidentified        0.005535\n",
      "yamadazyma_scolyti               0.005074\n",
      "oculimacula_yallundae-ccl031     0.004613\n",
      "candida_parapsilosis             0.004613\n",
      "zymoseptoria_tritici             0.003690\n",
      "oculimacula_yallundae-ccl029     0.003229\n",
      "cladophialophora_unidentified    0.002306\n",
      "pichia_kudriavzevii              0.002306\n",
      "aspergillus_unidentified         0.002306\n",
      "clavispora_lusitaniae            0.001845\n",
      "asteroma_ccl068                  0.001845\n",
      "fusarium_oxysporum               0.001384\n",
      "tuber_brumale                    0.001384\n",
      "puccinia_striiformis-tritici     0.000923\n",
      "diaporthe_unidentified           0.000461\n",
      "yarrowia_lipolytica              0.000461\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.259155\n",
      "cortinarius_globuliformis        0.169484\n",
      "quambalaria_cyanescens           0.167136\n",
      "yamadazyma_scolyti               0.108451\n",
      "blastobotrys_proliferans         0.052582\n",
      "saccharomyces_cerevisiae         0.023005\n",
      "clavispora_lusitaniae            0.023005\n",
      "meyerozyma_guillermondii         0.021127\n",
      "candida_unidentified             0.017840\n",
      "wickerhamomyces_anomalus         0.015493\n",
      "entoleuca_unidentified           0.013615\n",
      "candida_metapsilosis             0.012676\n",
      "dothiorella_vidmadera            0.010798\n",
      "debaryomyces_unidentified        0.009859\n",
      "rhodotorula_mucilaginosa         0.009859\n",
      "candida_orthopsilosis            0.009390\n",
      "puccinia_striiformis-tritici     0.007512\n",
      "candida_albicans                 0.007512\n",
      "kluyveromyces_unidentified       0.007042\n",
      "candida_parapsilosis             0.006573\n",
      "aspergillus_flavus               0.006573\n",
      "zymoseptoria_tritici             0.005164\n",
      "oculimacula_yallundae-ccl029     0.004695\n",
      "kluyveromyces_marxianus          0.004225\n",
      "aspergillus_niger                0.004225\n",
      "oculimacula_yallundae-ccl031     0.004225\n",
      "aspergillus_unidentified         0.003756\n",
      "cladophialophora_unidentified    0.002817\n",
      "zygoascus_hellenicus             0.002347\n",
      "kodamaea_ohmeri                  0.001408\n",
      "pyrenophora_tritici-repentis     0.001408\n",
      "yarrowia_lipolytica              0.001408\n",
      "yamadazyma_mexicana              0.000939\n",
      "penicillium_chrysogenum          0.000939\n",
      "asteroma_ccl060                  0.000939\n",
      "fusarium_oxysporum               0.000469\n",
      "tuber_brumale                    0.000469\n",
      "pichia_membranifaciens           0.000469\n",
      "diaporthe_ccl067                 0.000469\n",
      "diaporthe_unidentified           0.000469\n",
      "pichia_kudriavzevii              0.000469\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici AND Pyrenophora tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.349318\n",
      "candida_unidentified             0.167842\n",
      "cortinarius_globuliformis        0.126939\n",
      "kluyveromyces_unidentified       0.105313\n",
      "blastobotrys_proliferans         0.061119\n",
      "meyerozyma_guillermondii         0.021157\n",
      "saccharomyces_cerevisiae         0.017866\n",
      "aspergillus_flavus               0.014575\n",
      "quambalaria_cyanescens           0.012694\n",
      "rhodotorula_mucilaginosa         0.012224\n",
      "dothiorella_vidmadera            0.011283\n",
      "entoleuca_unidentified           0.010813\n",
      "puccinia_striiformis-tritici     0.009403\n",
      "candida_albicans                 0.007992\n",
      "candida_orthopsilosis            0.006112\n",
      "debaryomyces_unidentified        0.006112\n",
      "wickerhamomyces_anomalus         0.006112\n",
      "yarrowia_lipolytica              0.006112\n",
      "oculimacula_yallundae-ccl031     0.005642\n",
      "candida_metapsilosis             0.005172\n",
      "candida_parapsilosis             0.003761\n",
      "kluyveromyces_marxianus          0.003761\n",
      "yamadazyma_scolyti               0.003291\n",
      "clavispora_lusitaniae            0.002821\n",
      "cladophialophora_unidentified    0.002821\n",
      "oculimacula_yallundae-ccl029     0.002821\n",
      "aspergillus_niger                0.002821\n",
      "zymoseptoria_tritici             0.002821\n",
      "yamadazyma_mexicana              0.002351\n",
      "zygoascus_hellenicus             0.001881\n",
      "penicillium_chrysogenum          0.001881\n",
      "pichia_kudriavzevii              0.001410\n",
      "kodamaea_ohmeri                  0.000940\n",
      "tuber_brumale                    0.000940\n",
      "aspergillus_unidentified         0.000940\n",
      "diaporthe_unidentified           0.000470\n",
      "asteroma_ccl060                  0.000470\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici AND Zymoseptoria tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.368396\n",
      "cortinarius_globuliformis        0.181818\n",
      "kluyveromyces_marxianus          0.134698\n",
      "blastobotrys_proliferans         0.084246\n",
      "penicillium_chrysogenum          0.038553\n",
      "saccharomyces_cerevisiae         0.029510\n",
      "zymoseptoria_tritici             0.014755\n",
      "rhodotorula_mucilaginosa         0.014755\n",
      "wickerhamomyces_anomalus         0.013803\n",
      "entoleuca_unidentified           0.012375\n",
      "meyerozyma_guillermondii         0.012375\n",
      "aspergillus_flavus               0.011899\n",
      "quambalaria_cyanescens           0.010471\n",
      "candida_albicans                 0.009519\n",
      "candida_unidentified             0.008091\n",
      "debaryomyces_unidentified        0.005712\n",
      "dothiorella_vidmadera            0.005236\n",
      "candida_parapsilosis             0.005236\n",
      "kluyveromyces_unidentified       0.004760\n",
      "candida_orthopsilosis            0.004284\n",
      "oculimacula_yallundae-ccl031     0.003808\n",
      "yamadazyma_mexicana              0.003332\n",
      "cladophialophora_unidentified    0.003332\n",
      "tuber_brumale                    0.002380\n",
      "candida_metapsilosis             0.002380\n",
      "yamadazyma_scolyti               0.002380\n",
      "oculimacula_yallundae-ccl029     0.001904\n",
      "aspergillus_unidentified         0.001428\n",
      "pyrenophora_tritici-repentis     0.001428\n",
      "aspergillus_niger                0.000952\n",
      "galactomyces_geotrichum          0.000952\n",
      "clavispora_lusitaniae            0.000952\n",
      "diaporthe_ccl067                 0.000952\n",
      "zygoascus_hellenicus             0.000952\n",
      "puccinia_striiformis-tritici     0.000476\n",
      "yarrowia_lipolytica              0.000476\n",
      "asteroma_ccl068                  0.000476\n",
      "asteroma_ccl060                  0.000476\n",
      "pichia_kudriavzevii              0.000476\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Zymoseptoria tritici AND Puccinia striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.332065\n",
      "rhodotorula_mucilaginosa         0.162702\n",
      "cortinarius_globuliformis        0.136061\n",
      "blastobotrys_proliferans         0.064225\n",
      "zygoascus_hellenicus             0.056613\n",
      "zymoseptoria_tritici             0.026166\n",
      "saccharomyces_cerevisiae         0.026166\n",
      "pichia_kudriavzevii              0.022360\n",
      "aspergillus_flavus               0.018554\n",
      "meyerozyma_guillermondii         0.014748\n",
      "candida_unidentified             0.014748\n",
      "candida_albicans                 0.013796\n",
      "dothiorella_vidmadera            0.012369\n",
      "wickerhamomyces_anomalus         0.009515\n",
      "asteroma_ccl060                  0.008088\n",
      "quambalaria_cyanescens           0.008088\n",
      "debaryomyces_unidentified        0.007136\n",
      "candida_parapsilosis             0.005709\n",
      "candida_metapsilosis             0.005709\n",
      "kluyveromyces_unidentified       0.005233\n",
      "kluyveromyces_marxianus          0.005233\n",
      "candida_orthopsilosis            0.004282\n",
      "entoleuca_unidentified           0.004282\n",
      "yamadazyma_mexicana              0.004282\n",
      "yamadazyma_scolyti               0.003806\n",
      "pyrenophora_tritici-repentis     0.003330\n",
      "tuber_brumale                    0.003330\n",
      "aspergillus_niger                0.002854\n",
      "puccinia_striiformis-tritici     0.002854\n",
      "oculimacula_yallundae-ccl031     0.002379\n",
      "aspergillus_unidentified         0.002379\n",
      "yarrowia_lipolytica              0.001903\n",
      "penicillium_chrysogenum          0.001903\n",
      "oculimacula_yallundae-ccl029     0.001427\n",
      "diaporthe_unidentified           0.000951\n",
      "cladophialophora_unidentified    0.000951\n",
      "diaporthe_ccl067                 0.000951\n",
      "kodamaea_ohmeri                  0.000951\n",
      "asteroma_ccl068                  0.000476\n",
      "fusarium_oxysporum               0.000476\n",
      "pichia_membranifaciens           0.000476\n",
      "galactomyces_geotrichum          0.000476\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for species, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species, expected_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.640039\n",
      "blastobotrys_proliferans         0.159766\n",
      "saccharomyces_cerevisiae         0.073551\n",
      "cortinarius_globuliformis        0.041403\n",
      "rhodotorula_mucilaginosa         0.008281\n",
      "aspergillus_flavus               0.007793\n",
      "meyerozyma_guillermondii         0.006819\n",
      "puccinia_striiformis-tritici     0.005358\n",
      "kluyveromyces_unidentified       0.005358\n",
      "candida_albicans                 0.004384\n",
      "candida_unidentified             0.004384\n",
      "wickerhamomyces_anomalus         0.004384\n",
      "zymoseptoria_tritici             0.003897\n",
      "aspergillus_unidentified         0.003897\n",
      "quambalaria_cyanescens           0.003897\n",
      "entoleuca_unidentified           0.003897\n",
      "kluyveromyces_marxianus          0.003410\n",
      "pyrenophora_tritici-repentis     0.002923\n",
      "debaryomyces_unidentified        0.002923\n",
      "dothiorella_vidmadera            0.002435\n",
      "oculimacula_yallundae-ccl031     0.002435\n",
      "penicillium_chrysogenum          0.001461\n",
      "zygoascus_hellenicus             0.000974\n",
      "tuber_brumale                    0.000974\n",
      "cladophialophora_unidentified    0.000974\n",
      "pichia_kudriavzevii              0.000974\n",
      "fusarium_oxysporum               0.000487\n",
      "oculimacula_yallundae-ccl029     0.000487\n",
      "yamadazyma_scolyti               0.000487\n",
      "clavispora_lusitaniae            0.000487\n",
      "candida_parapsilosis             0.000487\n",
      "yarrowia_lipolytica              0.000487\n",
      "asteroma_ccl068                  0.000487\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Zymoseptoria tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zymoseptoria_tritici             0.519080\n",
      "debaryomyces_unidentified        0.124266\n",
      "cryptococcus_zero                0.109100\n",
      "cortinarius_globuliformis        0.061155\n",
      "pyrenophora_tritici-repentis     0.042564\n",
      "cladophialophora_unidentified    0.029843\n",
      "blastobotrys_proliferans         0.025440\n",
      "aspergillus_unidentified         0.024951\n",
      "quambalaria_cyanescens           0.008317\n",
      "aspergillus_flavus               0.006360\n",
      "saccharomyces_cerevisiae         0.006360\n",
      "candida_unidentified             0.005382\n",
      "rhodotorula_mucilaginosa         0.004403\n",
      "dothiorella_vidmadera            0.004403\n",
      "meyerozyma_guillermondii         0.004403\n",
      "candida_albicans                 0.003914\n",
      "kluyveromyces_unidentified       0.003914\n",
      "candida_parapsilosis             0.003425\n",
      "penicillium_chrysogenum          0.002446\n",
      "entoleuca_unidentified           0.001957\n",
      "wickerhamomyces_anomalus         0.001957\n",
      "kluyveromyces_marxianus          0.000978\n",
      "zygoascus_hellenicus             0.000978\n",
      "tuber_brumale                    0.000978\n",
      "diaporthe_unidentified           0.000489\n",
      "scedosporium_boydii              0.000489\n",
      "clavispora_lusitaniae            0.000489\n",
      "oculimacula_yallundae-ccl029     0.000489\n",
      "oculimacula_yallundae-ccl031     0.000489\n",
      "yamadazyma_mexicana              0.000489\n",
      "pichia_membranifaciens           0.000489\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Healthy wheat sample\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.344509\n",
      "candida_albicans                 0.275024\n",
      "cortinarius_globuliformis        0.137026\n",
      "blastobotrys_proliferans         0.087949\n",
      "saccharomyces_cerevisiae         0.014091\n",
      "galactomyces_geotrichum          0.014091\n",
      "aspergillus_flavus               0.013120\n",
      "meyerozyma_guillermondii         0.013120\n",
      "candida_unidentified             0.011176\n",
      "entoleuca_unidentified           0.011176\n",
      "rhodotorula_mucilaginosa         0.010204\n",
      "quambalaria_cyanescens           0.008260\n",
      "pyrenophora_tritici-repentis     0.006803\n",
      "diaporthe_ccl067                 0.005831\n",
      "dothiorella_vidmadera            0.005831\n",
      "zymoseptoria_tritici             0.005345\n",
      "zygoascus_hellenicus             0.003887\n",
      "wickerhamomyces_anomalus         0.003887\n",
      "kluyveromyces_unidentified       0.003401\n",
      "candida_parapsilosis             0.003401\n",
      "aspergillus_unidentified         0.002915\n",
      "oculimacula_yallundae-ccl029     0.002430\n",
      "debaryomyces_unidentified        0.001944\n",
      "oculimacula_yallundae-ccl031     0.001944\n",
      "yarrowia_lipolytica              0.001944\n",
      "kluyveromyces_marxianus          0.001944\n",
      "clavispora_lusitaniae            0.001458\n",
      "cladophialophora_unidentified    0.001458\n",
      "pichia_membranifaciens           0.000972\n",
      "puccinia_striiformis-tritici     0.000972\n",
      "penicillium_chrysogenum          0.000972\n",
      "yamadazyma_scolyti               0.000972\n",
      "tuber_brumale                    0.000972\n",
      "fusarium_oxysporum               0.000486\n",
      "yamadazyma_mexicana              0.000486\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici AND Zymoseptoria tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.339934\n",
      "cortinarius_globuliformis        0.214993\n",
      "kodamaea_ohmeri                  0.062706\n",
      "zymoseptoria_tritici             0.060349\n",
      "blastobotrys_proliferans         0.052334\n",
      "candida_albicans                 0.021216\n",
      "quambalaria_cyanescens           0.021216\n",
      "saccharomyces_cerevisiae         0.020273\n",
      "wickerhamomyces_anomalus         0.019331\n",
      "rhodotorula_mucilaginosa         0.017445\n",
      "pyrenophora_tritici-repentis     0.016502\n",
      "candida_parapsilosis             0.014144\n",
      "diaporthe_unidentified           0.013201\n",
      "entoleuca_unidentified           0.012258\n",
      "candida_unidentified             0.011787\n",
      "meyerozyma_guillermondii         0.010844\n",
      "aspergillus_flavus               0.009901\n",
      "debaryomyces_unidentified        0.009430\n",
      "kluyveromyces_unidentified       0.008487\n",
      "yamadazyma_scolyti               0.008015\n",
      "aspergillus_unidentified         0.007544\n",
      "cladophialophora_unidentified    0.007544\n",
      "dothiorella_vidmadera            0.007544\n",
      "kluyveromyces_marxianus          0.006601\n",
      "puccinia_striiformis-tritici     0.006129\n",
      "oculimacula_yallundae-ccl029     0.003772\n",
      "zygoascus_hellenicus             0.002829\n",
      "penicillium_chrysogenum          0.001886\n",
      "yamadazyma_mexicana              0.001886\n",
      "clavispora_lusitaniae            0.001414\n",
      "oculimacula_yallundae-ccl031     0.001414\n",
      "tuber_brumale                    0.001414\n",
      "fusarium_oxysporum               0.001414\n",
      "yarrowia_lipolytica              0.001414\n",
      "galactomyces_geotrichum          0.001414\n",
      "asteroma_ccl068                  0.000943\n",
      "pichia_kudriavzevii              0.000471\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pyrenophora tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pyrenophora_tritici-repentis     0.364768\n",
      "cryptococcus_zero                0.198181\n",
      "meyerozyma_guillermondii         0.130206\n",
      "cortinarius_globuliformis        0.083772\n",
      "blastobotrys_proliferans         0.040689\n",
      "oculimacula_yallundae-ccl031     0.033030\n",
      "candida_unidentified             0.015318\n",
      "tuber_brumale                    0.012925\n",
      "quambalaria_cyanescens           0.012925\n",
      "candida_albicans                 0.012446\n",
      "saccharomyces_cerevisiae         0.011967\n",
      "kluyveromyces_unidentified       0.008138\n",
      "aspergillus_flavus               0.007180\n",
      "wickerhamomyces_anomalus         0.006702\n",
      "rhodotorula_mucilaginosa         0.006702\n",
      "dothiorella_vidmadera            0.006702\n",
      "kluyveromyces_marxianus          0.005744\n",
      "zymoseptoria_tritici             0.005744\n",
      "entoleuca_unidentified           0.004787\n",
      "yamadazyma_scolyti               0.004787\n",
      "candida_parapsilosis             0.004308\n",
      "aspergillus_unidentified         0.004308\n",
      "oculimacula_yallundae-ccl029     0.003830\n",
      "cladophialophora_unidentified    0.003830\n",
      "debaryomyces_unidentified        0.003351\n",
      "galactomyces_geotrichum          0.001915\n",
      "yamadazyma_mexicana              0.001915\n",
      "pichia_kudriavzevii              0.000957\n",
      "yarrowia_lipolytica              0.000957\n",
      "zygoascus_hellenicus             0.000479\n",
      "clavispora_lusitaniae            0.000479\n",
      "asteroma_ccl060                  0.000479\n",
      "penicillium_chrysogenum          0.000479\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Healthy resistant wheat sample\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cortinarius_globuliformis        0.550024\n",
      "cryptococcus_zero                0.172547\n",
      "wickerhamomyces_anomalus         0.106815\n",
      "candida_parapsilosis             0.046883\n",
      "blastobotrys_proliferans         0.024166\n",
      "saccharomyces_cerevisiae         0.013533\n",
      "aspergillus_flavus               0.008217\n",
      "oculimacula_yallundae-ccl029     0.007733\n",
      "quambalaria_cyanescens           0.007250\n",
      "meyerozyma_guillermondii         0.005800\n",
      "debaryomyces_unidentified        0.005800\n",
      "candida_unidentified             0.005317\n",
      "dothiorella_vidmadera            0.004833\n",
      "entoleuca_unidentified           0.004350\n",
      "candida_albicans                 0.004350\n",
      "rhodotorula_mucilaginosa         0.004350\n",
      "zymoseptoria_tritici             0.003867\n",
      "cladophialophora_unidentified    0.003383\n",
      "kluyveromyces_unidentified       0.003383\n",
      "aspergillus_unidentified         0.003383\n",
      "yamadazyma_scolyti               0.002900\n",
      "kluyveromyces_marxianus          0.002417\n",
      "penicillium_chrysogenum          0.001933\n",
      "yarrowia_lipolytica              0.000967\n",
      "yamadazyma_mexicana              0.000967\n",
      "zygoascus_hellenicus             0.000967\n",
      "clavispora_lusitaniae            0.000483\n",
      "fusarium_oxysporum               0.000483\n",
      "tuber_brumale                    0.000483\n",
      "oculimacula_yallundae-ccl031     0.000483\n",
      "kodamaea_ohmeri                  0.000483\n",
      "diaporthe_ccl067                 0.000483\n",
      "diaporthe_unidentified           0.000483\n",
      "galactomyces_geotrichum          0.000483\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Healthy susceptible wheat sample\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.340720\n",
      "cortinarius_globuliformis        0.142659\n",
      "dothiorella_vidmadera            0.093259\n",
      "blastobotrys_proliferans         0.076177\n",
      "yamadazyma_mexicana              0.072484\n",
      "candida_unidentified             0.034164\n",
      "meyerozyma_guillermondii         0.033703\n",
      "saccharomyces_cerevisiae         0.033241\n",
      "rhodotorula_mucilaginosa         0.024469\n",
      "quambalaria_cyanescens           0.022622\n",
      "candida_albicans                 0.020776\n",
      "aspergillus_flavus               0.015235\n",
      "kluyveromyces_unidentified       0.011080\n",
      "wickerhamomyces_anomalus         0.010619\n",
      "entoleuca_unidentified           0.009234\n",
      "zygoascus_hellenicus             0.006464\n",
      "kluyveromyces_marxianus          0.006464\n",
      "candida_parapsilosis             0.005540\n",
      "oculimacula_yallundae-ccl031     0.005540\n",
      "yamadazyma_scolyti               0.005078\n",
      "debaryomyces_unidentified        0.005078\n",
      "aspergillus_unidentified         0.004155\n",
      "zymoseptoria_tritici             0.003693\n",
      "oculimacula_yallundae-ccl029     0.003232\n",
      "pichia_kudriavzevii              0.002770\n",
      "cladophialophora_unidentified    0.002308\n",
      "asteroma_ccl068                  0.001847\n",
      "clavispora_lusitaniae            0.001847\n",
      "fusarium_oxysporum               0.001385\n",
      "tuber_brumale                    0.001385\n",
      "puccinia_striiformis-tritici     0.000923\n",
      "penicillium_chrysogenum          0.000923\n",
      "diaporthe_unidentified           0.000462\n",
      "yarrowia_lipolytica              0.000462\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.263059\n",
      "cortinarius_globuliformis        0.174118\n",
      "quambalaria_cyanescens           0.168471\n",
      "yamadazyma_scolyti               0.108706\n",
      "blastobotrys_proliferans         0.051294\n",
      "saccharomyces_cerevisiae         0.025882\n",
      "clavispora_lusitaniae            0.023059\n",
      "meyerozyma_guillermondii         0.022118\n",
      "candida_unidentified             0.020706\n",
      "wickerhamomyces_anomalus         0.016000\n",
      "entoleuca_unidentified           0.014588\n",
      "dothiorella_vidmadera            0.011765\n",
      "debaryomyces_unidentified        0.010353\n",
      "rhodotorula_mucilaginosa         0.009882\n",
      "candida_parapsilosis             0.009882\n",
      "candida_albicans                 0.008000\n",
      "kluyveromyces_unidentified       0.008000\n",
      "aspergillus_flavus               0.007529\n",
      "puccinia_striiformis-tritici     0.007529\n",
      "zymoseptoria_tritici             0.005647\n",
      "kluyveromyces_marxianus          0.004235\n",
      "aspergillus_unidentified         0.004235\n",
      "oculimacula_yallundae-ccl029     0.004235\n",
      "oculimacula_yallundae-ccl031     0.004235\n",
      "cladophialophora_unidentified    0.003294\n",
      "zygoascus_hellenicus             0.002353\n",
      "yarrowia_lipolytica              0.001412\n",
      "yamadazyma_mexicana              0.001412\n",
      "pyrenophora_tritici-repentis     0.001412\n",
      "kodamaea_ohmeri                  0.001412\n",
      "penicillium_chrysogenum          0.001412\n",
      "asteroma_ccl060                  0.000941\n",
      "pichia_kudriavzevii              0.000471\n",
      "tuber_brumale                    0.000471\n",
      "diaporthe_ccl067                 0.000471\n",
      "diaporthe_unidentified           0.000471\n",
      "fusarium_oxysporum               0.000471\n",
      "pichia_membranifaciens           0.000471\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici AND Pyrenophora tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.351633\n",
      "candida_unidentified             0.170847\n",
      "cortinarius_globuliformis        0.129200\n",
      "kluyveromyces_unidentified       0.107430\n",
      "blastobotrys_proliferans         0.060104\n",
      "meyerozyma_guillermondii         0.021770\n",
      "saccharomyces_cerevisiae         0.018457\n",
      "aspergillus_flavus               0.016091\n",
      "quambalaria_cyanescens           0.012778\n",
      "rhodotorula_mucilaginosa         0.012305\n",
      "dothiorella_vidmadera            0.011832\n",
      "entoleuca_unidentified           0.010885\n",
      "puccinia_striiformis-tritici     0.009465\n",
      "candida_albicans                 0.008992\n",
      "wickerhamomyces_anomalus         0.006626\n",
      "yarrowia_lipolytica              0.006152\n",
      "oculimacula_yallundae-ccl031     0.006152\n",
      "debaryomyces_unidentified        0.006152\n",
      "kluyveromyces_marxianus          0.004259\n",
      "candida_parapsilosis             0.003313\n",
      "yamadazyma_scolyti               0.003313\n",
      "cladophialophora_unidentified    0.002840\n",
      "clavispora_lusitaniae            0.002840\n",
      "zymoseptoria_tritici             0.002840\n",
      "oculimacula_yallundae-ccl029     0.002840\n",
      "yamadazyma_mexicana              0.002366\n",
      "zygoascus_hellenicus             0.001893\n",
      "penicillium_chrysogenum          0.001420\n",
      "pichia_kudriavzevii              0.001420\n",
      "kodamaea_ohmeri                  0.000947\n",
      "tuber_brumale                    0.000947\n",
      "aspergillus_unidentified         0.000947\n",
      "asteroma_ccl060                  0.000473\n",
      "diaporthe_unidentified           0.000473\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia striiformis-tritici AND Zymoseptoria tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.369928\n",
      "cortinarius_globuliformis        0.182816\n",
      "kluyveromyces_marxianus          0.136993\n",
      "blastobotrys_proliferans         0.084010\n",
      "penicillium_chrysogenum          0.038663\n",
      "saccharomyces_cerevisiae         0.030072\n",
      "rhodotorula_mucilaginosa         0.015274\n",
      "zymoseptoria_tritici             0.014797\n",
      "wickerhamomyces_anomalus         0.014320\n",
      "meyerozyma_guillermondii         0.012888\n",
      "entoleuca_unidentified           0.011933\n",
      "aspergillus_flavus               0.011933\n",
      "quambalaria_cyanescens           0.010501\n",
      "candida_albicans                 0.010501\n",
      "candida_unidentified             0.007160\n",
      "debaryomyces_unidentified        0.006205\n",
      "candida_parapsilosis             0.006205\n",
      "dothiorella_vidmadera            0.004773\n",
      "kluyveromyces_unidentified       0.004773\n",
      "oculimacula_yallundae-ccl031     0.003819\n",
      "cladophialophora_unidentified    0.003341\n",
      "tuber_brumale                    0.002864\n",
      "yamadazyma_mexicana              0.002864\n",
      "yamadazyma_scolyti               0.002387\n",
      "oculimacula_yallundae-ccl029     0.001909\n",
      "pyrenophora_tritici-repentis     0.001432\n",
      "galactomyces_geotrichum          0.000955\n",
      "aspergillus_unidentified         0.000955\n",
      "zygoascus_hellenicus             0.000955\n",
      "clavispora_lusitaniae            0.000955\n",
      "diaporthe_ccl067                 0.000955\n",
      "pichia_kudriavzevii              0.000477\n",
      "puccinia_striiformis-tritici     0.000477\n",
      "yarrowia_lipolytica              0.000477\n",
      "asteroma_ccl068                  0.000477\n",
      "asteroma_ccl060                  0.000477\n",
      "pichia_membranifaciens           0.000477\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Zymoseptoria tritici AND Puccinia striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.334597\n",
      "rhodotorula_mucilaginosa         0.163033\n",
      "cortinarius_globuliformis        0.136967\n",
      "blastobotrys_proliferans         0.063981\n",
      "zygoascus_hellenicus             0.056398\n",
      "saccharomyces_cerevisiae         0.027488\n",
      "zymoseptoria_tritici             0.026540\n",
      "pichia_kudriavzevii              0.022275\n",
      "aspergillus_flavus               0.018483\n",
      "meyerozyma_guillermondii         0.015640\n",
      "candida_unidentified             0.015640\n",
      "candida_albicans                 0.014692\n",
      "dothiorella_vidmadera            0.012796\n",
      "wickerhamomyces_anomalus         0.009953\n",
      "asteroma_ccl060                  0.008531\n",
      "debaryomyces_unidentified        0.008531\n",
      "quambalaria_cyanescens           0.008057\n",
      "kluyveromyces_unidentified       0.006635\n",
      "candida_parapsilosis             0.006161\n",
      "kluyveromyces_marxianus          0.005687\n",
      "entoleuca_unidentified           0.004739\n",
      "yamadazyma_mexicana              0.004265\n",
      "yamadazyma_scolyti               0.003791\n",
      "pyrenophora_tritici-repentis     0.003318\n",
      "tuber_brumale                    0.002844\n",
      "aspergillus_unidentified         0.002844\n",
      "puccinia_striiformis-tritici     0.002844\n",
      "oculimacula_yallundae-ccl031     0.002370\n",
      "penicillium_chrysogenum          0.001896\n",
      "yarrowia_lipolytica              0.001896\n",
      "oculimacula_yallundae-ccl029     0.001422\n",
      "cladophialophora_unidentified    0.000948\n",
      "kodamaea_ohmeri                  0.000948\n",
      "diaporthe_ccl067                 0.000948\n",
      "diaporthe_unidentified           0.000948\n",
      "galactomyces_geotrichum          0.000474\n",
      "pichia_membranifaciens           0.000474\n",
      "asteroma_ccl068                  0.000474\n",
      "fusarium_oxysporum               0.000474\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running against a database that have ['Candida_orthopsilosis', 'Candida_metapsilosis', 'Aspergillus_niger'] deleted\n",
    "for species, hit_fn in sub_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species, expected_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
