{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook to analyzse the efficiency of minimap2 mapping for Oxford Nanopore MinION reads against Consensus Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv')\n",
    "max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta')\n",
    "taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_qiime.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#threads to use\n",
    "threads = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = os.path.abspath('../../analysis/Mapping_mock_gsref')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### list of species in the max database\n",
    "max_species = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']\n",
    "\n",
    "species_delete = [\n",
    "# 'Candida_orthopsilosis',\n",
    "#                  'Candida_metapsilosis',\n",
    "#                  'Aspergillus_niger'\n",
    "]\n",
    "\n",
    "mock_community = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = pd.read_csv(reference_dataframe_fn)\n",
    "ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['puccinia_striiformis-tritici',\n",
       " 'zymoseptoria_tritici',\n",
       " 'pyrenophora_tritici-repentis',\n",
       " 'fusarium_oxysporum',\n",
       " 'tuber_brumale',\n",
       " 'cortinarius_globuliformis',\n",
       " 'aspergillus_niger',\n",
       " 'clavispora_lusitaniae',\n",
       " 'kluyveromyces_unidentified',\n",
       " 'penicillium_chrysogenum',\n",
       " 'rhodotorula_mucilaginosa',\n",
       " 'scedosporium_boydii',\n",
       " 'blastobotrys_proliferans',\n",
       " 'debaryomyces_unidentified',\n",
       " 'galactomyces_geotrichum',\n",
       " 'kodamaea_ohmeri',\n",
       " 'meyerozyma_guilliermondii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'yamadazyma_mexicana',\n",
       " 'yamadazyma_scolyti',\n",
       " 'yarrowia_lipolytica',\n",
       " 'zygoascus_hellenicus',\n",
       " 'aspergillus_flavus',\n",
       " 'cryptococcus_zero',\n",
       " 'aspergillus_unidentified',\n",
       " 'diaporthe_ccl067',\n",
       " 'diaporthe_unidentified',\n",
       " 'oculimacula_yallundae-ccl031',\n",
       " 'oculimacula_yallundae-ccl029',\n",
       " 'dothiorella_vidmadera',\n",
       " 'quambalaria_cyanescens',\n",
       " 'entoleuca_unidentified',\n",
       " 'asteroma_ccl060',\n",
       " 'asteroma_ccl068',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'cladophialophora_unidentified',\n",
       " 'candida_albicans',\n",
       " 'candida_metapsilosis',\n",
       " 'candida_orthopsilosis',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_unidentified',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.name_species.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_fn = os.path.join(OUT_DIR, 'gsref.db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_list = []\n",
    "old_db_list = []\n",
    "for seq in SeqIO.parse(max_custom_database_fn, 'fasta'):\n",
    "    old_db_list.append(seq.id)\n",
    "    if seq.id.lower() in ref_df.name_species.tolist():\n",
    "        #print(seq.id)\n",
    "        seq.id = seq.name = seq.description = seq.id.lower()\n",
    "        new_db_list.append(seq)\n",
    "    else:\n",
    "        print(seq.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(new_db_list) == len(old_db_list):\n",
    "    SeqIO.write(new_db_list, new_db_fn, 'fasta')\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sub_db_fn = os.path.join(OUT_DIR, 'gsref.subdb.fasta')\n",
    "sub_db_list = []\n",
    "for seq in new_db_list:\n",
    "    if seq.id not in [x.lower() for x in species_delete]:\n",
    "        sub_db_list.append(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(sub_db_list) + len(species_delete) == len(new_db_list):\n",
    "    SeqIO.write(sub_db_list, sub_db_fn, 'fasta' )\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['puccinia_striiformis-tritici',\n",
       " 'zymoseptoria_tritici',\n",
       " 'pyrenophora_tritici-repentis',\n",
       " 'fusarium_oxysporum',\n",
       " 'tuber_brumale',\n",
       " 'cortinarius_globuliformis',\n",
       " 'aspergillus_niger',\n",
       " 'clavispora_lusitaniae',\n",
       " 'kluyveromyces_unidentified',\n",
       " 'penicillium_chrysogenum',\n",
       " 'rhodotorula_mucilaginosa',\n",
       " 'scedosporium_boydii',\n",
       " 'blastobotrys_proliferans',\n",
       " 'debaryomyces_unidentified',\n",
       " 'galactomyces_geotrichum',\n",
       " 'kodamaea_ohmeri',\n",
       " 'meyerozyma_guilliermondii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'yamadazyma_mexicana',\n",
       " 'yamadazyma_scolyti',\n",
       " 'yarrowia_lipolytica',\n",
       " 'zygoascus_hellenicus',\n",
       " 'aspergillus_flavus',\n",
       " 'cryptococcus_zero',\n",
       " 'aspergillus_unidentified',\n",
       " 'diaporthe_ccl067',\n",
       " 'diaporthe_unidentified',\n",
       " 'oculimacula_yallundae-ccl031',\n",
       " 'oculimacula_yallundae-ccl029',\n",
       " 'dothiorella_vidmadera',\n",
       " 'quambalaria_cyanescens',\n",
       " 'entoleuca_unidentified',\n",
       " 'asteroma_ccl060',\n",
       " 'asteroma_ccl068',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'cladophialophora_unidentified',\n",
       " 'candida_albicans',\n",
       " 'candida_metapsilosis',\n",
       " 'candida_orthopsilosis',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_unidentified',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x.id for x in sub_db_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = [x.lower() for x in mock_community]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['puccinia_striiformis-tritici',\n",
       " 'zymoseptoria_tritici',\n",
       " 'pyrenophora_tritici-repentis',\n",
       " 'fusarium_oxysporum',\n",
       " 'tuber_brumale',\n",
       " 'cortinarius_globuliformis',\n",
       " 'aspergillus_niger',\n",
       " 'clavispora_lusitaniae',\n",
       " 'kluyveromyces_unidentified',\n",
       " 'penicillium_chrysogenum',\n",
       " 'rhodotorula_mucilaginosa',\n",
       " 'scedosporium_boydii',\n",
       " 'blastobotrys_proliferans',\n",
       " 'debaryomyces_unidentified',\n",
       " 'galactomyces_geotrichum',\n",
       " 'kodamaea_ohmeri',\n",
       " 'meyerozyma_guilliermondii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'yamadazyma_mexicana',\n",
       " 'yamadazyma_scolyti',\n",
       " 'yarrowia_lipolytica',\n",
       " 'zygoascus_hellenicus',\n",
       " 'aspergillus_flavus',\n",
       " 'cryptococcus_zero',\n",
       " 'aspergillus_unidentified',\n",
       " 'diaporthe_ccl067',\n",
       " 'diaporthe_unidentified',\n",
       " 'oculimacula_yallundae-ccl031',\n",
       " 'oculimacula_yallundae-ccl029',\n",
       " 'dothiorella_vidmadera',\n",
       " 'quambalaria_cyanescens',\n",
       " 'entoleuca_unidentified',\n",
       " 'asteroma_ccl060',\n",
       " 'asteroma_ccl068',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'cladophialophora_unidentified',\n",
       " 'candida_albicans',\n",
       " 'candida_metapsilosis',\n",
       " 'candida_orthopsilosis',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_unidentified',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mock_community"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 15000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "MC_READ_DIR = os.path.join(OUT_DIR, 'MC_READS')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'species', 'genus', 'family', 'order', 'class', 'phylum',\n",
       "       'kingdom', '# raw reads', '# reads after homology filtering',\n",
       "       '# reads after length filtering', '# for use', 'path to raw reads',\n",
       "       'path to homology filtering', 'path to length filtering',\n",
       "       'path for use', 'name_species'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'puccinia_striiformis-tritici': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode01/length_restricted_for_use.fasta',\n",
       " 'zymoseptoria_tritici': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_for_use.fasta',\n",
       " 'pyrenophora_tritici-repentis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode03/length_restricted_for_use.fasta',\n",
       " 'fusarium_oxysporum': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode04/length_restricted_for_use.fasta',\n",
       " 'tuber_brumale': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode05/length_restricted_for_use.fasta',\n",
       " 'cortinarius_globuliformis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode06/length_restricted_for_use.fasta',\n",
       " 'aspergillus_niger': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta',\n",
       " 'clavispora_lusitaniae': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode08/length_restricted_for_use.fasta',\n",
       " 'kluyveromyces_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode09/length_restricted_for_use.fasta',\n",
       " 'penicillium_chrysogenum': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta',\n",
       " 'rhodotorula_mucilaginosa': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode11/length_restricted_for_use.fasta',\n",
       " 'scedosporium_boydii': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode12/length_restricted_for_use.fasta',\n",
       " 'blastobotrys_proliferans': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode01/length_restricted_for_use.fasta',\n",
       " 'debaryomyces_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode02/length_restricted_for_use.fasta',\n",
       " 'galactomyces_geotrichum': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_for_use.fasta',\n",
       " 'kodamaea_ohmeri': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode04/length_restricted_for_use.fasta',\n",
       " 'meyerozyma_guilliermondii': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode05/length_restricted_for_use.fasta',\n",
       " 'wickerhamomyces_anomalus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode06/length_restricted_for_use.fasta',\n",
       " 'yamadazyma_mexicana': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode07/length_restricted_for_use.fasta',\n",
       " 'yamadazyma_scolyti': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode08/length_restricted_for_use.fasta',\n",
       " 'yarrowia_lipolytica': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode09/length_restricted_for_use.fasta',\n",
       " 'zygoascus_hellenicus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode11/length_restricted_for_use.fasta',\n",
       " 'aspergillus_flavus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta',\n",
       " 'cryptococcus_zero': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode01/length_restricted_for_use.fasta',\n",
       " 'aspergillus_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode02/length_restricted_for_use.fasta',\n",
       " 'diaporthe_ccl067': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode03/length_restricted_for_use.fasta',\n",
       " 'diaporthe_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode04/length_restricted_for_use.fasta',\n",
       " 'oculimacula_yallundae-ccl031': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode05/length_restricted_for_use.fasta',\n",
       " 'oculimacula_yallundae-ccl029': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode06/length_restricted_for_use.fasta',\n",
       " 'dothiorella_vidmadera': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode07/length_restricted_for_use.fasta',\n",
       " 'quambalaria_cyanescens': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode08/length_restricted_for_use.fasta',\n",
       " 'entoleuca_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode09/length_restricted_for_use.fasta',\n",
       " 'asteroma_ccl060': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode11/length_restricted_for_use.fasta',\n",
       " 'asteroma_ccl068': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode12/length_restricted_for_use.fasta',\n",
       " 'saccharomyces_cerevisiae': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode01/length_restricted_for_use.fasta',\n",
       " 'cladophialophora_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode02/length_restricted_for_use.fasta',\n",
       " 'candida_albicans': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta',\n",
       " 'candida_metapsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode04/length_restricted_for_use.fasta',\n",
       " 'candida_orthopsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta',\n",
       " 'candida_parapsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode06/length_restricted_for_use.fasta',\n",
       " 'candida_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode09/length_restricted_for_use.fasta',\n",
       " 'kluyveromyces_marxianus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode10/length_restricted_for_use.fasta',\n",
       " 'pichia_kudriavzevii': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode11/length_restricted_for_use.fasta',\n",
       " 'pichia_membranifaciens': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode12/length_restricted_for_use.fasta'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode01/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/puccinia_striiformis-tritici.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zymoseptoria_tritici.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pyrenophora_tritici-repentis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/fusarium_oxysporum.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/tuber_brumale.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cortinarius_globuliformis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_niger.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode08/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/clavispora_lusitaniae.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode09/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/penicillium_chrysogenum.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/rhodotorula_mucilaginosa.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/scedosporium_boydii.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode01/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/blastobotrys_proliferans.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode02/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/debaryomyces_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/galactomyces_geotrichum.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kodamaea_ohmeri.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/meyerozyma_guilliermondii.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/wickerhamomyces_anomalus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode07/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_mexicana.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode08/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_scolyti.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode09/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yarrowia_lipolytica.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zygoascus_hellenicus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_flavus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode01/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cryptococcus_zero.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode02/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_ccl067.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl031.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl029.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode07/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/dothiorella_vidmadera.15000.fasta!!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode08/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/quambalaria_cyanescens.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode09/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/entoleuca_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl060.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl068.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode01/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/saccharomyces_cerevisiae.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode02/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cladophialophora_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_albicans.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_metapsilosis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_orthopsilosis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_parapsilosis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode09/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode10/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_marxianus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_kudriavzevii.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_membranifaciens.15000.fasta!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "    # Will print 'check one' if files already produced"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map with minimap against both databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb',\n",
       " '/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/puccinia_striiformis-tritici.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.puccinia_striiformis-tritici.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zymoseptoria_tritici.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.zymoseptoria_tritici.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pyrenophora_tritici-repentis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.pyrenophora_tritici-repentis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/fusarium_oxysporum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.fusarium_oxysporum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/tuber_brumale.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.tuber_brumale.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cortinarius_globuliformis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.cortinarius_globuliformis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_niger.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.aspergillus_niger.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/clavispora_lusitaniae.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.clavispora_lusitaniae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.kluyveromyces_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/penicillium_chrysogenum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.penicillium_chrysogenum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/rhodotorula_mucilaginosa.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.rhodotorula_mucilaginosa.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/scedosporium_boydii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.scedosporium_boydii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/blastobotrys_proliferans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.blastobotrys_proliferans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/debaryomyces_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.debaryomyces_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/galactomyces_geotrichum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.galactomyces_geotrichum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kodamaea_ohmeri.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.kodamaea_ohmeri.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/meyerozyma_guilliermondii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.meyerozyma_guilliermondii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/wickerhamomyces_anomalus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.wickerhamomyces_anomalus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_mexicana.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.yamadazyma_mexicana.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_scolyti.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.yamadazyma_scolyti.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yarrowia_lipolytica.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.yarrowia_lipolytica.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zygoascus_hellenicus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.zygoascus_hellenicus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_flavus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.aspergillus_flavus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cryptococcus_zero.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.cryptococcus_zero.minimap2.paf\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.aspergillus_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_ccl067.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.diaporthe_ccl067.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.diaporthe_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl031.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.oculimacula_yallundae-ccl031.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl029.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.oculimacula_yallundae-ccl029.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/dothiorella_vidmadera.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.dothiorella_vidmadera.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/quambalaria_cyanescens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.quambalaria_cyanescens.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/entoleuca_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.entoleuca_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl060.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.asteroma_ccl060.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl068.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.asteroma_ccl068.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/saccharomyces_cerevisiae.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.saccharomyces_cerevisiae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cladophialophora_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.cladophialophora_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_albicans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_albicans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_metapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_metapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_orthopsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_orthopsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_parapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_parapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_marxianus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.kluyveromyces_marxianus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_kudriavzevii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.pichia_kudriavzevii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_membranifaciens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.pichia_membranifaciens.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/puccinia_striiformis-tritici.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.puccinia_striiformis-tritici.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zymoseptoria_tritici.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.zymoseptoria_tritici.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pyrenophora_tritici-repentis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.pyrenophora_tritici-repentis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/fusarium_oxysporum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.fusarium_oxysporum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/tuber_brumale.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.tuber_brumale.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cortinarius_globuliformis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.cortinarius_globuliformis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_niger.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.aspergillus_niger.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/clavispora_lusitaniae.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.clavispora_lusitaniae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.kluyveromyces_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/penicillium_chrysogenum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.penicillium_chrysogenum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/rhodotorula_mucilaginosa.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.rhodotorula_mucilaginosa.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/scedosporium_boydii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.scedosporium_boydii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/blastobotrys_proliferans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.blastobotrys_proliferans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/debaryomyces_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.debaryomyces_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/galactomyces_geotrichum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.galactomyces_geotrichum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kodamaea_ohmeri.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.kodamaea_ohmeri.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/meyerozyma_guilliermondii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.meyerozyma_guilliermondii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/wickerhamomyces_anomalus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.wickerhamomyces_anomalus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_mexicana.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.yamadazyma_mexicana.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_scolyti.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.yamadazyma_scolyti.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yarrowia_lipolytica.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.yarrowia_lipolytica.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zygoascus_hellenicus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.zygoascus_hellenicus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_flavus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.aspergillus_flavus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cryptococcus_zero.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.cryptococcus_zero.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.aspergillus_unidentified.minimap2.paf\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_ccl067.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.diaporthe_ccl067.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.diaporthe_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl031.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.oculimacula_yallundae-ccl031.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl029.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.oculimacula_yallundae-ccl029.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/dothiorella_vidmadera.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.dothiorella_vidmadera.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/quambalaria_cyanescens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.quambalaria_cyanescens.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/entoleuca_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.entoleuca_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl060.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.asteroma_ccl060.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl068.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.asteroma_ccl068.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/saccharomyces_cerevisiae.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.saccharomyces_cerevisiae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cladophialophora_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.cladophialophora_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_albicans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_albicans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_metapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_metapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_orthopsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_orthopsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_parapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_parapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_marxianus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.kluyveromyces_marxianus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_kudriavzevii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.pichia_kudriavzevii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_membranifaciens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.pichia_membranifaciens.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results(fn, species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the query species: {species}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')\n",
    "    hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/custom_results/%s.json' % species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "    return hit_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getquery_taxfileid(refdf_fn, species):\n",
    "    \"\"\"\n",
    "    Takes the reference dataframe filename and the species name.\n",
    "    Returns the taxfileid, which is the date/flowcellid (column 0 value) of the ref_df.\n",
    "    \"\"\"\n",
    "    ref_df = pd.read_csv(refdf_fn)\n",
    "    ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']\n",
    "    return ref_df[ref_df.name_species == species].iloc[:,0].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxid_dict(taxid_fn, taxfileid):\n",
    "    \"\"\"\n",
    "    Takes a taxonomy assignment file filename in the Qiime format and a taxonomic identifier.\n",
    "    Returns the a dictionary with the taxonomic assignment at each rank.\n",
    "    \"\"\"\n",
    "    tax_dict = {}\n",
    "    with open(taxid_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(taxfileid):\n",
    "                taxrankids = line.rstrip().split('\\t')[1].split(';')\n",
    "                for taxrank in taxrankids:\n",
    "                    tax_dict[taxrank.split('__')[0]] = taxrank.split('__')[1]\n",
    "    return tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_taxranks_results(mapping_df, tax_fn, ref_df_fn = False):\n",
    "    \"\"\"\n",
    "    This function assigns the taxonomic ranks for each hit in the mapping results dataframe.\n",
    "    It takes a mapping_df, taxonomy assignment file, and if required a reference dataframe filename.\n",
    "    Returns the mapping dataframe with assignment. \n",
    "    \"\"\"\n",
    "    for tname in mapping_df.index:\n",
    "        if ref_df_fn:\n",
    "            tmp_taxfileid = getquery_taxfileid(ref_df_fn, tname)\n",
    "        else:\n",
    "            tmp_taxfileid = tname\n",
    "        tmp_tax_dict = get_taxid_dict(tax_fn, tmp_taxfileid)\n",
    "        for key, value in tmp_tax_dict.items():\n",
    "            mapping_df.loc[tname, key] = value\n",
    "    return mapping_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarieses the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = {}\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: puccinia_striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "puccinia_striiformis-tritici    0.994467\n",
      "pyrenophora_tritici-repentis    0.000533\n",
      "cortinarius_globuliformis       0.000467\n",
      "saccharomyces_cerevisiae        0.000400\n",
      "cryptococcus_zero               0.000400\n",
      "aspergillus_niger               0.000333\n",
      "yarrowia_lipolytica             0.000267\n",
      "blastobotrys_proliferans        0.000267\n",
      "kodamaea_ohmeri                 0.000267\n",
      "clavispora_lusitaniae           0.000200\n",
      "zymoseptoria_tritici            0.000200\n",
      "tuber_brumale                   0.000200\n",
      "yamadazyma_mexicana             0.000200\n",
      "scedosporium_boydii             0.000200\n",
      "penicillium_chrysogenum         0.000133\n",
      "oculimacula_yallundae-ccl031    0.000133\n",
      "kluyveromyces_marxianus         0.000133\n",
      "candida_metapsilosis            0.000133\n",
      "fusarium_oxysporum              0.000133\n",
      "candida_unidentified            0.000133\n",
      "rhodotorula_mucilaginosa        0.000133\n",
      "pichia_membranifaciens          0.000067\n",
      "aspergillus_unidentified        0.000067\n",
      "candida_parapsilosis            0.000067\n",
      "debaryomyces_unidentified       0.000067\n",
      "pichia_kudriavzevii             0.000067\n",
      "entoleuca_unidentified          0.000067\n",
      "galactomyces_geotrichum         0.000067\n",
      "quambalaria_cyanescens          0.000067\n",
      "zygoascus_hellenicus            0.000067\n",
      "oculimacula_yallundae-ccl029    0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: zymoseptoria_tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zymoseptoria_tritici             0.991542\n",
      "pyrenophora_tritici-repentis     0.001465\n",
      "penicillium_chrysogenum          0.000866\n",
      "dothiorella_vidmadera            0.000799\n",
      "oculimacula_yallundae-ccl029     0.000799\n",
      "tuber_brumale                    0.000466\n",
      "entoleuca_unidentified           0.000466\n",
      "fusarium_oxysporum               0.000400\n",
      "candida_metapsilosis             0.000333\n",
      "cladophialophora_unidentified    0.000266\n",
      "oculimacula_yallundae-ccl031     0.000266\n",
      "puccinia_striiformis-tritici     0.000200\n",
      "kluyveromyces_marxianus          0.000200\n",
      "asteroma_ccl060                  0.000200\n",
      "aspergillus_niger                0.000133\n",
      "saccharomyces_cerevisiae         0.000133\n",
      "aspergillus_unidentified         0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "candida_orthopsilosis            0.000133\n",
      "asteroma_ccl068                  0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "candida_unidentified             0.000067\n",
      "diaporthe_ccl067                 0.000067\n",
      "cortinarius_globuliformis        0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "diaporthe_unidentified           0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "meyerozyma_guilliermondii        0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pyrenophora_tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pyrenophora_tritici-repentis     0.956842\n",
      "zymoseptoria_tritici             0.033331\n",
      "galactomyces_geotrichum          0.001660\n",
      "oculimacula_yallundae-ccl029     0.000996\n",
      "diaporthe_ccl067                 0.000797\n",
      "candida_albicans                 0.000730\n",
      "dothiorella_vidmadera            0.000598\n",
      "entoleuca_unidentified           0.000465\n",
      "aspergillus_niger                0.000398\n",
      "penicillium_chrysogenum          0.000332\n",
      "candida_metapsilosis             0.000266\n",
      "aspergillus_unidentified         0.000199\n",
      "asteroma_ccl060                  0.000199\n",
      "asteroma_ccl068                  0.000199\n",
      "cladophialophora_unidentified    0.000199\n",
      "clavispora_lusitaniae            0.000199\n",
      "diaporthe_unidentified           0.000199\n",
      "fusarium_oxysporum               0.000199\n",
      "puccinia_striiformis-tritici     0.000199\n",
      "scedosporium_boydii              0.000199\n",
      "wickerhamomyces_anomalus         0.000199\n",
      "aspergillus_flavus               0.000199\n",
      "meyerozyma_guilliermondii        0.000133\n",
      "candida_parapsilosis             0.000133\n",
      "candida_orthopsilosis            0.000133\n",
      "kodamaea_ohmeri                  0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "oculimacula_yallundae-ccl031     0.000066\n",
      "debaryomyces_unidentified        0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "cortinarius_globuliformis        0.000066\n",
      "pichia_membranifaciens           0.000066\n",
      "yarrowia_lipolytica              0.000066\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "tuber_brumale                    0.000066\n",
      "yamadazyma_mexicana              0.000066\n",
      "yamadazyma_scolyti               0.000066\n",
      "kluyveromyces_marxianus          0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: fusarium_oxysporum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "fusarium_oxysporum               0.972563\n",
      "kodamaea_ohmeri                  0.009124\n",
      "diaporthe_unidentified           0.006104\n",
      "candida_metapsilosis             0.005579\n",
      "oculimacula_yallundae-ccl029     0.000722\n",
      "entoleuca_unidentified           0.000722\n",
      "scedosporium_boydii              0.000656\n",
      "asteroma_ccl060                  0.000591\n",
      "tuber_brumale                    0.000394\n",
      "zygoascus_hellenicus             0.000394\n",
      "puccinia_striiformis-tritici     0.000263\n",
      "diaporthe_ccl067                 0.000263\n",
      "asteroma_ccl068                  0.000263\n",
      "pyrenophora_tritici-repentis     0.000263\n",
      "saccharomyces_cerevisiae         0.000197\n",
      "candida_parapsilosis             0.000197\n",
      "penicillium_chrysogenum          0.000197\n",
      "dothiorella_vidmadera            0.000197\n",
      "clavispora_lusitaniae            0.000131\n",
      "zymoseptoria_tritici             0.000131\n",
      "kluyveromyces_unidentified       0.000131\n",
      "oculimacula_yallundae-ccl031     0.000131\n",
      "pichia_kudriavzevii              0.000131\n",
      "aspergillus_niger                0.000131\n",
      "galactomyces_geotrichum          0.000066\n",
      "meyerozyma_guilliermondii        0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "quambalaria_cyanescens           0.000066\n",
      "candida_orthopsilosis            0.000066\n",
      "rhodotorula_mucilaginosa         0.000066\n",
      "yamadazyma_mexicana              0.000066\n",
      "yarrowia_lipolytica              0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: tuber_brumale\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "tuber_brumale                    0.990028\n",
      "oculimacula_yallundae-ccl031     0.000931\n",
      "meyerozyma_guilliermondii        0.000864\n",
      "candida_orthopsilosis            0.000731\n",
      "zymoseptoria_tritici             0.000598\n",
      "yamadazyma_mexicana              0.000598\n",
      "clavispora_lusitaniae            0.000465\n",
      "yarrowia_lipolytica              0.000399\n",
      "yamadazyma_scolyti               0.000399\n",
      "penicillium_chrysogenum          0.000399\n",
      "wickerhamomyces_anomalus         0.000332\n",
      "kluyveromyces_marxianus          0.000266\n",
      "pichia_kudriavzevii              0.000266\n",
      "candida_parapsilosis             0.000199\n",
      "asteroma_ccl068                  0.000199\n",
      "quambalaria_cyanescens           0.000199\n",
      "cladophialophora_unidentified    0.000199\n",
      "kodamaea_ohmeri                  0.000199\n",
      "saccharomyces_cerevisiae         0.000199\n",
      "aspergillus_unidentified         0.000199\n",
      "entoleuca_unidentified           0.000199\n",
      "diaporthe_unidentified           0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "debaryomyces_unidentified        0.000133\n",
      "galactomyces_geotrichum          0.000133\n",
      "aspergillus_flavus               0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "rhodotorula_mucilaginosa         0.000133\n",
      "pyrenophora_tritici-repentis     0.000133\n",
      "zygoascus_hellenicus             0.000133\n",
      "oculimacula_yallundae-ccl029     0.000133\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "candida_metapsilosis             0.000066\n",
      "blastobotrys_proliferans         0.000066\n",
      "candida_albicans                 0.000066\n",
      "dothiorella_vidmadera            0.000066\n",
      "diaporthe_ccl067                 0.000066\n",
      "candida_unidentified             0.000066\n",
      "cortinarius_globuliformis        0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "fusarium_oxysporum               0.000066\n",
      "pichia_membranifaciens           0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cortinarius_globuliformis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cortinarius_globuliformis        0.990614\n",
      "pyrenophora_tritici-repentis     0.000732\n",
      "puccinia_striiformis-tritici     0.000666\n",
      "rhodotorula_mucilaginosa         0.000599\n",
      "clavispora_lusitaniae            0.000466\n",
      "penicillium_chrysogenum          0.000466\n",
      "zymoseptoria_tritici             0.000333\n",
      "aspergillus_niger                0.000333\n",
      "cryptococcus_zero                0.000333\n",
      "galactomyces_geotrichum          0.000333\n",
      "tuber_brumale                    0.000333\n",
      "debaryomyces_unidentified        0.000266\n",
      "fusarium_oxysporum               0.000266\n",
      "kluyveromyces_marxianus          0.000266\n",
      "kluyveromyces_unidentified       0.000266\n",
      "candida_metapsilosis             0.000266\n",
      "blastobotrys_proliferans         0.000200\n",
      "asteroma_ccl060                  0.000200\n",
      "scedosporium_boydii              0.000200\n",
      "cladophialophora_unidentified    0.000200\n",
      "yarrowia_lipolytica              0.000200\n",
      "dothiorella_vidmadera            0.000200\n",
      "yamadazyma_mexicana              0.000200\n",
      "quambalaria_cyanescens           0.000200\n",
      "asteroma_ccl068                  0.000133\n",
      "aspergillus_unidentified         0.000133\n",
      "yamadazyma_scolyti               0.000133\n",
      "diaporthe_ccl067                 0.000133\n",
      "saccharomyces_cerevisiae         0.000133\n",
      "kodamaea_ohmeri                  0.000133\n",
      "meyerozyma_guilliermondii        0.000133\n",
      "oculimacula_yallundae-ccl031     0.000133\n",
      "pichia_kudriavzevii              0.000133\n",
      "pichia_membranifaciens           0.000133\n",
      "diaporthe_unidentified           0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "candida_albicans                 0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_niger\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_niger                0.754601\n",
      "aspergillus_unidentified         0.223564\n",
      "aspergillus_flavus               0.008048\n",
      "yamadazyma_mexicana              0.004486\n",
      "dothiorella_vidmadera            0.004354\n",
      "penicillium_chrysogenum          0.001715\n",
      "entoleuca_unidentified           0.000462\n",
      "zymoseptoria_tritici             0.000330\n",
      "tuber_brumale                    0.000264\n",
      "galactomyces_geotrichum          0.000198\n",
      "scedosporium_boydii              0.000198\n",
      "puccinia_striiformis-tritici     0.000198\n",
      "cladophialophora_unidentified    0.000198\n",
      "pyrenophora_tritici-repentis     0.000198\n",
      "clavispora_lusitaniae            0.000198\n",
      "cortinarius_globuliformis        0.000132\n",
      "oculimacula_yallundae-ccl031     0.000132\n",
      "oculimacula_yallundae-ccl029     0.000132\n",
      "rhodotorula_mucilaginosa         0.000132\n",
      "kluyveromyces_marxianus          0.000132\n",
      "fusarium_oxysporum               0.000066\n",
      "candida_orthopsilosis            0.000066\n",
      "candida_metapsilosis             0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: clavispora_lusitaniae\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "clavispora_lusitaniae           0.984887\n",
      "candida_unidentified            0.002796\n",
      "kodamaea_ohmeri                 0.001931\n",
      "candida_orthopsilosis           0.000999\n",
      "yamadazyma_scolyti              0.000799\n",
      "candida_albicans                0.000666\n",
      "candida_metapsilosis            0.000666\n",
      "kluyveromyces_unidentified      0.000599\n",
      "saccharomyces_cerevisiae        0.000599\n",
      "kluyveromyces_marxianus         0.000533\n",
      "wickerhamomyces_anomalus        0.000466\n",
      "zymoseptoria_tritici            0.000399\n",
      "yamadazyma_mexicana             0.000399\n",
      "debaryomyces_unidentified       0.000333\n",
      "pichia_kudriavzevii             0.000333\n",
      "yarrowia_lipolytica             0.000333\n",
      "penicillium_chrysogenum         0.000266\n",
      "meyerozyma_guilliermondii       0.000266\n",
      "entoleuca_unidentified          0.000266\n",
      "puccinia_striiformis-tritici    0.000200\n",
      "blastobotrys_proliferans        0.000200\n",
      "zygoascus_hellenicus            0.000200\n",
      "pyrenophora_tritici-repentis    0.000200\n",
      "quambalaria_cyanescens          0.000200\n",
      "asteroma_ccl060                 0.000133\n",
      "candida_parapsilosis            0.000133\n",
      "scedosporium_boydii             0.000133\n",
      "aspergillus_flavus              0.000133\n",
      "fusarium_oxysporum              0.000133\n",
      "galactomyces_geotrichum         0.000133\n",
      "oculimacula_yallundae-ccl029    0.000133\n",
      "oculimacula_yallundae-ccl031    0.000133\n",
      "cryptococcus_zero               0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "diaporthe_ccl067                0.000067\n",
      "pichia_membranifaciens          0.000067\n",
      "tuber_brumale                   0.000067\n",
      "diaporthe_unidentified          0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kluyveromyces_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kluyveromyces_marxianus          0.536898\n",
      "kluyveromyces_unidentified       0.365759\n",
      "cryptococcus_zero                0.060773\n",
      "rhodotorula_mucilaginosa         0.004407\n",
      "entoleuca_unidentified           0.003946\n",
      "candida_unidentified             0.003552\n",
      "cortinarius_globuliformis        0.003223\n",
      "yarrowia_lipolytica              0.002697\n",
      "wickerhamomyces_anomalus         0.001644\n",
      "candida_orthopsilosis            0.001579\n",
      "saccharomyces_cerevisiae         0.001381\n",
      "oculimacula_yallundae-ccl029     0.001184\n",
      "yamadazyma_mexicana              0.001184\n",
      "meyerozyma_guilliermondii        0.001184\n",
      "candida_metapsilosis             0.001118\n",
      "quambalaria_cyanescens           0.001118\n",
      "zymoseptoria_tritici             0.000855\n",
      "clavispora_lusitaniae            0.000658\n",
      "candida_albicans                 0.000526\n",
      "pyrenophora_tritici-repentis     0.000526\n",
      "penicillium_chrysogenum          0.000460\n",
      "blastobotrys_proliferans         0.000460\n",
      "debaryomyces_unidentified        0.000460\n",
      "candida_parapsilosis             0.000460\n",
      "tuber_brumale                    0.000460\n",
      "cladophialophora_unidentified    0.000395\n",
      "aspergillus_niger                0.000395\n",
      "kodamaea_ohmeri                  0.000395\n",
      "yamadazyma_scolyti               0.000263\n",
      "zygoascus_hellenicus             0.000263\n",
      "fusarium_oxysporum               0.000263\n",
      "puccinia_striiformis-tritici     0.000263\n",
      "asteroma_ccl060                  0.000197\n",
      "oculimacula_yallundae-ccl031     0.000197\n",
      "aspergillus_flavus               0.000197\n",
      "dothiorella_vidmadera            0.000132\n",
      "asteroma_ccl068                  0.000132\n",
      "pichia_kudriavzevii              0.000132\n",
      "galactomyces_geotrichum          0.000132\n",
      "pichia_membranifaciens           0.000066\n",
      "aspergillus_unidentified         0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: penicillium_chrysogenum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "penicillium_chrysogenum         0.986958\n",
      "aspergillus_flavus              0.002462\n",
      "aspergillus_unidentified        0.002196\n",
      "aspergillus_niger               0.002129\n",
      "kluyveromyces_marxianus         0.000932\n",
      "zymoseptoria_tritici            0.000799\n",
      "kluyveromyces_unidentified      0.000799\n",
      "cortinarius_globuliformis       0.000466\n",
      "tuber_brumale                   0.000466\n",
      "dothiorella_vidmadera           0.000333\n",
      "diaporthe_unidentified          0.000200\n",
      "kodamaea_ohmeri                 0.000200\n",
      "fusarium_oxysporum              0.000133\n",
      "asteroma_ccl068                 0.000133\n",
      "blastobotrys_proliferans        0.000133\n",
      "clavispora_lusitaniae           0.000133\n",
      "saccharomyces_cerevisiae        0.000133\n",
      "scedosporium_boydii             0.000133\n",
      "oculimacula_yallundae-ccl029    0.000133\n",
      "puccinia_striiformis-tritici    0.000133\n",
      "entoleuca_unidentified          0.000133\n",
      "zygoascus_hellenicus            0.000067\n",
      "pichia_kudriavzevii             0.000067\n",
      "pichia_membranifaciens          0.000067\n",
      "diaporthe_ccl067                0.000067\n",
      "debaryomyces_unidentified       0.000067\n",
      "cryptococcus_zero               0.000067\n",
      "pyrenophora_tritici-repentis    0.000067\n",
      "candida_unidentified            0.000067\n",
      "quambalaria_cyanescens          0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "asteroma_ccl060                 0.000067\n",
      "yamadazyma_mexicana             0.000067\n",
      "yamadazyma_scolyti              0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: rhodotorula_mucilaginosa\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "rhodotorula_mucilaginosa         0.982583\n",
      "pichia_kudriavzevii              0.004948\n",
      "asteroma_ccl060                  0.003497\n",
      "zygoascus_hellenicus             0.003101\n",
      "cortinarius_globuliformis        0.000858\n",
      "penicillium_chrysogenum          0.000726\n",
      "aspergillus_niger                0.000462\n",
      "cryptococcus_zero                0.000462\n",
      "pyrenophora_tritici-repentis     0.000396\n",
      "tuber_brumale                    0.000330\n",
      "entoleuca_unidentified           0.000264\n",
      "zymoseptoria_tritici             0.000198\n",
      "cladophialophora_unidentified    0.000198\n",
      "puccinia_striiformis-tritici     0.000198\n",
      "candida_metapsilosis             0.000132\n",
      "kodamaea_ohmeri                  0.000132\n",
      "yamadazyma_scolyti               0.000132\n",
      "quambalaria_cyanescens           0.000132\n",
      "scedosporium_boydii              0.000132\n",
      "oculimacula_yallundae-ccl031     0.000132\n",
      "oculimacula_yallundae-ccl029     0.000066\n",
      "wickerhamomyces_anomalus         0.000066\n",
      "yarrowia_lipolytica              0.000066\n",
      "aspergillus_unidentified         0.000066\n",
      "asteroma_ccl068                  0.000066\n",
      "blastobotrys_proliferans         0.000066\n",
      "candida_albicans                 0.000066\n",
      "candida_parapsilosis             0.000066\n",
      "candida_orthopsilosis            0.000066\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "debaryomyces_unidentified        0.000066\n",
      "diaporthe_ccl067                 0.000066\n",
      "dothiorella_vidmadera            0.000066\n",
      "pichia_membranifaciens           0.000066\n",
      "aspergillus_flavus               0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: scedosporium_boydii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "scedosporium_boydii              0.991613\n",
      "fusarium_oxysporum               0.001531\n",
      "oculimacula_yallundae-ccl029     0.000865\n",
      "entoleuca_unidentified           0.000599\n",
      "pyrenophora_tritici-repentis     0.000466\n",
      "rhodotorula_mucilaginosa         0.000399\n",
      "asteroma_ccl068                  0.000399\n",
      "penicillium_chrysogenum          0.000399\n",
      "aspergillus_unidentified         0.000266\n",
      "asteroma_ccl060                  0.000266\n",
      "cryptococcus_zero                0.000266\n",
      "diaporthe_ccl067                 0.000266\n",
      "aspergillus_flavus               0.000266\n",
      "tuber_brumale                    0.000266\n",
      "clavispora_lusitaniae            0.000200\n",
      "oculimacula_yallundae-ccl031     0.000200\n",
      "pichia_membranifaciens           0.000200\n",
      "zymoseptoria_tritici             0.000200\n",
      "diaporthe_unidentified           0.000200\n",
      "dothiorella_vidmadera            0.000200\n",
      "cortinarius_globuliformis        0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "galactomyces_geotrichum          0.000133\n",
      "kodamaea_ohmeri                  0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "kluyveromyces_marxianus          0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: blastobotrys_proliferans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "blastobotrys_proliferans         0.995003\n",
      "meyerozyma_guilliermondii        0.000866\n",
      "debaryomyces_unidentified        0.000733\n",
      "entoleuca_unidentified           0.000466\n",
      "candida_orthopsilosis            0.000400\n",
      "candida_albicans                 0.000267\n",
      "aspergillus_flavus               0.000200\n",
      "yamadazyma_mexicana              0.000200\n",
      "wickerhamomyces_anomalus         0.000200\n",
      "tuber_brumale                    0.000200\n",
      "candida_metapsilosis             0.000200\n",
      "oculimacula_yallundae-ccl029     0.000200\n",
      "yarrowia_lipolytica              0.000133\n",
      "yamadazyma_scolyti               0.000133\n",
      "kodamaea_ohmeri                  0.000133\n",
      "galactomyces_geotrichum          0.000067\n",
      "fusarium_oxysporum               0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: debaryomyces_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "debaryomyces_unidentified       0.984419\n",
      "clavispora_lusitaniae           0.004994\n",
      "meyerozyma_guilliermondii       0.002464\n",
      "yamadazyma_mexicana             0.001665\n",
      "wickerhamomyces_anomalus        0.000599\n",
      "candida_metapsilosis            0.000533\n",
      "candida_parapsilosis            0.000533\n",
      "tuber_brumale                   0.000466\n",
      "yamadazyma_scolyti              0.000466\n",
      "candida_unidentified            0.000466\n",
      "oculimacula_yallundae-ccl029    0.000400\n",
      "kluyveromyces_marxianus         0.000333\n",
      "entoleuca_unidentified          0.000333\n",
      "candida_orthopsilosis           0.000266\n",
      "kluyveromyces_unidentified      0.000266\n",
      "kodamaea_ohmeri                 0.000200\n",
      "zygoascus_hellenicus            0.000200\n",
      "candida_albicans                0.000133\n",
      "blastobotrys_proliferans        0.000133\n",
      "zymoseptoria_tritici            0.000133\n",
      "galactomyces_geotrichum         0.000133\n",
      "aspergillus_flavus              0.000133\n",
      "diaporthe_ccl067                0.000067\n",
      "diaporthe_unidentified          0.000067\n",
      "dothiorella_vidmadera           0.000067\n",
      "fusarium_oxysporum              0.000067\n",
      "penicillium_chrysogenum         0.000067\n",
      "pyrenophora_tritici-repentis    0.000067\n",
      "quambalaria_cyanescens          0.000067\n",
      "saccharomyces_cerevisiae        0.000067\n",
      "yarrowia_lipolytica             0.000067\n",
      "asteroma_ccl068                 0.000067\n",
      "aspergillus_niger               0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: galactomyces_geotrichum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "galactomyces_geotrichum          0.984758\n",
      "candida_metapsilosis             0.000877\n",
      "pyrenophora_tritici-repentis     0.000809\n",
      "entoleuca_unidentified           0.000809\n",
      "diaporthe_ccl067                 0.000809\n",
      "candida_albicans                 0.000809\n",
      "yamadazyma_mexicana              0.000674\n",
      "wickerhamomyces_anomalus         0.000674\n",
      "kodamaea_ohmeri                  0.000607\n",
      "debaryomyces_unidentified        0.000540\n",
      "kluyveromyces_unidentified       0.000540\n",
      "blastobotrys_proliferans         0.000540\n",
      "yarrowia_lipolytica              0.000540\n",
      "zygoascus_hellenicus             0.000472\n",
      "pichia_kudriavzevii              0.000472\n",
      "saccharomyces_cerevisiae         0.000472\n",
      "clavispora_lusitaniae            0.000405\n",
      "yamadazyma_scolyti               0.000405\n",
      "candida_parapsilosis             0.000405\n",
      "dothiorella_vidmadera            0.000337\n",
      "aspergillus_flavus               0.000337\n",
      "tuber_brumale                    0.000337\n",
      "pichia_membranifaciens           0.000337\n",
      "cortinarius_globuliformis        0.000270\n",
      "candida_unidentified             0.000270\n",
      "quambalaria_cyanescens           0.000270\n",
      "candida_orthopsilosis            0.000202\n",
      "aspergillus_unidentified         0.000202\n",
      "oculimacula_yallundae-ccl029     0.000202\n",
      "diaporthe_unidentified           0.000202\n",
      "asteroma_ccl060                  0.000202\n",
      "penicillium_chrysogenum          0.000202\n",
      "kluyveromyces_marxianus          0.000135\n",
      "rhodotorula_mucilaginosa         0.000135\n",
      "cladophialophora_unidentified    0.000135\n",
      "cryptococcus_zero                0.000135\n",
      "meyerozyma_guilliermondii        0.000135\n",
      "fusarium_oxysporum               0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "aspergillus_niger                0.000067\n",
      "zymoseptoria_tritici             0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kodamaea_ohmeri\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kodamaea_ohmeri                  0.946206\n",
      "candida_metapsilosis             0.018825\n",
      "diaporthe_unidentified           0.018697\n",
      "fusarium_oxysporum               0.013209\n",
      "clavispora_lusitaniae            0.000383\n",
      "entoleuca_unidentified           0.000319\n",
      "saccharomyces_cerevisiae         0.000255\n",
      "yarrowia_lipolytica              0.000191\n",
      "yamadazyma_mexicana              0.000191\n",
      "wickerhamomyces_anomalus         0.000191\n",
      "tuber_brumale                    0.000191\n",
      "oculimacula_yallundae-ccl029     0.000191\n",
      "cortinarius_globuliformis        0.000128\n",
      "candida_albicans                 0.000128\n",
      "yamadazyma_scolyti               0.000128\n",
      "meyerozyma_guilliermondii        0.000128\n",
      "oculimacula_yallundae-ccl031     0.000128\n",
      "pichia_membranifaciens           0.000128\n",
      "kluyveromyces_marxianus          0.000064\n",
      "debaryomyces_unidentified        0.000064\n",
      "cladophialophora_unidentified    0.000064\n",
      "candida_orthopsilosis            0.000064\n",
      "puccinia_striiformis-tritici     0.000064\n",
      "kluyveromyces_unidentified       0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: meyerozyma_guilliermondii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "meyerozyma_guilliermondii        0.988285\n",
      "debaryomyces_unidentified        0.001464\n",
      "yamadazyma_scolyti               0.000998\n",
      "yamadazyma_mexicana              0.000932\n",
      "candida_metapsilosis             0.000932\n",
      "candida_orthopsilosis            0.000799\n",
      "candida_parapsilosis             0.000666\n",
      "wickerhamomyces_anomalus         0.000599\n",
      "candida_albicans                 0.000599\n",
      "oculimacula_yallundae-ccl029     0.000466\n",
      "kluyveromyces_unidentified       0.000399\n",
      "candida_unidentified             0.000399\n",
      "kluyveromyces_marxianus          0.000399\n",
      "galactomyces_geotrichum          0.000333\n",
      "clavispora_lusitaniae            0.000333\n",
      "zygoascus_hellenicus             0.000266\n",
      "kodamaea_ohmeri                  0.000266\n",
      "yarrowia_lipolytica              0.000200\n",
      "saccharomyces_cerevisiae         0.000200\n",
      "aspergillus_flavus               0.000200\n",
      "entoleuca_unidentified           0.000200\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "oculimacula_yallundae-ccl031     0.000133\n",
      "tuber_brumale                    0.000133\n",
      "cryptococcus_zero                0.000067\n",
      "cortinarius_globuliformis        0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "aspergillus_niger                0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: wickerhamomyces_anomalus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "wickerhamomyces_anomalus        0.992273\n",
      "debaryomyces_unidentified       0.000999\n",
      "meyerozyma_guilliermondii       0.000866\n",
      "yamadazyma_mexicana             0.000733\n",
      "blastobotrys_proliferans        0.000600\n",
      "aspergillus_flavus              0.000400\n",
      "kluyveromyces_marxianus         0.000400\n",
      "candida_metapsilosis            0.000333\n",
      "entoleuca_unidentified          0.000333\n",
      "zygoascus_hellenicus            0.000333\n",
      "kluyveromyces_unidentified      0.000266\n",
      "yarrowia_lipolytica             0.000266\n",
      "candida_parapsilosis            0.000266\n",
      "tuber_brumale                   0.000200\n",
      "galactomyces_geotrichum         0.000200\n",
      "candida_albicans                0.000200\n",
      "saccharomyces_cerevisiae        0.000200\n",
      "oculimacula_yallundae-ccl029    0.000200\n",
      "kodamaea_ohmeri                 0.000133\n",
      "yamadazyma_scolyti              0.000133\n",
      "candida_unidentified            0.000133\n",
      "candida_orthopsilosis           0.000133\n",
      "aspergillus_niger               0.000067\n",
      "pichia_membranifaciens          0.000067\n",
      "quambalaria_cyanescens          0.000067\n",
      "pyrenophora_tritici-repentis    0.000067\n",
      "puccinia_striiformis-tritici    0.000067\n",
      "clavispora_lusitaniae           0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yamadazyma_mexicana\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yamadazyma_mexicana              0.979692\n",
      "yamadazyma_scolyti               0.013050\n",
      "meyerozyma_guilliermondii        0.001265\n",
      "debaryomyces_unidentified        0.000866\n",
      "kodamaea_ohmeri                  0.000799\n",
      "wickerhamomyces_anomalus         0.000333\n",
      "entoleuca_unidentified           0.000333\n",
      "candida_parapsilosis             0.000333\n",
      "oculimacula_yallundae-ccl029     0.000266\n",
      "dothiorella_vidmadera            0.000266\n",
      "candida_orthopsilosis            0.000266\n",
      "aspergillus_flavus               0.000266\n",
      "candida_metapsilosis             0.000266\n",
      "saccharomyces_cerevisiae         0.000266\n",
      "tuber_brumale                    0.000266\n",
      "candida_unidentified             0.000200\n",
      "yarrowia_lipolytica              0.000200\n",
      "aspergillus_niger                0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "kluyveromyces_marxianus          0.000133\n",
      "zygoascus_hellenicus             0.000133\n",
      "cladophialophora_unidentified    0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "zymoseptoria_tritici             0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yamadazyma_scolyti\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yamadazyma_scolyti              0.987078\n",
      "yamadazyma_mexicana             0.007527\n",
      "debaryomyces_unidentified       0.001332\n",
      "meyerozyma_guilliermondii       0.000799\n",
      "candida_parapsilosis            0.000466\n",
      "candida_metapsilosis            0.000333\n",
      "candida_albicans                0.000266\n",
      "clavispora_lusitaniae           0.000266\n",
      "entoleuca_unidentified          0.000266\n",
      "candida_orthopsilosis           0.000200\n",
      "quambalaria_cyanescens          0.000200\n",
      "yarrowia_lipolytica             0.000200\n",
      "candida_unidentified            0.000133\n",
      "galactomyces_geotrichum         0.000133\n",
      "kluyveromyces_marxianus         0.000133\n",
      "blastobotrys_proliferans        0.000133\n",
      "wickerhamomyces_anomalus        0.000133\n",
      "kluyveromyces_unidentified      0.000133\n",
      "zygoascus_hellenicus            0.000067\n",
      "oculimacula_yallundae-ccl029    0.000067\n",
      "saccharomyces_cerevisiae        0.000067\n",
      "aspergillus_flavus              0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yarrowia_lipolytica\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yarrowia_lipolytica             0.998867\n",
      "yamadazyma_scolyti              0.000133\n",
      "yamadazyma_mexicana             0.000133\n",
      "kluyveromyces_unidentified      0.000133\n",
      "entoleuca_unidentified          0.000133\n",
      "candida_unidentified            0.000133\n",
      "aspergillus_flavus              0.000133\n",
      "wickerhamomyces_anomalus        0.000067\n",
      "oculimacula_yallundae-ccl029    0.000067\n",
      "galactomyces_geotrichum         0.000067\n",
      "clavispora_lusitaniae           0.000067\n",
      "blastobotrys_proliferans        0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: zygoascus_hellenicus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zygoascus_hellenicus            0.973031\n",
      "pichia_kudriavzevii             0.008685\n",
      "rhodotorula_mucilaginosa        0.006922\n",
      "asteroma_ccl060                 0.006399\n",
      "oculimacula_yallundae-ccl029    0.000522\n",
      "entoleuca_unidentified          0.000457\n",
      "yamadazyma_scolyti              0.000392\n",
      "tuber_brumale                   0.000326\n",
      "candida_metapsilosis            0.000326\n",
      "galactomyces_geotrichum         0.000326\n",
      "yamadazyma_mexicana             0.000261\n",
      "candida_albicans                0.000261\n",
      "wickerhamomyces_anomalus        0.000196\n",
      "candida_orthopsilosis           0.000196\n",
      "meyerozyma_guilliermondii       0.000196\n",
      "yarrowia_lipolytica             0.000196\n",
      "debaryomyces_unidentified       0.000131\n",
      "kluyveromyces_unidentified      0.000131\n",
      "blastobotrys_proliferans        0.000131\n",
      "kodamaea_ohmeri                 0.000131\n",
      "saccharomyces_cerevisiae        0.000131\n",
      "candida_unidentified            0.000131\n",
      "aspergillus_unidentified        0.000065\n",
      "zymoseptoria_tritici            0.000065\n",
      "cortinarius_globuliformis       0.000065\n",
      "fusarium_oxysporum              0.000065\n",
      "kluyveromyces_marxianus         0.000065\n",
      "oculimacula_yallundae-ccl031    0.000065\n",
      "quambalaria_cyanescens          0.000065\n",
      "aspergillus_flavus              0.000065\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_flavus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_flavus               0.989142\n",
      "aspergillus_niger                0.003464\n",
      "aspergillus_unidentified         0.003397\n",
      "penicillium_chrysogenum          0.001399\n",
      "entoleuca_unidentified           0.000333\n",
      "asteroma_ccl068                  0.000266\n",
      "pichia_membranifaciens           0.000266\n",
      "tuber_brumale                    0.000200\n",
      "kodamaea_ohmeri                  0.000200\n",
      "diaporthe_ccl067                 0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "candida_metapsilosis             0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "zymoseptoria_tritici             0.000067\n",
      "dothiorella_vidmadera            0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "pyrenophora_tritici-repentis     0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cryptococcus_zero\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.994200\n",
      "quambalaria_cyanescens           0.000733\n",
      "entoleuca_unidentified           0.000600\n",
      "rhodotorula_mucilaginosa         0.000533\n",
      "oculimacula_yallundae-ccl029     0.000467\n",
      "cortinarius_globuliformis        0.000467\n",
      "saccharomyces_cerevisiae         0.000333\n",
      "diaporthe_unidentified           0.000267\n",
      "aspergillus_niger                0.000267\n",
      "zymoseptoria_tritici             0.000200\n",
      "diaporthe_ccl067                 0.000200\n",
      "asteroma_ccl060                  0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "fusarium_oxysporum               0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "kluyveromyces_marxianus          0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "candida_unidentified             0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "puccinia_striiformis-tritici     0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_unidentified         0.667932\n",
      "aspergillus_niger                0.319217\n",
      "aspergillus_flavus               0.006526\n",
      "penicillium_chrysogenum          0.001065\n",
      "entoleuca_unidentified           0.000533\n",
      "oculimacula_yallundae-ccl031     0.000466\n",
      "zymoseptoria_tritici             0.000400\n",
      "dothiorella_vidmadera            0.000333\n",
      "diaporthe_unidentified           0.000333\n",
      "saccharomyces_cerevisiae         0.000266\n",
      "tuber_brumale                    0.000266\n",
      "asteroma_ccl068                  0.000266\n",
      "cladophialophora_unidentified    0.000266\n",
      "oculimacula_yallundae-ccl029     0.000266\n",
      "diaporthe_ccl067                 0.000200\n",
      "debaryomyces_unidentified        0.000200\n",
      "kluyveromyces_marxianus          0.000200\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "quambalaria_cyanescens           0.000133\n",
      "fusarium_oxysporum               0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "zygoascus_hellenicus             0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "candida_unidentified             0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: diaporthe_ccl067\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "asteroma_ccl068                  0.528731\n",
      "diaporthe_ccl067                 0.455928\n",
      "galactomyces_geotrichum          0.003967\n",
      "pyrenophora_tritici-repentis     0.002380\n",
      "candida_albicans                 0.001984\n",
      "asteroma_ccl060                  0.001785\n",
      "diaporthe_unidentified           0.001719\n",
      "entoleuca_unidentified           0.000926\n",
      "cryptococcus_zero                0.000661\n",
      "oculimacula_yallundae-ccl029     0.000529\n",
      "dothiorella_vidmadera            0.000198\n",
      "saccharomyces_cerevisiae         0.000132\n",
      "fusarium_oxysporum               0.000132\n",
      "kluyveromyces_marxianus          0.000132\n",
      "oculimacula_yallundae-ccl031     0.000132\n",
      "aspergillus_niger                0.000132\n",
      "quambalaria_cyanescens           0.000066\n",
      "kluyveromyces_unidentified       0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "candida_unidentified             0.000066\n",
      "candida_parapsilosis             0.000066\n",
      "candida_orthopsilosis            0.000066\n",
      "meyerozyma_guilliermondii        0.000066\n",
      "penicillium_chrysogenum          0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: diaporthe_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "diaporthe_unidentified          0.552707\n",
      "asteroma_ccl060                 0.418716\n",
      "kodamaea_ohmeri                 0.009417\n",
      "candida_metapsilosis            0.006801\n",
      "fusarium_oxysporum              0.005166\n",
      "asteroma_ccl068                 0.001700\n",
      "entoleuca_unidentified          0.001635\n",
      "diaporthe_ccl067                0.001439\n",
      "oculimacula_yallundae-ccl029    0.000327\n",
      "dothiorella_vidmadera           0.000327\n",
      "scedosporium_boydii             0.000262\n",
      "aspergillus_unidentified        0.000196\n",
      "oculimacula_yallundae-ccl031    0.000196\n",
      "cryptococcus_zero               0.000131\n",
      "zymoseptoria_tritici            0.000131\n",
      "pyrenophora_tritici-repentis    0.000131\n",
      "quambalaria_cyanescens          0.000131\n",
      "rhodotorula_mucilaginosa        0.000131\n",
      "yarrowia_lipolytica             0.000065\n",
      "clavispora_lusitaniae           0.000065\n",
      "saccharomyces_cerevisiae        0.000065\n",
      "tuber_brumale                   0.000065\n",
      "wickerhamomyces_anomalus        0.000065\n",
      "yamadazyma_scolyti              0.000065\n",
      "aspergillus_flavus              0.000065\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: oculimacula_yallundae-ccl031\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "oculimacula_yallundae-ccl031     0.885486\n",
      "oculimacula_yallundae-ccl029     0.106591\n",
      "entoleuca_unidentified           0.001198\n",
      "quambalaria_cyanescens           0.000866\n",
      "dothiorella_vidmadera            0.000732\n",
      "tuber_brumale                    0.000666\n",
      "meyerozyma_guilliermondii        0.000466\n",
      "cladophialophora_unidentified    0.000333\n",
      "yamadazyma_mexicana              0.000266\n",
      "saccharomyces_cerevisiae         0.000266\n",
      "asteroma_ccl060                  0.000200\n",
      "asteroma_ccl068                  0.000200\n",
      "clavispora_lusitaniae            0.000200\n",
      "cryptococcus_zero                0.000200\n",
      "diaporthe_ccl067                 0.000200\n",
      "diaporthe_unidentified           0.000200\n",
      "zymoseptoria_tritici             0.000200\n",
      "pyrenophora_tritici-repentis     0.000200\n",
      "rhodotorula_mucilaginosa         0.000200\n",
      "yarrowia_lipolytica              0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "aspergillus_niger                0.000133\n",
      "debaryomyces_unidentified        0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "cortinarius_globuliformis        0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "candida_unidentified             0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "candida_albicans                 0.000067\n",
      "kluyveromyces_marxianus          0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "fusarium_oxysporum               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: oculimacula_yallundae-ccl029\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "oculimacula_yallundae-ccl029     0.895094\n",
      "oculimacula_yallundae-ccl031     0.093671\n",
      "dothiorella_vidmadera            0.000864\n",
      "diaporthe_ccl067                 0.000598\n",
      "aspergillus_unidentified         0.000532\n",
      "blastobotrys_proliferans         0.000532\n",
      "cladophialophora_unidentified    0.000532\n",
      "quambalaria_cyanescens           0.000465\n",
      "diaporthe_unidentified           0.000465\n",
      "zymoseptoria_tritici             0.000465\n",
      "cryptococcus_zero                0.000465\n",
      "yarrowia_lipolytica              0.000465\n",
      "wickerhamomyces_anomalus         0.000399\n",
      "asteroma_ccl060                  0.000399\n",
      "pyrenophora_tritici-repentis     0.000332\n",
      "pichia_membranifaciens           0.000332\n",
      "zygoascus_hellenicus             0.000266\n",
      "entoleuca_unidentified           0.000266\n",
      "galactomyces_geotrichum          0.000266\n",
      "kluyveromyces_unidentified       0.000266\n",
      "kodamaea_ohmeri                  0.000266\n",
      "penicillium_chrysogenum          0.000266\n",
      "scedosporium_boydii              0.000266\n",
      "aspergillus_niger                0.000199\n",
      "fusarium_oxysporum               0.000199\n",
      "pichia_kudriavzevii              0.000199\n",
      "yamadazyma_mexicana              0.000199\n",
      "saccharomyces_cerevisiae         0.000199\n",
      "kluyveromyces_marxianus          0.000199\n",
      "tuber_brumale                    0.000199\n",
      "candida_metapsilosis             0.000133\n",
      "yamadazyma_scolyti               0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "aspergillus_flavus               0.000133\n",
      "rhodotorula_mucilaginosa         0.000133\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "candida_orthopsilosis            0.000066\n",
      "candida_unidentified             0.000066\n",
      "cortinarius_globuliformis        0.000066\n",
      "meyerozyma_guilliermondii        0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: dothiorella_vidmadera\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "dothiorella_vidmadera            0.986621\n",
      "oculimacula_yallundae-ccl029     0.001797\n",
      "zymoseptoria_tritici             0.001664\n",
      "entoleuca_unidentified           0.001464\n",
      "oculimacula_yallundae-ccl031     0.000865\n",
      "pyrenophora_tritici-repentis     0.000799\n",
      "asteroma_ccl060                  0.000599\n",
      "diaporthe_ccl067                 0.000533\n",
      "tuber_brumale                    0.000466\n",
      "fusarium_oxysporum               0.000466\n",
      "diaporthe_unidentified           0.000466\n",
      "quambalaria_cyanescens           0.000399\n",
      "saccharomyces_cerevisiae         0.000333\n",
      "cladophialophora_unidentified    0.000266\n",
      "cryptococcus_zero                0.000266\n",
      "aspergillus_unidentified         0.000266\n",
      "candida_parapsilosis             0.000266\n",
      "asteroma_ccl068                  0.000200\n",
      "candida_metapsilosis             0.000200\n",
      "cortinarius_globuliformis        0.000200\n",
      "galactomyces_geotrichum          0.000200\n",
      "kodamaea_ohmeri                  0.000200\n",
      "kluyveromyces_marxianus          0.000200\n",
      "kluyveromyces_unidentified       0.000133\n",
      "aspergillus_niger                0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "candida_orthopsilosis            0.000133\n",
      "zygoascus_hellenicus             0.000067\n",
      "candida_albicans                 0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "puccinia_striiformis-tritici     0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "candida_unidentified             0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: quambalaria_cyanescens\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "quambalaria_cyanescens           0.992534\n",
      "oculimacula_yallundae-ccl029     0.001000\n",
      "entoleuca_unidentified           0.000733\n",
      "asteroma_ccl068                  0.000400\n",
      "saccharomyces_cerevisiae         0.000400\n",
      "rhodotorula_mucilaginosa         0.000400\n",
      "cortinarius_globuliformis        0.000400\n",
      "dothiorella_vidmadera            0.000333\n",
      "kodamaea_ohmeri                  0.000333\n",
      "cryptococcus_zero                0.000333\n",
      "oculimacula_yallundae-ccl031     0.000333\n",
      "tuber_brumale                    0.000267\n",
      "diaporthe_unidentified           0.000267\n",
      "yarrowia_lipolytica              0.000200\n",
      "galactomyces_geotrichum          0.000200\n",
      "clavispora_lusitaniae            0.000200\n",
      "puccinia_striiformis-tritici     0.000200\n",
      "diaporthe_ccl067                 0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "aspergillus_niger                0.000133\n",
      "pichia_membranifaciens           0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "fusarium_oxysporum               0.000067\n",
      "kluyveromyces_marxianus          0.000067\n",
      "meyerozyma_guilliermondii        0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "pyrenophora_tritici-repentis     0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_unidentified             0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: entoleuca_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "entoleuca_unidentified           0.988216\n",
      "asteroma_ccl068                  0.000999\n",
      "zymoseptoria_tritici             0.000799\n",
      "oculimacula_yallundae-ccl031     0.000799\n",
      "diaporthe_ccl067                 0.000599\n",
      "fusarium_oxysporum               0.000599\n",
      "diaporthe_unidentified           0.000533\n",
      "cladophialophora_unidentified    0.000533\n",
      "asteroma_ccl060                  0.000466\n",
      "scedosporium_boydii              0.000399\n",
      "wickerhamomyces_anomalus         0.000399\n",
      "dothiorella_vidmadera            0.000399\n",
      "cryptococcus_zero                0.000399\n",
      "clavispora_lusitaniae            0.000333\n",
      "pichia_kudriavzevii              0.000333\n",
      "zygoascus_hellenicus             0.000333\n",
      "pyrenophora_tritici-repentis     0.000333\n",
      "aspergillus_unidentified         0.000333\n",
      "penicillium_chrysogenum          0.000266\n",
      "oculimacula_yallundae-ccl029     0.000266\n",
      "blastobotrys_proliferans         0.000266\n",
      "rhodotorula_mucilaginosa         0.000266\n",
      "candida_parapsilosis             0.000200\n",
      "candida_orthopsilosis            0.000200\n",
      "galactomyces_geotrichum          0.000200\n",
      "kluyveromyces_unidentified       0.000200\n",
      "aspergillus_niger                0.000200\n",
      "quambalaria_cyanescens           0.000200\n",
      "yarrowia_lipolytica              0.000200\n",
      "aspergillus_flavus               0.000133\n",
      "candida_albicans                 0.000133\n",
      "candida_metapsilosis             0.000133\n",
      "pichia_membranifaciens           0.000133\n",
      "tuber_brumale                    0.000067\n",
      "meyerozyma_guilliermondii        0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: asteroma_ccl060\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "diaporthe_unidentified           0.519429\n",
      "asteroma_ccl060                  0.446994\n",
      "zygoascus_hellenicus             0.010301\n",
      "rhodotorula_mucilaginosa         0.010106\n",
      "pichia_kudriavzevii              0.005542\n",
      "diaporthe_ccl067                 0.001695\n",
      "entoleuca_unidentified           0.001108\n",
      "asteroma_ccl068                  0.000913\n",
      "oculimacula_yallundae-ccl029     0.000848\n",
      "fusarium_oxysporum               0.000456\n",
      "oculimacula_yallundae-ccl031     0.000391\n",
      "scedosporium_boydii              0.000326\n",
      "zymoseptoria_tritici             0.000261\n",
      "quambalaria_cyanescens           0.000261\n",
      "tuber_brumale                    0.000196\n",
      "dothiorella_vidmadera            0.000130\n",
      "cryptococcus_zero                0.000130\n",
      "candida_parapsilosis             0.000130\n",
      "aspergillus_unidentified         0.000130\n",
      "penicillium_chrysogenum          0.000065\n",
      "puccinia_striiformis-tritici     0.000065\n",
      "pyrenophora_tritici-repentis     0.000065\n",
      "cortinarius_globuliformis        0.000065\n",
      "clavispora_lusitaniae            0.000065\n",
      "cladophialophora_unidentified    0.000065\n",
      "candida_albicans                 0.000065\n",
      "blastobotrys_proliferans         0.000065\n",
      "saccharomyces_cerevisiae         0.000065\n",
      "aspergillus_flavus               0.000065\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: asteroma_ccl068\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "asteroma_ccl068                  0.556854\n",
      "diaporthe_ccl067                 0.431663\n",
      "asteroma_ccl060                  0.001792\n",
      "aspergillus_flavus               0.001394\n",
      "pichia_membranifaciens           0.001328\n",
      "diaporthe_unidentified           0.001062\n",
      "entoleuca_unidentified           0.001062\n",
      "oculimacula_yallundae-ccl029     0.000797\n",
      "oculimacula_yallundae-ccl031     0.000664\n",
      "scedosporium_boydii              0.000597\n",
      "quambalaria_cyanescens           0.000465\n",
      "cryptococcus_zero                0.000398\n",
      "aspergillus_unidentified         0.000398\n",
      "dothiorella_vidmadera            0.000332\n",
      "fusarium_oxysporum               0.000266\n",
      "aspergillus_niger                0.000266\n",
      "saccharomyces_cerevisiae         0.000133\n",
      "galactomyces_geotrichum          0.000133\n",
      "penicillium_chrysogenum          0.000066\n",
      "rhodotorula_mucilaginosa         0.000066\n",
      "candida_unidentified             0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "tuber_brumale                    0.000066\n",
      "yamadazyma_mexicana              0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: saccharomyces_cerevisiae\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "saccharomyces_cerevisiae         0.987414\n",
      "kluyveromyces_unidentified       0.002730\n",
      "kluyveromyces_marxianus          0.001132\n",
      "oculimacula_yallundae-ccl029     0.000733\n",
      "candida_metapsilosis             0.000599\n",
      "clavispora_lusitaniae            0.000599\n",
      "pichia_membranifaciens           0.000533\n",
      "debaryomyces_unidentified        0.000466\n",
      "entoleuca_unidentified           0.000466\n",
      "kodamaea_ohmeri                  0.000466\n",
      "wickerhamomyces_anomalus         0.000400\n",
      "yamadazyma_mexicana              0.000400\n",
      "candida_albicans                 0.000333\n",
      "pichia_kudriavzevii              0.000266\n",
      "candida_orthopsilosis            0.000266\n",
      "tuber_brumale                    0.000266\n",
      "dothiorella_vidmadera            0.000266\n",
      "blastobotrys_proliferans         0.000266\n",
      "meyerozyma_guilliermondii        0.000266\n",
      "cladophialophora_unidentified    0.000200\n",
      "yamadazyma_scolyti               0.000200\n",
      "candida_parapsilosis             0.000200\n",
      "galactomyces_geotrichum          0.000200\n",
      "asteroma_ccl060                  0.000133\n",
      "candida_unidentified             0.000133\n",
      "zymoseptoria_tritici             0.000133\n",
      "oculimacula_yallundae-ccl031     0.000133\n",
      "pyrenophora_tritici-repentis     0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "aspergillus_flavus               0.000133\n",
      "diaporthe_ccl067                 0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "aspergillus_niger                0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cladophialophora_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cladophialophora_unidentified    0.994075\n",
      "zymoseptoria_tritici             0.000666\n",
      "oculimacula_yallundae-ccl029     0.000466\n",
      "aspergillus_unidentified         0.000333\n",
      "tuber_brumale                    0.000333\n",
      "pyrenophora_tritici-repentis     0.000333\n",
      "clavispora_lusitaniae            0.000333\n",
      "kluyveromyces_unidentified       0.000266\n",
      "debaryomyces_unidentified        0.000266\n",
      "candida_albicans                 0.000200\n",
      "candida_metapsilosis             0.000200\n",
      "candida_orthopsilosis            0.000200\n",
      "dothiorella_vidmadera            0.000200\n",
      "entoleuca_unidentified           0.000200\n",
      "fusarium_oxysporum               0.000200\n",
      "oculimacula_yallundae-ccl031     0.000200\n",
      "rhodotorula_mucilaginosa         0.000133\n",
      "penicillium_chrysogenum          0.000133\n",
      "cryptococcus_zero                0.000133\n",
      "galactomyces_geotrichum          0.000133\n",
      "kluyveromyces_marxianus          0.000067\n",
      "aspergillus_niger                0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "candida_unidentified             0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "puccinia_striiformis-tritici     0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "diaporthe_unidentified           0.000067\n",
      "meyerozyma_guilliermondii        0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_albicans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_unidentified             0.424446\n",
      "candida_albicans                 0.361687\n",
      "candida_orthopsilosis            0.098020\n",
      "candida_metapsilosis             0.060062\n",
      "candida_parapsilosis             0.032169\n",
      "galactomyces_geotrichum          0.003618\n",
      "pyrenophora_tritici-repentis     0.003421\n",
      "diaporthe_ccl067                 0.002895\n",
      "debaryomyces_unidentified        0.002039\n",
      "meyerozyma_guilliermondii        0.001447\n",
      "yamadazyma_scolyti               0.001184\n",
      "kluyveromyces_unidentified       0.001184\n",
      "yamadazyma_mexicana              0.000987\n",
      "kodamaea_ohmeri                  0.000855\n",
      "kluyveromyces_marxianus          0.000855\n",
      "saccharomyces_cerevisiae         0.000592\n",
      "wickerhamomyces_anomalus         0.000526\n",
      "clavispora_lusitaniae            0.000526\n",
      "oculimacula_yallundae-ccl029     0.000460\n",
      "zygoascus_hellenicus             0.000460\n",
      "blastobotrys_proliferans         0.000395\n",
      "entoleuca_unidentified           0.000395\n",
      "tuber_brumale                    0.000197\n",
      "cortinarius_globuliformis        0.000197\n",
      "asteroma_ccl068                  0.000132\n",
      "zymoseptoria_tritici             0.000132\n",
      "oculimacula_yallundae-ccl031     0.000132\n",
      "pichia_kudriavzevii              0.000132\n",
      "pichia_membranifaciens           0.000132\n",
      "quambalaria_cyanescens           0.000132\n",
      "rhodotorula_mucilaginosa         0.000132\n",
      "yarrowia_lipolytica              0.000132\n",
      "cryptococcus_zero                0.000066\n",
      "fusarium_oxysporum               0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "scedosporium_boydii              0.000066\n",
      "aspergillus_flavus               0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_metapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_orthopsilosis            0.557372\n",
      "candida_metapsilosis             0.353986\n",
      "candida_parapsilosis             0.045290\n",
      "kodamaea_ohmeri                  0.016152\n",
      "diaporthe_unidentified           0.011888\n",
      "fusarium_oxysporum               0.006202\n",
      "candida_albicans                 0.001357\n",
      "candida_unidentified             0.001292\n",
      "meyerozyma_guilliermondii        0.000711\n",
      "debaryomyces_unidentified        0.000646\n",
      "saccharomyces_cerevisiae         0.000581\n",
      "kluyveromyces_unidentified       0.000581\n",
      "yamadazyma_mexicana              0.000517\n",
      "oculimacula_yallundae-ccl029     0.000388\n",
      "yamadazyma_scolyti               0.000388\n",
      "clavispora_lusitaniae            0.000323\n",
      "entoleuca_unidentified           0.000323\n",
      "kluyveromyces_marxianus          0.000258\n",
      "pichia_kudriavzevii              0.000194\n",
      "pyrenophora_tritici-repentis     0.000194\n",
      "zygoascus_hellenicus             0.000194\n",
      "cladophialophora_unidentified    0.000194\n",
      "yarrowia_lipolytica              0.000129\n",
      "galactomyces_geotrichum          0.000129\n",
      "wickerhamomyces_anomalus         0.000129\n",
      "blastobotrys_proliferans         0.000065\n",
      "asteroma_ccl060                  0.000065\n",
      "asteroma_ccl068                  0.000065\n",
      "zymoseptoria_tritici             0.000065\n",
      "oculimacula_yallundae-ccl031     0.000065\n",
      "pichia_membranifaciens           0.000065\n",
      "puccinia_striiformis-tritici     0.000065\n",
      "tuber_brumale                    0.000065\n",
      "aspergillus_flavus               0.000065\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_orthopsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_orthopsilosis            0.567648\n",
      "candida_metapsilosis             0.366399\n",
      "candida_parapsilosis             0.046207\n",
      "meyerozyma_guilliermondii        0.009441\n",
      "candida_unidentified             0.002194\n",
      "candida_albicans                 0.001263\n",
      "debaryomyces_unidentified        0.001197\n",
      "saccharomyces_cerevisiae         0.000665\n",
      "kluyveromyces_marxianus          0.000465\n",
      "clavispora_lusitaniae            0.000399\n",
      "entoleuca_unidentified           0.000332\n",
      "yamadazyma_scolyti               0.000332\n",
      "kluyveromyces_unidentified       0.000266\n",
      "wickerhamomyces_anomalus         0.000266\n",
      "yamadazyma_mexicana              0.000266\n",
      "blastobotrys_proliferans         0.000199\n",
      "zygoascus_hellenicus             0.000199\n",
      "galactomyces_geotrichum          0.000199\n",
      "oculimacula_yallundae-ccl029     0.000199\n",
      "oculimacula_yallundae-ccl031     0.000199\n",
      "quambalaria_cyanescens           0.000199\n",
      "pichia_membranifaciens           0.000199\n",
      "aspergillus_flavus               0.000133\n",
      "aspergillus_unidentified         0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "puccinia_striiformis-tritici     0.000066\n",
      "aspergillus_niger                0.000066\n",
      "asteroma_ccl068                  0.000066\n",
      "tuber_brumale                    0.000066\n",
      "scedosporium_boydii              0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "yarrowia_lipolytica              0.000066\n",
      "penicillium_chrysogenum          0.000066\n",
      "cortinarius_globuliformis        0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "diaporthe_unidentified           0.000066\n",
      "dothiorella_vidmadera            0.000066\n",
      "pichia_kudriavzevii              0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_parapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_parapsilosis             0.828866\n",
      "candida_orthopsilosis            0.090323\n",
      "candida_metapsilosis             0.072098\n",
      "candida_unidentified             0.001330\n",
      "meyerozyma_guilliermondii        0.000998\n",
      "debaryomyces_unidentified        0.000865\n",
      "candida_albicans                 0.000732\n",
      "kluyveromyces_unidentified       0.000466\n",
      "wickerhamomyces_anomalus         0.000399\n",
      "oculimacula_yallundae-ccl029     0.000399\n",
      "entoleuca_unidentified           0.000333\n",
      "cortinarius_globuliformis        0.000333\n",
      "kluyveromyces_marxianus          0.000333\n",
      "saccharomyces_cerevisiae         0.000266\n",
      "yamadazyma_mexicana              0.000266\n",
      "tuber_brumale                    0.000200\n",
      "yamadazyma_scolyti               0.000200\n",
      "clavispora_lusitaniae            0.000133\n",
      "cryptococcus_zero                0.000133\n",
      "zymoseptoria_tritici             0.000133\n",
      "kodamaea_ohmeri                  0.000133\n",
      "pyrenophora_tritici-repentis     0.000133\n",
      "galactomyces_geotrichum          0.000067\n",
      "dothiorella_vidmadera            0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "aspergillus_niger                0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_unidentified             0.818551\n",
      "candida_albicans                 0.166079\n",
      "candida_metapsilosis             0.002861\n",
      "candida_orthopsilosis            0.002063\n",
      "meyerozyma_guilliermondii        0.001530\n",
      "debaryomyces_unidentified        0.001131\n",
      "candida_parapsilosis             0.000865\n",
      "yamadazyma_mexicana              0.000798\n",
      "saccharomyces_cerevisiae         0.000732\n",
      "kluyveromyces_unidentified       0.000665\n",
      "oculimacula_yallundae-ccl029     0.000665\n",
      "entoleuca_unidentified           0.000599\n",
      "blastobotrys_proliferans         0.000466\n",
      "kodamaea_ohmeri                  0.000466\n",
      "yarrowia_lipolytica              0.000333\n",
      "wickerhamomyces_anomalus         0.000266\n",
      "yamadazyma_scolyti               0.000266\n",
      "pichia_membranifaciens           0.000200\n",
      "tuber_brumale                    0.000200\n",
      "oculimacula_yallundae-ccl031     0.000133\n",
      "zygoascus_hellenicus             0.000133\n",
      "diaporthe_unidentified           0.000133\n",
      "cryptococcus_zero                0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "kluyveromyces_marxianus          0.000133\n",
      "zymoseptoria_tritici             0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "cortinarius_globuliformis        0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kluyveromyces_marxianus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kluyveromyces_marxianus          0.633453\n",
      "kluyveromyces_unidentified       0.352464\n",
      "saccharomyces_cerevisiae         0.002657\n",
      "candida_metapsilosis             0.000930\n",
      "wickerhamomyces_anomalus         0.000864\n",
      "entoleuca_unidentified           0.000664\n",
      "yarrowia_lipolytica              0.000664\n",
      "oculimacula_yallundae-ccl029     0.000598\n",
      "candida_parapsilosis             0.000598\n",
      "debaryomyces_unidentified        0.000531\n",
      "meyerozyma_guilliermondii        0.000465\n",
      "yamadazyma_scolyti               0.000465\n",
      "yamadazyma_mexicana              0.000465\n",
      "penicillium_chrysogenum          0.000399\n",
      "tuber_brumale                    0.000399\n",
      "clavispora_lusitaniae            0.000332\n",
      "candida_unidentified             0.000332\n",
      "candida_albicans                 0.000332\n",
      "zymoseptoria_tritici             0.000266\n",
      "kodamaea_ohmeri                  0.000266\n",
      "pichia_membranifaciens           0.000266\n",
      "cladophialophora_unidentified    0.000266\n",
      "zygoascus_hellenicus             0.000199\n",
      "pichia_kudriavzevii              0.000199\n",
      "dothiorella_vidmadera            0.000199\n",
      "quambalaria_cyanescens           0.000199\n",
      "blastobotrys_proliferans         0.000199\n",
      "cryptococcus_zero                0.000199\n",
      "candida_orthopsilosis            0.000133\n",
      "galactomyces_geotrichum          0.000133\n",
      "cortinarius_globuliformis        0.000133\n",
      "diaporthe_ccl067                 0.000066\n",
      "diaporthe_unidentified           0.000066\n",
      "oculimacula_yallundae-ccl031     0.000066\n",
      "puccinia_striiformis-tritici     0.000066\n",
      "rhodotorula_mucilaginosa         0.000066\n",
      "scedosporium_boydii              0.000066\n",
      "asteroma_ccl068                  0.000066\n",
      "asteroma_ccl060                  0.000066\n",
      "aspergillus_unidentified         0.000066\n",
      "aspergillus_niger                0.000066\n",
      "aspergillus_flavus               0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_kudriavzevii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_kudriavzevii              0.944954\n",
      "zygoascus_hellenicus             0.018135\n",
      "rhodotorula_mucilaginosa         0.017943\n",
      "asteroma_ccl060                  0.012047\n",
      "pichia_membranifaciens           0.002371\n",
      "kodamaea_ohmeri                  0.000449\n",
      "kluyveromyces_unidentified       0.000384\n",
      "candida_unidentified             0.000384\n",
      "candida_orthopsilosis            0.000384\n",
      "saccharomyces_cerevisiae         0.000320\n",
      "yamadazyma_scolyti               0.000256\n",
      "oculimacula_yallundae-ccl029     0.000256\n",
      "yamadazyma_mexicana              0.000192\n",
      "yarrowia_lipolytica              0.000192\n",
      "tuber_brumale                    0.000192\n",
      "kluyveromyces_marxianus          0.000192\n",
      "candida_parapsilosis             0.000192\n",
      "cortinarius_globuliformis        0.000192\n",
      "candida_albicans                 0.000128\n",
      "candida_metapsilosis             0.000128\n",
      "meyerozyma_guilliermondii        0.000128\n",
      "clavispora_lusitaniae            0.000128\n",
      "quambalaria_cyanescens           0.000128\n",
      "cladophialophora_unidentified    0.000064\n",
      "debaryomyces_unidentified        0.000064\n",
      "penicillium_chrysogenum          0.000064\n",
      "scedosporium_boydii              0.000064\n",
      "wickerhamomyces_anomalus         0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_membranifaciens\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_membranifaciens           0.987299\n",
      "pichia_kudriavzevii              0.003591\n",
      "asteroma_ccl068                  0.000731\n",
      "scedosporium_boydii              0.000598\n",
      "aspergillus_flavus               0.000598\n",
      "saccharomyces_cerevisiae         0.000532\n",
      "kluyveromyces_marxianus          0.000532\n",
      "kodamaea_ohmeri                  0.000465\n",
      "kluyveromyces_unidentified       0.000465\n",
      "oculimacula_yallundae-ccl029     0.000399\n",
      "candida_metapsilosis             0.000332\n",
      "candida_albicans                 0.000332\n",
      "debaryomyces_unidentified        0.000332\n",
      "yamadazyma_scolyti               0.000332\n",
      "candida_unidentified             0.000332\n",
      "candida_orthopsilosis            0.000266\n",
      "candida_parapsilosis             0.000266\n",
      "meyerozyma_guilliermondii        0.000266\n",
      "clavispora_lusitaniae            0.000199\n",
      "cortinarius_globuliformis        0.000199\n",
      "zygoascus_hellenicus             0.000199\n",
      "entoleuca_unidentified           0.000199\n",
      "tuber_brumale                    0.000199\n",
      "yamadazyma_mexicana              0.000199\n",
      "cryptococcus_zero                0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "quambalaria_cyanescens           0.000133\n",
      "wickerhamomyces_anomalus         0.000133\n",
      "galactomyces_geotrichum          0.000133\n",
      "dothiorella_vidmadera            0.000066\n",
      "diaporthe_ccl067                 0.000066\n",
      "fusarium_oxysporum               0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "oculimacula_yallundae-ccl031     0.000066\n",
      "puccinia_striiformis-tritici     0.000066\n",
      "aspergillus_unidentified         0.000066\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "### this is running the reads against the full database\n",
    "### These results show which species each sample match to and the\n",
    "### proportion of reads matching to each species\n",
    "for species, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "puccinia_striiformis-tritici\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9955339288094921,\n",
      " \"c\": 0.9944674043460872,\n",
      " \"o\": 0.9944674043460872,\n",
      " \"f\": 0.9944674043460872,\n",
      " \"g\": 0.9944674043460872,\n",
      " \"s\": 0.9944674043460872\n",
      "}\n",
      "zymoseptoria_tritici\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9996670218433671,\n",
      " \"c\": 0.9938066062866276,\n",
      " \"o\": 0.9915423548215238,\n",
      " \"f\": 0.9915423548215238,\n",
      " \"g\": 0.9915423548215238,\n",
      " \"s\": 0.9915423548215238\n",
      "}\n",
      "pyrenophora_tritici-repentis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9996680167319567,\n",
      " \"c\": 0.9907708651483965,\n",
      " \"o\": 0.9568421751543722,\n",
      " \"f\": 0.9568421751543722,\n",
      " \"g\": 0.9568421751543722,\n",
      " \"s\": 0.9568421751543722\n",
      "}\n",
      "fusarium_oxysporum\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9996061700032819,\n",
      " \"c\": 0.9811617984903184,\n",
      " \"o\": 0.9725631768953069,\n",
      " \"f\": 0.9725631768953069,\n",
      " \"g\": 0.9725631768953069,\n",
      " \"s\": 0.9725631768953069\n",
      "}\n",
      "tuber_brumale\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9994016753091344,\n",
      " \"c\": 0.9900279218189071,\n",
      " \"o\": 0.9900279218189071,\n",
      " \"f\": 0.9900279218189071,\n",
      " \"g\": 0.9900279218189071,\n",
      " \"s\": 0.9900279218189071\n",
      "}\n",
      "cortinarius_globuliformis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9924116354922452,\n",
      " \"c\": 0.9906143912667243,\n",
      " \"o\": 0.9906143912667243,\n",
      " \"f\": 0.9906143912667243,\n",
      " \"g\": 0.9906143912667243,\n",
      " \"s\": 0.9906143912667243\n",
      "}\n",
      "aspergillus_niger\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9994722607032126,\n",
      " \"c\": 0.9881258658222838,\n",
      " \"o\": 0.9879279635859886,\n",
      " \"f\": 0.9879279635859886,\n",
      " \"g\": 0.9862128108714295,\n",
      " \"s\": 0.754601226993865\n",
      "}\n",
      "clavispora_lusitaniae\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9994673768308922,\n",
      " \"c\": 0.9973368841544608,\n",
      " \"o\": 0.9973368841544608,\n",
      " \"f\": 0.9868175765645806,\n",
      " \"g\": 0.9848868175765646,\n",
      " \"s\": 0.9848868175765646\n",
      "}\n",
      "kluyveromyces_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9302157327019206,\n",
      " \"c\": 0.9208103130755064,\n",
      " \"o\": 0.9208103130755064,\n",
      " \"f\": 0.9040384109444883,\n",
      " \"g\": 0.902657195474875,\n",
      " \"s\": 0.374177847934754\n",
      "}\n",
      "penicillium_chrysogenum\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9992014905509715,\n",
      " \"c\": 0.9937450093159436,\n",
      " \"o\": 0.9937450093159436,\n",
      " \"f\": 0.9937450093159436,\n",
      " \"g\": 0.9869576789992015,\n",
      " \"s\": 0.9869576789992015\n",
      "}\n",
      "rhodotorula_mucilaginosa\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9842327483836918,\n",
      " \"c\": 0.9825834542815675,\n",
      " \"o\": 0.9825834542815675,\n",
      " \"f\": 0.9825834542815675,\n",
      " \"g\": 0.9825834542815675,\n",
      " \"s\": 0.9825834542815675\n",
      "}\n",
      "scedosporium_boydii\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9992012247886574,\n",
      " \"c\": 0.9948745257272182,\n",
      " \"o\": 0.9916128602809026,\n",
      " \"f\": 0.9916128602809026,\n",
      " \"g\": 0.9916128602809026,\n",
      " \"s\": 0.9916128602809026\n",
      "}\n",
      "blastobotrys_proliferans\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9999333733093477,\n",
      " \"c\": 0.9985342128056499,\n",
      " \"o\": 0.9985342128056499,\n",
      " \"f\": 0.9950029982010794,\n",
      " \"g\": 0.9950029982010794,\n",
      " \"s\": 0.9950029982010794\n",
      "}\n",
      "debaryomyces_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9999334132374484,\n",
      " \"c\": 0.9979358103609003,\n",
      " \"o\": 0.9979358103609003,\n",
      " \"f\": 0.9890131841789852,\n",
      " \"g\": 0.9844186975629244,\n",
      " \"s\": 0.9855506725263018\n",
      "}\n",
      "galactomyces_geotrichum\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9991906656774803,\n",
      " \"c\": 0.9942672152154852,\n",
      " \"o\": 0.9942672152154852,\n",
      " \"f\": 0.9852970931408916,\n",
      " \"g\": 0.9847575369258784,\n",
      " \"s\": 0.9847575369258784\n",
      "}\n",
      "kodamaea_ohmeri\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9998085635887946,\n",
      " \"c\": 0.9670091251356008,\n",
      " \"o\": 0.9670091251356008,\n",
      " \"f\": 0.9465892412736903,\n",
      " \"g\": 0.9462063684512795,\n",
      " \"s\": 0.9462063684512795\n",
      "}\n",
      "meyerozyma_guilliermondii\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9996671991480298,\n",
      " \"c\": 0.998069755058573,\n",
      " \"o\": 0.998069755058573,\n",
      " \"f\": 0.9916799787007454,\n",
      " \"g\": 0.9882854100106496,\n",
      " \"s\": 0.9882854100106496\n",
      "}\n",
      "wickerhamomyces_anomalus\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9998667732480682,\n",
      " \"c\": 0.9986011191047163,\n",
      " \"o\": 0.9986011191047163,\n",
      " \"f\": 0.9922728483879563,\n",
      " \"g\": 0.9922728483879563,\n",
      " \"s\": 0.9922728483879563\n",
      "}\n",
      "yamadazyma_mexicana\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9999334176709501,\n",
      " \"c\": 0.9981356947866037,\n",
      " \"o\": 0.9981356947866037,\n",
      " \"f\": 0.99487316066316,\n",
      " \"g\": 0.9927425261335642,\n",
      " \"s\": 0.9796923896397896\n",
      "}\n",
      "yamadazyma_scolyti\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9998001731832412,\n",
      " \"c\": 0.9994005195497235,\n",
      " \"o\": 0.9994005195497235,\n",
      " \"f\": 0.9967361619929395,\n",
      " \"g\": 0.9946046759475121,\n",
      " \"s\": 0.987077865849597\n",
      "}\n",
      "yarrowia_lipolytica\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 1.0,\n",
      " \"c\": 0.9996667777407531,\n",
      " \"o\": 0.9996667777407531,\n",
      " \"f\": 0.9989336887704099,\n",
      " \"g\": 0.9988670443185604,\n",
      " \"s\": 0.9988670443185604\n",
      "}\n",
      "zygoascus_hellenicus\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9929476296199556,\n",
      " \"c\": 0.9849157633537939,\n",
      " \"o\": 0.9849157633537939,\n",
      " \"f\": 0.9731618127203866,\n",
      " \"g\": 0.9730312132689043,\n",
      " \"s\": 0.9730312132689043\n",
      "}\n",
      "aspergillus_flavus\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9998667732480682,\n",
      " \"c\": 0.997468691713296,\n",
      " \"o\": 0.9974020783373302,\n",
      " \"f\": 0.9974020783373302,\n",
      " \"g\": 0.9960031974420464,\n",
      " \"s\": 0.9891420197175593\n",
      "}\n",
      "cryptococcus_zero\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9960002666488901,\n",
      " \"c\": 0.9942003866408906,\n",
      " \"o\": 0.9942003866408906,\n",
      " \"f\": 0.9942003866408906,\n",
      " \"g\": 0.9942003866408906,\n",
      " \"s\": 0.9942003866408906\n",
      "}\n",
      "aspergillus_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9997336529497935,\n",
      " \"c\": 0.9950059928086297,\n",
      " \"o\": 0.9947396457584232,\n",
      " \"f\": 0.9947396457584232,\n",
      " \"g\": 0.9936742575575975,\n",
      " \"s\": 0.669463310693834\n",
      "}\n",
      "diaporthe_ccl067\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9992726310917146,\n",
      " \"c\": 0.9892217152681346,\n",
      " \"o\": 0.9881637241288104,\n",
      " \"f\": 0.4576472922039278,\n",
      " \"g\": 0.4576472922039278,\n",
      " \"s\": 0.45592805660252594\n",
      "}\n",
      "diaporthe_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9996076379806434,\n",
      " \"c\": 0.981624378760136,\n",
      " \"o\": 0.9745618624117185,\n",
      " \"f\": 0.5541459586712006,\n",
      " \"g\": 0.5541459586712006,\n",
      " \"s\": 0.5545383206905572\n",
      "}\n",
      "oculimacula_yallundae-ccl031\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9986684420772304,\n",
      " \"c\": 0.9920772303595207,\n",
      " \"o\": 0.9920772303595207,\n",
      " \"f\": 0.9920772303595207,\n",
      " \"g\": 0.9920772303595207,\n",
      " \"s\": 0.9920772303595207\n",
      "}\n",
      "oculimacula_yallundae-ccl029\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9987368700970616,\n",
      " \"c\": 0.9887647919159687,\n",
      " \"o\": 0.9887647919159687,\n",
      " \"f\": 0.9887647919159687,\n",
      " \"g\": 0.9887647919159687,\n",
      " \"s\": 0.9887647919159687\n",
      "}\n",
      "dothiorella_vidmadera\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9990015309858218,\n",
      " \"c\": 0.9890834054449844,\n",
      " \"o\": 0.9866205152100113,\n",
      " \"f\": 0.9866205152100113,\n",
      " \"g\": 0.9866205152100113,\n",
      " \"s\": 0.9866205152100113\n",
      "}\n",
      "quambalaria_cyanescens\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9938674843354219,\n",
      " \"c\": 0.9925343287561659,\n",
      " \"o\": 0.9925343287561659,\n",
      " \"f\": 0.9925343287561659,\n",
      " \"g\": 0.9925343287561659,\n",
      " \"s\": 0.9925343287561659\n",
      "}\n",
      "entoleuca_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9991345449703748,\n",
      " \"c\": 0.9918114639504694,\n",
      " \"o\": 0.9882164969043339,\n",
      " \"f\": 0.9882164969043339,\n",
      " \"g\": 0.9882164969043339,\n",
      " \"s\": 0.9898142600359496\n",
      "}\n",
      "asteroma_ccl060\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9893727995827357,\n",
      " \"c\": 0.9709218933368106,\n",
      " \"o\": 0.9690311644282175,\n",
      " \"f\": 0.44790715869083325,\n",
      " \"g\": 0.44790715869083325,\n",
      " \"s\": 0.4469943930108228\n",
      "}\n",
      "asteroma_ccl068\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9990706936608031,\n",
      " \"c\": 0.9932957185529373,\n",
      " \"o\": 0.9913707268503154,\n",
      " \"f\": 0.5586458679057418,\n",
      " \"g\": 0.5586458679057418,\n",
      " \"s\": 0.5568536342515765\n",
      "}\n",
      "saccharomyces_cerevisiae\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 1.0,\n",
      " \"c\": 0.9970699873476726,\n",
      " \"o\": 0.9970699873476726,\n",
      " \"f\": 0.9912765532396617,\n",
      " \"g\": 0.9874142638343211,\n",
      " \"s\": 0.9874142638343211\n",
      "}\n",
      "cladophialophora_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9996671326809133,\n",
      " \"c\": 0.9946741228946142,\n",
      " \"o\": 0.9940749617202583,\n",
      " \"f\": 0.9940749617202583,\n",
      " \"g\": 0.9940749617202583,\n",
      " \"s\": 0.9952732840689701\n",
      "}\n",
      "candida_albicans\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9994737188342873,\n",
      " \"c\": 0.9914479310571673,\n",
      " \"o\": 0.9914479310571673,\n",
      " \"f\": 0.9763831326886389,\n",
      " \"g\": 0.9763831326886389,\n",
      " \"s\": 0.36168673113610944\n",
      "}\n",
      "candida_metapsilosis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.999935392169531,\n",
      " \"c\": 0.9803592195374079,\n",
      " \"o\": 0.9803592195374079,\n",
      " \"f\": 0.9592970668044967,\n",
      " \"g\": 0.9592970668044967,\n",
      " \"s\": 0.35398630313994056\n",
      "}\n",
      "candida_orthopsilosis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.999601090353035,\n",
      " \"c\": 0.9980054517651752,\n",
      " \"o\": 0.9980054517651752,\n",
      " \"f\": 0.9837111894155973,\n",
      " \"g\": 0.9837111894155973,\n",
      " \"s\": 0.5676484276311415\n",
      "}\n",
      "candida_parapsilosis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9994013967409379,\n",
      " \"c\": 0.9977386099102095,\n",
      " \"o\": 0.9977386099102095,\n",
      " \"f\": 0.9933488526770868,\n",
      " \"g\": 0.9933488526770868,\n",
      " \"s\": 0.8288659793814434\n",
      "}\n",
      "candida_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9996673098675893,\n",
      " \"c\": 0.9977377070996074,\n",
      " \"o\": 0.9977377070996074,\n",
      " \"f\": 0.9904185241865726,\n",
      " \"g\": 0.9904185241865726,\n",
      " \"s\": 0.8211457848160224\n",
      "}\n",
      "kluyveromyces_marxianus\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9993357247243257,\n",
      " \"c\": 0.9959479208183871,\n",
      " \"o\": 0.9959479208183871,\n",
      " \"f\": 0.9885744652584031,\n",
      " \"g\": 0.9859173641557061,\n",
      " \"s\": 0.6334529028829547\n",
      "}\n",
      "pichia_kudriavzevii\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9817366228772829,\n",
      " \"c\": 0.969048381928869,\n",
      " \"o\": 0.969048381928869,\n",
      " \"f\": 0.9473245754565844,\n",
      " \"g\": 0.9473245754565844,\n",
      " \"s\": 0.9449535405318809\n",
      "}\n",
      "pichia_membranifaciens\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9994680143636122,\n",
      " \"c\": 0.9963425987498338,\n",
      " \"o\": 0.9963425987498338,\n",
      " \"f\": 0.9908897459768586,\n",
      " \"g\": 0.9908897459768586,\n",
      " \"s\": 0.9872988429312408\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "### These results show which the proportion of reads that match\n",
    "### the sample label at each taxonomic rank\n",
    "\n",
    "for species in mock_community:\n",
    "    print(species)\n",
    "    mapping_results = pull_mapping_results(sub_db_mapping_fn[species])\n",
    "    \n",
    "    mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "    taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "    query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "    sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "    print(json.dumps(sensitivity_dict, indent=1))\n",
    "    \n",
    "    with open('/media/MassStorage/tmp/TE/honours/analysis/Mapping/custom_results/tax_rank/%s.json' % species,'w+') as json_file:\n",
    "        json.dump(sensitivity_dict,json_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook to analyzse the efficiency of minimap2 mapping for Oxford Nanopore MinION reads against Qiime2 Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/media/MassStorage/tmp/TE/honours/scripts/Notebooks'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.abspath(os.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "puccinia_striiformis-tritici\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Pucciniomycetes', 'o': 'Pucciniales', 'f': 'Pucciniaceae', 'g': 'Puccinia', 's': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9562130177514793,\n",
      " \"c\": 0.8740959894806049,\n",
      " \"o\": 0.8740959894806049,\n",
      " \"f\": 0.8723208415516108,\n",
      " \"g\": 0.8619986850756082,\n",
      " \"s\": 0.801051939513478\n",
      "}\n",
      "zymoseptoria_tritici\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Dothideomycetes', 'o': 'Capnodiales', 'f': 'Mycosphaerellaceae', 'g': 'Zymoseptoria', 's': 'Zymoseptoria_tritici'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8341285085816965,\n",
      " \"c\": 0.35993555982402875,\n",
      " \"o\": 0.18811574446991758,\n",
      " \"f\": 0.1839023483487205,\n",
      " \"g\": 0.16159613358944172,\n",
      " \"s\": 0.0\n",
      "}\n",
      "pyrenophora_tritici-repentis\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Dothideomycetes', 'o': 'Pleosporales', 'f': 'Pleosporaceae', 'g': 'Pyrenophora', 's': 'Pyrenophora_tritici-repentis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.844637829536729,\n",
      " \"c\": 0.6033401586895316,\n",
      " \"o\": 0.5959175838239058,\n",
      " \"f\": 0.35865113898131556,\n",
      " \"g\": 0.3045175326337343,\n",
      " \"s\": 0.2966470437675966\n",
      "}\n",
      "fusarium_oxysporum\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Sordariomycetes', 'o': 'Hypocreales', 'f': 'Nectriaceae', 'g': 'Fusarium', 's': 'Fusarium_oxysporum'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9889788015687832,\n",
      " \"c\": 0.8369656952787569,\n",
      " \"o\": 0.5967830015389962,\n",
      " \"f\": 0.5790100779427096,\n",
      " \"g\": 0.49262771186019955,\n",
      " \"s\": 0.10599215608399941\n",
      "}\n",
      "tuber_brumale\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Pezizomycetes', 'o': 'Pezizales', 'f': 'Tuberaceae', 'g': 'Tuber', 's': 'Tuber_brumale'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9242064217388656,\n",
      " \"c\": 0.48333637969901905,\n",
      " \"o\": 0.48333637969901905,\n",
      " \"f\": 0.48333637969901905,\n",
      " \"g\": 0.48333637969901905,\n",
      " \"s\": 0.48327545238530434\n",
      "}\n",
      "cortinarius_globuliformis\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Agaricomycetes', 'o': 'Agaricales', 'f': 'Cortinariaceae', 'g': 'Cortinarius', 's': 'Cortinarius_globuliformis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9101073345259392,\n",
      " \"c\": 0.8448121645796064,\n",
      " \"o\": 0.7964477383082035,\n",
      " \"f\": 0.4217352415026834,\n",
      " \"g\": 0.40780730897009965,\n",
      " \"s\": 0.1348709430104779\n",
      "}\n",
      "aspergillus_niger\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Eurotiomycetes', 'o': 'Eurotiales', 'f': 'Aspergillaceae', 'g': 'Aspergillus', 's': 'Aspergillus_niger'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.94044933682216,\n",
      " \"c\": 0.619146440494451,\n",
      " \"o\": 0.6179734728864026,\n",
      " \"f\": 0.6177479021925472,\n",
      " \"g\": 0.4781196426960209,\n",
      " \"s\": 0.15614003428674547\n",
      "}\n",
      "clavispora_lusitaniae\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Metschnikowiaceae', 'g': 'Clavispora', 's': 'Clavispora_lusitaniae'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9166620552266062,\n",
      " \"c\": 0.5796580156051132,\n",
      " \"o\": 0.5796580156051132,\n",
      " \"f\": 0.17547451718222568,\n",
      " \"g\": 0.1746997952520613,\n",
      " \"s\": 0.1722096176193902\n",
      "}\n",
      "kluyveromyces_unidentified\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetaceae', 'g': 'Kluyveromyces', 's': 'unidentified'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8710018668326074,\n",
      " \"c\": 0.7006845052893591,\n",
      " \"o\": 0.7006845052893591,\n",
      " \"f\": 0.6092097075295582,\n",
      " \"g\": 0.4976975731176105,\n",
      " \"s\": 0.0\n",
      "}\n",
      "penicillium_chrysogenum\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Eurotiomycetes', 'o': 'Eurotiales', 'f': 'Aspergillaceae', 'g': 'Penicillium', 's': 'Penicillium_chrysogenum'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9554882091107112,\n",
      " \"c\": 0.6781418694952174,\n",
      " \"o\": 0.6507718533952079,\n",
      " \"f\": 0.6482621460365565,\n",
      " \"g\": 0.5539350317264893,\n",
      " \"s\": 0.045885026991192346\n",
      "}\n",
      "rhodotorula_mucilaginosa\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Microbotryomycetes', 'o': 'Sporidiobolales', 'f': 'Sporidiobolaceae', 'g': 'Rhodotorula', 's': 'Rhodotorula_mucilaginosa'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9285580815945188,\n",
      " \"c\": 0.8092805979445655,\n",
      " \"o\": 0.8088445966988477,\n",
      " \"f\": 0.8088445966988477,\n",
      " \"g\": 0.7986919962628465,\n",
      " \"s\": 0.6029274369355341\n",
      "}\n",
      "scedosporium_boydii\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Sordariomycetes', 'o': 'Microascales', 'f': 'Microascaceae', 'g': 'Scedosporium', 's': 'Scedosporium_boydii'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9904975812024879,\n",
      " \"c\": 0.8690969822621516,\n",
      " \"o\": 0.7303040774015204,\n",
      " \"f\": 0.49550794747753973,\n",
      " \"g\": 0.2988366735775167,\n",
      " \"s\": 0.28656991476618293\n",
      "}\n",
      "blastobotrys_proliferans\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Trichomonascaceae', 'g': 'Blastobotrys', 's': 'Blastobotrys_proliferans'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9189696136051019,\n",
      " \"c\": 0.664124046517444,\n",
      " \"o\": 0.664124046517444,\n",
      " \"f\": 0.540515193197449,\n",
      " \"g\": 0.531824434162811,\n",
      " \"s\": 0.5312617231461798\n",
      "}\n",
      "debaryomyces_unidentified\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Debaryomycetaceae', 'g': 'Debaryomyces', 's': 'unidentified'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9712056827551214,\n",
      " \"c\": 0.884125071351557,\n",
      " \"o\": 0.884125071351557,\n",
      " \"f\": 0.02257880383078582,\n",
      " \"g\": 0.0013953193378575506,\n",
      " \"s\": 0.0\n",
      "}\n",
      "galactomyces_geotrichum\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Dipodascaceae', 'g': 'Galactomyces', 's': 'Galactomyces_geotrichum'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9411825298722702,\n",
      " \"c\": 0.7701380304903173,\n",
      " \"o\": 0.7575710754017305,\n",
      " \"f\": 0.7259476720230738,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "kodamaea_ohmeri\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Metschnikowiaceae', 'g': 'Kodamaea', 's': 'Kodamaea_ohmeri'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8881764555014107,\n",
      " \"c\": 0.625673249551167,\n",
      " \"o\": 0.625673249551167,\n",
      " \"f\": 0.45800205180815595,\n",
      " \"g\": 0.45800205180815595,\n",
      " \"s\": 0.45800205180815595\n",
      "}\n",
      "meyerozyma_guilliermondii\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Debaryomycetaceae', 'g': 'Meyerozyma', 's': 'Meyerozyma_guilliermondii'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.963316304548532,\n",
      " \"c\": 0.7987936234381732,\n",
      " \"o\": 0.7987936234381732,\n",
      " \"f\": 0.5183110728134425,\n",
      " \"g\": 0.5098787468455714,\n",
      " \"s\": 0.44759032436757556\n",
      "}\n",
      "wickerhamomyces_anomalus\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Phaffomycetaceae', 'g': 'Wickerhamomyces', 's': 'Wickerhamomyces_anomalus'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9676904412234212,\n",
      " \"c\": 0.8233190728561395,\n",
      " \"o\": 0.8233190728561395,\n",
      " \"f\": 0.6517463763488921,\n",
      " \"g\": 0.6511078475193155,\n",
      " \"s\": 0.6206500223485091\n",
      "}\n",
      "yamadazyma_mexicana\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Debaryomycetaceae', 'g': 'Yamadazyma', 's': 'Yamadazyma_mexicana'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9470909552845529,\n",
      " \"c\": 0.7951600609756098,\n",
      " \"o\": 0.7951600609756098,\n",
      " \"f\": 0.6092479674796748,\n",
      " \"g\": 0.5994029471544715,\n",
      " \"s\": 0.5383638211382114\n",
      "}\n",
      "yamadazyma_scolyti\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Debaryomycetaceae', 'g': 'Yamadazyma', 's': 'Yamadazyma_scolyti'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9493841416474211,\n",
      " \"c\": 0.8085706954067231,\n",
      " \"o\": 0.8085706954067231,\n",
      " \"f\": 0.45008981267641773,\n",
      " \"g\": 0.44194251988709266,\n",
      " \"s\": 0.42596869386707725\n",
      "}\n",
      "yarrowia_lipolytica\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Dipodascaceae', 'g': 'Yarrowia', 's': 'Yarrowia_lipolytica'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8699395964528981,\n",
      " \"c\": 0.20376558282997045,\n",
      " \"o\": 0.20376558282997045,\n",
      " \"f\": 0.00025703637064644646,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "zygoascus_hellenicus\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Trichomonascaceae', 'g': 'Zygoascus', 's': 'Zygoascus_hellenicus'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9421318498625232,\n",
      " \"c\": 0.794935737579129,\n",
      " \"o\": 0.794935737579129,\n",
      " \"f\": 0.7159664940213568,\n",
      " \"g\": 0.6739561353027688,\n",
      " \"s\": 0.6504252190037726\n",
      "}\n",
      "aspergillus_flavus\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Eurotiomycetes', 'o': 'Eurotiales', 'f': 'Aspergillaceae', 'g': 'Aspergillus', 's': 'Aspergillus_flavus'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.951682519382943,\n",
      " \"c\": 0.7709215890016785,\n",
      " \"o\": 0.7702821517064983,\n",
      " \"f\": 0.7702421868755495,\n",
      " \"g\": 0.6473902965390457,\n",
      " \"s\": 0.5963552074174726\n",
      "}\n",
      "cryptococcus_zero\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Tremellomycetes', 'o': 'Tremellales', 'f': 'Tremellaceae', 'g': 'Cryptococcus', 's': 'Cryptococcus_zero'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8877897252090801,\n",
      " \"c\": 0.773142174432497,\n",
      " \"o\": 0.0021983273596176823,\n",
      " \"f\": 0.00023894862604540023,\n",
      " \"g\": 0.0021983273596176823,\n",
      " \"s\": 0.0\n",
      "}\n",
      "aspergillus_unidentified\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Eurotiomycetes', 'o': 'Eurotiales', 'f': 'Aspergillaceae', 'g': 'Aspergillus', 's': 'unidentified'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9483128364479846,\n",
      " \"c\": 0.6535078121580813,\n",
      " \"o\": 0.6524574379622741,\n",
      " \"f\": 0.6523699067792901,\n",
      " \"g\": 0.5191036806862445,\n",
      " \"s\": 0.03221147533808919\n",
      "}\n",
      "diaporthe_ccl067\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Sordariomycetes', 'o': 'Diaporthales', 'f': 'Diaporthaceae', 'g': 'Diaporthe', 's': 'Diaporthe_CCL067'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9729810508697472,\n",
      " \"c\": 0.7845657241915237,\n",
      " \"o\": 0.47934724131747264,\n",
      " \"f\": 0.004722338454181361,\n",
      " \"g\": 0.004722338454181361,\n",
      " \"s\": 0.0\n",
      "}\n",
      "diaporthe_unidentified\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Sordariomycetes', 'o': 'Diaporthales', 'f': 'Diaporthaceae', 'g': 'Diaporthe', 's': 'unidentified'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9779748794571106,\n",
      " \"c\": 0.7123638311804275,\n",
      " \"o\": 0.311387582594202,\n",
      " \"f\": 0.27424251443538306,\n",
      " \"g\": 0.2722781117923686,\n",
      " \"s\": 0.024822906125364606\n",
      "}\n",
      "oculimacula_yallundae-ccl031\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Leotiomycetes', 'o': 'Helotiales', 'f': 'Dermateaceae', 'g': 'Oculimacula', 's': 'Oculimacula_yallundae'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9573016064742118,\n",
      " \"c\": 0.8018480492813141,\n",
      " \"o\": 0.6465152796231429,\n",
      " \"f\": 0.12018359705278416,\n",
      " \"g\": 0.10321294842372267,\n",
      " \"s\": 0.0521198212344486\n",
      "}\n",
      "oculimacula_yallundae-ccl029\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Leotiomycetes', 'o': 'Helotiales', 'f': 'Dermateaceae', 'g': 'Oculimacula', 's': 'Oculimacula_yallundae'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9545668549905838,\n",
      " \"c\": 0.8302142184557438,\n",
      " \"o\": 0.5726224105461394,\n",
      " \"f\": 0.1832627118644068,\n",
      " \"g\": 0.16148775894538606,\n",
      " \"s\": 0.09080743879472693\n",
      "}\n",
      "dothiorella_vidmadera\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Dothideomycetes', 'o': 'Botryosphaeriales', 'f': 'Botryosphaeriaceae', 'g': 'Dothiorella', 's': 'Dothiorella_vidmadera'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9229026260748315,\n",
      " \"c\": 0.5569370206832442,\n",
      " \"o\": 0.3101324657215896,\n",
      " \"f\": 0.3095514757146177,\n",
      " \"g\": 0.28526609342319315,\n",
      " \"s\": 0.0\n",
      "}\n",
      "quambalaria_cyanescens\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Exobasidiomycetes', 'o': 'Microstromatales', 'f': 'Quambalariaceae', 'g': 'Quambalaria', 's': 'Quambalaria_cyanescens'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9052947283423693,\n",
      " \"c\": 0.7707924789479755,\n",
      " \"o\": 0.770561771830661,\n",
      " \"f\": 0.764101972545853,\n",
      " \"g\": 0.764101972545853,\n",
      " \"s\": 0.7110393355635022\n",
      "}\n",
      "entoleuca_unidentified\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Sordariomycetes', 'o': 'Xylariales', 'f': 'Xylariaceae', 'g': 'Entoleuca', 's': 'unidentified'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9679527604196526,\n",
      " \"c\": 0.8599438170039557,\n",
      " \"o\": 0.6274150088860861,\n",
      " \"f\": 0.595425098893539,\n",
      " \"g\": 0.03147394370234478,\n",
      " \"s\": 0.015192340767069885\n",
      "}\n",
      "asteroma_ccl060\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Sordariomycetes', 'o': 'Diaporthales', 'f': 'Gnomoniaceae', 'g': 'Asteroma', 's': 'Asteroma_CCL060'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9763606049779683,\n",
      " \"c\": 0.7185304275336429,\n",
      " \"o\": 0.3088603072525902,\n",
      " \"f\": 0.00011909015124449208,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "asteroma_ccl068\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Sordariomycetes', 'o': 'Diaporthales', 'f': 'Gnomoniaceae', 'g': 'Asteroma', 's': 'Asteroma_CCL068'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9735481543825899,\n",
      " \"c\": 0.7914512444391006,\n",
      " \"o\": 0.47457015750871706,\n",
      " \"f\": 0.4475171335818204,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "saccharomyces_cerevisiae\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetaceae', 'g': 'Saccharomyces', 's': 'Saccharomyces_cerevisiae'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9554016441193605,\n",
      " \"c\": 0.7639329406434073,\n",
      " \"o\": 0.7639329406434073,\n",
      " \"f\": 0.7048352644184089,\n",
      " \"g\": 0.6630849893196971,\n",
      " \"s\": 0.4988025114894168\n",
      "}\n",
      "cladophialophora_unidentified\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Eurotiomycetes', 'o': 'Chaetothyriales', 'f': 'Herpotrichiellaceae', 'g': 'Cladophialophora', 's': 'unidentified'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9231377799415774,\n",
      " \"c\": 0.43433544303797467,\n",
      " \"o\": 0.4000730282375852,\n",
      " \"f\": 0.3869279454722493,\n",
      " \"g\": 0.3591163583252191,\n",
      " \"s\": 0.3337390457643622\n",
      "}\n",
      "candida_albicans\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetales_fam_Incertae_sedis', 'g': 'Candida', 's': 'Candida_albicans'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9345638619667881,\n",
      " \"c\": 0.5976294833014384,\n",
      " \"o\": 0.5976294833014384,\n",
      " \"f\": 0.3740971664917587,\n",
      " \"g\": 0.3740971664917587,\n",
      " \"s\": 0.30551268596826964\n",
      "}\n",
      "candida_metapsilosis\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetales_fam_Incertae_sedis', 'g': 'Candida', 's': 'Candida_metapsilosis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9502836576587568,\n",
      " \"c\": 0.6719941438418837,\n",
      " \"o\": 0.6719941438418837,\n",
      " \"f\": 0.6510705789056305,\n",
      " \"g\": 0.651009577258586,\n",
      " \"s\": 0.017995485878118708\n",
      "}\n",
      "candida_orthopsilosis\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetales_fam_Incertae_sedis', 'g': 'Candida', 's': 'Candida_orthopsilosis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9510279226756674,\n",
      " \"c\": 0.6717397974838908,\n",
      " \"o\": 0.6717397974838908,\n",
      " \"f\": 0.6462104940165695,\n",
      " \"g\": 0.6460263884627186,\n",
      " \"s\": 0.3096655415771709\n",
      "}\n",
      "candida_parapsilosis\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetales_fam_Incertae_sedis', 'g': 'Candida', 's': 'Candida_parapsilosis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9498905375820968,\n",
      " \"c\": 0.6724641206519095,\n",
      " \"o\": 0.6724641206519095,\n",
      " \"f\": 0.6569569447822914,\n",
      " \"g\": 0.6569569447822914,\n",
      " \"s\": 0.1899172950620287\n",
      "}\n",
      "candida_unidentified\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetales_fam_Incertae_sedis', 'g': 'Candida', 's': 'unidentified'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.937395296891481,\n",
      " \"c\": 0.5846621579698456,\n",
      " \"o\": 0.5846621579698456,\n",
      " \"f\": 0.2989390085003413,\n",
      " \"g\": 0.2989390085003413,\n",
      " \"s\": 0.00043432400570825834\n",
      "}\n",
      "kluyveromyces_marxianus\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetaceae', 'g': 'Kluyveromyces', 's': 'Kluyveromyces_marxianus'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9490740740740741,\n",
      " \"c\": 0.7819634703196348,\n",
      " \"o\": 0.7819634703196348,\n",
      " \"f\": 0.6944444444444444,\n",
      " \"g\": 0.5779426686960933,\n",
      " \"s\": 0.5154743784880771\n",
      "}\n",
      "pichia_kudriavzevii\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Pichiaceae', 'g': 'Pichia', 's': 'Pichia_kudriavzevii'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9504913480025635,\n",
      " \"c\": 0.8423413800469985,\n",
      " \"o\": 0.8423413800469985,\n",
      " \"f\": 0.20732749412518692,\n",
      " \"g\": 0.20679342020935698,\n",
      " \"s\": 0.0\n",
      "}\n",
      "pichia_membranifaciens\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Pichiaceae', 'g': 'Pichia', 's': 'Pichia_membranifaciens'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9436147516001219,\n",
      " \"c\": 0.7956110941786041,\n",
      " \"o\": 0.7956110941786041,\n",
      " \"f\": 0.748796098750381,\n",
      " \"g\": 0.747698872295032,\n",
      " \"s\": 0.6996647363608656\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from collections import OrderedDict\n",
    "\n",
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarises the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = OrderedDict()\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        tmps_df = pd.DataFrame(data=None)\n",
    "        if tax_rank == 's':\n",
    "            for index, row in mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]].iterrows():\n",
    "                if row['s'] == 'unidentified' and row['g'] != query_tax_dict['g']:\n",
    "                    mapping_df.drop(index, axis=0, inplace=True)\n",
    "                else:\n",
    "                    continue\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        else:\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict\n",
    "\n",
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df\n",
    "    \n",
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "test_species_list = []\n",
    "for entry in ref_df.name_species.tolist():\n",
    "#     if entry[-7:] != '-ccl031' and entry[-7:] != '-ccl029':\n",
    "#         test_species_list.append(entry)\n",
    "#     else:\n",
    "#         test_species_list.append(entry[:-7])\n",
    "#         print(entry[:-7])\n",
    "    test_species_list.append(entry)\n",
    "    \n",
    "for test_species in test_species_list:\n",
    "    \n",
    "    print(test_species)\n",
    "    \n",
    "    #subsample tests species\n",
    "    fn_subsampling = {}\n",
    "    test_species = [test_species]\n",
    "    for x in test_species:\n",
    "        fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "        fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "    sub_reads_fn = {}\n",
    "    n_reads = 15000\n",
    "    for key, value in fn_subsampling.items():\n",
    "        species = key\n",
    "        in_fn = value\n",
    "        out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "        subsamplereads(in_fn, out_fn, n_reads)\n",
    "        sub_reads_fn[species] = out_fn\n",
    "        \n",
    "    ###Map the reads\n",
    "    db_fn = qiime_db_fn\n",
    "    sub_db_mapping_fn = {}\n",
    "    for species, fasta_fn in sub_reads_fn.items():\n",
    "        db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "        out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "        sub_db_mapping_fn[species] = out_fn\n",
    "        minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "        \n",
    "    ###Test out the summary results statistic for a single mapping result\n",
    "    species = test_species[0]\n",
    "    mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "    mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "    taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "    \n",
    "    query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "    print(query_tax_dict)\n",
    "    \n",
    "    sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "        \n",
    "    print(json.dumps(sensitivity_dict, indent=1))\n",
    "    \n",
    "    with open('/media/MassStorage/tmp/TE/honours/analysis/Mapping/qiime_results/tax_rank/%s.json' % species,'w+') as json_file:\n",
    "        json.dump(sensitivity_dict,json_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>barcode</th>\n",
       "      <th># raw reads</th>\n",
       "      <th># reads after homology filtering</th>\n",
       "      <th># reads after length filtering</th>\n",
       "      <th># for use</th>\n",
       "      <th>path to raw reads</th>\n",
       "      <th>path to homology filtering</th>\n",
       "      <th>path to length filtering</th>\n",
       "      <th>path for use</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>115052</td>\n",
       "      <td>23477</td>\n",
       "      <td>21590</td>\n",
       "      <td>21590.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode01/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode01/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode01/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode01/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>barcode02</td>\n",
       "      <td>barcode02</td>\n",
       "      <td>151005</td>\n",
       "      <td>20416</td>\n",
       "      <td>18100</td>\n",
       "      <td>18100.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode02/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode02/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode02/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode02/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>barcode03</td>\n",
       "      <td>barcode03</td>\n",
       "      <td>92927</td>\n",
       "      <td>23616</td>\n",
       "      <td>21919</td>\n",
       "      <td>21919.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode03/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode03/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode03/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode03/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>barcode04</td>\n",
       "      <td>barcode04</td>\n",
       "      <td>115407</td>\n",
       "      <td>22716</td>\n",
       "      <td>20568</td>\n",
       "      <td>20568.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode04/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode04/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode04/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode04/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>134094</td>\n",
       "      <td>24785</td>\n",
       "      <td>22433</td>\n",
       "      <td>22433.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode05/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode05/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode05/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode05/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>barcode06</td>\n",
       "      <td>150739</td>\n",
       "      <td>25462</td>\n",
       "      <td>23769</td>\n",
       "      <td>23769.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode06/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode06/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode06/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode06/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>barcode07</td>\n",
       "      <td>109201</td>\n",
       "      <td>24688</td>\n",
       "      <td>22931</td>\n",
       "      <td>22931.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode07/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode07/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode07/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode07/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>barcode08</td>\n",
       "      <td>barcode08</td>\n",
       "      <td>114215</td>\n",
       "      <td>21725</td>\n",
       "      <td>20105</td>\n",
       "      <td>20105.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode08/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode08/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode08/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode08/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>barcode09</td>\n",
       "      <td>barcode09</td>\n",
       "      <td>106976</td>\n",
       "      <td>22012</td>\n",
       "      <td>20451</td>\n",
       "      <td>20451.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode09/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode09/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode09/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode09/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>barcode10</td>\n",
       "      <td>barcode10</td>\n",
       "      <td>108583</td>\n",
       "      <td>21514</td>\n",
       "      <td>19810</td>\n",
       "      <td>19810.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode10/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode10/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode10/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode10/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>barcode11</td>\n",
       "      <td>barcode11</td>\n",
       "      <td>118345</td>\n",
       "      <td>22462</td>\n",
       "      <td>20832</td>\n",
       "      <td>20832.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode11/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode11/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode11/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode11/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>barcode12</td>\n",
       "      <td>barcode12</td>\n",
       "      <td>260</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode12/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode12/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode12/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode12/lengt...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0    barcode  # raw reads  # reads after homology filtering  \\\n",
       "0   barcode01  barcode01       115052                             23477   \n",
       "1   barcode02  barcode02       151005                             20416   \n",
       "2   barcode03  barcode03        92927                             23616   \n",
       "3   barcode04  barcode04       115407                             22716   \n",
       "4   barcode05  barcode05       134094                             24785   \n",
       "5   barcode06  barcode06       150739                             25462   \n",
       "6   barcode07  barcode07       109201                             24688   \n",
       "7   barcode08  barcode08       114215                             21725   \n",
       "8   barcode09  barcode09       106976                             22012   \n",
       "9   barcode10  barcode10       108583                             21514   \n",
       "10  barcode11  barcode11       118345                             22462   \n",
       "11  barcode12  barcode12          260                                 3   \n",
       "\n",
       "    # reads after length filtering  # for use  \\\n",
       "0                            21590    21590.0   \n",
       "1                            18100    18100.0   \n",
       "2                            21919    21919.0   \n",
       "3                            20568    20568.0   \n",
       "4                            22433    22433.0   \n",
       "5                            23769    23769.0   \n",
       "6                            22931    22931.0   \n",
       "7                            20105    20105.0   \n",
       "8                            20451    20451.0   \n",
       "9                            19810    19810.0   \n",
       "10                           20832    20832.0   \n",
       "11                               3        3.0   \n",
       "\n",
       "                                    path to raw reads  \\\n",
       "0   analysis/Concatenated/wheat/barcode01/merged.f...   \n",
       "1   analysis/Concatenated/wheat/barcode02/merged.f...   \n",
       "2   analysis/Concatenated/wheat/barcode03/merged.f...   \n",
       "3   analysis/Concatenated/wheat/barcode04/merged.f...   \n",
       "4   analysis/Concatenated/wheat/barcode05/merged.f...   \n",
       "5   analysis/Concatenated/wheat/barcode06/merged.f...   \n",
       "6   analysis/Concatenated/wheat/barcode07/merged.f...   \n",
       "7   analysis/Concatenated/wheat/barcode08/merged.f...   \n",
       "8   analysis/Concatenated/wheat/barcode09/merged.f...   \n",
       "9   analysis/Concatenated/wheat/barcode10/merged.f...   \n",
       "10  analysis/Concatenated/wheat/barcode11/merged.f...   \n",
       "11  analysis/Concatenated/wheat/barcode12/merged.f...   \n",
       "\n",
       "                           path to homology filtering  \\\n",
       "0   analysis/Python_Processing/wheat/barcode01/com...   \n",
       "1   analysis/Python_Processing/wheat/barcode02/com...   \n",
       "2   analysis/Python_Processing/wheat/barcode03/com...   \n",
       "3   analysis/Python_Processing/wheat/barcode04/com...   \n",
       "4   analysis/Python_Processing/wheat/barcode05/com...   \n",
       "5   analysis/Python_Processing/wheat/barcode06/com...   \n",
       "6   analysis/Python_Processing/wheat/barcode07/com...   \n",
       "7   analysis/Python_Processing/wheat/barcode08/com...   \n",
       "8   analysis/Python_Processing/wheat/barcode09/com...   \n",
       "9   analysis/Python_Processing/wheat/barcode10/com...   \n",
       "10  analysis/Python_Processing/wheat/barcode11/com...   \n",
       "11  analysis/Python_Processing/wheat/barcode12/com...   \n",
       "\n",
       "                             path to length filtering  \\\n",
       "0   analysis/Length_Filtered/wheat/barcode01/lengt...   \n",
       "1   analysis/Length_Filtered/wheat/barcode02/lengt...   \n",
       "2   analysis/Length_Filtered/wheat/barcode03/lengt...   \n",
       "3   analysis/Length_Filtered/wheat/barcode04/lengt...   \n",
       "4   analysis/Length_Filtered/wheat/barcode05/lengt...   \n",
       "5   analysis/Length_Filtered/wheat/barcode06/lengt...   \n",
       "6   analysis/Length_Filtered/wheat/barcode07/lengt...   \n",
       "7   analysis/Length_Filtered/wheat/barcode08/lengt...   \n",
       "8   analysis/Length_Filtered/wheat/barcode09/lengt...   \n",
       "9   analysis/Length_Filtered/wheat/barcode10/lengt...   \n",
       "10  analysis/Length_Filtered/wheat/barcode11/lengt...   \n",
       "11  analysis/Length_Filtered/wheat/barcode12/lengt...   \n",
       "\n",
       "                                         path for use  \n",
       "0   analysis/Length_Filtered/wheat/barcode01/lengt...  \n",
       "1   analysis/Length_Filtered/wheat/barcode02/lengt...  \n",
       "2   analysis/Length_Filtered/wheat/barcode03/lengt...  \n",
       "3   analysis/Length_Filtered/wheat/barcode04/lengt...  \n",
       "4   analysis/Length_Filtered/wheat/barcode05/lengt...  \n",
       "5   analysis/Length_Filtered/wheat/barcode06/lengt...  \n",
       "6   analysis/Length_Filtered/wheat/barcode07/lengt...  \n",
       "7   analysis/Length_Filtered/wheat/barcode08/lengt...  \n",
       "8   analysis/Length_Filtered/wheat/barcode09/lengt...  \n",
       "9   analysis/Length_Filtered/wheat/barcode10/lengt...  \n",
       "10  analysis/Length_Filtered/wheat/barcode11/lengt...  \n",
       "11  analysis/Length_Filtered/wheat/barcode12/lengt...  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "\n",
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')\n",
    "subsampling_dir = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/subsample_reads')\n",
    "mapping_dir = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/qiime_results')\n",
    "\n",
    "wheat_reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/wheat_reference_dataframe_v3.csv')\n",
    "wheat_max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/wheat_database_labelled.fasta')\n",
    "wheat_taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/wheat_taxonomy_file_qiime.csv')\n",
    "\n",
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "\n",
    "wheat_ref_df = pd.read_csv(wheat_reference_dataframe_fn)\n",
    "wheat_ref_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = ['barcode01','barcode02',\n",
    "                  'barcode03','barcode04',\n",
    "                 'barcode05','barcode06',\n",
    "                  'barcode07','barcode08',\n",
    "                  'barcode09','barcode10',\n",
    "                  'barcode11','barcode12',\n",
    "                 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = os.path.abspath('../../analysis/Mapping/wheat')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)\n",
    "MC_READ_DIR = os.path.join(OUT_DIR, 'MC_READS')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)\n",
    "sub_db_fn = os.path.join('../../analysis/Mapping/gsref.subdb.fasta')\n",
    "new_db_fn = os.path.join('../../analysis/Mapping/gsref.db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'barcode', '# raw reads',\n",
       "       '# reads after homology filtering', '# reads after length filtering',\n",
       "       '# for use', 'path to raw reads', 'path to homology filtering',\n",
       "       'path to length filtering', 'path for use'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wheat_ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode01\n",
      "barcode02\n",
      "barcode03\n",
      "barcode04\n",
      "barcode05\n",
      "barcode06\n",
      "barcode07\n",
      "barcode08\n",
      "barcode09\n",
      "barcode10\n",
      "barcode11\n",
      "barcode12\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'barcode01': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta',\n",
       " 'barcode02': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta',\n",
       " 'barcode03': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode03/length_restricted_reads.fasta',\n",
       " 'barcode04': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode04/length_restricted_reads.fasta',\n",
       " 'barcode05': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta',\n",
       " 'barcode06': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode06/length_restricted_reads.fasta',\n",
       " 'barcode07': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode07/length_restricted_reads.fasta',\n",
       " 'barcode08': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode08/length_restricted_reads.fasta',\n",
       " 'barcode09': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode09/length_restricted_reads.fasta',\n",
       " 'barcode10': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode10/length_restricted_reads.fasta',\n",
       " 'barcode11': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode11/length_restricted_reads.fasta',\n",
       " 'barcode12': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode12/length_restricted_reads.fasta'}"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    print(x)\n",
    "    fn_subsampling[x] = wheat_ref_df[wheat_ref_df['Unnamed: 0'] == x]['path for use'].tolist()[0]\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode01\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta\n",
      ":(check one reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.200.fasta!!\n",
      "\n",
      "barcode02\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta\n",
      ":(check one reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.200.fasta!!\n",
      "\n",
      "barcode03\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode03/length_restricted_reads.fasta\n",
      ":(check one reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode03/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode03.200.fasta!!\n",
      "\n",
      "barcode04\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode04/length_restricted_reads.fasta\n",
      ":(check one reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode04/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode04.200.fasta!!\n",
      "\n",
      "barcode05\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta\n",
      ":(check one reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.200.fasta!!\n",
      "\n",
      "barcode06\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode06/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode06/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode06.200.fasta\n",
      "\n",
      "barcode07\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode07/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode07/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode07.200.fasta\n",
      "\n",
      "barcode08\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode08/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode08/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode08.200.fasta\n",
      "\n",
      "barcode09\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode09/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode09/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode09.200.fasta\n",
      "\n",
      "barcode10\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode10/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode10/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode10.200.fasta\n",
      "\n",
      "barcode11\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode11/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode11/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode11.200.fasta\n",
      "\n",
      "barcode12\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode12/length_restricted_reads.fasta\n",
      ":)Completed reformat.sh samplereadstarget=200 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode12/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode12.200.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    print(key)\n",
    "    print(value)\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../analysis/Mapping/gsref.subdb.fasta\n",
      "../../analysis/Mapping/gsref.db.fasta\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'../../analysis/Mapping/gsref.subdb.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb',\n",
       " '../../analysis/Mapping/gsref.db.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db'}"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    print(x)\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode01.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode02.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode03.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode03.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode04.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode04.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode05.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode06.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode06.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode07.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode07.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode08.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode08.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode09.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode09.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode10.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode10.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode11.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode11.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode12.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode12.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode01.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode02.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode03.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode03.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode04.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode04.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode05.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode06.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode06.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode07.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode07.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode08.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode08.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode09.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode09.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode10.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode10.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode11.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode11.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode12.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode12.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results_v2(fn, species,expected_species, tax_rank):\n",
    "    import numpy as np\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    if tax_rank == 's' or tax_rank == 'species':\n",
    "        new_new_hit_series = hit_series\n",
    "        new_new_hit_series.sort_values(ascending=False, inplace=True)\n",
    "    else:\n",
    "        new_hit_series = pd.Series()\n",
    "        new_new_hit_series = pd.Series()\n",
    "        for index in hit_series.index.unique():\n",
    "            new_hit_series.at[index] = data=np.sum(hit_series[index])\n",
    "        for index in new_hit_series.index.unique():\n",
    "            new_new_hit_series.at[index] = data=np.sum(new_hit_series[index])\n",
    "        new_new_hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the sample type expected:\", barcode,\"\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(new_new_hit_series,'\\n')\n",
    "    new_new_hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/custom_results/%s_%s.json' % (tax_rank,expected_species[species]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "expected_rank_s = {}\n",
    "expected_rank_g = {}\n",
    "expected_rank_f = {}\n",
    "expected_rank_o = {}\n",
    "expected_rank_c = {}\n",
    "expected_rank_p = {}\n",
    "\n",
    "for column, row in wheat_ref_df.iterrows():\n",
    "    expected_rank_s[row['Unnamed: 0']] = row['barcode']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode01 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.582090\n",
      "blastobotrys_proliferans         0.208955\n",
      "cortinarius_globuliformis        0.029851\n",
      "saccharomyces_cerevisiae         0.024876\n",
      "aspergillus_unidentified         0.014925\n",
      "cladophialophora_unidentified    0.014925\n",
      "entoleuca_unidentified           0.009950\n",
      "dothiorella_vidmadera            0.009950\n",
      "candida_orthopsilosis            0.009950\n",
      "candida_unidentified             0.009950\n",
      "debaryomyces_unidentified        0.009950\n",
      "aspergillus_flavus               0.009950\n",
      "quambalaria_cyanescens           0.009950\n",
      "yamadazyma_mexicana              0.009950\n",
      "yarrowia_lipolytica              0.004975\n",
      "meyerozyma_guilliermondii        0.004975\n",
      "oculimacula_yallundae-ccl029     0.004975\n",
      "penicillium_chrysogenum          0.004975\n",
      "pichia_membranifaciens           0.004975\n",
      "rhodotorula_mucilaginosa         0.004975\n",
      "candida_metapsilosis             0.004975\n",
      "tuber_brumale                    0.004975\n",
      "zymoseptoria_tritici             0.004975\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode02 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zymoseptoria_tritici             0.515\n",
      "cryptococcus_zero                0.190\n",
      "debaryomyces_unidentified        0.050\n",
      "pyrenophora_tritici-repentis     0.045\n",
      "blastobotrys_proliferans         0.040\n",
      "entoleuca_unidentified           0.030\n",
      "cortinarius_globuliformis        0.030\n",
      "cladophialophora_unidentified    0.015\n",
      "aspergillus_unidentified         0.010\n",
      "candida_albicans                 0.010\n",
      "dothiorella_vidmadera            0.010\n",
      "wickerhamomyces_anomalus         0.010\n",
      "candida_unidentified             0.010\n",
      "candida_parapsilosis             0.005\n",
      "candida_orthopsilosis            0.005\n",
      "quambalaria_cyanescens           0.005\n",
      "meyerozyma_guilliermondii        0.005\n",
      "oculimacula_yallundae-ccl029     0.005\n",
      "oculimacula_yallundae-ccl031     0.005\n",
      "rhodotorula_mucilaginosa         0.005\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode03 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.411765\n",
      "candida_albicans                 0.186275\n",
      "cortinarius_globuliformis        0.068627\n",
      "blastobotrys_proliferans         0.063725\n",
      "galactomyces_geotrichum          0.053922\n",
      "aspergillus_flavus               0.024510\n",
      "rhodotorula_mucilaginosa         0.019608\n",
      "pyrenophora_tritici-repentis     0.019608\n",
      "dothiorella_vidmadera            0.014706\n",
      "candida_orthopsilosis            0.014706\n",
      "candida_parapsilosis             0.014706\n",
      "diaporthe_ccl067                 0.014706\n",
      "quambalaria_cyanescens           0.014706\n",
      "saccharomyces_cerevisiae         0.009804\n",
      "zymoseptoria_tritici             0.009804\n",
      "entoleuca_unidentified           0.004902\n",
      "yamadazyma_scolyti               0.004902\n",
      "debaryomyces_unidentified        0.004902\n",
      "meyerozyma_guilliermondii        0.004902\n",
      "oculimacula_yallundae-ccl029     0.004902\n",
      "clavispora_lusitaniae            0.004902\n",
      "cladophialophora_unidentified    0.004902\n",
      "candida_unidentified             0.004902\n",
      "candida_metapsilosis             0.004902\n",
      "tuber_brumale                    0.004902\n",
      "wickerhamomyces_anomalus         0.004902\n",
      "asteroma_ccl060                  0.004902\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode04 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.415459\n",
      "cortinarius_globuliformis        0.149758\n",
      "blastobotrys_proliferans         0.096618\n",
      "kodamaea_ohmeri                  0.057971\n",
      "zymoseptoria_tritici             0.053140\n",
      "candida_metapsilosis             0.048309\n",
      "rhodotorula_mucilaginosa         0.024155\n",
      "diaporthe_unidentified           0.019324\n",
      "wickerhamomyces_anomalus         0.014493\n",
      "candida_albicans                 0.014493\n",
      "meyerozyma_guilliermondii        0.014493\n",
      "candida_unidentified             0.009662\n",
      "aspergillus_flavus               0.009662\n",
      "kluyveromyces_marxianus          0.009662\n",
      "oculimacula_yallundae-ccl029     0.009662\n",
      "entoleuca_unidentified           0.004831\n",
      "fusarium_oxysporum               0.004831\n",
      "debaryomyces_unidentified        0.004831\n",
      "kluyveromyces_unidentified       0.004831\n",
      "clavispora_lusitaniae            0.004831\n",
      "cladophialophora_unidentified    0.004831\n",
      "candida_orthopsilosis            0.004831\n",
      "pyrenophora_tritici-repentis     0.004831\n",
      "quambalaria_cyanescens           0.004831\n",
      "tuber_brumale                    0.004831\n",
      "dothiorella_vidmadera            0.004831\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode05 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pyrenophora_tritici-repentis     0.374384\n",
      "cryptococcus_zero                0.206897\n",
      "cortinarius_globuliformis        0.078818\n",
      "meyerozyma_guilliermondii        0.073892\n",
      "candida_orthopsilosis            0.059113\n",
      "blastobotrys_proliferans         0.049261\n",
      "tuber_brumale                    0.019704\n",
      "oculimacula_yallundae-ccl031     0.019704\n",
      "candida_metapsilosis             0.014778\n",
      "candida_unidentified             0.014778\n",
      "aspergillus_flavus               0.009852\n",
      "rhodotorula_mucilaginosa         0.009852\n",
      "candida_albicans                 0.009852\n",
      "dothiorella_vidmadera            0.009852\n",
      "cladophialophora_unidentified    0.004926\n",
      "aspergillus_unidentified         0.004926\n",
      "zymoseptoria_tritici             0.004926\n",
      "yamadazyma_mexicana              0.004926\n",
      "oculimacula_yallundae-ccl029     0.004926\n",
      "penicillium_chrysogenum          0.004926\n",
      "pichia_kudriavzevii              0.004926\n",
      "quambalaria_cyanescens           0.004926\n",
      "wickerhamomyces_anomalus         0.004926\n",
      "kluyveromyces_unidentified       0.004926\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode06 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cortinarius_globuliformis        0.333333\n",
      "cryptococcus_zero                0.308458\n",
      "candida_parapsilosis             0.089552\n",
      "wickerhamomyces_anomalus         0.079602\n",
      "blastobotrys_proliferans         0.059701\n",
      "candida_albicans                 0.019900\n",
      "zymoseptoria_tritici             0.009950\n",
      "dothiorella_vidmadera            0.009950\n",
      "rhodotorula_mucilaginosa         0.009950\n",
      "cladophialophora_unidentified    0.009950\n",
      "meyerozyma_guilliermondii        0.009950\n",
      "aspergillus_unidentified         0.004975\n",
      "clavispora_lusitaniae            0.004975\n",
      "kluyveromyces_unidentified       0.004975\n",
      "entoleuca_unidentified           0.004975\n",
      "yarrowia_lipolytica              0.004975\n",
      "oculimacula_yallundae-ccl029     0.004975\n",
      "oculimacula_yallundae-ccl031     0.004975\n",
      "penicillium_chrysogenum          0.004975\n",
      "quambalaria_cyanescens           0.004975\n",
      "saccharomyces_cerevisiae         0.004975\n",
      "yamadazyma_scolyti               0.004975\n",
      "aspergillus_niger                0.004975\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode07 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.415\n",
      "cortinarius_globuliformis        0.155\n",
      "blastobotrys_proliferans         0.110\n",
      "dothiorella_vidmadera            0.065\n",
      "yamadazyma_mexicana              0.040\n",
      "rhodotorula_mucilaginosa         0.035\n",
      "meyerozyma_guilliermondii        0.025\n",
      "candida_unidentified             0.020\n",
      "candida_orthopsilosis            0.015\n",
      "candida_parapsilosis             0.015\n",
      "candida_albicans                 0.010\n",
      "aspergillus_flavus               0.010\n",
      "kluyveromyces_unidentified       0.010\n",
      "quambalaria_cyanescens           0.010\n",
      "saccharomyces_cerevisiae         0.010\n",
      "entoleuca_unidentified           0.005\n",
      "yamadazyma_scolyti               0.005\n",
      "diaporthe_unidentified           0.005\n",
      "oculimacula_yallundae-ccl029     0.005\n",
      "cladophialophora_unidentified    0.005\n",
      "oculimacula_yallundae-ccl031     0.005\n",
      "pichia_kudriavzevii              0.005\n",
      "wickerhamomyces_anomalus         0.005\n",
      "aspergillus_unidentified         0.005\n",
      "aspergillus_niger                0.005\n",
      "zymoseptoria_tritici             0.005\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode08 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.415\n",
      "cortinarius_globuliformis        0.125\n",
      "blastobotrys_proliferans         0.080\n",
      "quambalaria_cyanescens           0.080\n",
      "yamadazyma_scolyti               0.070\n",
      "clavispora_lusitaniae            0.050\n",
      "rhodotorula_mucilaginosa         0.030\n",
      "candida_unidentified             0.020\n",
      "entoleuca_unidentified           0.015\n",
      "puccinia_striiformis             0.015\n",
      "cladophialophora_unidentified    0.010\n",
      "aspergillus_flavus               0.010\n",
      "kluyveromyces_unidentified       0.010\n",
      "meyerozyma_guilliermondii        0.010\n",
      "wickerhamomyces_anomalus         0.010\n",
      "fusarium_oxysporum               0.005\n",
      "dothiorella_vidmadera            0.005\n",
      "debaryomyces_unidentified        0.005\n",
      "galactomyces_geotrichum          0.005\n",
      "oculimacula_yallundae-ccl031     0.005\n",
      "pyrenophora_tritici-repentis     0.005\n",
      "candida_orthopsilosis            0.005\n",
      "candida_metapsilosis             0.005\n",
      "candida_albicans                 0.005\n",
      "zymoseptoria_tritici             0.005\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode09 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.445545\n",
      "candida_unidentified             0.168317\n",
      "blastobotrys_proliferans         0.094059\n",
      "cortinarius_globuliformis        0.069307\n",
      "kluyveromyces_unidentified       0.039604\n",
      "meyerozyma_guilliermondii        0.024752\n",
      "yarrowia_lipolytica              0.019802\n",
      "candida_albicans                 0.014851\n",
      "entoleuca_unidentified           0.014851\n",
      "dothiorella_vidmadera            0.014851\n",
      "wickerhamomyces_anomalus         0.009901\n",
      "penicillium_chrysogenum          0.009901\n",
      "aspergillus_niger                0.004950\n",
      "candida_metapsilosis             0.004950\n",
      "cladophialophora_unidentified    0.004950\n",
      "clavispora_lusitaniae            0.004950\n",
      "zymoseptoria_tritici             0.004950\n",
      "debaryomyces_unidentified        0.004950\n",
      "oculimacula_yallundae-ccl031     0.004950\n",
      "puccinia_striiformis             0.004950\n",
      "quambalaria_cyanescens           0.004950\n",
      "rhodotorula_mucilaginosa         0.004950\n",
      "saccharomyces_cerevisiae         0.004950\n",
      "tuber_brumale                    0.004950\n",
      "yamadazyma_mexicana              0.004950\n",
      "yamadazyma_scolyti               0.004950\n",
      "aspergillus_flavus               0.004950\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode10 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.447761\n",
      "blastobotrys_proliferans         0.124378\n",
      "cortinarius_globuliformis        0.089552\n",
      "kluyveromyces_marxianus          0.069652\n",
      "penicillium_chrysogenum          0.034826\n",
      "zymoseptoria_tritici             0.024876\n",
      "rhodotorula_mucilaginosa         0.024876\n",
      "quambalaria_cyanescens           0.014925\n",
      "meyerozyma_guilliermondii        0.014925\n",
      "candida_metapsilosis             0.014925\n",
      "candida_unidentified             0.014925\n",
      "dothiorella_vidmadera            0.014925\n",
      "saccharomyces_cerevisiae         0.014925\n",
      "entoleuca_unidentified           0.009950\n",
      "candida_albicans                 0.009950\n",
      "wickerhamomyces_anomalus         0.009950\n",
      "kluyveromyces_unidentified       0.004975\n",
      "yarrowia_lipolytica              0.004975\n",
      "oculimacula_yallundae-ccl029     0.004975\n",
      "puccinia_striiformis             0.004975\n",
      "tuber_brumale                    0.004975\n",
      "clavispora_lusitaniae            0.004975\n",
      "cladophialophora_unidentified    0.004975\n",
      "oculimacula_yallundae-ccl031     0.004975\n",
      "candida_parapsilosis             0.004975\n",
      "candida_orthopsilosis            0.004975\n",
      "pichia_kudriavzevii              0.004975\n",
      "asteroma_ccl068                  0.004975\n",
      "aspergillus_unidentified         0.004975\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode11 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero               0.403941\n",
      "cortinarius_globuliformis       0.118227\n",
      "rhodotorula_mucilaginosa        0.088670\n",
      "blastobotrys_proliferans        0.083744\n",
      "zymoseptoria_tritici            0.039409\n",
      "zygoascus_hellenicus            0.039409\n",
      "dothiorella_vidmadera           0.029557\n",
      "aspergillus_flavus              0.029557\n",
      "pichia_kudriavzevii             0.024631\n",
      "candida_unidentified            0.024631\n",
      "asteroma_ccl060                 0.024631\n",
      "meyerozyma_guilliermondii       0.019704\n",
      "quambalaria_cyanescens          0.014778\n",
      "candida_albicans                0.014778\n",
      "oculimacula_yallundae-ccl029    0.009852\n",
      "wickerhamomyces_anomalus        0.009852\n",
      "oculimacula_yallundae-ccl031    0.004926\n",
      "penicillium_chrysogenum         0.004926\n",
      "clavispora_lusitaniae           0.004926\n",
      "candida_parapsilosis            0.004926\n",
      "candida_metapsilosis            0.004926\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: barcode12 \n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero            0.666667\n",
      "cortinarius_globuliformis    0.333333\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 's')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
