{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%writefile ../alignment_to_database.py\n",
    "\n",
    "# \"\"\"\n",
    "# Script aims to read in the reference_dataframe file, select a taxonomic\n",
    "# level and group, and read the path to the location of that data. It then\n",
    "# prepares data for machine learning by converting base pair coding to numerical\n",
    "# encoding, pads it out and then runs the algorithm\n",
    "# \"\"\"\n",
    "\n",
    "# import pandas as pd\n",
    "# from Bio import SeqIO\n",
    "# import numpy as np\n",
    "# import os\n",
    "# import random\n",
    "# import argparse\n",
    "# import math\n",
    "# import matplotlib.pyplot as plt\n",
    "# import pandas as pd\n",
    "\n",
    "# parser = argparse.ArgumentParser(description=\"\"\"\n",
    "# Script aims to read in the reference_dataframe file, and for each\n",
    "# sample, randomly subsample n_reads number of reads and save the\n",
    "# file, and save the keys used as as separate file. The script ends there.\n",
    "\n",
    "# Outside the script these files will be utilised in minimap2 to\n",
    "# map against a specified database, produce an output paf file\n",
    "# and then remove the file to conserve space. \n",
    "# \"\"\")\n",
    "# parser.add_argument(\"ref_df_fn\", help=\"File path to the reference dataframe\")\n",
    "# parser.add_argument(\"data_root\", help=\"Root folder for analysis/\")\n",
    "# parser.add_argument(\"--n_reads\", \"-c\", help=\"count of reads per class\")\n",
    "# parser.add_argument(\"--d_type\", \"-d\", help=\"type of database - CUSTOM or UNITE\")\n",
    "# group = parser.add_mutually_exclusive_group()\n",
    "# group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "# group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "# args = parser.parse_args()\n",
    "\n",
    "# # assign required arguments to variables\n",
    "# ref_df_fn = args.ref_df_fn\n",
    "# data_root = args.data_root\n",
    "\n",
    "# # assign a number of reads per class\n",
    "# n_reads = int(args.n_reads)\n",
    "\n",
    "# # test to make sure both required file paths are input\n",
    "# try:\n",
    "#     os.path.exists(ref_df_fn)\n",
    "# except:\n",
    "#     print('Cannot find %s' % ref_df_fn)\n",
    "# try:\n",
    "#     os.path.exists(data_root)\n",
    "# except:\n",
    "#     print('Cannot find %s' % data_root)\n",
    "    \n",
    "# if args.verbose:\n",
    "#     print('\\033[1;34m' + \"Reference dataframe is at \" + ref_df_fn + '\\033[0m')\n",
    "#     print('\\033[1;34m' + \"Root directory is at \" + data_root + '\\033[0m')\n",
    "#     print('\\033[1;34m' + \"Count of reads per sample is\", n_reads,'\\033[0m')\n",
    "    \n",
    "# # read in the reference dataframe from the argument path\n",
    "# ref_df = pd.read_csv(ref_df_fn, index_col=None)\n",
    "\n",
    "# # check whether the reference dataframe implies there are enough reads\n",
    "# # to continue given n_reads\n",
    "# try:\n",
    "#     if ref_df[ref_df[\"# for use\"] \\\n",
    "#               < n_reads].shape[0] > 0 :\n",
    "#         print(\"These species need more reads.\")\n",
    "#         print(ref_df[ref_df[\"# for use\"] \\\n",
    "#               < n_reads])\n",
    "#         exit()\n",
    "# except:\n",
    "#     print('Check %s to have the wanted column names' % ref_df_fn)\n",
    "   \n",
    "\n",
    "# for index, row in ref_df.iterrows():\n",
    "#     fasta_dict = SeqIO.to_dict(SeqIO.parse(row['path for use'], \"fasta\"))\n",
    "#     key_list = []\n",
    "#     for key in fasta_dict:\n",
    "#         key_list.append(key)\n",
    "#     print(len(key_list))\n",
    "#     keys_list = random.sample(key_list,k=n_reads)\n",
    "#     new_dict = {}\n",
    "#     for key in keys_list:\n",
    "#         new_dict[key] = fasta_dict[key]\n",
    "#     if args.d_type.upper() == 'CUSTOM':\n",
    "#         SeqIO.write(new_dict.values(), data_root+\"database_mapping/custom/%s_%s.fasta\" % (row['genus'],row['species']), \"fasta\")\n",
    "#         with open(data_root+'database_mapping/custom/%s_%s_keys.csv'% (row['genus'],row['species']), 'w+') as f:\n",
    "#             for key in keys_list:\n",
    "#                 f.write(\"%s\\n\"%(key))\n",
    "#     elif args.d_type.upper() == 'UNITE':\n",
    "#         SeqIO.write(new_dict.values(), \"analysis/database_mapping/unite/%s_%s.fasta\" % (row['genus'],row['species']), \"fasta\")\n",
    "#         with open(data_root+'database_mapping/unite/%s_%s_keys.csv'% (row['genus'],row['species']), 'w+') as f:\n",
    "#             for key in keys_list:\n",
    "#                 f.write(\"%s\\n\"%(key))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%writefile ../custom_minimap_result.py\n",
    "\n",
    "# import Bio\n",
    "# from Bio import SeqIO\n",
    "# from Bio.Seq import Seq\n",
    "# from Bio.SeqRecord import SeqRecord as SR\n",
    "# from Bio.Blast import NCBIXML\n",
    "# from Bio import pairwise2\n",
    "# from Bio.pairwise2 import format_alignment\n",
    "# from Bio import AlignIO\n",
    "# from Bio.Align import AlignInfo\n",
    "# import numpy as np\n",
    "# import pandas as pd\n",
    "# from pandas import DataFrame as df\n",
    "# # import seaborn as sns\n",
    "# import csv\n",
    "# import matplotlib\n",
    "# import matplotlib.pyplot as plt\n",
    "# import subprocess\n",
    "# import os\n",
    "# from shutil import copy\n",
    "# import random\n",
    "# import warnings\n",
    "# from Bio import pairwise2\n",
    "# from Bio.pairwise2 import format_alignment\n",
    "# # from ipysankeywidget import SankeyWidget\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "# import argparse\n",
    "\n",
    "\n",
    "# parser = argparse.ArgumentParser(description=\"\"\"\n",
    "# This program extracts a specified number of reads from a fasta\n",
    "# file and saves them to a new fasta file\n",
    "# \"\"\")\n",
    "# group = parser.add_mutually_exclusive_group()\n",
    "# group.add_argument(\"--verbose\", \"-v\", \"--v\", action=\"store_true\")\n",
    "# group.add_argument(\"--quiet\", \"-q\", \"--q\", action=\"store_true\")\n",
    "# parser.add_argument(\"input_file\", help=\"The input file for extraction\")\n",
    "# args = parser.parse_args()\n",
    "\n",
    "# if args.verbose:\n",
    "#     print(\"Input file is \" + args.input_file + \"\\n\")\n",
    "\n",
    "# f=open(args.input_file,\"r\")\n",
    "# if f.mode == \"r\":\n",
    "#     contents=f.read()\n",
    "\n",
    "# tmp=contents.replace(\"\\t\",\",\").split('\\n')\n",
    "# tmp_dict = {}\n",
    "# for line in tmp[:-1]:\n",
    "#     tmp_dict[line.split(\",\")[0]] = str(line.split(\",\")[5])\n",
    "\n",
    "# count_dict = {}\n",
    "# correct = 0\n",
    "# incorrect = 0\n",
    "# for item in tmp_dict:\n",
    "#     if tmp_dict[item] not in count_dict:\n",
    "#         count_dict[tmp_dict[item]] = 1\n",
    "#     else:\n",
    "#         count_dict[tmp_dict[item]] = count_dict[tmp_dict[item]] + 1\n",
    "        \n",
    "# tmp = pd.DataFrame.from_dict(count_dict,orient='index',columns=[\"Count\"])\n",
    "# tmp.index.names = [args.input_file[33:-4]]\n",
    "# tmp = tmp.sort_values(by=\"Count\",ascending=False)\n",
    "\n",
    "# if args.verbose:\n",
    "#     print(tmp)\n",
    "\n",
    "# tmp.to_csv(args.input_file[:-4]+'_match_distribution.csv',sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting points from Tavish  \n",
    "\n",
    "reference_dataframe at '/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv' \n",
    "\n",
    "custom_database at '/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta'\n",
    "\n",
    "taxonomy_file at '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### workflow\n",
    "\n",
    "* Two databases\n",
    "* subsample 15000 reads per each mock community species. Save those out.\n",
    "* map reads against both databases with minimap safe out data in paf format.\n",
    "* get best hit per species (see what this means while looking at the data).\n",
    "* add the full taxonomy to each best match using the taxonomy file.\n",
    "* summarize data at different taxonomic ranks for each species.\n",
    "* pull this all together somehow (summary across all the samples? focus on species of interest e.g. deleted from analyis?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv')\n",
    "max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta')\n",
    "taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#threads to use\n",
    "threads = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')\n",
    "\n",
    "OUT_DIR = os.path.abspath('../../analysis/Mapping')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "### list of species in the max database\n",
    "max_species = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guillermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Removed from second test database\n",
    "species_delete = ['Candida_orthopsilosis',\n",
    "                 'Candida_metapsilosis',\n",
    "                 'Aspergillus_niger']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "###species to be searched against both databases\n",
    "mock_community = ['Penicillium_chrysogenum',\n",
    " 'Aspergillus_flavus',\n",
    " 'Aspergillus_niger',\n",
    " 'Pichia_kudriavzevii',\n",
    " 'Pichia_membranifaciens',\n",
    " 'Candida_albicans',\n",
    " 'Candida_parapsilosis',\n",
    " 'Candida_orthopsilosis',\n",
    " 'Candida_metapsilosis']\n",
    "\n",
    "# mock_community = ['Puccinia_striiformis-tritici',\n",
    "#              'Zymoseptoria_tritici',\n",
    "#              'Pyrenophora_tritici-repentis',\n",
    "#              'Fusarium_oxysporum',\n",
    "#              'Tuber_brumale',\n",
    "#              'Cortinarius_globuliformis',\n",
    "#              'Aspergillus_niger',\n",
    "#              'Clavispora_lusitaniae',\n",
    "#              'Kluyveromyces_unidentified',\n",
    "#              'Penicillium_chrysogenum',\n",
    "#              'Rhodotorula_mucilaginosa',\n",
    "#              'Scedosporium_boydii',\n",
    "#              'Blastobotrys_proliferans',\n",
    "#              'Debaryomyces_unidentified',\n",
    "#              'Galactomyces_geotrichum',\n",
    "#              'Kodamaea_ohmeri',\n",
    "#              'Meyerozyma_guillermondii',\n",
    "#              'Wickerhamomyces_anomalus',\n",
    "#              'Yamadazyma_mexicana',\n",
    "#              'Yamadazyma_scolyti',\n",
    "#              'Yarrowia_lipolytica',\n",
    "#              'Zygoascus_hellenicus',\n",
    "#              'Aspergillus_flavus',\n",
    "#              'Cryptococcus_zero',\n",
    "#              'Aspergillus_unidentified',\n",
    "#              'Diaporthe_CCL067',\n",
    "#              'Diaporthe_unidentified',\n",
    "#              'Oculimacula_yallundae-CCL031',\n",
    "#              'Oculimacula_yallundae-CCL029',\n",
    "#              'Dothiorella_vidmadera',\n",
    "#              'Quambalaria_cyanescens',\n",
    "#              'Entoleuca_unidentified',\n",
    "#              'Asteroma_CCL060',\n",
    "#              'Asteroma_CCL068',\n",
    "#              'Saccharomyces_cerevisiae',\n",
    "#              'Cladophialophora_unidentified',\n",
    "#              'Candida_albicans',\n",
    "#              'Candida_metapsilosis',\n",
    "#              'Candida_orthopsilosis',\n",
    "#              'Candida_parapsilosis',\n",
    "#              'Candida_unidentified',\n",
    "#              'Kluyveromyces_marxianus',\n",
    "#              'Pichia_kudriavzevii',\n",
    "#              'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = pd.read_csv(reference_dataframe_fn)\n",
    "ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['puccinia_striiformis-tritici',\n",
       " 'zymoseptoria_tritici',\n",
       " 'pyrenophora_tritici-repentis',\n",
       " 'fusarium_oxysporum',\n",
       " 'tuber_brumale',\n",
       " 'cortinarius_globuliformis',\n",
       " 'aspergillus_niger',\n",
       " 'clavispora_lusitaniae',\n",
       " 'kluyveromyces_unidentified',\n",
       " 'penicillium_chrysogenum',\n",
       " 'rhodotorula_mucilaginosa',\n",
       " 'scedosporium_boydii',\n",
       " 'blastobotrys_proliferans',\n",
       " 'debaryomyces_unidentified',\n",
       " 'galactomyces_geotrichum',\n",
       " 'kodamaea_ohmeri',\n",
       " 'meyerozyma_guillermondii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'yamadazyma_mexicana',\n",
       " 'yamadazyma_scolyti',\n",
       " 'yarrowia_lipolytica',\n",
       " 'zygoascus_hellenicus',\n",
       " 'aspergillus_flavus',\n",
       " 'cryptococcus_zero',\n",
       " 'aspergillus_unidentified',\n",
       " 'diaporthe_ccl067',\n",
       " 'diaporthe_unidentified',\n",
       " 'oculimacula_yallundae-ccl031',\n",
       " 'oculimacula_yallundae-ccl029',\n",
       " 'dothiorella_vidmadera',\n",
       " 'quambalaria_cyanescens',\n",
       " 'entoleuca_unidentified',\n",
       " 'asteroma_ccl060',\n",
       " 'asteroma_ccl068',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'cladophialophora_unidentified',\n",
       " 'candida_albicans',\n",
       " 'candida_metapsilosis',\n",
       " 'candida_orthopsilosis',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_unidentified',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.name_species.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_fn = os.path.join(OUT_DIR, 'gsref.db.fasta')\n",
    "\n",
    "new_db_list = []\n",
    "old_db_list = []\n",
    "for seq in SeqIO.parse(max_custom_database_fn, 'fasta'):\n",
    "    old_db_list.append(seq.id)\n",
    "    if seq.id.lower() in ref_df.name_species.tolist():\n",
    "        #print(seq.id)\n",
    "        seq.id = seq.name = seq.description = seq.id.lower()\n",
    "        new_db_list.append(seq)\n",
    "    else:\n",
    "        print(seq.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(new_db_list) == len(old_db_list):\n",
    "    SeqIO.write(new_db_list, new_db_fn, 'fasta')\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_db_fn = os.path.join(OUT_DIR, 'gsref.subdb.fasta')\n",
    "sub_db_list = []\n",
    "for seq in new_db_list:\n",
    "    if seq.id not in [x.lower() for x in species_delete]:\n",
    "        sub_db_list.append(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(sub_db_list) + len(species_delete) == len(new_db_list):\n",
    "    SeqIO.write(sub_db_list, sub_db_fn, 'fasta' )\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['puccinia_striiformis-tritici',\n",
       " 'zymoseptoria_tritici',\n",
       " 'pyrenophora_tritici-repentis',\n",
       " 'fusarium_oxysporum',\n",
       " 'tuber_brumale',\n",
       " 'cortinarius_globuliformis',\n",
       " 'clavispora_lusitaniae',\n",
       " 'kluyveromyces_unidentified',\n",
       " 'penicillium_chrysogenum',\n",
       " 'rhodotorula_mucilaginosa',\n",
       " 'scedosporium_boydii',\n",
       " 'blastobotrys_proliferans',\n",
       " 'debaryomyces_unidentified',\n",
       " 'galactomyces_geotrichum',\n",
       " 'kodamaea_ohmeri',\n",
       " 'meyerozyma_guillermondii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'yamadazyma_mexicana',\n",
       " 'yamadazyma_scolyti',\n",
       " 'yarrowia_lipolytica',\n",
       " 'zygoascus_hellenicus',\n",
       " 'aspergillus_flavus',\n",
       " 'cryptococcus_zero',\n",
       " 'aspergillus_unidentified',\n",
       " 'diaporthe_ccl067',\n",
       " 'diaporthe_unidentified',\n",
       " 'oculimacula_yallundae-ccl031',\n",
       " 'oculimacula_yallundae-ccl029',\n",
       " 'dothiorella_vidmadera',\n",
       " 'quambalaria_cyanescens',\n",
       " 'entoleuca_unidentified',\n",
       " 'asteroma_ccl060',\n",
       " 'asteroma_ccl068',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'cladophialophora_unidentified',\n",
       " 'candida_albicans',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_unidentified',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens']"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x.id for x in sub_db_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = [x.lower() for x in mock_community]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['penicillium_chrysogenum',\n",
       " 'aspergillus_flavus',\n",
       " 'aspergillus_niger',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens',\n",
       " 'candida_albicans',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_orthopsilosis',\n",
       " 'candida_metapsilosis']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mock_community"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 15000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "MC_READ_DIR = os.path.join(OUT_DIR, 'MC_READS')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'species', 'genus', 'family', 'order', 'class', 'phylum',\n",
       "       'kingdom', '# raw reads', '# reads after homology filtering',\n",
       "       '# reads after length filtering', '# for use', 'path to raw reads',\n",
       "       'path to homology filtering', 'path to length filtering',\n",
       "       'path for use', 'name_species'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'penicillium_chrysogenum': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta',\n",
       " 'aspergillus_flavus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta',\n",
       " 'aspergillus_niger': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta',\n",
       " 'pichia_kudriavzevii': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode11/length_restricted_for_use.fasta',\n",
       " 'pichia_membranifaciens': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode12/length_restricted_for_use.fasta',\n",
       " 'candida_albicans': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta',\n",
       " 'candida_parapsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode06/length_restricted_for_use.fasta',\n",
       " 'candida_orthopsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta',\n",
       " 'candida_metapsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode04/length_restricted_for_use.fasta'}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/penicillium_chrysogenum.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/aspergillus_flavus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/aspergillus_niger.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/pichia_kudriavzevii.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/pichia_membranifaciens.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_albicans.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_parapsilosis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_orthopsilosis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_metapsilosis.15000.fasta!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map with minimap against both databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn):\n",
    "    command = F\"minimap2 -x map-ont -t 6 {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb',\n",
       " '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db'}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/penicillium_chrysogenum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.penicillium_chrysogenum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/aspergillus_flavus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.aspergillus_flavus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/aspergillus_niger.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.aspergillus_niger.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/pichia_kudriavzevii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.pichia_kudriavzevii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/pichia_membranifaciens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.pichia_membranifaciens.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_albicans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.candida_albicans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_parapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.candida_parapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_orthopsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.candida_orthopsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_metapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.candida_metapsilosis.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'penicillium_chrysogenum': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.penicillium_chrysogenum.minimap2.paf',\n",
       " 'aspergillus_flavus': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.aspergillus_flavus.minimap2.paf',\n",
       " 'aspergillus_niger': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.aspergillus_niger.minimap2.paf',\n",
       " 'pichia_kudriavzevii': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.pichia_kudriavzevii.minimap2.paf',\n",
       " 'pichia_membranifaciens': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.pichia_membranifaciens.minimap2.paf',\n",
       " 'candida_albicans': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.candida_albicans.minimap2.paf',\n",
       " 'candida_parapsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.candida_parapsilosis.minimap2.paf',\n",
       " 'candida_orthopsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.candida_orthopsilosis.minimap2.paf',\n",
       " 'candida_metapsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_subdb/gsref.subdb.candida_metapsilosis.minimap2.paf'}"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_db_mapping_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/penicillium_chrysogenum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.penicillium_chrysogenum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/aspergillus_flavus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.aspergillus_flavus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/aspergillus_niger.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.aspergillus_niger.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/pichia_kudriavzevii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.pichia_kudriavzevii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/pichia_membranifaciens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.pichia_membranifaciens.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_albicans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.candida_albicans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_parapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.candida_parapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_orthopsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.candida_orthopsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/MC_READS/candida_metapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/gsref_db/gsref.db.candida_metapsilosis.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results(fn, species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[tmp_df['nmatch'] == tmp_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the query species: {species}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tavish/anaconda3/envs/honours1/lib/python3.7/site-packages/ipykernel_launcher.py:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  \"\"\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##########\n",
      "\n",
      "This was the query species: penicillium_chrysogenum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "penicillium_chrysogenum          0.987255\n",
      "aspergillus_niger                0.002048\n",
      "aspergillus_flavus               0.001972\n",
      "aspergillus_unidentified         0.001745\n",
      "kluyveromyces_marxianus          0.000986\n",
      "kluyveromyces_unidentified       0.000986\n",
      "zymoseptoria_tritici             0.000759\n",
      "tuber_brumale                    0.000531\n",
      "cortinarius_globuliformis        0.000531\n",
      "cladophialophora_unidentified    0.000379\n",
      "rhodotorula_mucilaginosa         0.000379\n",
      "kodamaea_ohmeri                  0.000228\n",
      "yamadazyma_scolyti               0.000152\n",
      "puccinia_striiformis-tritici     0.000152\n",
      "scedosporium_boydii              0.000152\n",
      "asteroma_ccl060                  0.000152\n",
      "clavispora_lusitaniae            0.000152\n",
      "zygoascus_hellenicus             0.000152\n",
      "dothiorella_vidmadera            0.000152\n",
      "yamadazyma_mexicana              0.000076\n",
      "blastobotrys_proliferans         0.000076\n",
      "candida_metapsilosis             0.000076\n",
      "candida_orthopsilosis            0.000076\n",
      "candida_unidentified             0.000076\n",
      "diaporthe_unidentified           0.000076\n",
      "cryptococcus_zero                0.000076\n",
      "debaryomyces_unidentified        0.000076\n",
      "quambalaria_cyanescens           0.000076\n",
      "fusarium_oxysporum               0.000076\n",
      "wickerhamomyces_anomalus         0.000076\n",
      "oculimacula_yallundae-ccl029     0.000076\n",
      "oculimacula_yallundae-ccl031     0.000076\n",
      "pyrenophora_tritici-repentis     0.000076\n",
      "galactomyces_geotrichum          0.000076\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_flavus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_flavus              0.989737\n",
      "aspergillus_niger               0.003277\n",
      "aspergillus_unidentified        0.003277\n",
      "penicillium_chrysogenum         0.002242\n",
      "dothiorella_vidmadera           0.000172\n",
      "zygoascus_hellenicus            0.000172\n",
      "zymoseptoria_tritici            0.000172\n",
      "yamadazyma_scolyti              0.000172\n",
      "oculimacula_yallundae-ccl031    0.000086\n",
      "pichia_membranifaciens          0.000086\n",
      "tuber_brumale                   0.000086\n",
      "debaryomyces_unidentified       0.000086\n",
      "cryptococcus_zero               0.000086\n",
      "candida_parapsilosis            0.000086\n",
      "blastobotrys_proliferans        0.000086\n",
      "wickerhamomyces_anomalus        0.000086\n",
      "entoleuca_unidentified          0.000086\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_niger\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_niger                0.773966\n",
      "aspergillus_unidentified         0.167765\n",
      "aspergillus_flavus               0.044038\n",
      "penicillium_chrysogenum          0.007639\n",
      "zymoseptoria_tritici             0.000899\n",
      "yamadazyma_mexicana              0.000749\n",
      "entoleuca_unidentified           0.000449\n",
      "yamadazyma_scolyti               0.000300\n",
      "asteroma_ccl068                  0.000300\n",
      "pichia_membranifaciens           0.000300\n",
      "oculimacula_yallundae-ccl031     0.000300\n",
      "fusarium_oxysporum               0.000300\n",
      "rhodotorula_mucilaginosa         0.000300\n",
      "saccharomyces_cerevisiae         0.000150\n",
      "clavispora_lusitaniae            0.000150\n",
      "candida_albicans                 0.000150\n",
      "candida_metapsilosis             0.000150\n",
      "candida_orthopsilosis            0.000150\n",
      "candida_parapsilosis             0.000150\n",
      "cladophialophora_unidentified    0.000150\n",
      "dothiorella_vidmadera            0.000150\n",
      "cortinarius_globuliformis        0.000150\n",
      "debaryomyces_unidentified        0.000150\n",
      "yarrowia_lipolytica              0.000150\n",
      "oculimacula_yallundae-ccl029     0.000150\n",
      "scedosporium_boydii              0.000150\n",
      "pichia_kudriavzevii              0.000150\n",
      "puccinia_striiformis-tritici     0.000150\n",
      "pyrenophora_tritici-repentis     0.000150\n",
      "quambalaria_cyanescens           0.000150\n",
      "meyerozyma_guillermondii         0.000150\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_kudriavzevii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_kudriavzevii              0.992285\n",
      "pichia_membranifaciens           0.003116\n",
      "candida_unidentified             0.000668\n",
      "candida_albicans                 0.000371\n",
      "candida_orthopsilosis            0.000371\n",
      "candida_metapsilosis             0.000297\n",
      "candida_parapsilosis             0.000297\n",
      "yamadazyma_mexicana              0.000223\n",
      "cladophialophora_unidentified    0.000223\n",
      "kodamaea_ohmeri                  0.000223\n",
      "yarrowia_lipolytica              0.000223\n",
      "tuber_brumale                    0.000148\n",
      "asteroma_ccl068                  0.000148\n",
      "yamadazyma_scolyti               0.000148\n",
      "oculimacula_yallundae-ccl029     0.000148\n",
      "kluyveromyces_marxianus          0.000148\n",
      "aspergillus_unidentified         0.000074\n",
      "asteroma_ccl060                  0.000074\n",
      "zymoseptoria_tritici             0.000074\n",
      "kluyveromyces_unidentified       0.000074\n",
      "clavispora_lusitaniae            0.000074\n",
      "cortinarius_globuliformis        0.000074\n",
      "galactomyces_geotrichum          0.000074\n",
      "oculimacula_yallundae-ccl031     0.000074\n",
      "penicillium_chrysogenum          0.000074\n",
      "pyrenophora_tritici-repentis     0.000074\n",
      "saccharomyces_cerevisiae         0.000074\n",
      "wickerhamomyces_anomalus         0.000074\n",
      "aspergillus_niger                0.000074\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_membranifaciens\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_membranifaciens           0.988318\n",
      "pichia_kudriavzevii              0.002279\n",
      "kodamaea_ohmeri                  0.000997\n",
      "candida_orthopsilosis            0.000926\n",
      "clavispora_lusitaniae            0.000712\n",
      "candida_metapsilosis             0.000641\n",
      "debaryomyces_unidentified        0.000570\n",
      "galactomyces_geotrichum          0.000570\n",
      "kluyveromyces_unidentified       0.000570\n",
      "meyerozyma_guillermondii         0.000570\n",
      "candida_parapsilosis             0.000499\n",
      "wickerhamomyces_anomalus         0.000356\n",
      "saccharomyces_cerevisiae         0.000356\n",
      "kluyveromyces_marxianus          0.000356\n",
      "zygoascus_hellenicus             0.000285\n",
      "candida_unidentified             0.000285\n",
      "yamadazyma_mexicana              0.000214\n",
      "candida_albicans                 0.000214\n",
      "yarrowia_lipolytica              0.000214\n",
      "yamadazyma_scolyti               0.000214\n",
      "blastobotrys_proliferans         0.000142\n",
      "rhodotorula_mucilaginosa         0.000142\n",
      "aspergillus_niger                0.000071\n",
      "zymoseptoria_tritici             0.000071\n",
      "cladophialophora_unidentified    0.000071\n",
      "oculimacula_yallundae-ccl029     0.000071\n",
      "penicillium_chrysogenum          0.000071\n",
      "quambalaria_cyanescens           0.000071\n",
      "scedosporium_boydii              0.000071\n",
      "aspergillus_flavus               0.000071\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: candida_albicans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_albicans                 0.423162\n",
      "candida_unidentified             0.276921\n",
      "candida_orthopsilosis            0.121179\n",
      "candida_metapsilosis             0.076701\n",
      "candida_parapsilosis             0.072156\n",
      "meyerozyma_guillermondii         0.006885\n",
      "debaryomyces_unidentified        0.004406\n",
      "galactomyces_geotrichum          0.001928\n",
      "kluyveromyces_marxianus          0.001790\n",
      "yamadazyma_mexicana              0.001652\n",
      "yamadazyma_scolyti               0.001652\n",
      "kodamaea_ohmeri                  0.001515\n",
      "saccharomyces_cerevisiae         0.001377\n",
      "kluyveromyces_unidentified       0.001239\n",
      "tuber_brumale                    0.000826\n",
      "entoleuca_unidentified           0.000826\n",
      "blastobotrys_proliferans         0.000689\n",
      "clavispora_lusitaniae            0.000689\n",
      "zygoascus_hellenicus             0.000689\n",
      "wickerhamomyces_anomalus         0.000689\n",
      "pyrenophora_tritici-repentis     0.000551\n",
      "pichia_membranifaciens           0.000413\n",
      "quambalaria_cyanescens           0.000413\n",
      "diaporthe_ccl067                 0.000275\n",
      "pichia_kudriavzevii              0.000275\n",
      "puccinia_striiformis-tritici     0.000275\n",
      "diaporthe_unidentified           0.000138\n",
      "fusarium_oxysporum               0.000138\n",
      "cryptococcus_zero                0.000138\n",
      "oculimacula_yallundae-ccl031     0.000138\n",
      "cladophialophora_unidentified    0.000138\n",
      "aspergillus_niger                0.000138\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: candida_parapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_parapsilosis             0.822293\n",
      "candida_orthopsilosis            0.072828\n",
      "candida_metapsilosis             0.070157\n",
      "candida_unidentified             0.010328\n",
      "candida_albicans                 0.006944\n",
      "meyerozyma_guillermondii         0.003383\n",
      "yamadazyma_scolyti               0.001959\n",
      "oculimacula_yallundae-ccl029     0.001603\n",
      "debaryomyces_unidentified        0.001603\n",
      "yamadazyma_mexicana              0.001068\n",
      "kodamaea_ohmeri                  0.001068\n",
      "clavispora_lusitaniae            0.000890\n",
      "galactomyces_geotrichum          0.000712\n",
      "wickerhamomyces_anomalus         0.000712\n",
      "tuber_brumale                    0.000534\n",
      "zymoseptoria_tritici             0.000534\n",
      "entoleuca_unidentified           0.000356\n",
      "blastobotrys_proliferans         0.000356\n",
      "kluyveromyces_unidentified       0.000356\n",
      "quambalaria_cyanescens           0.000356\n",
      "fusarium_oxysporum               0.000178\n",
      "kluyveromyces_marxianus          0.000178\n",
      "cortinarius_globuliformis        0.000178\n",
      "zygoascus_hellenicus             0.000178\n",
      "cladophialophora_unidentified    0.000178\n",
      "pichia_membranifaciens           0.000178\n",
      "pyrenophora_tritici-repentis     0.000178\n",
      "rhodotorula_mucilaginosa         0.000178\n",
      "saccharomyces_cerevisiae         0.000178\n",
      "scedosporium_boydii              0.000178\n",
      "aspergillus_niger                0.000178\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: candida_orthopsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_orthopsilosis           0.593027\n",
      "candida_metapsilosis            0.214988\n",
      "candida_parapsilosis            0.137297\n",
      "candida_unidentified            0.017361\n",
      "candida_albicans                0.012731\n",
      "meyerozyma_guillermondii        0.011140\n",
      "debaryomyces_unidentified       0.003183\n",
      "yamadazyma_scolyti              0.001447\n",
      "kluyveromyces_marxianus         0.001157\n",
      "galactomyces_geotrichum         0.000868\n",
      "wickerhamomyces_anomalus        0.000723\n",
      "clavispora_lusitaniae           0.000723\n",
      "zygoascus_hellenicus            0.000579\n",
      "saccharomyces_cerevisiae        0.000579\n",
      "kodamaea_ohmeri                 0.000434\n",
      "blastobotrys_proliferans        0.000434\n",
      "yamadazyma_mexicana             0.000434\n",
      "entoleuca_unidentified          0.000434\n",
      "kluyveromyces_unidentified      0.000289\n",
      "tuber_brumale                   0.000289\n",
      "pichia_membranifaciens          0.000289\n",
      "oculimacula_yallundae-ccl029    0.000289\n",
      "yarrowia_lipolytica             0.000289\n",
      "pichia_kudriavzevii             0.000145\n",
      "penicillium_chrysogenum         0.000145\n",
      "oculimacula_yallundae-ccl031    0.000145\n",
      "asteroma_ccl068                 0.000145\n",
      "asteroma_ccl060                 0.000145\n",
      "aspergillus_unidentified        0.000145\n",
      "aspergillus_flavus              0.000145\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: candida_metapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_metapsilosis             0.422978\n",
      "candida_orthopsilosis            0.353746\n",
      "candida_parapsilosis             0.170167\n",
      "candida_unidentified             0.019239\n",
      "candida_albicans                 0.009890\n",
      "kodamaea_ohmeri                  0.005419\n",
      "meyerozyma_guillermondii         0.003658\n",
      "debaryomyces_unidentified        0.003387\n",
      "diaporthe_unidentified           0.001761\n",
      "yamadazyma_scolyti               0.001355\n",
      "clavispora_lusitaniae            0.001219\n",
      "oculimacula_yallundae-ccl029     0.000948\n",
      "yamadazyma_mexicana              0.000948\n",
      "saccharomyces_cerevisiae         0.000813\n",
      "fusarium_oxysporum               0.000677\n",
      "kluyveromyces_marxianus          0.000542\n",
      "kluyveromyces_unidentified       0.000406\n",
      "tuber_brumale                    0.000406\n",
      "blastobotrys_proliferans         0.000406\n",
      "yarrowia_lipolytica              0.000271\n",
      "galactomyces_geotrichum          0.000271\n",
      "wickerhamomyces_anomalus         0.000271\n",
      "cladophialophora_unidentified    0.000135\n",
      "entoleuca_unidentified           0.000135\n",
      "penicillium_chrysogenum          0.000135\n",
      "pichia_kudriavzevii              0.000135\n",
      "rhodotorula_mucilaginosa         0.000135\n",
      "scedosporium_boydii              0.000135\n",
      "asteroma_ccl060                  0.000135\n",
      "aspergillus_unidentified         0.000135\n",
      "aspergillus_flavus               0.000135\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for species, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tavish/anaconda3/envs/honours1/lib/python3.7/site-packages/ipykernel_launcher.py:5: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  \"\"\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##########\n",
      "\n",
      "This was the query species: penicillium_chrysogenum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "penicillium_chrysogenum          0.990067\n",
      "aspergillus_unidentified         0.001779\n",
      "aspergillus_flavus               0.001631\n",
      "kluyveromyces_unidentified       0.000964\n",
      "kluyveromyces_marxianus          0.000889\n",
      "zymoseptoria_tritici             0.000593\n",
      "tuber_brumale                    0.000519\n",
      "cortinarius_globuliformis        0.000445\n",
      "rhodotorula_mucilaginosa         0.000296\n",
      "cladophialophora_unidentified    0.000296\n",
      "pyrenophora_tritici-repentis     0.000148\n",
      "yamadazyma_scolyti               0.000148\n",
      "asteroma_ccl060                  0.000148\n",
      "blastobotrys_proliferans         0.000148\n",
      "candida_unidentified             0.000148\n",
      "clavispora_lusitaniae            0.000148\n",
      "debaryomyces_unidentified        0.000148\n",
      "dothiorella_vidmadera            0.000148\n",
      "fusarium_oxysporum               0.000148\n",
      "zygoascus_hellenicus             0.000148\n",
      "yamadazyma_mexicana              0.000148\n",
      "kodamaea_ohmeri                  0.000148\n",
      "puccinia_striiformis-tritici     0.000148\n",
      "scedosporium_boydii              0.000074\n",
      "galactomyces_geotrichum          0.000074\n",
      "cryptococcus_zero                0.000074\n",
      "oculimacula_yallundae-ccl029     0.000074\n",
      "candida_albicans                 0.000074\n",
      "oculimacula_yallundae-ccl031     0.000074\n",
      "wickerhamomyces_anomalus         0.000074\n",
      "diaporthe_unidentified           0.000074\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_flavus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_flavus              0.994542\n",
      "aspergillus_unidentified        0.002653\n",
      "penicillium_chrysogenum         0.001365\n",
      "dothiorella_vidmadera           0.000303\n",
      "zymoseptoria_tritici            0.000227\n",
      "zygoascus_hellenicus            0.000152\n",
      "tuber_brumale                   0.000152\n",
      "yamadazyma_scolyti              0.000076\n",
      "saccharomyces_cerevisiae        0.000076\n",
      "oculimacula_yallundae-ccl031    0.000076\n",
      "kodamaea_ohmeri                 0.000076\n",
      "kluyveromyces_marxianus         0.000076\n",
      "entoleuca_unidentified          0.000076\n",
      "candida_parapsilosis            0.000076\n",
      "blastobotrys_proliferans        0.000076\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_niger\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_unidentified         0.937173\n",
      "aspergillus_flavus               0.042779\n",
      "penicillium_chrysogenum          0.011748\n",
      "rhodotorula_mucilaginosa         0.000894\n",
      "dothiorella_vidmadera            0.000766\n",
      "oculimacula_yallundae-ccl031     0.000766\n",
      "zymoseptoria_tritici             0.000766\n",
      "tuber_brumale                    0.000766\n",
      "yamadazyma_mexicana              0.000638\n",
      "cladophialophora_unidentified    0.000511\n",
      "fusarium_oxysporum               0.000383\n",
      "oculimacula_yallundae-ccl029     0.000255\n",
      "meyerozyma_guillermondii         0.000255\n",
      "kluyveromyces_marxianus          0.000255\n",
      "quambalaria_cyanescens           0.000255\n",
      "entoleuca_unidentified           0.000255\n",
      "diaporthe_ccl067                 0.000128\n",
      "yarrowia_lipolytica              0.000128\n",
      "asteroma_ccl068                  0.000128\n",
      "clavispora_lusitaniae            0.000128\n",
      "cortinarius_globuliformis        0.000128\n",
      "scedosporium_boydii              0.000128\n",
      "zygoascus_hellenicus             0.000128\n",
      "wickerhamomyces_anomalus         0.000128\n",
      "pichia_membranifaciens           0.000128\n",
      "puccinia_striiformis-tritici     0.000128\n",
      "pyrenophora_tritici-repentis     0.000128\n",
      "galactomyces_geotrichum          0.000128\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_kudriavzevii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_kudriavzevii              0.992741\n",
      "pichia_membranifaciens           0.003185\n",
      "candida_parapsilosis             0.000741\n",
      "candida_unidentified             0.000519\n",
      "kodamaea_ohmeri                  0.000222\n",
      "candida_albicans                 0.000222\n",
      "clavispora_lusitaniae            0.000222\n",
      "yamadazyma_mexicana              0.000222\n",
      "asteroma_ccl068                  0.000148\n",
      "tuber_brumale                    0.000148\n",
      "oculimacula_yallundae-ccl029     0.000148\n",
      "yamadazyma_scolyti               0.000148\n",
      "galactomyces_geotrichum          0.000148\n",
      "blastobotrys_proliferans         0.000074\n",
      "cortinarius_globuliformis        0.000074\n",
      "asteroma_ccl060                  0.000074\n",
      "cladophialophora_unidentified    0.000074\n",
      "zymoseptoria_tritici             0.000074\n",
      "kluyveromyces_unidentified       0.000074\n",
      "debaryomyces_unidentified        0.000074\n",
      "dothiorella_vidmadera            0.000074\n",
      "kluyveromyces_marxianus          0.000074\n",
      "yarrowia_lipolytica              0.000074\n",
      "oculimacula_yallundae-ccl031     0.000074\n",
      "penicillium_chrysogenum          0.000074\n",
      "pyrenophora_tritici-repentis     0.000074\n",
      "saccharomyces_cerevisiae         0.000074\n",
      "wickerhamomyces_anomalus         0.000074\n",
      "aspergillus_unidentified         0.000074\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_membranifaciens\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_membranifaciens           0.988630\n",
      "pichia_kudriavzevii              0.002487\n",
      "kodamaea_ohmeri                  0.000995\n",
      "candida_parapsilosis             0.000853\n",
      "clavispora_lusitaniae            0.000711\n",
      "kluyveromyces_unidentified       0.000640\n",
      "meyerozyma_guillermondii         0.000640\n",
      "galactomyces_geotrichum          0.000569\n",
      "debaryomyces_unidentified        0.000569\n",
      "candida_albicans                 0.000497\n",
      "candida_unidentified             0.000497\n",
      "wickerhamomyces_anomalus         0.000426\n",
      "kluyveromyces_marxianus          0.000426\n",
      "yamadazyma_scolyti               0.000355\n",
      "saccharomyces_cerevisiae         0.000284\n",
      "yamadazyma_mexicana              0.000213\n",
      "zygoascus_hellenicus             0.000213\n",
      "yarrowia_lipolytica              0.000213\n",
      "rhodotorula_mucilaginosa         0.000142\n",
      "blastobotrys_proliferans         0.000071\n",
      "cladophialophora_unidentified    0.000071\n",
      "zymoseptoria_tritici             0.000071\n",
      "cortinarius_globuliformis        0.000071\n",
      "oculimacula_yallundae-ccl029     0.000071\n",
      "penicillium_chrysogenum          0.000071\n",
      "quambalaria_cyanescens           0.000071\n",
      "scedosporium_boydii              0.000071\n",
      "aspergillus_flavus               0.000071\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: candida_albicans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_albicans                 0.488268\n",
      "candida_unidentified             0.318877\n",
      "candida_parapsilosis             0.156341\n",
      "meyerozyma_guillermondii         0.007514\n",
      "debaryomyces_unidentified        0.006459\n",
      "kluyveromyces_marxianus          0.002373\n",
      "yamadazyma_scolyti               0.002241\n",
      "yamadazyma_mexicana              0.002109\n",
      "kluyveromyces_unidentified       0.001977\n",
      "galactomyces_geotrichum          0.001977\n",
      "saccharomyces_cerevisiae         0.001846\n",
      "kodamaea_ohmeri                  0.001450\n",
      "wickerhamomyces_anomalus         0.001186\n",
      "tuber_brumale                    0.001186\n",
      "clavispora_lusitaniae            0.001186\n",
      "blastobotrys_proliferans         0.000923\n",
      "entoleuca_unidentified           0.000659\n",
      "pichia_kudriavzevii              0.000527\n",
      "zygoascus_hellenicus             0.000527\n",
      "cryptococcus_zero                0.000395\n",
      "pichia_membranifaciens           0.000264\n",
      "pyrenophora_tritici-repentis     0.000264\n",
      "diaporthe_ccl067                 0.000264\n",
      "cladophialophora_unidentified    0.000132\n",
      "asteroma_ccl068                  0.000132\n",
      "zymoseptoria_tritici             0.000132\n",
      "diaporthe_unidentified           0.000132\n",
      "fusarium_oxysporum               0.000132\n",
      "oculimacula_yallundae-ccl029     0.000132\n",
      "quambalaria_cyanescens           0.000132\n",
      "yarrowia_lipolytica              0.000132\n",
      "aspergillus_flavus               0.000132\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: candida_parapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_parapsilosis            0.984587\n",
      "candida_unidentified            0.002213\n",
      "candida_albicans                0.001897\n",
      "debaryomyces_unidentified       0.001344\n",
      "meyerozyma_guillermondii        0.001107\n",
      "kodamaea_ohmeri                 0.001107\n",
      "pichia_membranifaciens          0.000711\n",
      "yamadazyma_scolyti              0.000711\n",
      "blastobotrys_proliferans        0.000632\n",
      "kluyveromyces_unidentified      0.000632\n",
      "kluyveromyces_marxianus         0.000632\n",
      "saccharomyces_cerevisiae        0.000553\n",
      "galactomyces_geotrichum         0.000553\n",
      "wickerhamomyces_anomalus        0.000474\n",
      "clavispora_lusitaniae           0.000474\n",
      "yamadazyma_mexicana             0.000474\n",
      "tuber_brumale                   0.000395\n",
      "zymoseptoria_tritici            0.000316\n",
      "pichia_kudriavzevii             0.000316\n",
      "zygoascus_hellenicus            0.000158\n",
      "pyrenophora_tritici-repentis    0.000158\n",
      "fusarium_oxysporum              0.000079\n",
      "entoleuca_unidentified          0.000079\n",
      "cortinarius_globuliformis       0.000079\n",
      "oculimacula_yallundae-ccl029    0.000079\n",
      "puccinia_striiformis-tritici    0.000079\n",
      "quambalaria_cyanescens          0.000079\n",
      "aspergillus_flavus              0.000079\n",
      "dtype: float64 \n",
      "\n",
      "##########\n",
      "\n",
      "This was the query species: candida_orthopsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_parapsilosis             0.940327\n",
      "meyerozyma_guillermondii         0.016757\n",
      "candida_albicans                 0.009264\n",
      "candida_unidentified             0.009128\n",
      "debaryomyces_unidentified        0.004768\n",
      "yamadazyma_scolyti               0.003134\n",
      "wickerhamomyces_anomalus         0.002452\n",
      "kodamaea_ohmeri                  0.001907\n",
      "yamadazyma_mexicana              0.001635\n",
      "tuber_brumale                    0.001362\n",
      "kluyveromyces_unidentified       0.001226\n",
      "galactomyces_geotrichum          0.001226\n",
      "kluyveromyces_marxianus          0.001090\n",
      "entoleuca_unidentified           0.000954\n",
      "blastobotrys_proliferans         0.000954\n",
      "clavispora_lusitaniae            0.000681\n",
      "saccharomyces_cerevisiae         0.000681\n",
      "zygoascus_hellenicus             0.000272\n",
      "pichia_kudriavzevii              0.000272\n",
      "pichia_membranifaciens           0.000272\n",
      "asteroma_ccl060                  0.000136\n",
      "aspergillus_unidentified         0.000136\n",
      "cladophialophora_unidentified    0.000136\n",
      "zymoseptoria_tritici             0.000136\n",
      "cortinarius_globuliformis        0.000136\n",
      "fusarium_oxysporum               0.000136\n",
      "oculimacula_yallundae-ccl029     0.000136\n",
      "oculimacula_yallundae-ccl031     0.000136\n",
      "penicillium_chrysogenum          0.000136\n",
      "rhodotorula_mucilaginosa         0.000136\n",
      "scedosporium_boydii              0.000136\n",
      "aspergillus_flavus               0.000136\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##########\n",
      "\n",
      "This was the query species: candida_metapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_parapsilosis             0.959939\n",
      "candida_albicans                 0.007885\n",
      "candida_unidentified             0.007376\n",
      "meyerozyma_guillermondii         0.005469\n",
      "debaryomyces_unidentified        0.004324\n",
      "kodamaea_ohmeri                  0.001780\n",
      "yamadazyma_scolyti               0.001780\n",
      "wickerhamomyces_anomalus         0.001780\n",
      "clavispora_lusitaniae            0.001653\n",
      "yamadazyma_mexicana              0.001272\n",
      "kluyveromyces_unidentified       0.000763\n",
      "blastobotrys_proliferans         0.000763\n",
      "galactomyces_geotrichum          0.000763\n",
      "kluyveromyces_marxianus          0.000763\n",
      "fusarium_oxysporum               0.000509\n",
      "zymoseptoria_tritici             0.000382\n",
      "pichia_membranifaciens           0.000382\n",
      "saccharomyces_cerevisiae         0.000382\n",
      "zygoascus_hellenicus             0.000254\n",
      "oculimacula_yallundae-ccl031     0.000254\n",
      "cladophialophora_unidentified    0.000254\n",
      "tuber_brumale                    0.000254\n",
      "asteroma_ccl068                  0.000254\n",
      "pichia_kudriavzevii              0.000127\n",
      "entoleuca_unidentified           0.000127\n",
      "quambalaria_cyanescens           0.000127\n",
      "rhodotorula_mucilaginosa         0.000127\n",
      "yarrowia_lipolytica              0.000127\n",
      "aspergillus_flavus               0.000127\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running against a database that have ['Candida_orthopsilosis', 'Candida_metapsilosis', 'Aspergillus_niger'] deleted\n",
    "for species, hit_fn in sub_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
