{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook to analyzse the efficiency of minimap mapping against a mock community"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting points from Tavish  \n",
    "reference_dataframe at '/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv'  \n",
    "custom_database at '/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta'  \n",
    "taxonomy_file at '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### workflow\n",
    "\n",
    "* Two databases\n",
    "* subsample 15000 reads per each mock community species. Save those out.\n",
    "* map reads against both databases with minimap safe out data in paf format.\n",
    "* get best hit per species (see what this means while looking at the data).\n",
    "* add the full taxonomy to each best match using the taxonomy file.\n",
    "* summarize data at different taxonomic ranks for each species.\n",
    "* pull this all together somehow (summary across all the samples? focus on species of interest e.g. deleted from analyis?)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### requirment\n",
    "\n",
    "* Bbmap (conda install bbmap https://anaconda.org/bioconda/bbmap)  \n",
    "* minimap2 (conda install minimap2 https://anaconda.org/bioconda/minimap2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fixes on the command line\n",
    "\n",
    "* fixed the taxonomy file to fit with the quiime format\n",
    "\n",
    "cat /media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file.csv | sed 's/,/\\t/' > /media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v2.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fix the taxonomy_file_v2.csv more to reflect Qiime style\n",
    "\n",
    "* This requires to make the species name genus_species and not _species.... hope this makes sense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "old_taxonomy_file_fn = '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v2.csv'\n",
    "new_taxonomy_file_fn = '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v3.csv'\n",
    "with open(new_taxonomy_file_fn, 'w') as out_fh:\n",
    "    with open(old_taxonomy_file_fn, 'r') as in_fh:\n",
    "        for line in in_fh:\n",
    "            line = line.rstrip()\n",
    "            #print(line)\n",
    "            first_half = line.split('s__')[0]\n",
    "            second_half = line.split('s__')[1]\n",
    "            pattern = re.compile(r'g__\\w+;')\n",
    "            genus = re.findall(pattern, first_half)[0].replace('g__','').replace(';','')\n",
    "            new_line = F\"{first_half}s__{genus}_{second_half}\"\n",
    "            print(new_line,file=out_fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20171103_FAH15473/barcode01\tk__Fungi;p__Basidiomycota;c__Pucciniomycetes;o__Pucciniales;f__Pucciniaceae;g__Puccinia;s__striiformis-tritici\r\n",
      "20171103_FAH15473/barcode02\tk__Fungi;p__Ascomycota;c__Dothideomycetes;o__Capnodiales;f__Mycosphaerellaceae;g__Zymoseptoria;s__tritici\r\n"
     ]
    }
   ],
   "source": [
    "!head -2 {old_taxonomy_file_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20171103_FAH15473/barcode01\tk__Fungi;p__Basidiomycota;c__Pucciniomycetes;o__Pucciniales;f__Pucciniaceae;g__Puccinia;s__Puccinia_striiformis-tritici\r\n",
      "20171103_FAH15473/barcode02\tk__Fungi;p__Ascomycota;c__Dothideomycetes;o__Capnodiales;f__Mycosphaerellaceae;g__Zymoseptoria;s__Zymoseptoria_tritici\r\n"
     ]
    }
   ],
   "source": [
    "!head -2 {new_taxonomy_file_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Initial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv')\n",
    "max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta')\n",
    "taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_qiime.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "#threads to use\n",
    "threads = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = os.path.abspath('../../analysis/Mapping_mock_gsref')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "### list of species in the max database\n",
    "max_species = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ###Removed from second test databes\n",
    "species_delete = [\n",
    "# 'Candida_orthopsilosis',\n",
    "#                  'Candida_metapsilosis',\n",
    "#                  'Aspergillus_niger'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "###species to be searched against both databases\n",
    "# mock_community = ['Penicillium_chrysogenum',\n",
    "#  'Aspergillus_flavus',\n",
    "#  'Aspergillus_niger',\n",
    "#  'Pichia_kudriavzevii',\n",
    "#  'Pichia_membranifaciens',\n",
    "#  'Candida_albicans',\n",
    "#  'Candida_parapsilosis',\n",
    "#  'Candida_orthopsilosis',\n",
    "#  'Candida_metapsilosis']\n",
    "\n",
    "mock_community = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fixed_old_names = ['Kluyveromyces_lactis',\n",
    "#                    'Candida_zeylanoides',\n",
    "#                    'Cladophialophora_sp.',\n",
    "#                    'Diaporthe_sp.',\n",
    "#                    'CCL060',\n",
    "#                    'CCL068',\n",
    "#                    'CCL067',\n",
    "#                    'Aspergillus_sp.',\n",
    "#                    'Entoleuca_sp.',\n",
    "#                    'Tapesia_yallundae_CCL029',\n",
    "#                    'Tapesia_yallundae_CCL031',\n",
    "#                    'Cryptococcus_neoformans']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fixed_new_names = ['candida_unidentified',\n",
    "#                    'debaryomyces_unidentified',\n",
    "#                    'cladophialophora_unidentified',\n",
    "#                    'diaporthe_unidentified',\n",
    "#                    'asteroma_ccl060',\n",
    "#                    'asteroma_ccl068',\n",
    "#                    'diaporthe_ccl067',\n",
    "#                    'aspergillus_unidentified',\n",
    "#                    'entoleuca_unidentified',\n",
    "#                    'oculimacula_yallundae-ccl029',\n",
    "#                    'oculimacula_yallundae-ccl031',\n",
    "#                    'kluyveromyces_unidentified']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old_to_new_names = dict(zip(fixed_old_names, fixed_new_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old_to_new_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fix databases and names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = pd.read_csv(reference_dataframe_fn)\n",
    "ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['puccinia_striiformis-tritici',\n",
       " 'zymoseptoria_tritici',\n",
       " 'pyrenophora_tritici-repentis',\n",
       " 'fusarium_oxysporum',\n",
       " 'tuber_brumale',\n",
       " 'cortinarius_globuliformis',\n",
       " 'aspergillus_niger',\n",
       " 'clavispora_lusitaniae',\n",
       " 'kluyveromyces_unidentified',\n",
       " 'penicillium_chrysogenum',\n",
       " 'rhodotorula_mucilaginosa',\n",
       " 'scedosporium_boydii',\n",
       " 'blastobotrys_proliferans',\n",
       " 'debaryomyces_unidentified',\n",
       " 'galactomyces_geotrichum',\n",
       " 'kodamaea_ohmeri',\n",
       " 'meyerozyma_guilliermondii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'yamadazyma_mexicana',\n",
       " 'yamadazyma_scolyti',\n",
       " 'yarrowia_lipolytica',\n",
       " 'zygoascus_hellenicus',\n",
       " 'aspergillus_flavus',\n",
       " 'cryptococcus_zero',\n",
       " 'aspergillus_unidentified',\n",
       " 'diaporthe_ccl067',\n",
       " 'diaporthe_unidentified',\n",
       " 'oculimacula_yallundae-ccl031',\n",
       " 'oculimacula_yallundae-ccl029',\n",
       " 'dothiorella_vidmadera',\n",
       " 'quambalaria_cyanescens',\n",
       " 'entoleuca_unidentified',\n",
       " 'asteroma_ccl060',\n",
       " 'asteroma_ccl068',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'cladophialophora_unidentified',\n",
       " 'candida_albicans',\n",
       " 'candida_metapsilosis',\n",
       " 'candida_orthopsilosis',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_unidentified',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens']"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.name_species.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_fn = os.path.join(OUT_DIR, 'gsref.db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_list = []\n",
    "old_db_list = []\n",
    "for seq in SeqIO.parse(max_custom_database_fn, 'fasta'):\n",
    "    old_db_list.append(seq.id)\n",
    "    if seq.id.lower() in ref_df.name_species.tolist():\n",
    "        #print(seq.id)\n",
    "        seq.id = seq.name = seq.description = seq.id.lower()\n",
    "        new_db_list.append(seq)\n",
    "    else:\n",
    "        print(seq.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(new_db_list) == len(old_db_list):\n",
    "    SeqIO.write(new_db_list, new_db_fn, 'fasta')\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_db_fn = os.path.join(OUT_DIR, 'gsref.subdb.fasta')\n",
    "sub_db_list = []\n",
    "for seq in new_db_list:\n",
    "    if seq.id not in [x.lower() for x in species_delete]:\n",
    "        sub_db_list.append(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(sub_db_list) + len(species_delete) == len(new_db_list):\n",
    "    SeqIO.write(sub_db_list, sub_db_fn, 'fasta' )\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['puccinia_striiformis-tritici',\n",
       " 'zymoseptoria_tritici',\n",
       " 'pyrenophora_tritici-repentis',\n",
       " 'fusarium_oxysporum',\n",
       " 'tuber_brumale',\n",
       " 'cortinarius_globuliformis',\n",
       " 'aspergillus_niger',\n",
       " 'clavispora_lusitaniae',\n",
       " 'kluyveromyces_unidentified',\n",
       " 'penicillium_chrysogenum',\n",
       " 'rhodotorula_mucilaginosa',\n",
       " 'scedosporium_boydii',\n",
       " 'blastobotrys_proliferans',\n",
       " 'debaryomyces_unidentified',\n",
       " 'galactomyces_geotrichum',\n",
       " 'kodamaea_ohmeri',\n",
       " 'meyerozyma_guilliermondii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'yamadazyma_mexicana',\n",
       " 'yamadazyma_scolyti',\n",
       " 'yarrowia_lipolytica',\n",
       " 'zygoascus_hellenicus',\n",
       " 'aspergillus_flavus',\n",
       " 'cryptococcus_zero',\n",
       " 'aspergillus_unidentified',\n",
       " 'diaporthe_ccl067',\n",
       " 'diaporthe_unidentified',\n",
       " 'oculimacula_yallundae-ccl031',\n",
       " 'oculimacula_yallundae-ccl029',\n",
       " 'dothiorella_vidmadera',\n",
       " 'quambalaria_cyanescens',\n",
       " 'entoleuca_unidentified',\n",
       " 'asteroma_ccl060',\n",
       " 'asteroma_ccl068',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'cladophialophora_unidentified',\n",
       " 'candida_albicans',\n",
       " 'candida_metapsilosis',\n",
       " 'candida_orthopsilosis',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_unidentified',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens']"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x.id for x in sub_db_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = [x.lower() for x in mock_community]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['puccinia_striiformis-tritici',\n",
       " 'zymoseptoria_tritici',\n",
       " 'pyrenophora_tritici-repentis',\n",
       " 'fusarium_oxysporum',\n",
       " 'tuber_brumale',\n",
       " 'cortinarius_globuliformis',\n",
       " 'aspergillus_niger',\n",
       " 'clavispora_lusitaniae',\n",
       " 'kluyveromyces_unidentified',\n",
       " 'penicillium_chrysogenum',\n",
       " 'rhodotorula_mucilaginosa',\n",
       " 'scedosporium_boydii',\n",
       " 'blastobotrys_proliferans',\n",
       " 'debaryomyces_unidentified',\n",
       " 'galactomyces_geotrichum',\n",
       " 'kodamaea_ohmeri',\n",
       " 'meyerozyma_guilliermondii',\n",
       " 'wickerhamomyces_anomalus',\n",
       " 'yamadazyma_mexicana',\n",
       " 'yamadazyma_scolyti',\n",
       " 'yarrowia_lipolytica',\n",
       " 'zygoascus_hellenicus',\n",
       " 'aspergillus_flavus',\n",
       " 'cryptococcus_zero',\n",
       " 'aspergillus_unidentified',\n",
       " 'diaporthe_ccl067',\n",
       " 'diaporthe_unidentified',\n",
       " 'oculimacula_yallundae-ccl031',\n",
       " 'oculimacula_yallundae-ccl029',\n",
       " 'dothiorella_vidmadera',\n",
       " 'quambalaria_cyanescens',\n",
       " 'entoleuca_unidentified',\n",
       " 'asteroma_ccl060',\n",
       " 'asteroma_ccl068',\n",
       " 'saccharomyces_cerevisiae',\n",
       " 'cladophialophora_unidentified',\n",
       " 'candida_albicans',\n",
       " 'candida_metapsilosis',\n",
       " 'candida_orthopsilosis',\n",
       " 'candida_parapsilosis',\n",
       " 'candida_unidentified',\n",
       " 'kluyveromyces_marxianus',\n",
       " 'pichia_kudriavzevii',\n",
       " 'pichia_membranifaciens']"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mock_community"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subsample reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 15000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "MC_READ_DIR = os.path.join(OUT_DIR, 'MC_READS')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'species', 'genus', 'family', 'order', 'class', 'phylum',\n",
       "       'kingdom', '# raw reads', '# reads after homology filtering',\n",
       "       '# reads after length filtering', '# for use', 'path to raw reads',\n",
       "       'path to homology filtering', 'path to length filtering',\n",
       "       'path for use', 'name_species'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'puccinia_striiformis-tritici': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode01/length_restricted_for_use.fasta',\n",
       " 'zymoseptoria_tritici': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_for_use.fasta',\n",
       " 'pyrenophora_tritici-repentis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode03/length_restricted_for_use.fasta',\n",
       " 'fusarium_oxysporum': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode04/length_restricted_for_use.fasta',\n",
       " 'tuber_brumale': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode05/length_restricted_for_use.fasta',\n",
       " 'cortinarius_globuliformis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode06/length_restricted_for_use.fasta',\n",
       " 'aspergillus_niger': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta',\n",
       " 'clavispora_lusitaniae': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode08/length_restricted_for_use.fasta',\n",
       " 'kluyveromyces_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode09/length_restricted_for_use.fasta',\n",
       " 'penicillium_chrysogenum': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta',\n",
       " 'rhodotorula_mucilaginosa': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode11/length_restricted_for_use.fasta',\n",
       " 'scedosporium_boydii': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode12/length_restricted_for_use.fasta',\n",
       " 'blastobotrys_proliferans': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode01/length_restricted_for_use.fasta',\n",
       " 'debaryomyces_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode02/length_restricted_for_use.fasta',\n",
       " 'galactomyces_geotrichum': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_for_use.fasta',\n",
       " 'kodamaea_ohmeri': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode04/length_restricted_for_use.fasta',\n",
       " 'meyerozyma_guilliermondii': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode05/length_restricted_for_use.fasta',\n",
       " 'wickerhamomyces_anomalus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode06/length_restricted_for_use.fasta',\n",
       " 'yamadazyma_mexicana': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode07/length_restricted_for_use.fasta',\n",
       " 'yamadazyma_scolyti': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode08/length_restricted_for_use.fasta',\n",
       " 'yarrowia_lipolytica': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode09/length_restricted_for_use.fasta',\n",
       " 'zygoascus_hellenicus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode11/length_restricted_for_use.fasta',\n",
       " 'aspergillus_flavus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta',\n",
       " 'cryptococcus_zero': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode01/length_restricted_for_use.fasta',\n",
       " 'aspergillus_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode02/length_restricted_for_use.fasta',\n",
       " 'diaporthe_ccl067': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode03/length_restricted_for_use.fasta',\n",
       " 'diaporthe_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode04/length_restricted_for_use.fasta',\n",
       " 'oculimacula_yallundae-ccl031': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode05/length_restricted_for_use.fasta',\n",
       " 'oculimacula_yallundae-ccl029': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode06/length_restricted_for_use.fasta',\n",
       " 'dothiorella_vidmadera': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode07/length_restricted_for_use.fasta',\n",
       " 'quambalaria_cyanescens': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode08/length_restricted_for_use.fasta',\n",
       " 'entoleuca_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode09/length_restricted_for_use.fasta',\n",
       " 'asteroma_ccl060': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode11/length_restricted_for_use.fasta',\n",
       " 'asteroma_ccl068': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode12/length_restricted_for_use.fasta',\n",
       " 'saccharomyces_cerevisiae': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode01/length_restricted_for_use.fasta',\n",
       " 'cladophialophora_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode02/length_restricted_for_use.fasta',\n",
       " 'candida_albicans': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta',\n",
       " 'candida_metapsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode04/length_restricted_for_use.fasta',\n",
       " 'candida_orthopsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta',\n",
       " 'candida_parapsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode06/length_restricted_for_use.fasta',\n",
       " 'candida_unidentified': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode09/length_restricted_for_use.fasta',\n",
       " 'kluyveromyces_marxianus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode10/length_restricted_for_use.fasta',\n",
       " 'pichia_kudriavzevii': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode11/length_restricted_for_use.fasta',\n",
       " 'pichia_membranifaciens': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode12/length_restricted_for_use.fasta'}"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode01/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/puccinia_striiformis-tritici.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zymoseptoria_tritici.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pyrenophora_tritici-repentis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/fusarium_oxysporum.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/tuber_brumale.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cortinarius_globuliformis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_niger.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode08/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/clavispora_lusitaniae.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode09/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/penicillium_chrysogenum.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/rhodotorula_mucilaginosa.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/scedosporium_boydii.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode01/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/blastobotrys_proliferans.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode02/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/debaryomyces_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/galactomyces_geotrichum.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kodamaea_ohmeri.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/meyerozyma_guilliermondii.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/wickerhamomyces_anomalus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode07/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_mexicana.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode08/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_scolyti.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode09/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yarrowia_lipolytica.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zygoascus_hellenicus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_flavus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode01/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cryptococcus_zero.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode02/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_ccl067.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl031.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl029.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode07/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/dothiorella_vidmadera.15000.fasta!!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode08/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/quambalaria_cyanescens.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode09/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/entoleuca_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl060.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171212_FAH18688/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl068.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode01/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/saccharomyces_cerevisiae.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode02/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cladophialophora_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_albicans.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode04/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_metapsilosis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_orthopsilosis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_parapsilosis.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode09/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_unidentified.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode10/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_marxianus.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode11/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_kudriavzevii.15000.fasta!!\n",
      "\n",
      ":(check one reformat.sh samplereadstarget=15000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode12/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_membranifaciens.15000.fasta!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map with minimap against both databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb',\n",
       " '/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db'}"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/puccinia_striiformis-tritici.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.puccinia_striiformis-tritici.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zymoseptoria_tritici.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.zymoseptoria_tritici.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pyrenophora_tritici-repentis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.pyrenophora_tritici-repentis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/fusarium_oxysporum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.fusarium_oxysporum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/tuber_brumale.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.tuber_brumale.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cortinarius_globuliformis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.cortinarius_globuliformis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_niger.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.aspergillus_niger.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/clavispora_lusitaniae.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.clavispora_lusitaniae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.kluyveromyces_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/penicillium_chrysogenum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.penicillium_chrysogenum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/rhodotorula_mucilaginosa.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.rhodotorula_mucilaginosa.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/scedosporium_boydii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.scedosporium_boydii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/blastobotrys_proliferans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.blastobotrys_proliferans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/debaryomyces_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.debaryomyces_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/galactomyces_geotrichum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.galactomyces_geotrichum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kodamaea_ohmeri.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.kodamaea_ohmeri.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/meyerozyma_guilliermondii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.meyerozyma_guilliermondii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/wickerhamomyces_anomalus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.wickerhamomyces_anomalus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_mexicana.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.yamadazyma_mexicana.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_scolyti.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.yamadazyma_scolyti.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yarrowia_lipolytica.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.yarrowia_lipolytica.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zygoascus_hellenicus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.zygoascus_hellenicus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_flavus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.aspergillus_flavus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cryptococcus_zero.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.cryptococcus_zero.minimap2.paf\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.aspergillus_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_ccl067.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.diaporthe_ccl067.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.diaporthe_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl031.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.oculimacula_yallundae-ccl031.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl029.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.oculimacula_yallundae-ccl029.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/dothiorella_vidmadera.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.dothiorella_vidmadera.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/quambalaria_cyanescens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.quambalaria_cyanescens.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/entoleuca_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.entoleuca_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl060.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.asteroma_ccl060.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl068.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.asteroma_ccl068.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/saccharomyces_cerevisiae.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.saccharomyces_cerevisiae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cladophialophora_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.cladophialophora_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_albicans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_albicans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_metapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_metapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_orthopsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_orthopsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_parapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_parapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.candida_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_marxianus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.kluyveromyces_marxianus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_kudriavzevii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.pichia_kudriavzevii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_membranifaciens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_subdb/gsref.subdb.pichia_membranifaciens.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/puccinia_striiformis-tritici.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.puccinia_striiformis-tritici.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zymoseptoria_tritici.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.zymoseptoria_tritici.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pyrenophora_tritici-repentis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.pyrenophora_tritici-repentis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/fusarium_oxysporum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.fusarium_oxysporum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/tuber_brumale.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.tuber_brumale.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cortinarius_globuliformis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.cortinarius_globuliformis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_niger.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.aspergillus_niger.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/clavispora_lusitaniae.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.clavispora_lusitaniae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.kluyveromyces_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/penicillium_chrysogenum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.penicillium_chrysogenum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/rhodotorula_mucilaginosa.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.rhodotorula_mucilaginosa.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/scedosporium_boydii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.scedosporium_boydii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/blastobotrys_proliferans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.blastobotrys_proliferans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/debaryomyces_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.debaryomyces_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/galactomyces_geotrichum.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.galactomyces_geotrichum.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kodamaea_ohmeri.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.kodamaea_ohmeri.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/meyerozyma_guilliermondii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.meyerozyma_guilliermondii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/wickerhamomyces_anomalus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.wickerhamomyces_anomalus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_mexicana.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.yamadazyma_mexicana.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yamadazyma_scolyti.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.yamadazyma_scolyti.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/yarrowia_lipolytica.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.yarrowia_lipolytica.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/zygoascus_hellenicus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.zygoascus_hellenicus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_flavus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.aspergillus_flavus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cryptococcus_zero.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.cryptococcus_zero.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/aspergillus_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.aspergillus_unidentified.minimap2.paf\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_ccl067.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.diaporthe_ccl067.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/diaporthe_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.diaporthe_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl031.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.oculimacula_yallundae-ccl031.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/oculimacula_yallundae-ccl029.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.oculimacula_yallundae-ccl029.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/dothiorella_vidmadera.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.dothiorella_vidmadera.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/quambalaria_cyanescens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.quambalaria_cyanescens.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/entoleuca_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.entoleuca_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl060.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.asteroma_ccl060.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/asteroma_ccl068.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.asteroma_ccl068.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/saccharomyces_cerevisiae.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.saccharomyces_cerevisiae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/cladophialophora_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.cladophialophora_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_albicans.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_albicans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_metapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_metapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_orthopsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_orthopsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_parapsilosis.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_parapsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/candida_unidentified.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.candida_unidentified.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/kluyveromyces_marxianus.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.kluyveromyces_marxianus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_kudriavzevii.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.pichia_kudriavzevii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 6 /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/MC_READS/pichia_membranifaciens.15000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping_mock_gsref/gsref_db/gsref.db.pichia_membranifaciens.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results(fn, species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the query species: {species}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: puccinia_striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "puccinia_striiformis-tritici    0.998531\n",
      "pyrenophora_tritici-repentis    0.000467\n",
      "zymoseptoria_tritici            0.000134\n",
      "penicillium_chrysogenum         0.000134\n",
      "cortinarius_globuliformis       0.000134\n",
      "clavispora_lusitaniae           0.000134\n",
      "aspergillus_niger               0.000134\n",
      "wickerhamomyces_anomalus        0.000067\n",
      "tuber_brumale                   0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "meyerozyma_guilliermondii       0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: zymoseptoria_tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zymoseptoria_tritici             0.996068\n",
      "pyrenophora_tritici-repentis     0.001266\n",
      "penicillium_chrysogenum          0.000533\n",
      "tuber_brumale                    0.000267\n",
      "fusarium_oxysporum               0.000200\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "candida_metapsilosis             0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_unidentified             0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "diaporthe_unidentified           0.000067\n",
      "diaporthe_ccl067                 0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "kluyveromyces_marxianus          0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "aspergillus_niger                0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pyrenophora_tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pyrenophora_tritici-repentis     0.963710\n",
      "zymoseptoria_tritici             0.033360\n",
      "aspergillus_niger                0.000333\n",
      "scedosporium_boydii              0.000200\n",
      "meyerozyma_guilliermondii        0.000133\n",
      "aspergillus_unidentified         0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "fusarium_oxysporum               0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "aspergillus_flavus               0.000133\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "oculimacula_yallundae-ccl031     0.000133\n",
      "penicillium_chrysogenum          0.000133\n",
      "candida_orthopsilosis            0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "candida_albicans                 0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "diaporthe_ccl067                 0.000067\n",
      "diaporthe_unidentified           0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "candida_unidentified             0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: fusarium_oxysporum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "fusarium_oxysporum               0.996669\n",
      "scedosporium_boydii              0.000466\n",
      "pyrenophora_tritici-repentis     0.000466\n",
      "diaporthe_unidentified           0.000200\n",
      "tuber_brumale                    0.000200\n",
      "puccinia_striiformis-tritici     0.000200\n",
      "clavispora_lusitaniae            0.000200\n",
      "diaporthe_ccl067                 0.000200\n",
      "aspergillus_niger                0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "zymoseptoria_tritici             0.000133\n",
      "oculimacula_yallundae-ccl029     0.000133\n",
      "dothiorella_vidmadera            0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "candida_albicans                 0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: tuber_brumale\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "tuber_brumale                   0.997202\n",
      "zymoseptoria_tritici            0.000466\n",
      "penicillium_chrysogenum         0.000466\n",
      "kluyveromyces_marxianus         0.000200\n",
      "saccharomyces_cerevisiae        0.000200\n",
      "puccinia_striiformis-tritici    0.000133\n",
      "clavispora_lusitaniae           0.000133\n",
      "debaryomyces_unidentified       0.000133\n",
      "wickerhamomyces_anomalus        0.000133\n",
      "quambalaria_cyanescens          0.000067\n",
      "pichia_kudriavzevii             0.000067\n",
      "yamadazyma_scolyti              0.000067\n",
      "kodamaea_ohmeri                 0.000067\n",
      "yamadazyma_mexicana             0.000067\n",
      "entoleuca_unidentified          0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "cryptococcus_zero               0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "candida_unidentified            0.000067\n",
      "candida_parapsilosis            0.000067\n",
      "candida_orthopsilosis           0.000067\n",
      "aspergillus_unidentified        0.000067\n",
      "aspergillus_flavus              0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cortinarius_globuliformis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cortinarius_globuliformis        0.995337\n",
      "pyrenophora_tritici-repentis     0.000466\n",
      "penicillium_chrysogenum          0.000400\n",
      "kluyveromyces_unidentified       0.000333\n",
      "puccinia_striiformis-tritici     0.000333\n",
      "clavispora_lusitaniae            0.000266\n",
      "fusarium_oxysporum               0.000266\n",
      "kluyveromyces_marxianus          0.000200\n",
      "aspergillus_unidentified         0.000200\n",
      "tuber_brumale                    0.000200\n",
      "cryptococcus_zero                0.000200\n",
      "rhodotorula_mucilaginosa         0.000200\n",
      "debaryomyces_unidentified        0.000200\n",
      "zymoseptoria_tritici             0.000133\n",
      "zygoascus_hellenicus             0.000133\n",
      "meyerozyma_guilliermondii        0.000133\n",
      "oculimacula_yallundae-ccl031     0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "yamadazyma_mexicana              0.000133\n",
      "aspergillus_niger                0.000133\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "candida_albicans                 0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_niger\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_niger                0.843215\n",
      "aspergillus_unidentified         0.144137\n",
      "aspergillus_flavus               0.007312\n",
      "penicillium_chrysogenum          0.001120\n",
      "entoleuca_unidentified           0.000593\n",
      "yamadazyma_scolyti               0.000264\n",
      "yamadazyma_mexicana              0.000264\n",
      "scedosporium_boydii              0.000264\n",
      "cladophialophora_unidentified    0.000264\n",
      "oculimacula_yallundae-ccl029     0.000264\n",
      "zymoseptoria_tritici             0.000198\n",
      "puccinia_striiformis-tritici     0.000198\n",
      "tuber_brumale                    0.000198\n",
      "rhodotorula_mucilaginosa         0.000198\n",
      "diaporthe_unidentified           0.000132\n",
      "candida_unidentified             0.000132\n",
      "cortinarius_globuliformis        0.000132\n",
      "galactomyces_geotrichum          0.000132\n",
      "quambalaria_cyanescens           0.000066\n",
      "yarrowia_lipolytica              0.000066\n",
      "diaporthe_ccl067                 0.000066\n",
      "debaryomyces_unidentified        0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "pyrenophora_tritici-repentis     0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "kluyveromyces_marxianus          0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "candida_parapsilosis             0.000066\n",
      "candida_albicans                 0.000066\n",
      "blastobotrys_proliferans         0.000066\n",
      "asteroma_ccl068                  0.000066\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "meyerozyma_guilliermondii        0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: clavispora_lusitaniae\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "clavispora_lusitaniae           0.991673\n",
      "candida_unidentified            0.002398\n",
      "kodamaea_ohmeri                 0.000733\n",
      "candida_albicans                0.000533\n",
      "candida_orthopsilosis           0.000533\n",
      "meyerozyma_guilliermondii       0.000466\n",
      "zymoseptoria_tritici            0.000400\n",
      "yamadazyma_scolyti              0.000400\n",
      "galactomyces_geotrichum         0.000333\n",
      "candida_metapsilosis            0.000266\n",
      "candida_parapsilosis            0.000266\n",
      "debaryomyces_unidentified       0.000266\n",
      "kluyveromyces_unidentified      0.000200\n",
      "scedosporium_boydii             0.000200\n",
      "penicillium_chrysogenum         0.000200\n",
      "kluyveromyces_marxianus         0.000133\n",
      "fusarium_oxysporum              0.000133\n",
      "yamadazyma_mexicana             0.000133\n",
      "puccinia_striiformis-tritici    0.000133\n",
      "pyrenophora_tritici-repentis    0.000133\n",
      "tuber_brumale                   0.000067\n",
      "zygoascus_hellenicus            0.000067\n",
      "saccharomyces_cerevisiae        0.000067\n",
      "wickerhamomyces_anomalus        0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "yarrowia_lipolytica             0.000067\n",
      "blastobotrys_proliferans        0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kluyveromyces_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kluyveromyces_unidentified       0.488380\n",
      "kluyveromyces_marxianus          0.425123\n",
      "cryptococcus_zero                0.059735\n",
      "cortinarius_globuliformis        0.004610\n",
      "rhodotorula_mucilaginosa         0.002625\n",
      "entoleuca_unidentified           0.002049\n",
      "quambalaria_cyanescens           0.001473\n",
      "meyerozyma_guilliermondii        0.001409\n",
      "candida_unidentified             0.001345\n",
      "saccharomyces_cerevisiae         0.001280\n",
      "wickerhamomyces_anomalus         0.000960\n",
      "yarrowia_lipolytica              0.000960\n",
      "oculimacula_yallundae-ccl029     0.000960\n",
      "candida_metapsilosis             0.000768\n",
      "candida_orthopsilosis            0.000704\n",
      "clavispora_lusitaniae            0.000704\n",
      "yamadazyma_mexicana              0.000704\n",
      "zymoseptoria_tritici             0.000576\n",
      "candida_albicans                 0.000512\n",
      "debaryomyces_unidentified        0.000512\n",
      "candida_parapsilosis             0.000384\n",
      "penicillium_chrysogenum          0.000384\n",
      "puccinia_striiformis-tritici     0.000384\n",
      "aspergillus_niger                0.000320\n",
      "yamadazyma_scolyti               0.000320\n",
      "pyrenophora_tritici-repentis     0.000320\n",
      "cladophialophora_unidentified    0.000256\n",
      "asteroma_ccl068                  0.000256\n",
      "aspergillus_flavus               0.000256\n",
      "zygoascus_hellenicus             0.000256\n",
      "fusarium_oxysporum               0.000256\n",
      "scedosporium_boydii              0.000192\n",
      "tuber_brumale                    0.000192\n",
      "blastobotrys_proliferans         0.000192\n",
      "galactomyces_geotrichum          0.000192\n",
      "kodamaea_ohmeri                  0.000192\n",
      "pichia_kudriavzevii              0.000128\n",
      "pichia_membranifaciens           0.000064\n",
      "asteroma_ccl060                  0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: penicillium_chrysogenum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "penicillium_chrysogenum          0.991548\n",
      "aspergillus_flavus               0.001730\n",
      "aspergillus_niger                0.001198\n",
      "kluyveromyces_unidentified       0.000799\n",
      "aspergillus_unidentified         0.000732\n",
      "kluyveromyces_marxianus          0.000732\n",
      "tuber_brumale                    0.000532\n",
      "cortinarius_globuliformis        0.000466\n",
      "zymoseptoria_tritici             0.000466\n",
      "cladophialophora_unidentified    0.000200\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "kodamaea_ohmeri                  0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "scedosporium_boydii              0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_unidentified             0.000067\n",
      "cryptococcus_zero                0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "dothiorella_vidmadera            0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "fusarium_oxysporum               0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "pyrenophora_tritici-repentis     0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: rhodotorula_mucilaginosa\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "rhodotorula_mucilaginosa         0.997001\n",
      "penicillium_chrysogenum          0.000667\n",
      "pyrenophora_tritici-repentis     0.000467\n",
      "aspergillus_niger                0.000467\n",
      "cortinarius_globuliformis        0.000333\n",
      "tuber_brumale                    0.000267\n",
      "scedosporium_boydii              0.000133\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "cryptococcus_zero                0.000133\n",
      "zymoseptoria_tritici             0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_albicans                 0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: scedosporium_boydii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "scedosporium_boydii              0.995737\n",
      "pyrenophora_tritici-repentis     0.000533\n",
      "fusarium_oxysporum               0.000466\n",
      "rhodotorula_mucilaginosa         0.000400\n",
      "penicillium_chrysogenum          0.000400\n",
      "aspergillus_unidentified         0.000333\n",
      "cladophialophora_unidentified    0.000200\n",
      "clavispora_lusitaniae            0.000200\n",
      "zymoseptoria_tritici             0.000133\n",
      "diaporthe_unidentified           0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "cortinarius_globuliformis        0.000133\n",
      "debaryomyces_unidentified        0.000133\n",
      "aspergillus_niger                0.000133\n",
      "tuber_brumale                    0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "cryptococcus_zero                0.000067\n",
      "kluyveromyces_marxianus          0.000067\n",
      "meyerozyma_guilliermondii        0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "candida_unidentified             0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "dothiorella_vidmadera            0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: blastobotrys_proliferans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "blastobotrys_proliferans         0.996736\n",
      "meyerozyma_guilliermondii        0.000999\n",
      "debaryomyces_unidentified        0.000533\n",
      "candida_albicans                 0.000200\n",
      "candida_orthopsilosis            0.000200\n",
      "yamadazyma_scolyti               0.000133\n",
      "candida_metapsilosis             0.000133\n",
      "candida_parapsilosis             0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "tuber_brumale                    0.000133\n",
      "wickerhamomyces_anomalus         0.000133\n",
      "aspergillus_flavus               0.000133\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_unidentified             0.000067\n",
      "diaporthe_unidentified           0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: debaryomyces_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "debaryomyces_unidentified       0.989691\n",
      "clavispora_lusitaniae           0.004789\n",
      "meyerozyma_guilliermondii       0.001530\n",
      "yamadazyma_mexicana             0.000931\n",
      "yamadazyma_scolyti              0.000399\n",
      "wickerhamomyces_anomalus        0.000333\n",
      "candida_unidentified            0.000266\n",
      "kluyveromyces_unidentified      0.000266\n",
      "kluyveromyces_marxianus         0.000200\n",
      "blastobotrys_proliferans        0.000200\n",
      "candida_parapsilosis            0.000200\n",
      "candida_metapsilosis            0.000133\n",
      "candida_orthopsilosis           0.000133\n",
      "aspergillus_flavus              0.000133\n",
      "galactomyces_geotrichum         0.000133\n",
      "saccharomyces_cerevisiae        0.000133\n",
      "zygoascus_hellenicus            0.000067\n",
      "kodamaea_ohmeri                 0.000067\n",
      "oculimacula_yallundae-ccl031    0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "candida_albicans                0.000067\n",
      "yarrowia_lipolytica             0.000067\n",
      "aspergillus_unidentified        0.000067\n",
      "zymoseptoria_tritici            0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: galactomyces_geotrichum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "galactomyces_geotrichum          0.996215\n",
      "kodamaea_ohmeri                  0.000744\n",
      "clavispora_lusitaniae            0.000541\n",
      "blastobotrys_proliferans         0.000473\n",
      "yarrowia_lipolytica              0.000406\n",
      "aspergillus_flavus               0.000338\n",
      "zygoascus_hellenicus             0.000135\n",
      "debaryomyces_unidentified        0.000135\n",
      "yamadazyma_mexicana              0.000135\n",
      "scedosporium_boydii              0.000068\n",
      "saccharomyces_cerevisiae         0.000068\n",
      "penicillium_chrysogenum          0.000068\n",
      "oculimacula_yallundae-ccl031     0.000068\n",
      "wickerhamomyces_anomalus         0.000068\n",
      "fusarium_oxysporum               0.000068\n",
      "dothiorella_vidmadera            0.000068\n",
      "cladophialophora_unidentified    0.000068\n",
      "candida_orthopsilosis            0.000068\n",
      "candida_metapsilosis             0.000068\n",
      "yamadazyma_scolyti               0.000068\n",
      "asteroma_ccl060                  0.000068\n",
      "zymoseptoria_tritici             0.000068\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kodamaea_ohmeri\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kodamaea_ohmeri                 0.998734\n",
      "yarrowia_lipolytica             0.000133\n",
      "clavispora_lusitaniae           0.000133\n",
      "candida_orthopsilosis           0.000133\n",
      "candida_albicans                0.000133\n",
      "yamadazyma_scolyti              0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "quambalaria_cyanescens          0.000067\n",
      "pichia_kudriavzevii             0.000067\n",
      "oculimacula_yallundae-ccl029    0.000067\n",
      "meyerozyma_guilliermondii       0.000067\n",
      "kluyveromyces_unidentified      0.000067\n",
      "entoleuca_unidentified          0.000067\n",
      "debaryomyces_unidentified       0.000067\n",
      "cortinarius_globuliformis       0.000067\n",
      "candida_metapsilosis            0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: meyerozyma_guilliermondii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "meyerozyma_guilliermondii       0.994081\n",
      "debaryomyces_unidentified       0.000931\n",
      "yamadazyma_mexicana             0.000665\n",
      "yamadazyma_scolyti              0.000532\n",
      "candida_orthopsilosis           0.000399\n",
      "wickerhamomyces_anomalus        0.000333\n",
      "candida_metapsilosis            0.000333\n",
      "zygoascus_hellenicus            0.000266\n",
      "candida_albicans                0.000266\n",
      "candida_unidentified            0.000266\n",
      "kodamaea_ohmeri                 0.000266\n",
      "kluyveromyces_unidentified      0.000200\n",
      "candida_parapsilosis            0.000200\n",
      "scedosporium_boydii             0.000200\n",
      "pichia_kudriavzevii             0.000133\n",
      "yarrowia_lipolytica             0.000133\n",
      "kluyveromyces_marxianus         0.000133\n",
      "galactomyces_geotrichum         0.000133\n",
      "aspergillus_flavus              0.000133\n",
      "puccinia_striiformis-tritici    0.000067\n",
      "pyrenophora_tritici-repentis    0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "saccharomyces_cerevisiae        0.000067\n",
      "asteroma_ccl068                 0.000067\n",
      "asteroma_ccl060                 0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: wickerhamomyces_anomalus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "wickerhamomyces_anomalus         0.995606\n",
      "meyerozyma_guilliermondii        0.000732\n",
      "yamadazyma_mexicana              0.000732\n",
      "debaryomyces_unidentified        0.000666\n",
      "aspergillus_flavus               0.000399\n",
      "blastobotrys_proliferans         0.000333\n",
      "candida_parapsilosis             0.000266\n",
      "zygoascus_hellenicus             0.000266\n",
      "candida_unidentified             0.000133\n",
      "yamadazyma_scolyti               0.000133\n",
      "kluyveromyces_marxianus          0.000133\n",
      "kodamaea_ohmeri                  0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yamadazyma_mexicana\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yamadazyma_mexicana              0.984399\n",
      "yamadazyma_scolyti               0.009626\n",
      "debaryomyces_unidentified        0.001062\n",
      "meyerozyma_guilliermondii        0.000797\n",
      "kodamaea_ohmeri                  0.000730\n",
      "candida_metapsilosis             0.000465\n",
      "aspergillus_flavus               0.000332\n",
      "candida_orthopsilosis            0.000332\n",
      "candida_unidentified             0.000199\n",
      "zygoascus_hellenicus             0.000199\n",
      "wickerhamomyces_anomalus         0.000199\n",
      "penicillium_chrysogenum          0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "candida_albicans                 0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "tuber_brumale                    0.000133\n",
      "candida_parapsilosis             0.000133\n",
      "oculimacula_yallundae-ccl029     0.000133\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "puccinia_striiformis-tritici     0.000066\n",
      "diaporthe_ccl067                 0.000066\n",
      "entoleuca_unidentified           0.000066\n",
      "pichia_membranifaciens           0.000066\n",
      "pichia_kudriavzevii              0.000066\n",
      "kluyveromyces_marxianus          0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yamadazyma_scolyti\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yamadazyma_scolyti               0.991761\n",
      "yamadazyma_mexicana              0.004120\n",
      "debaryomyces_unidentified        0.000598\n",
      "meyerozyma_guilliermondii        0.000532\n",
      "candida_unidentified             0.000465\n",
      "candida_metapsilosis             0.000399\n",
      "candida_parapsilosis             0.000399\n",
      "candida_orthopsilosis            0.000332\n",
      "candida_albicans                 0.000266\n",
      "zygoascus_hellenicus             0.000133\n",
      "pyrenophora_tritici-repentis     0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "galactomyces_geotrichum          0.000133\n",
      "blastobotrys_proliferans         0.000066\n",
      "entoleuca_unidentified           0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "cortinarius_globuliformis        0.000066\n",
      "yarrowia_lipolytica              0.000066\n",
      "kluyveromyces_marxianus          0.000066\n",
      "pichia_membranifaciens           0.000066\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "aspergillus_flavus               0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yarrowia_lipolytica\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yarrowia_lipolytica         0.999467\n",
      "yamadazyma_scolyti          0.000133\n",
      "aspergillus_flavus          0.000133\n",
      "yamadazyma_mexicana         0.000067\n",
      "wickerhamomyces_anomalus    0.000067\n",
      "galactomyces_geotrichum     0.000067\n",
      "blastobotrys_proliferans    0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: zygoascus_hellenicus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zygoascus_hellenicus          0.998268\n",
      "yamadazyma_mexicana           0.000400\n",
      "blastobotrys_proliferans      0.000267\n",
      "yamadazyma_scolyti            0.000133\n",
      "meyerozyma_guilliermondii     0.000133\n",
      "galactomyces_geotrichum       0.000133\n",
      "zymoseptoria_tritici          0.000067\n",
      "wickerhamomyces_anomalus      0.000067\n",
      "kodamaea_ohmeri               0.000067\n",
      "kluyveromyces_unidentified    0.000067\n",
      "diaporthe_unidentified        0.000067\n",
      "candida_orthopsilosis         0.000067\n",
      "candida_metapsilosis          0.000067\n",
      "candida_albicans              0.000067\n",
      "asteroma_ccl060               0.000067\n",
      "aspergillus_flavus            0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_flavus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_flavus               0.994210\n",
      "aspergillus_niger                0.002263\n",
      "aspergillus_unidentified         0.001797\n",
      "penicillium_chrysogenum          0.000466\n",
      "yamadazyma_scolyti               0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "cryptococcus_zero                0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_unidentified             0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "zymoseptoria_tritici             0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "pyrenophora_tritici-repentis     0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cryptococcus_zero\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.995871\n",
      "quambalaria_cyanescens           0.000466\n",
      "cortinarius_globuliformis        0.000333\n",
      "entoleuca_unidentified           0.000266\n",
      "aspergillus_niger                0.000266\n",
      "diaporthe_ccl067                 0.000266\n",
      "yamadazyma_mexicana              0.000200\n",
      "aspergillus_unidentified         0.000200\n",
      "diaporthe_unidentified           0.000200\n",
      "zymoseptoria_tritici             0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "debaryomyces_unidentified        0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "kluyveromyces_marxianus          0.000133\n",
      "meyerozyma_guilliermondii        0.000133\n",
      "rhodotorula_mucilaginosa         0.000133\n",
      "wickerhamomyces_anomalus         0.000133\n",
      "yarrowia_lipolytica              0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "clavispora_lusitaniae            0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_unidentified         0.802217\n",
      "aspergillus_niger                0.187690\n",
      "aspergillus_flavus               0.004486\n",
      "penicillium_chrysogenum          0.000726\n",
      "entoleuca_unidentified           0.000528\n",
      "oculimacula_yallundae-ccl029     0.000462\n",
      "dothiorella_vidmadera            0.000396\n",
      "oculimacula_yallundae-ccl031     0.000396\n",
      "tuber_brumale                    0.000330\n",
      "asteroma_ccl068                  0.000330\n",
      "zymoseptoria_tritici             0.000330\n",
      "quambalaria_cyanescens           0.000198\n",
      "diaporthe_unidentified           0.000198\n",
      "diaporthe_ccl067                 0.000198\n",
      "blastobotrys_proliferans         0.000198\n",
      "kluyveromyces_marxianus          0.000198\n",
      "asteroma_ccl060                  0.000198\n",
      "saccharomyces_cerevisiae         0.000198\n",
      "fusarium_oxysporum               0.000132\n",
      "galactomyces_geotrichum          0.000132\n",
      "debaryomyces_unidentified        0.000066\n",
      "kluyveromyces_unidentified       0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "candida_orthopsilosis            0.000066\n",
      "scedosporium_boydii              0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: diaporthe_ccl067\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "diaporthe_ccl067                 0.625385\n",
      "asteroma_ccl068                  0.367346\n",
      "diaporthe_unidentified           0.001375\n",
      "asteroma_ccl060                  0.001244\n",
      "entoleuca_unidentified           0.001179\n",
      "oculimacula_yallundae-ccl029     0.000655\n",
      "cryptococcus_zero                0.000589\n",
      "pyrenophora_tritici-repentis     0.000327\n",
      "oculimacula_yallundae-ccl031     0.000262\n",
      "dothiorella_vidmadera            0.000262\n",
      "galactomyces_geotrichum          0.000196\n",
      "quambalaria_cyanescens           0.000196\n",
      "fusarium_oxysporum               0.000131\n",
      "tuber_brumale                    0.000131\n",
      "cortinarius_globuliformis        0.000131\n",
      "aspergillus_unidentified         0.000065\n",
      "cladophialophora_unidentified    0.000065\n",
      "candida_albicans                 0.000065\n",
      "zymoseptoria_tritici             0.000065\n",
      "wickerhamomyces_anomalus         0.000065\n",
      "kluyveromyces_marxianus          0.000065\n",
      "kluyveromyces_unidentified       0.000065\n",
      "puccinia_striiformis-tritici     0.000065\n",
      "aspergillus_flavus               0.000065\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: diaporthe_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "diaporthe_unidentified          0.702311\n",
      "asteroma_ccl060                 0.288997\n",
      "entoleuca_unidentified          0.001646\n",
      "asteroma_ccl068                 0.001119\n",
      "diaporthe_ccl067                0.001119\n",
      "kodamaea_ohmeri                 0.000922\n",
      "fusarium_oxysporum              0.000593\n",
      "oculimacula_yallundae-ccl029    0.000461\n",
      "dothiorella_vidmadera           0.000329\n",
      "tuber_brumale                   0.000329\n",
      "cryptococcus_zero               0.000329\n",
      "scedosporium_boydii             0.000263\n",
      "quambalaria_cyanescens          0.000263\n",
      "oculimacula_yallundae-ccl031    0.000263\n",
      "aspergillus_unidentified        0.000198\n",
      "kluyveromyces_unidentified      0.000132\n",
      "yamadazyma_scolyti              0.000132\n",
      "saccharomyces_cerevisiae        0.000132\n",
      "puccinia_striiformis-tritici    0.000066\n",
      "pyrenophora_tritici-repentis    0.000066\n",
      "cortinarius_globuliformis       0.000066\n",
      "clavispora_lusitaniae           0.000066\n",
      "candida_metapsilosis            0.000066\n",
      "rhodotorula_mucilaginosa        0.000066\n",
      "zymoseptoria_tritici            0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: oculimacula_yallundae-ccl031\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "oculimacula_yallundae-ccl031     0.963964\n",
      "oculimacula_yallundae-ccl029     0.027756\n",
      "quambalaria_cyanescens           0.000795\n",
      "entoleuca_unidentified           0.000530\n",
      "dothiorella_vidmadera            0.000530\n",
      "asteroma_ccl060                  0.000530\n",
      "yamadazyma_mexicana              0.000464\n",
      "tuber_brumale                    0.000464\n",
      "pyrenophora_tritici-repentis     0.000397\n",
      "asteroma_ccl068                  0.000331\n",
      "clavispora_lusitaniae            0.000331\n",
      "zygoascus_hellenicus             0.000331\n",
      "fusarium_oxysporum               0.000265\n",
      "cladophialophora_unidentified    0.000265\n",
      "cryptococcus_zero                0.000265\n",
      "diaporthe_unidentified           0.000199\n",
      "diaporthe_ccl067                 0.000199\n",
      "aspergillus_niger                0.000199\n",
      "scedosporium_boydii              0.000199\n",
      "yarrowia_lipolytica              0.000199\n",
      "galactomyces_geotrichum          0.000132\n",
      "aspergillus_unidentified         0.000132\n",
      "blastobotrys_proliferans         0.000132\n",
      "candida_albicans                 0.000132\n",
      "kluyveromyces_marxianus          0.000132\n",
      "zymoseptoria_tritici             0.000132\n",
      "cortinarius_globuliformis        0.000132\n",
      "penicillium_chrysogenum          0.000132\n",
      "meyerozyma_guilliermondii        0.000132\n",
      "wickerhamomyces_anomalus         0.000066\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "candida_parapsilosis             0.000066\n",
      "candida_orthopsilosis            0.000066\n",
      "candida_metapsilosis             0.000066\n",
      "rhodotorula_mucilaginosa         0.000066\n",
      "pichia_kudriavzevii              0.000066\n",
      "yamadazyma_scolyti               0.000066\n",
      "candida_unidentified             0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: oculimacula_yallundae-ccl029\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "oculimacula_yallundae-ccl029     0.984795\n",
      "oculimacula_yallundae-ccl031     0.010823\n",
      "entoleuca_unidentified           0.000465\n",
      "aspergillus_unidentified         0.000398\n",
      "dothiorella_vidmadera            0.000332\n",
      "cryptococcus_zero                0.000266\n",
      "saccharomyces_cerevisiae         0.000266\n",
      "diaporthe_ccl067                 0.000266\n",
      "fusarium_oxysporum               0.000199\n",
      "diaporthe_unidentified           0.000199\n",
      "wickerhamomyces_anomalus         0.000199\n",
      "aspergillus_niger                0.000199\n",
      "pyrenophora_tritici-repentis     0.000199\n",
      "candida_unidentified             0.000133\n",
      "tuber_brumale                    0.000133\n",
      "quambalaria_cyanescens           0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "candida_orthopsilosis            0.000066\n",
      "asteroma_ccl060                  0.000066\n",
      "candida_albicans                 0.000066\n",
      "candida_metapsilosis             0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "candida_parapsilosis             0.000066\n",
      "meyerozyma_guilliermondii        0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "cortinarius_globuliformis        0.000066\n",
      "puccinia_striiformis-tritici     0.000066\n",
      "yamadazyma_scolyti               0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: dothiorella_vidmadera\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "dothiorella_vidmadera            0.993807\n",
      "zymoseptoria_tritici             0.001199\n",
      "diaporthe_ccl067                 0.000533\n",
      "oculimacula_yallundae-ccl031     0.000466\n",
      "entoleuca_unidentified           0.000400\n",
      "oculimacula_yallundae-ccl029     0.000333\n",
      "diaporthe_unidentified           0.000266\n",
      "quambalaria_cyanescens           0.000266\n",
      "cryptococcus_zero                0.000266\n",
      "pyrenophora_tritici-repentis     0.000200\n",
      "aspergillus_niger                0.000200\n",
      "aspergillus_unidentified         0.000200\n",
      "asteroma_ccl060                  0.000200\n",
      "clavispora_lusitaniae            0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "fusarium_oxysporum               0.000133\n",
      "candida_unidentified             0.000133\n",
      "rhodotorula_mucilaginosa         0.000133\n",
      "candida_parapsilosis             0.000133\n",
      "candida_metapsilosis             0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "tuber_brumale                    0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: quambalaria_cyanescens\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "quambalaria_cyanescens          0.997399\n",
      "entoleuca_unidentified          0.000467\n",
      "dothiorella_vidmadera           0.000400\n",
      "oculimacula_yallundae-ccl031    0.000333\n",
      "asteroma_ccl068                 0.000267\n",
      "diaporthe_unidentified          0.000200\n",
      "cryptococcus_zero               0.000200\n",
      "rhodotorula_mucilaginosa        0.000133\n",
      "diaporthe_ccl067                0.000133\n",
      "zymoseptoria_tritici            0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "saccharomyces_cerevisiae        0.000067\n",
      "oculimacula_yallundae-ccl029    0.000067\n",
      "cortinarius_globuliformis       0.000067\n",
      "asteroma_ccl060                 0.000067\n",
      "aspergillus_niger               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: entoleuca_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "entoleuca_unidentified          0.996870\n",
      "oculimacula_yallundae-ccl031    0.000466\n",
      "diaporthe_ccl067                0.000400\n",
      "diaporthe_unidentified          0.000266\n",
      "oculimacula_yallundae-ccl029    0.000266\n",
      "asteroma_ccl068                 0.000200\n",
      "fusarium_oxysporum              0.000200\n",
      "dothiorella_vidmadera           0.000200\n",
      "cryptococcus_zero               0.000200\n",
      "aspergillus_unidentified        0.000133\n",
      "asteroma_ccl060                 0.000133\n",
      "candida_orthopsilosis           0.000067\n",
      "candida_albicans                0.000067\n",
      "candida_metapsilosis            0.000067\n",
      "yamadazyma_scolyti              0.000067\n",
      "clavispora_lusitaniae           0.000067\n",
      "yamadazyma_mexicana             0.000067\n",
      "meyerozyma_guilliermondii       0.000067\n",
      "quambalaria_cyanescens          0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "aspergillus_niger               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: asteroma_ccl060\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "asteroma_ccl060                 0.616911\n",
      "diaporthe_unidentified          0.375074\n",
      "diaporthe_ccl067                0.001314\n",
      "asteroma_ccl068                 0.000985\n",
      "entoleuca_unidentified          0.000985\n",
      "oculimacula_yallundae-ccl029    0.000920\n",
      "rhodotorula_mucilaginosa        0.000328\n",
      "zymoseptoria_tritici            0.000263\n",
      "cortinarius_globuliformis       0.000263\n",
      "fusarium_oxysporum              0.000263\n",
      "quambalaria_cyanescens          0.000263\n",
      "tuber_brumale                   0.000263\n",
      "pyrenophora_tritici-repentis    0.000263\n",
      "aspergillus_unidentified        0.000197\n",
      "clavispora_lusitaniae           0.000197\n",
      "scedosporium_boydii             0.000197\n",
      "oculimacula_yallundae-ccl031    0.000197\n",
      "penicillium_chrysogenum         0.000131\n",
      "cryptococcus_zero               0.000131\n",
      "dothiorella_vidmadera           0.000131\n",
      "zygoascus_hellenicus            0.000131\n",
      "galactomyces_geotrichum         0.000131\n",
      "aspergillus_flavus              0.000131\n",
      "puccinia_striiformis-tritici    0.000066\n",
      "candida_parapsilosis            0.000066\n",
      "candida_metapsilosis            0.000066\n",
      "yamadazyma_mexicana             0.000066\n",
      "pichia_kudriavzevii             0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: asteroma_ccl068\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "asteroma_ccl068                  0.711335\n",
      "diaporthe_ccl067                 0.278568\n",
      "entoleuca_unidentified           0.001573\n",
      "asteroma_ccl060                  0.001311\n",
      "oculimacula_yallundae-ccl029     0.001115\n",
      "diaporthe_unidentified           0.001049\n",
      "aspergillus_unidentified         0.000524\n",
      "quambalaria_cyanescens           0.000459\n",
      "oculimacula_yallundae-ccl031     0.000459\n",
      "cryptococcus_zero                0.000393\n",
      "fusarium_oxysporum               0.000393\n",
      "dothiorella_vidmadera            0.000328\n",
      "pichia_membranifaciens           0.000262\n",
      "galactomyces_geotrichum          0.000262\n",
      "aspergillus_niger                0.000262\n",
      "rhodotorula_mucilaginosa         0.000197\n",
      "saccharomyces_cerevisiae         0.000197\n",
      "scedosporium_boydii              0.000197\n",
      "pyrenophora_tritici-repentis     0.000131\n",
      "tuber_brumale                    0.000131\n",
      "cladophialophora_unidentified    0.000131\n",
      "candida_parapsilosis             0.000131\n",
      "candida_orthopsilosis            0.000066\n",
      "candida_albicans                 0.000066\n",
      "zymoseptoria_tritici             0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "zygoascus_hellenicus             0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "penicillium_chrysogenum          0.000066\n",
      "wickerhamomyces_anomalus         0.000066\n",
      "aspergillus_flavus               0.000066\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: saccharomyces_cerevisiae\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "saccharomyces_cerevisiae         0.992092\n",
      "kluyveromyces_unidentified       0.002259\n",
      "kluyveromyces_marxianus          0.002126\n",
      "debaryomyces_unidentified        0.000532\n",
      "wickerhamomyces_anomalus         0.000465\n",
      "candida_metapsilosis             0.000332\n",
      "yamadazyma_mexicana              0.000266\n",
      "candida_orthopsilosis            0.000266\n",
      "pichia_membranifaciens           0.000199\n",
      "meyerozyma_guilliermondii        0.000199\n",
      "candida_parapsilosis             0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "zygoascus_hellenicus             0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "dothiorella_vidmadera            0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "yamadazyma_scolyti               0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "candida_unidentified             0.000066\n",
      "pichia_kudriavzevii              0.000066\n",
      "rhodotorula_mucilaginosa         0.000066\n",
      "tuber_brumale                    0.000066\n",
      "candida_albicans                 0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cladophialophora_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cladophialophora_unidentified    0.997734\n",
      "zymoseptoria_tritici             0.000267\n",
      "candida_orthopsilosis            0.000267\n",
      "candida_albicans                 0.000200\n",
      "kluyveromyces_unidentified       0.000200\n",
      "fusarium_oxysporum               0.000133\n",
      "candida_metapsilosis             0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "candida_parapsilosis             0.000067\n",
      "candida_unidentified             0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "tuber_brumale                    0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_albicans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_albicans                 0.495014\n",
      "candida_unidentified             0.300521\n",
      "candida_orthopsilosis            0.105192\n",
      "candida_metapsilosis             0.052242\n",
      "candida_parapsilosis             0.032555\n",
      "meyerozyma_guilliermondii        0.001544\n",
      "debaryomyces_unidentified        0.001544\n",
      "galactomyces_geotrichum          0.001222\n",
      "entoleuca_unidentified           0.000965\n",
      "yamadazyma_mexicana              0.000836\n",
      "kodamaea_ohmeri                  0.000836\n",
      "yamadazyma_scolyti               0.000772\n",
      "oculimacula_yallundae-ccl029     0.000708\n",
      "kluyveromyces_unidentified       0.000708\n",
      "tuber_brumale                    0.000579\n",
      "kluyveromyces_marxianus          0.000579\n",
      "clavispora_lusitaniae            0.000450\n",
      "pyrenophora_tritici-repentis     0.000450\n",
      "saccharomyces_cerevisiae         0.000450\n",
      "wickerhamomyces_anomalus         0.000257\n",
      "blastobotrys_proliferans         0.000257\n",
      "pichia_membranifaciens           0.000193\n",
      "asteroma_ccl060                  0.000193\n",
      "fusarium_oxysporum               0.000193\n",
      "quambalaria_cyanescens           0.000193\n",
      "dothiorella_vidmadera            0.000193\n",
      "diaporthe_ccl067                 0.000193\n",
      "zygoascus_hellenicus             0.000129\n",
      "diaporthe_unidentified           0.000129\n",
      "yarrowia_lipolytica              0.000129\n",
      "pichia_kudriavzevii              0.000129\n",
      "rhodotorula_mucilaginosa         0.000129\n",
      "cortinarius_globuliformis        0.000064\n",
      "cryptococcus_zero                0.000064\n",
      "cladophialophora_unidentified    0.000064\n",
      "oculimacula_yallundae-ccl031     0.000064\n",
      "penicillium_chrysogenum          0.000064\n",
      "asteroma_ccl068                  0.000064\n",
      "scedosporium_boydii              0.000064\n",
      "aspergillus_unidentified         0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_metapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_metapsilosis             0.515065\n",
      "candida_orthopsilosis            0.420290\n",
      "candida_parapsilosis             0.051487\n",
      "kodamaea_ohmeri                  0.003432\n",
      "candida_unidentified             0.001335\n",
      "oculimacula_yallundae-ccl029     0.001017\n",
      "meyerozyma_guilliermondii        0.000826\n",
      "candida_albicans                 0.000763\n",
      "clavispora_lusitaniae            0.000572\n",
      "debaryomyces_unidentified        0.000445\n",
      "yamadazyma_scolyti               0.000381\n",
      "galactomyces_geotrichum          0.000318\n",
      "kluyveromyces_marxianus          0.000318\n",
      "tuber_brumale                    0.000318\n",
      "yarrowia_lipolytica              0.000254\n",
      "entoleuca_unidentified           0.000254\n",
      "cladophialophora_unidentified    0.000254\n",
      "quambalaria_cyanescens           0.000254\n",
      "wickerhamomyces_anomalus         0.000254\n",
      "blastobotrys_proliferans         0.000254\n",
      "saccharomyces_cerevisiae         0.000191\n",
      "fusarium_oxysporum               0.000191\n",
      "kluyveromyces_unidentified       0.000191\n",
      "zygoascus_hellenicus             0.000127\n",
      "pichia_kudriavzevii              0.000127\n",
      "pichia_membranifaciens           0.000127\n",
      "pyrenophora_tritici-repentis     0.000127\n",
      "rhodotorula_mucilaginosa         0.000127\n",
      "yamadazyma_mexicana              0.000127\n",
      "aspergillus_flavus               0.000127\n",
      "diaporthe_unidentified           0.000064\n",
      "cryptococcus_zero                0.000064\n",
      "penicillium_chrysogenum          0.000064\n",
      "puccinia_striiformis-tritici     0.000064\n",
      "scedosporium_boydii              0.000064\n",
      "asteroma_ccl068                  0.000064\n",
      "aspergillus_niger                0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_orthopsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_orthopsilosis            0.735129\n",
      "candida_metapsilosis             0.202352\n",
      "candida_parapsilosis             0.041848\n",
      "meyerozyma_guilliermondii        0.009988\n",
      "candida_unidentified             0.001644\n",
      "candida_albicans                 0.000948\n",
      "yamadazyma_mexicana              0.000632\n",
      "tuber_brumale                    0.000632\n",
      "entoleuca_unidentified           0.000569\n",
      "saccharomyces_cerevisiae         0.000569\n",
      "debaryomyces_unidentified        0.000569\n",
      "oculimacula_yallundae-ccl029     0.000443\n",
      "yamadazyma_scolyti               0.000379\n",
      "kodamaea_ohmeri                  0.000379\n",
      "galactomyces_geotrichum          0.000316\n",
      "blastobotrys_proliferans         0.000253\n",
      "wickerhamomyces_anomalus         0.000253\n",
      "clavispora_lusitaniae            0.000253\n",
      "kluyveromyces_marxianus          0.000253\n",
      "zygoascus_hellenicus             0.000253\n",
      "cladophialophora_unidentified    0.000190\n",
      "pichia_membranifaciens           0.000190\n",
      "cryptococcus_zero                0.000190\n",
      "aspergillus_unidentified         0.000126\n",
      "aspergillus_niger                0.000126\n",
      "aspergillus_flavus               0.000126\n",
      "dothiorella_vidmadera            0.000126\n",
      "fusarium_oxysporum               0.000126\n",
      "kluyveromyces_unidentified       0.000126\n",
      "oculimacula_yallundae-ccl031     0.000126\n",
      "pyrenophora_tritici-repentis     0.000126\n",
      "scedosporium_boydii              0.000126\n",
      "yarrowia_lipolytica              0.000126\n",
      "diaporthe_unidentified           0.000063\n",
      "cortinarius_globuliformis        0.000063\n",
      "pichia_kudriavzevii              0.000063\n",
      "quambalaria_cyanescens           0.000063\n",
      "rhodotorula_mucilaginosa         0.000063\n",
      "asteroma_ccl068                  0.000063\n",
      "asteroma_ccl060                  0.000063\n",
      "zymoseptoria_tritici             0.000063\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_parapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_parapsilosis             0.896477\n",
      "candida_orthopsilosis            0.052642\n",
      "candida_metapsilosis             0.041748\n",
      "candida_unidentified             0.001239\n",
      "oculimacula_yallundae-ccl029     0.000848\n",
      "kodamaea_ohmeri                  0.000718\n",
      "meyerozyma_guilliermondii        0.000718\n",
      "candida_albicans                 0.000718\n",
      "tuber_brumale                    0.000457\n",
      "entoleuca_unidentified           0.000326\n",
      "kluyveromyces_unidentified       0.000326\n",
      "debaryomyces_unidentified        0.000326\n",
      "pichia_membranifaciens           0.000261\n",
      "blastobotrys_proliferans         0.000261\n",
      "yamadazyma_mexicana              0.000261\n",
      "quambalaria_cyanescens           0.000196\n",
      "saccharomyces_cerevisiae         0.000196\n",
      "kluyveromyces_marxianus          0.000196\n",
      "yamadazyma_scolyti               0.000196\n",
      "asteroma_ccl060                  0.000196\n",
      "galactomyces_geotrichum          0.000196\n",
      "cryptococcus_zero                0.000130\n",
      "clavispora_lusitaniae            0.000130\n",
      "cortinarius_globuliformis        0.000130\n",
      "aspergillus_flavus               0.000130\n",
      "zygoascus_hellenicus             0.000130\n",
      "wickerhamomyces_anomalus         0.000130\n",
      "yarrowia_lipolytica              0.000130\n",
      "diaporthe_ccl067                 0.000065\n",
      "diaporthe_unidentified           0.000065\n",
      "cladophialophora_unidentified    0.000065\n",
      "fusarium_oxysporum               0.000065\n",
      "pichia_kudriavzevii              0.000065\n",
      "puccinia_striiformis-tritici     0.000065\n",
      "pyrenophora_tritici-repentis     0.000065\n",
      "aspergillus_niger                0.000065\n",
      "zymoseptoria_tritici             0.000065\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_unidentified             0.882645\n",
      "candida_albicans                 0.100478\n",
      "candida_orthopsilosis            0.003140\n",
      "candida_metapsilosis             0.002355\n",
      "candida_parapsilosis             0.001635\n",
      "oculimacula_yallundae-ccl029     0.001047\n",
      "entoleuca_unidentified           0.001047\n",
      "meyerozyma_guilliermondii        0.000981\n",
      "kodamaea_ohmeri                  0.000654\n",
      "yamadazyma_scolyti               0.000654\n",
      "debaryomyces_unidentified        0.000589\n",
      "yamadazyma_mexicana              0.000589\n",
      "wickerhamomyces_anomalus         0.000589\n",
      "saccharomyces_cerevisiae         0.000589\n",
      "kluyveromyces_unidentified       0.000392\n",
      "clavispora_lusitaniae            0.000327\n",
      "tuber_brumale                    0.000262\n",
      "yarrowia_lipolytica              0.000262\n",
      "blastobotrys_proliferans         0.000196\n",
      "kluyveromyces_marxianus          0.000131\n",
      "zygoascus_hellenicus             0.000131\n",
      "pichia_membranifaciens           0.000131\n",
      "quambalaria_cyanescens           0.000131\n",
      "rhodotorula_mucilaginosa         0.000131\n",
      "cryptococcus_zero                0.000131\n",
      "galactomyces_geotrichum          0.000131\n",
      "aspergillus_unidentified         0.000065\n",
      "zymoseptoria_tritici             0.000065\n",
      "cladophialophora_unidentified    0.000065\n",
      "diaporthe_ccl067                 0.000065\n",
      "diaporthe_unidentified           0.000065\n",
      "dothiorella_vidmadera            0.000065\n",
      "fusarium_oxysporum               0.000065\n",
      "penicillium_chrysogenum          0.000065\n",
      "scedosporium_boydii              0.000065\n",
      "aspergillus_flavus               0.000065\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kluyveromyces_marxianus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kluyveromyces_marxianus          0.740035\n",
      "kluyveromyces_unidentified       0.249117\n",
      "candida_metapsilosis             0.000963\n",
      "oculimacula_yallundae-ccl029     0.000963\n",
      "entoleuca_unidentified           0.000706\n",
      "wickerhamomyces_anomalus         0.000706\n",
      "yamadazyma_scolyti               0.000578\n",
      "clavispora_lusitaniae            0.000514\n",
      "debaryomyces_unidentified        0.000514\n",
      "saccharomyces_cerevisiae         0.000514\n",
      "meyerozyma_guilliermondii        0.000514\n",
      "yamadazyma_mexicana              0.000449\n",
      "candida_orthopsilosis            0.000449\n",
      "candida_albicans                 0.000321\n",
      "yarrowia_lipolytica              0.000321\n",
      "candida_unidentified             0.000321\n",
      "candida_parapsilosis             0.000257\n",
      "tuber_brumale                    0.000257\n",
      "oculimacula_yallundae-ccl031     0.000257\n",
      "pichia_kudriavzevii              0.000257\n",
      "galactomyces_geotrichum          0.000193\n",
      "rhodotorula_mucilaginosa         0.000193\n",
      "cladophialophora_unidentified    0.000193\n",
      "kodamaea_ohmeri                  0.000193\n",
      "blastobotrys_proliferans         0.000128\n",
      "cryptococcus_zero                0.000128\n",
      "cortinarius_globuliformis        0.000128\n",
      "quambalaria_cyanescens           0.000128\n",
      "asteroma_ccl068                  0.000128\n",
      "pichia_membranifaciens           0.000128\n",
      "asteroma_ccl060                  0.000064\n",
      "zymoseptoria_tritici             0.000064\n",
      "dothiorella_vidmadera            0.000064\n",
      "zygoascus_hellenicus             0.000064\n",
      "puccinia_striiformis-tritici     0.000064\n",
      "scedosporium_boydii              0.000064\n",
      "aspergillus_flavus               0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_kudriavzevii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_kudriavzevii              0.996867\n",
      "pichia_membranifaciens           0.001533\n",
      "candida_orthopsilosis            0.000333\n",
      "candida_unidentified             0.000133\n",
      "tuber_brumale                    0.000133\n",
      "yamadazyma_scolyti               0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "yarrowia_lipolytica              0.000067\n",
      "candida_albicans                 0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "pyrenophora_tritici-repentis     0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_membranifaciens\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_membranifaciens           0.994269\n",
      "pichia_kudriavzevii              0.001933\n",
      "candida_orthopsilosis            0.000400\n",
      "meyerozyma_guilliermondii        0.000400\n",
      "kodamaea_ohmeri                  0.000333\n",
      "candida_metapsilosis             0.000333\n",
      "candida_parapsilosis             0.000267\n",
      "saccharomyces_cerevisiae         0.000200\n",
      "clavispora_lusitaniae            0.000200\n",
      "kluyveromyces_unidentified       0.000200\n",
      "galactomyces_geotrichum          0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "debaryomyces_unidentified        0.000133\n",
      "zymoseptoria_tritici             0.000133\n",
      "kluyveromyces_marxianus          0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "quambalaria_cyanescens           0.000133\n",
      "wickerhamomyces_anomalus         0.000133\n",
      "candida_albicans                 0.000133\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "candida_unidentified             0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "tuber_brumale                    0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for species, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: puccinia_striiformis-tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "puccinia_striiformis-tritici    0.998531\n",
      "pyrenophora_tritici-repentis    0.000467\n",
      "zymoseptoria_tritici            0.000134\n",
      "penicillium_chrysogenum         0.000134\n",
      "cortinarius_globuliformis       0.000134\n",
      "clavispora_lusitaniae           0.000134\n",
      "aspergillus_niger               0.000134\n",
      "wickerhamomyces_anomalus        0.000067\n",
      "tuber_brumale                   0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "meyerozyma_guilliermondii       0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: zymoseptoria_tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zymoseptoria_tritici             0.996068\n",
      "pyrenophora_tritici-repentis     0.001266\n",
      "penicillium_chrysogenum          0.000533\n",
      "tuber_brumale                    0.000267\n",
      "fusarium_oxysporum               0.000200\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "candida_metapsilosis             0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_unidentified             0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "diaporthe_unidentified           0.000067\n",
      "diaporthe_ccl067                 0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "kluyveromyces_marxianus          0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "scedosporium_boydii              0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "aspergillus_niger                0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pyrenophora_tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pyrenophora_tritici-repentis     0.963710\n",
      "zymoseptoria_tritici             0.033360\n",
      "aspergillus_niger                0.000333\n",
      "scedosporium_boydii              0.000200\n",
      "meyerozyma_guilliermondii        0.000133\n",
      "aspergillus_unidentified         0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "fusarium_oxysporum               0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "aspergillus_flavus               0.000133\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "oculimacula_yallundae-ccl031     0.000133\n",
      "penicillium_chrysogenum          0.000133\n",
      "candida_orthopsilosis            0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "candida_albicans                 0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "diaporthe_ccl067                 0.000067\n",
      "diaporthe_unidentified           0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "candida_unidentified             0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: fusarium_oxysporum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "fusarium_oxysporum               0.996669\n",
      "scedosporium_boydii              0.000466\n",
      "pyrenophora_tritici-repentis     0.000466\n",
      "diaporthe_unidentified           0.000200\n",
      "tuber_brumale                    0.000200\n",
      "puccinia_striiformis-tritici     0.000200\n",
      "clavispora_lusitaniae            0.000200\n",
      "diaporthe_ccl067                 0.000200\n",
      "aspergillus_niger                0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "zymoseptoria_tritici             0.000133\n",
      "oculimacula_yallundae-ccl029     0.000133\n",
      "dothiorella_vidmadera            0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "candida_albicans                 0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: tuber_brumale\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "tuber_brumale                   0.997202\n",
      "zymoseptoria_tritici            0.000466\n",
      "penicillium_chrysogenum         0.000466\n",
      "kluyveromyces_marxianus         0.000200\n",
      "saccharomyces_cerevisiae        0.000200\n",
      "puccinia_striiformis-tritici    0.000133\n",
      "clavispora_lusitaniae           0.000133\n",
      "debaryomyces_unidentified       0.000133\n",
      "wickerhamomyces_anomalus        0.000133\n",
      "quambalaria_cyanescens          0.000067\n",
      "pichia_kudriavzevii             0.000067\n",
      "yamadazyma_scolyti              0.000067\n",
      "kodamaea_ohmeri                 0.000067\n",
      "yamadazyma_mexicana             0.000067\n",
      "entoleuca_unidentified          0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "cryptococcus_zero               0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "candida_unidentified            0.000067\n",
      "candida_parapsilosis            0.000067\n",
      "candida_orthopsilosis           0.000067\n",
      "aspergillus_unidentified        0.000067\n",
      "aspergillus_flavus              0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cortinarius_globuliformis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cortinarius_globuliformis        0.995337\n",
      "pyrenophora_tritici-repentis     0.000466\n",
      "penicillium_chrysogenum          0.000400\n",
      "kluyveromyces_unidentified       0.000333\n",
      "puccinia_striiformis-tritici     0.000333\n",
      "clavispora_lusitaniae            0.000266\n",
      "fusarium_oxysporum               0.000266\n",
      "kluyveromyces_marxianus          0.000200\n",
      "aspergillus_unidentified         0.000200\n",
      "tuber_brumale                    0.000200\n",
      "cryptococcus_zero                0.000200\n",
      "rhodotorula_mucilaginosa         0.000200\n",
      "debaryomyces_unidentified        0.000200\n",
      "zymoseptoria_tritici             0.000133\n",
      "zygoascus_hellenicus             0.000133\n",
      "meyerozyma_guilliermondii        0.000133\n",
      "oculimacula_yallundae-ccl031     0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "yamadazyma_mexicana              0.000133\n",
      "aspergillus_niger                0.000133\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "candida_albicans                 0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_niger\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_niger                0.843215\n",
      "aspergillus_unidentified         0.144137\n",
      "aspergillus_flavus               0.007312\n",
      "penicillium_chrysogenum          0.001120\n",
      "entoleuca_unidentified           0.000593\n",
      "yamadazyma_scolyti               0.000264\n",
      "yamadazyma_mexicana              0.000264\n",
      "scedosporium_boydii              0.000264\n",
      "cladophialophora_unidentified    0.000264\n",
      "oculimacula_yallundae-ccl029     0.000264\n",
      "zymoseptoria_tritici             0.000198\n",
      "puccinia_striiformis-tritici     0.000198\n",
      "tuber_brumale                    0.000198\n",
      "rhodotorula_mucilaginosa         0.000198\n",
      "diaporthe_unidentified           0.000132\n",
      "candida_unidentified             0.000132\n",
      "cortinarius_globuliformis        0.000132\n",
      "galactomyces_geotrichum          0.000132\n",
      "quambalaria_cyanescens           0.000066\n",
      "yarrowia_lipolytica              0.000066\n",
      "diaporthe_ccl067                 0.000066\n",
      "debaryomyces_unidentified        0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "pyrenophora_tritici-repentis     0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "kluyveromyces_marxianus          0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "candida_parapsilosis             0.000066\n",
      "candida_albicans                 0.000066\n",
      "blastobotrys_proliferans         0.000066\n",
      "asteroma_ccl068                  0.000066\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "meyerozyma_guilliermondii        0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: clavispora_lusitaniae\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "clavispora_lusitaniae           0.991673\n",
      "candida_unidentified            0.002398\n",
      "kodamaea_ohmeri                 0.000733\n",
      "candida_albicans                0.000533\n",
      "candida_orthopsilosis           0.000533\n",
      "meyerozyma_guilliermondii       0.000466\n",
      "zymoseptoria_tritici            0.000400\n",
      "yamadazyma_scolyti              0.000400\n",
      "galactomyces_geotrichum         0.000333\n",
      "candida_metapsilosis            0.000266\n",
      "candida_parapsilosis            0.000266\n",
      "debaryomyces_unidentified       0.000266\n",
      "kluyveromyces_unidentified      0.000200\n",
      "scedosporium_boydii             0.000200\n",
      "penicillium_chrysogenum         0.000200\n",
      "kluyveromyces_marxianus         0.000133\n",
      "fusarium_oxysporum              0.000133\n",
      "yamadazyma_mexicana             0.000133\n",
      "puccinia_striiformis-tritici    0.000133\n",
      "pyrenophora_tritici-repentis    0.000133\n",
      "tuber_brumale                   0.000067\n",
      "zygoascus_hellenicus            0.000067\n",
      "saccharomyces_cerevisiae        0.000067\n",
      "wickerhamomyces_anomalus        0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "yarrowia_lipolytica             0.000067\n",
      "blastobotrys_proliferans        0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kluyveromyces_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kluyveromyces_unidentified       0.488380\n",
      "kluyveromyces_marxianus          0.425123\n",
      "cryptococcus_zero                0.059735\n",
      "cortinarius_globuliformis        0.004610\n",
      "rhodotorula_mucilaginosa         0.002625\n",
      "entoleuca_unidentified           0.002049\n",
      "quambalaria_cyanescens           0.001473\n",
      "meyerozyma_guilliermondii        0.001409\n",
      "candida_unidentified             0.001345\n",
      "saccharomyces_cerevisiae         0.001280\n",
      "wickerhamomyces_anomalus         0.000960\n",
      "yarrowia_lipolytica              0.000960\n",
      "oculimacula_yallundae-ccl029     0.000960\n",
      "candida_metapsilosis             0.000768\n",
      "candida_orthopsilosis            0.000704\n",
      "clavispora_lusitaniae            0.000704\n",
      "yamadazyma_mexicana              0.000704\n",
      "zymoseptoria_tritici             0.000576\n",
      "candida_albicans                 0.000512\n",
      "debaryomyces_unidentified        0.000512\n",
      "candida_parapsilosis             0.000384\n",
      "penicillium_chrysogenum          0.000384\n",
      "puccinia_striiformis-tritici     0.000384\n",
      "aspergillus_niger                0.000320\n",
      "yamadazyma_scolyti               0.000320\n",
      "pyrenophora_tritici-repentis     0.000320\n",
      "cladophialophora_unidentified    0.000256\n",
      "asteroma_ccl068                  0.000256\n",
      "aspergillus_flavus               0.000256\n",
      "zygoascus_hellenicus             0.000256\n",
      "fusarium_oxysporum               0.000256\n",
      "scedosporium_boydii              0.000192\n",
      "tuber_brumale                    0.000192\n",
      "blastobotrys_proliferans         0.000192\n",
      "galactomyces_geotrichum          0.000192\n",
      "kodamaea_ohmeri                  0.000192\n",
      "pichia_kudriavzevii              0.000128\n",
      "pichia_membranifaciens           0.000064\n",
      "asteroma_ccl060                  0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: penicillium_chrysogenum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "penicillium_chrysogenum          0.991548\n",
      "aspergillus_flavus               0.001730\n",
      "aspergillus_niger                0.001198\n",
      "kluyveromyces_unidentified       0.000799\n",
      "aspergillus_unidentified         0.000732\n",
      "kluyveromyces_marxianus          0.000732\n",
      "tuber_brumale                    0.000532\n",
      "cortinarius_globuliformis        0.000466\n",
      "zymoseptoria_tritici             0.000466\n",
      "cladophialophora_unidentified    0.000200\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "kodamaea_ohmeri                  0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "scedosporium_boydii              0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_unidentified             0.000067\n",
      "cryptococcus_zero                0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "dothiorella_vidmadera            0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "fusarium_oxysporum               0.000067\n",
      "zygoascus_hellenicus             0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "pyrenophora_tritici-repentis     0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: rhodotorula_mucilaginosa\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "rhodotorula_mucilaginosa         0.997001\n",
      "penicillium_chrysogenum          0.000667\n",
      "pyrenophora_tritici-repentis     0.000467\n",
      "aspergillus_niger                0.000467\n",
      "cortinarius_globuliformis        0.000333\n",
      "tuber_brumale                    0.000267\n",
      "scedosporium_boydii              0.000133\n",
      "puccinia_striiformis-tritici     0.000133\n",
      "cryptococcus_zero                0.000133\n",
      "zymoseptoria_tritici             0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_albicans                 0.000067\n",
      "aspergillus_unidentified         0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: scedosporium_boydii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "scedosporium_boydii              0.995737\n",
      "pyrenophora_tritici-repentis     0.000533\n",
      "fusarium_oxysporum               0.000466\n",
      "rhodotorula_mucilaginosa         0.000400\n",
      "penicillium_chrysogenum          0.000400\n",
      "aspergillus_unidentified         0.000333\n",
      "cladophialophora_unidentified    0.000200\n",
      "clavispora_lusitaniae            0.000200\n",
      "zymoseptoria_tritici             0.000133\n",
      "diaporthe_unidentified           0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "cortinarius_globuliformis        0.000133\n",
      "debaryomyces_unidentified        0.000133\n",
      "aspergillus_niger                0.000133\n",
      "tuber_brumale                    0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "cryptococcus_zero                0.000067\n",
      "kluyveromyces_marxianus          0.000067\n",
      "meyerozyma_guilliermondii        0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "candida_unidentified             0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "asteroma_ccl068                  0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "dothiorella_vidmadera            0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: blastobotrys_proliferans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "blastobotrys_proliferans         0.996736\n",
      "meyerozyma_guilliermondii        0.000999\n",
      "debaryomyces_unidentified        0.000533\n",
      "candida_albicans                 0.000200\n",
      "candida_orthopsilosis            0.000200\n",
      "yamadazyma_scolyti               0.000133\n",
      "candida_metapsilosis             0.000133\n",
      "candida_parapsilosis             0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "tuber_brumale                    0.000133\n",
      "wickerhamomyces_anomalus         0.000133\n",
      "aspergillus_flavus               0.000133\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_unidentified             0.000067\n",
      "diaporthe_unidentified           0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: debaryomyces_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "debaryomyces_unidentified       0.989691\n",
      "clavispora_lusitaniae           0.004789\n",
      "meyerozyma_guilliermondii       0.001530\n",
      "yamadazyma_mexicana             0.000931\n",
      "yamadazyma_scolyti              0.000399\n",
      "wickerhamomyces_anomalus        0.000333\n",
      "candida_unidentified            0.000266\n",
      "kluyveromyces_unidentified      0.000266\n",
      "kluyveromyces_marxianus         0.000200\n",
      "blastobotrys_proliferans        0.000200\n",
      "candida_parapsilosis            0.000200\n",
      "candida_metapsilosis            0.000133\n",
      "candida_orthopsilosis           0.000133\n",
      "aspergillus_flavus              0.000133\n",
      "galactomyces_geotrichum         0.000133\n",
      "saccharomyces_cerevisiae        0.000133\n",
      "zygoascus_hellenicus            0.000067\n",
      "kodamaea_ohmeri                 0.000067\n",
      "oculimacula_yallundae-ccl031    0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "candida_albicans                0.000067\n",
      "yarrowia_lipolytica             0.000067\n",
      "aspergillus_unidentified        0.000067\n",
      "zymoseptoria_tritici            0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: galactomyces_geotrichum\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "galactomyces_geotrichum          0.996215\n",
      "kodamaea_ohmeri                  0.000744\n",
      "clavispora_lusitaniae            0.000541\n",
      "blastobotrys_proliferans         0.000473\n",
      "yarrowia_lipolytica              0.000406\n",
      "aspergillus_flavus               0.000338\n",
      "zygoascus_hellenicus             0.000135\n",
      "debaryomyces_unidentified        0.000135\n",
      "yamadazyma_mexicana              0.000135\n",
      "scedosporium_boydii              0.000068\n",
      "saccharomyces_cerevisiae         0.000068\n",
      "penicillium_chrysogenum          0.000068\n",
      "oculimacula_yallundae-ccl031     0.000068\n",
      "wickerhamomyces_anomalus         0.000068\n",
      "fusarium_oxysporum               0.000068\n",
      "dothiorella_vidmadera            0.000068\n",
      "cladophialophora_unidentified    0.000068\n",
      "candida_orthopsilosis            0.000068\n",
      "candida_metapsilosis             0.000068\n",
      "yamadazyma_scolyti               0.000068\n",
      "asteroma_ccl060                  0.000068\n",
      "zymoseptoria_tritici             0.000068\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kodamaea_ohmeri\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kodamaea_ohmeri                 0.998734\n",
      "yarrowia_lipolytica             0.000133\n",
      "clavispora_lusitaniae           0.000133\n",
      "candida_orthopsilosis           0.000133\n",
      "candida_albicans                0.000133\n",
      "yamadazyma_scolyti              0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "quambalaria_cyanescens          0.000067\n",
      "pichia_kudriavzevii             0.000067\n",
      "oculimacula_yallundae-ccl029    0.000067\n",
      "meyerozyma_guilliermondii       0.000067\n",
      "kluyveromyces_unidentified      0.000067\n",
      "entoleuca_unidentified          0.000067\n",
      "debaryomyces_unidentified       0.000067\n",
      "cortinarius_globuliformis       0.000067\n",
      "candida_metapsilosis            0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: meyerozyma_guilliermondii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "meyerozyma_guilliermondii       0.994081\n",
      "debaryomyces_unidentified       0.000931\n",
      "yamadazyma_mexicana             0.000665\n",
      "yamadazyma_scolyti              0.000532\n",
      "candida_orthopsilosis           0.000399\n",
      "wickerhamomyces_anomalus        0.000333\n",
      "candida_metapsilosis            0.000333\n",
      "zygoascus_hellenicus            0.000266\n",
      "candida_albicans                0.000266\n",
      "candida_unidentified            0.000266\n",
      "kodamaea_ohmeri                 0.000266\n",
      "kluyveromyces_unidentified      0.000200\n",
      "candida_parapsilosis            0.000200\n",
      "scedosporium_boydii             0.000200\n",
      "pichia_kudriavzevii             0.000133\n",
      "yarrowia_lipolytica             0.000133\n",
      "kluyveromyces_marxianus         0.000133\n",
      "galactomyces_geotrichum         0.000133\n",
      "aspergillus_flavus              0.000133\n",
      "puccinia_striiformis-tritici    0.000067\n",
      "pyrenophora_tritici-repentis    0.000067\n",
      "rhodotorula_mucilaginosa        0.000067\n",
      "saccharomyces_cerevisiae        0.000067\n",
      "asteroma_ccl068                 0.000067\n",
      "asteroma_ccl060                 0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: wickerhamomyces_anomalus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "wickerhamomyces_anomalus         0.995606\n",
      "meyerozyma_guilliermondii        0.000732\n",
      "yamadazyma_mexicana              0.000732\n",
      "debaryomyces_unidentified        0.000666\n",
      "aspergillus_flavus               0.000399\n",
      "blastobotrys_proliferans         0.000333\n",
      "candida_parapsilosis             0.000266\n",
      "zygoascus_hellenicus             0.000266\n",
      "candida_unidentified             0.000133\n",
      "yamadazyma_scolyti               0.000133\n",
      "kluyveromyces_marxianus          0.000133\n",
      "kodamaea_ohmeri                  0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yamadazyma_mexicana\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yamadazyma_mexicana              0.984399\n",
      "yamadazyma_scolyti               0.009626\n",
      "debaryomyces_unidentified        0.001062\n",
      "meyerozyma_guilliermondii        0.000797\n",
      "kodamaea_ohmeri                  0.000730\n",
      "candida_metapsilosis             0.000465\n",
      "aspergillus_flavus               0.000332\n",
      "candida_orthopsilosis            0.000332\n",
      "candida_unidentified             0.000199\n",
      "zygoascus_hellenicus             0.000199\n",
      "wickerhamomyces_anomalus         0.000199\n",
      "penicillium_chrysogenum          0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "candida_albicans                 0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "tuber_brumale                    0.000133\n",
      "candida_parapsilosis             0.000133\n",
      "oculimacula_yallundae-ccl029     0.000133\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "puccinia_striiformis-tritici     0.000066\n",
      "diaporthe_ccl067                 0.000066\n",
      "entoleuca_unidentified           0.000066\n",
      "pichia_membranifaciens           0.000066\n",
      "pichia_kudriavzevii              0.000066\n",
      "kluyveromyces_marxianus          0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yamadazyma_scolyti\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yamadazyma_scolyti               0.991761\n",
      "yamadazyma_mexicana              0.004120\n",
      "debaryomyces_unidentified        0.000598\n",
      "meyerozyma_guilliermondii        0.000532\n",
      "candida_unidentified             0.000465\n",
      "candida_metapsilosis             0.000399\n",
      "candida_parapsilosis             0.000399\n",
      "candida_orthopsilosis            0.000332\n",
      "candida_albicans                 0.000266\n",
      "zygoascus_hellenicus             0.000133\n",
      "pyrenophora_tritici-repentis     0.000133\n",
      "kluyveromyces_unidentified       0.000133\n",
      "galactomyces_geotrichum          0.000133\n",
      "blastobotrys_proliferans         0.000066\n",
      "entoleuca_unidentified           0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "cortinarius_globuliformis        0.000066\n",
      "yarrowia_lipolytica              0.000066\n",
      "kluyveromyces_marxianus          0.000066\n",
      "pichia_membranifaciens           0.000066\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "aspergillus_flavus               0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: yarrowia_lipolytica\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "yarrowia_lipolytica         0.999467\n",
      "yamadazyma_scolyti          0.000133\n",
      "aspergillus_flavus          0.000133\n",
      "yamadazyma_mexicana         0.000067\n",
      "wickerhamomyces_anomalus    0.000067\n",
      "galactomyces_geotrichum     0.000067\n",
      "blastobotrys_proliferans    0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: zygoascus_hellenicus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zygoascus_hellenicus          0.998268\n",
      "yamadazyma_mexicana           0.000400\n",
      "blastobotrys_proliferans      0.000267\n",
      "yamadazyma_scolyti            0.000133\n",
      "meyerozyma_guilliermondii     0.000133\n",
      "galactomyces_geotrichum       0.000133\n",
      "zymoseptoria_tritici          0.000067\n",
      "wickerhamomyces_anomalus      0.000067\n",
      "kodamaea_ohmeri               0.000067\n",
      "kluyveromyces_unidentified    0.000067\n",
      "diaporthe_unidentified        0.000067\n",
      "candida_orthopsilosis         0.000067\n",
      "candida_metapsilosis          0.000067\n",
      "candida_albicans              0.000067\n",
      "asteroma_ccl060               0.000067\n",
      "aspergillus_flavus            0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_flavus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_flavus               0.994210\n",
      "aspergillus_niger                0.002263\n",
      "aspergillus_unidentified         0.001797\n",
      "penicillium_chrysogenum          0.000466\n",
      "yamadazyma_scolyti               0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "cryptococcus_zero                0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "candida_unidentified             0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "zymoseptoria_tritici             0.000067\n",
      "debaryomyces_unidentified        0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "pyrenophora_tritici-repentis     0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cryptococcus_zero\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.995871\n",
      "quambalaria_cyanescens           0.000466\n",
      "cortinarius_globuliformis        0.000333\n",
      "entoleuca_unidentified           0.000266\n",
      "aspergillus_niger                0.000266\n",
      "diaporthe_ccl067                 0.000266\n",
      "yamadazyma_mexicana              0.000200\n",
      "aspergillus_unidentified         0.000200\n",
      "diaporthe_unidentified           0.000200\n",
      "zymoseptoria_tritici             0.000133\n",
      "asteroma_ccl060                  0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "debaryomyces_unidentified        0.000133\n",
      "dothiorella_vidmadera            0.000133\n",
      "kluyveromyces_marxianus          0.000133\n",
      "meyerozyma_guilliermondii        0.000133\n",
      "rhodotorula_mucilaginosa         0.000133\n",
      "wickerhamomyces_anomalus         0.000133\n",
      "yarrowia_lipolytica              0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "clavispora_lusitaniae            0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "yamadazyma_scolyti               0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: aspergillus_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_unidentified         0.802217\n",
      "aspergillus_niger                0.187690\n",
      "aspergillus_flavus               0.004486\n",
      "penicillium_chrysogenum          0.000726\n",
      "entoleuca_unidentified           0.000528\n",
      "oculimacula_yallundae-ccl029     0.000462\n",
      "dothiorella_vidmadera            0.000396\n",
      "oculimacula_yallundae-ccl031     0.000396\n",
      "tuber_brumale                    0.000330\n",
      "asteroma_ccl068                  0.000330\n",
      "zymoseptoria_tritici             0.000330\n",
      "quambalaria_cyanescens           0.000198\n",
      "diaporthe_unidentified           0.000198\n",
      "diaporthe_ccl067                 0.000198\n",
      "blastobotrys_proliferans         0.000198\n",
      "kluyveromyces_marxianus          0.000198\n",
      "asteroma_ccl060                  0.000198\n",
      "saccharomyces_cerevisiae         0.000198\n",
      "fusarium_oxysporum               0.000132\n",
      "galactomyces_geotrichum          0.000132\n",
      "debaryomyces_unidentified        0.000066\n",
      "kluyveromyces_unidentified       0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "candida_orthopsilosis            0.000066\n",
      "scedosporium_boydii              0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: diaporthe_ccl067\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "diaporthe_ccl067                 0.625385\n",
      "asteroma_ccl068                  0.367346\n",
      "diaporthe_unidentified           0.001375\n",
      "asteroma_ccl060                  0.001244\n",
      "entoleuca_unidentified           0.001179\n",
      "oculimacula_yallundae-ccl029     0.000655\n",
      "cryptococcus_zero                0.000589\n",
      "pyrenophora_tritici-repentis     0.000327\n",
      "oculimacula_yallundae-ccl031     0.000262\n",
      "dothiorella_vidmadera            0.000262\n",
      "galactomyces_geotrichum          0.000196\n",
      "quambalaria_cyanescens           0.000196\n",
      "fusarium_oxysporum               0.000131\n",
      "tuber_brumale                    0.000131\n",
      "cortinarius_globuliformis        0.000131\n",
      "aspergillus_unidentified         0.000065\n",
      "cladophialophora_unidentified    0.000065\n",
      "candida_albicans                 0.000065\n",
      "zymoseptoria_tritici             0.000065\n",
      "wickerhamomyces_anomalus         0.000065\n",
      "kluyveromyces_marxianus          0.000065\n",
      "kluyveromyces_unidentified       0.000065\n",
      "puccinia_striiformis-tritici     0.000065\n",
      "aspergillus_flavus               0.000065\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: diaporthe_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "diaporthe_unidentified          0.702311\n",
      "asteroma_ccl060                 0.288997\n",
      "entoleuca_unidentified          0.001646\n",
      "asteroma_ccl068                 0.001119\n",
      "diaporthe_ccl067                0.001119\n",
      "kodamaea_ohmeri                 0.000922\n",
      "fusarium_oxysporum              0.000593\n",
      "oculimacula_yallundae-ccl029    0.000461\n",
      "dothiorella_vidmadera           0.000329\n",
      "tuber_brumale                   0.000329\n",
      "cryptococcus_zero               0.000329\n",
      "scedosporium_boydii             0.000263\n",
      "quambalaria_cyanescens          0.000263\n",
      "oculimacula_yallundae-ccl031    0.000263\n",
      "aspergillus_unidentified        0.000198\n",
      "kluyveromyces_unidentified      0.000132\n",
      "yamadazyma_scolyti              0.000132\n",
      "saccharomyces_cerevisiae        0.000132\n",
      "puccinia_striiformis-tritici    0.000066\n",
      "pyrenophora_tritici-repentis    0.000066\n",
      "cortinarius_globuliformis       0.000066\n",
      "clavispora_lusitaniae           0.000066\n",
      "candida_metapsilosis            0.000066\n",
      "rhodotorula_mucilaginosa        0.000066\n",
      "zymoseptoria_tritici            0.000066\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: oculimacula_yallundae-ccl031\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "oculimacula_yallundae-ccl031     0.963964\n",
      "oculimacula_yallundae-ccl029     0.027756\n",
      "quambalaria_cyanescens           0.000795\n",
      "entoleuca_unidentified           0.000530\n",
      "dothiorella_vidmadera            0.000530\n",
      "asteroma_ccl060                  0.000530\n",
      "yamadazyma_mexicana              0.000464\n",
      "tuber_brumale                    0.000464\n",
      "pyrenophora_tritici-repentis     0.000397\n",
      "asteroma_ccl068                  0.000331\n",
      "clavispora_lusitaniae            0.000331\n",
      "zygoascus_hellenicus             0.000331\n",
      "fusarium_oxysporum               0.000265\n",
      "cladophialophora_unidentified    0.000265\n",
      "cryptococcus_zero                0.000265\n",
      "diaporthe_unidentified           0.000199\n",
      "diaporthe_ccl067                 0.000199\n",
      "aspergillus_niger                0.000199\n",
      "scedosporium_boydii              0.000199\n",
      "yarrowia_lipolytica              0.000199\n",
      "galactomyces_geotrichum          0.000132\n",
      "aspergillus_unidentified         0.000132\n",
      "blastobotrys_proliferans         0.000132\n",
      "candida_albicans                 0.000132\n",
      "kluyveromyces_marxianus          0.000132\n",
      "zymoseptoria_tritici             0.000132\n",
      "cortinarius_globuliformis        0.000132\n",
      "penicillium_chrysogenum          0.000132\n",
      "meyerozyma_guilliermondii        0.000132\n",
      "wickerhamomyces_anomalus         0.000066\n",
      "saccharomyces_cerevisiae         0.000066\n",
      "candida_parapsilosis             0.000066\n",
      "candida_orthopsilosis            0.000066\n",
      "candida_metapsilosis             0.000066\n",
      "rhodotorula_mucilaginosa         0.000066\n",
      "pichia_kudriavzevii              0.000066\n",
      "yamadazyma_scolyti               0.000066\n",
      "candida_unidentified             0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: oculimacula_yallundae-ccl029\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "oculimacula_yallundae-ccl029     0.984795\n",
      "oculimacula_yallundae-ccl031     0.010823\n",
      "entoleuca_unidentified           0.000465\n",
      "aspergillus_unidentified         0.000398\n",
      "dothiorella_vidmadera            0.000332\n",
      "cryptococcus_zero                0.000266\n",
      "saccharomyces_cerevisiae         0.000266\n",
      "diaporthe_ccl067                 0.000266\n",
      "fusarium_oxysporum               0.000199\n",
      "diaporthe_unidentified           0.000199\n",
      "wickerhamomyces_anomalus         0.000199\n",
      "aspergillus_niger                0.000199\n",
      "pyrenophora_tritici-repentis     0.000199\n",
      "candida_unidentified             0.000133\n",
      "tuber_brumale                    0.000133\n",
      "quambalaria_cyanescens           0.000133\n",
      "scedosporium_boydii              0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "candida_orthopsilosis            0.000066\n",
      "asteroma_ccl060                  0.000066\n",
      "candida_albicans                 0.000066\n",
      "candida_metapsilosis             0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "candida_parapsilosis             0.000066\n",
      "meyerozyma_guilliermondii        0.000066\n",
      "cladophialophora_unidentified    0.000066\n",
      "cortinarius_globuliformis        0.000066\n",
      "puccinia_striiformis-tritici     0.000066\n",
      "yamadazyma_scolyti               0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: dothiorella_vidmadera\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "dothiorella_vidmadera            0.993807\n",
      "zymoseptoria_tritici             0.001199\n",
      "diaporthe_ccl067                 0.000533\n",
      "oculimacula_yallundae-ccl031     0.000466\n",
      "entoleuca_unidentified           0.000400\n",
      "oculimacula_yallundae-ccl029     0.000333\n",
      "diaporthe_unidentified           0.000266\n",
      "quambalaria_cyanescens           0.000266\n",
      "cryptococcus_zero                0.000266\n",
      "pyrenophora_tritici-repentis     0.000200\n",
      "aspergillus_niger                0.000200\n",
      "aspergillus_unidentified         0.000200\n",
      "asteroma_ccl060                  0.000200\n",
      "clavispora_lusitaniae            0.000133\n",
      "asteroma_ccl068                  0.000133\n",
      "fusarium_oxysporum               0.000133\n",
      "candida_unidentified             0.000133\n",
      "rhodotorula_mucilaginosa         0.000133\n",
      "candida_parapsilosis             0.000133\n",
      "candida_metapsilosis             0.000067\n",
      "candida_orthopsilosis            0.000067\n",
      "blastobotrys_proliferans         0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "tuber_brumale                    0.000067\n",
      "yarrowia_lipolytica              0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "kluyveromyces_unidentified       0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "pichia_kudriavzevii              0.000067\n",
      "aspergillus_flavus               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: quambalaria_cyanescens\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "quambalaria_cyanescens          0.997399\n",
      "entoleuca_unidentified          0.000467\n",
      "dothiorella_vidmadera           0.000400\n",
      "oculimacula_yallundae-ccl031    0.000333\n",
      "asteroma_ccl068                 0.000267\n",
      "diaporthe_unidentified          0.000200\n",
      "cryptococcus_zero               0.000200\n",
      "rhodotorula_mucilaginosa        0.000133\n",
      "diaporthe_ccl067                0.000133\n",
      "zymoseptoria_tritici            0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "saccharomyces_cerevisiae        0.000067\n",
      "oculimacula_yallundae-ccl029    0.000067\n",
      "cortinarius_globuliformis       0.000067\n",
      "asteroma_ccl060                 0.000067\n",
      "aspergillus_niger               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: entoleuca_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "entoleuca_unidentified          0.996870\n",
      "oculimacula_yallundae-ccl031    0.000466\n",
      "diaporthe_ccl067                0.000400\n",
      "diaporthe_unidentified          0.000266\n",
      "oculimacula_yallundae-ccl029    0.000266\n",
      "asteroma_ccl068                 0.000200\n",
      "fusarium_oxysporum              0.000200\n",
      "dothiorella_vidmadera           0.000200\n",
      "cryptococcus_zero               0.000200\n",
      "aspergillus_unidentified        0.000133\n",
      "asteroma_ccl060                 0.000133\n",
      "candida_orthopsilosis           0.000067\n",
      "candida_albicans                0.000067\n",
      "candida_metapsilosis            0.000067\n",
      "yamadazyma_scolyti              0.000067\n",
      "clavispora_lusitaniae           0.000067\n",
      "yamadazyma_mexicana             0.000067\n",
      "meyerozyma_guilliermondii       0.000067\n",
      "quambalaria_cyanescens          0.000067\n",
      "scedosporium_boydii             0.000067\n",
      "aspergillus_niger               0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: asteroma_ccl060\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "asteroma_ccl060                 0.616911\n",
      "diaporthe_unidentified          0.375074\n",
      "diaporthe_ccl067                0.001314\n",
      "asteroma_ccl068                 0.000985\n",
      "entoleuca_unidentified          0.000985\n",
      "oculimacula_yallundae-ccl029    0.000920\n",
      "rhodotorula_mucilaginosa        0.000328\n",
      "zymoseptoria_tritici            0.000263\n",
      "cortinarius_globuliformis       0.000263\n",
      "fusarium_oxysporum              0.000263\n",
      "quambalaria_cyanescens          0.000263\n",
      "tuber_brumale                   0.000263\n",
      "pyrenophora_tritici-repentis    0.000263\n",
      "aspergillus_unidentified        0.000197\n",
      "clavispora_lusitaniae           0.000197\n",
      "scedosporium_boydii             0.000197\n",
      "oculimacula_yallundae-ccl031    0.000197\n",
      "penicillium_chrysogenum         0.000131\n",
      "cryptococcus_zero               0.000131\n",
      "dothiorella_vidmadera           0.000131\n",
      "zygoascus_hellenicus            0.000131\n",
      "galactomyces_geotrichum         0.000131\n",
      "aspergillus_flavus              0.000131\n",
      "puccinia_striiformis-tritici    0.000066\n",
      "candida_parapsilosis            0.000066\n",
      "candida_metapsilosis            0.000066\n",
      "yamadazyma_mexicana             0.000066\n",
      "pichia_kudriavzevii             0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: asteroma_ccl068\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "asteroma_ccl068                  0.711335\n",
      "diaporthe_ccl067                 0.278568\n",
      "entoleuca_unidentified           0.001573\n",
      "asteroma_ccl060                  0.001311\n",
      "oculimacula_yallundae-ccl029     0.001115\n",
      "diaporthe_unidentified           0.001049\n",
      "aspergillus_unidentified         0.000524\n",
      "quambalaria_cyanescens           0.000459\n",
      "oculimacula_yallundae-ccl031     0.000459\n",
      "cryptococcus_zero                0.000393\n",
      "fusarium_oxysporum               0.000393\n",
      "dothiorella_vidmadera            0.000328\n",
      "pichia_membranifaciens           0.000262\n",
      "galactomyces_geotrichum          0.000262\n",
      "aspergillus_niger                0.000262\n",
      "rhodotorula_mucilaginosa         0.000197\n",
      "saccharomyces_cerevisiae         0.000197\n",
      "scedosporium_boydii              0.000197\n",
      "pyrenophora_tritici-repentis     0.000131\n",
      "tuber_brumale                    0.000131\n",
      "cladophialophora_unidentified    0.000131\n",
      "candida_parapsilosis             0.000131\n",
      "candida_orthopsilosis            0.000066\n",
      "candida_albicans                 0.000066\n",
      "zymoseptoria_tritici             0.000066\n",
      "clavispora_lusitaniae            0.000066\n",
      "zygoascus_hellenicus             0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "penicillium_chrysogenum          0.000066\n",
      "wickerhamomyces_anomalus         0.000066\n",
      "aspergillus_flavus               0.000066\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: saccharomyces_cerevisiae\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "saccharomyces_cerevisiae         0.992092\n",
      "kluyveromyces_unidentified       0.002259\n",
      "kluyveromyces_marxianus          0.002126\n",
      "debaryomyces_unidentified        0.000532\n",
      "wickerhamomyces_anomalus         0.000465\n",
      "candida_metapsilosis             0.000332\n",
      "yamadazyma_mexicana              0.000266\n",
      "candida_orthopsilosis            0.000266\n",
      "pichia_membranifaciens           0.000199\n",
      "meyerozyma_guilliermondii        0.000199\n",
      "candida_parapsilosis             0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "zygoascus_hellenicus             0.000133\n",
      "blastobotrys_proliferans         0.000133\n",
      "dothiorella_vidmadera            0.000066\n",
      "cryptococcus_zero                0.000066\n",
      "yamadazyma_scolyti               0.000066\n",
      "kodamaea_ohmeri                  0.000066\n",
      "candida_unidentified             0.000066\n",
      "pichia_kudriavzevii              0.000066\n",
      "rhodotorula_mucilaginosa         0.000066\n",
      "tuber_brumale                    0.000066\n",
      "candida_albicans                 0.000066\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: cladophialophora_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cladophialophora_unidentified    0.997734\n",
      "zymoseptoria_tritici             0.000267\n",
      "candida_orthopsilosis            0.000267\n",
      "candida_albicans                 0.000200\n",
      "kluyveromyces_unidentified       0.000200\n",
      "fusarium_oxysporum               0.000133\n",
      "candida_metapsilosis             0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "candida_parapsilosis             0.000067\n",
      "candida_unidentified             0.000067\n",
      "entoleuca_unidentified           0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "oculimacula_yallundae-ccl031     0.000067\n",
      "penicillium_chrysogenum          0.000067\n",
      "pichia_membranifaciens           0.000067\n",
      "quambalaria_cyanescens           0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "saccharomyces_cerevisiae         0.000067\n",
      "tuber_brumale                    0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_albicans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_albicans                 0.495014\n",
      "candida_unidentified             0.300521\n",
      "candida_orthopsilosis            0.105192\n",
      "candida_metapsilosis             0.052242\n",
      "candida_parapsilosis             0.032555\n",
      "meyerozyma_guilliermondii        0.001544\n",
      "debaryomyces_unidentified        0.001544\n",
      "galactomyces_geotrichum          0.001222\n",
      "entoleuca_unidentified           0.000965\n",
      "yamadazyma_mexicana              0.000836\n",
      "kodamaea_ohmeri                  0.000836\n",
      "yamadazyma_scolyti               0.000772\n",
      "oculimacula_yallundae-ccl029     0.000708\n",
      "kluyveromyces_unidentified       0.000708\n",
      "tuber_brumale                    0.000579\n",
      "kluyveromyces_marxianus          0.000579\n",
      "clavispora_lusitaniae            0.000450\n",
      "pyrenophora_tritici-repentis     0.000450\n",
      "saccharomyces_cerevisiae         0.000450\n",
      "wickerhamomyces_anomalus         0.000257\n",
      "blastobotrys_proliferans         0.000257\n",
      "pichia_membranifaciens           0.000193\n",
      "asteroma_ccl060                  0.000193\n",
      "fusarium_oxysporum               0.000193\n",
      "quambalaria_cyanescens           0.000193\n",
      "dothiorella_vidmadera            0.000193\n",
      "diaporthe_ccl067                 0.000193\n",
      "zygoascus_hellenicus             0.000129\n",
      "diaporthe_unidentified           0.000129\n",
      "yarrowia_lipolytica              0.000129\n",
      "pichia_kudriavzevii              0.000129\n",
      "rhodotorula_mucilaginosa         0.000129\n",
      "cortinarius_globuliformis        0.000064\n",
      "cryptococcus_zero                0.000064\n",
      "cladophialophora_unidentified    0.000064\n",
      "oculimacula_yallundae-ccl031     0.000064\n",
      "penicillium_chrysogenum          0.000064\n",
      "asteroma_ccl068                  0.000064\n",
      "scedosporium_boydii              0.000064\n",
      "aspergillus_unidentified         0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_metapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_metapsilosis             0.515065\n",
      "candida_orthopsilosis            0.420290\n",
      "candida_parapsilosis             0.051487\n",
      "kodamaea_ohmeri                  0.003432\n",
      "candida_unidentified             0.001335\n",
      "oculimacula_yallundae-ccl029     0.001017\n",
      "meyerozyma_guilliermondii        0.000826\n",
      "candida_albicans                 0.000763\n",
      "clavispora_lusitaniae            0.000572\n",
      "debaryomyces_unidentified        0.000445\n",
      "yamadazyma_scolyti               0.000381\n",
      "galactomyces_geotrichum          0.000318\n",
      "kluyveromyces_marxianus          0.000318\n",
      "tuber_brumale                    0.000318\n",
      "yarrowia_lipolytica              0.000254\n",
      "entoleuca_unidentified           0.000254\n",
      "cladophialophora_unidentified    0.000254\n",
      "quambalaria_cyanescens           0.000254\n",
      "wickerhamomyces_anomalus         0.000254\n",
      "blastobotrys_proliferans         0.000254\n",
      "saccharomyces_cerevisiae         0.000191\n",
      "fusarium_oxysporum               0.000191\n",
      "kluyveromyces_unidentified       0.000191\n",
      "zygoascus_hellenicus             0.000127\n",
      "pichia_kudriavzevii              0.000127\n",
      "pichia_membranifaciens           0.000127\n",
      "pyrenophora_tritici-repentis     0.000127\n",
      "rhodotorula_mucilaginosa         0.000127\n",
      "yamadazyma_mexicana              0.000127\n",
      "aspergillus_flavus               0.000127\n",
      "diaporthe_unidentified           0.000064\n",
      "cryptococcus_zero                0.000064\n",
      "penicillium_chrysogenum          0.000064\n",
      "puccinia_striiformis-tritici     0.000064\n",
      "scedosporium_boydii              0.000064\n",
      "asteroma_ccl068                  0.000064\n",
      "aspergillus_niger                0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_orthopsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_orthopsilosis            0.735129\n",
      "candida_metapsilosis             0.202352\n",
      "candida_parapsilosis             0.041848\n",
      "meyerozyma_guilliermondii        0.009988\n",
      "candida_unidentified             0.001644\n",
      "candida_albicans                 0.000948\n",
      "yamadazyma_mexicana              0.000632\n",
      "tuber_brumale                    0.000632\n",
      "entoleuca_unidentified           0.000569\n",
      "saccharomyces_cerevisiae         0.000569\n",
      "debaryomyces_unidentified        0.000569\n",
      "oculimacula_yallundae-ccl029     0.000443\n",
      "yamadazyma_scolyti               0.000379\n",
      "kodamaea_ohmeri                  0.000379\n",
      "galactomyces_geotrichum          0.000316\n",
      "blastobotrys_proliferans         0.000253\n",
      "wickerhamomyces_anomalus         0.000253\n",
      "clavispora_lusitaniae            0.000253\n",
      "kluyveromyces_marxianus          0.000253\n",
      "zygoascus_hellenicus             0.000253\n",
      "cladophialophora_unidentified    0.000190\n",
      "pichia_membranifaciens           0.000190\n",
      "cryptococcus_zero                0.000190\n",
      "aspergillus_unidentified         0.000126\n",
      "aspergillus_niger                0.000126\n",
      "aspergillus_flavus               0.000126\n",
      "dothiorella_vidmadera            0.000126\n",
      "fusarium_oxysporum               0.000126\n",
      "kluyveromyces_unidentified       0.000126\n",
      "oculimacula_yallundae-ccl031     0.000126\n",
      "pyrenophora_tritici-repentis     0.000126\n",
      "scedosporium_boydii              0.000126\n",
      "yarrowia_lipolytica              0.000126\n",
      "diaporthe_unidentified           0.000063\n",
      "cortinarius_globuliformis        0.000063\n",
      "pichia_kudriavzevii              0.000063\n",
      "quambalaria_cyanescens           0.000063\n",
      "rhodotorula_mucilaginosa         0.000063\n",
      "asteroma_ccl068                  0.000063\n",
      "asteroma_ccl060                  0.000063\n",
      "zymoseptoria_tritici             0.000063\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_parapsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_parapsilosis             0.896477\n",
      "candida_orthopsilosis            0.052642\n",
      "candida_metapsilosis             0.041748\n",
      "candida_unidentified             0.001239\n",
      "oculimacula_yallundae-ccl029     0.000848\n",
      "kodamaea_ohmeri                  0.000718\n",
      "meyerozyma_guilliermondii        0.000718\n",
      "candida_albicans                 0.000718\n",
      "tuber_brumale                    0.000457\n",
      "entoleuca_unidentified           0.000326\n",
      "kluyveromyces_unidentified       0.000326\n",
      "debaryomyces_unidentified        0.000326\n",
      "pichia_membranifaciens           0.000261\n",
      "blastobotrys_proliferans         0.000261\n",
      "yamadazyma_mexicana              0.000261\n",
      "quambalaria_cyanescens           0.000196\n",
      "saccharomyces_cerevisiae         0.000196\n",
      "kluyveromyces_marxianus          0.000196\n",
      "yamadazyma_scolyti               0.000196\n",
      "asteroma_ccl060                  0.000196\n",
      "galactomyces_geotrichum          0.000196\n",
      "cryptococcus_zero                0.000130\n",
      "clavispora_lusitaniae            0.000130\n",
      "cortinarius_globuliformis        0.000130\n",
      "aspergillus_flavus               0.000130\n",
      "zygoascus_hellenicus             0.000130\n",
      "wickerhamomyces_anomalus         0.000130\n",
      "yarrowia_lipolytica              0.000130\n",
      "diaporthe_ccl067                 0.000065\n",
      "diaporthe_unidentified           0.000065\n",
      "cladophialophora_unidentified    0.000065\n",
      "fusarium_oxysporum               0.000065\n",
      "pichia_kudriavzevii              0.000065\n",
      "puccinia_striiformis-tritici     0.000065\n",
      "pyrenophora_tritici-repentis     0.000065\n",
      "aspergillus_niger                0.000065\n",
      "zymoseptoria_tritici             0.000065\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: candida_unidentified\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_unidentified             0.882645\n",
      "candida_albicans                 0.100478\n",
      "candida_orthopsilosis            0.003140\n",
      "candida_metapsilosis             0.002355\n",
      "candida_parapsilosis             0.001635\n",
      "oculimacula_yallundae-ccl029     0.001047\n",
      "entoleuca_unidentified           0.001047\n",
      "meyerozyma_guilliermondii        0.000981\n",
      "kodamaea_ohmeri                  0.000654\n",
      "yamadazyma_scolyti               0.000654\n",
      "debaryomyces_unidentified        0.000589\n",
      "yamadazyma_mexicana              0.000589\n",
      "wickerhamomyces_anomalus         0.000589\n",
      "saccharomyces_cerevisiae         0.000589\n",
      "kluyveromyces_unidentified       0.000392\n",
      "clavispora_lusitaniae            0.000327\n",
      "tuber_brumale                    0.000262\n",
      "yarrowia_lipolytica              0.000262\n",
      "blastobotrys_proliferans         0.000196\n",
      "kluyveromyces_marxianus          0.000131\n",
      "zygoascus_hellenicus             0.000131\n",
      "pichia_membranifaciens           0.000131\n",
      "quambalaria_cyanescens           0.000131\n",
      "rhodotorula_mucilaginosa         0.000131\n",
      "cryptococcus_zero                0.000131\n",
      "galactomyces_geotrichum          0.000131\n",
      "aspergillus_unidentified         0.000065\n",
      "zymoseptoria_tritici             0.000065\n",
      "cladophialophora_unidentified    0.000065\n",
      "diaporthe_ccl067                 0.000065\n",
      "diaporthe_unidentified           0.000065\n",
      "dothiorella_vidmadera            0.000065\n",
      "fusarium_oxysporum               0.000065\n",
      "penicillium_chrysogenum          0.000065\n",
      "scedosporium_boydii              0.000065\n",
      "aspergillus_flavus               0.000065\n",
      "dtype: float64 \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: kluyveromyces_marxianus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "kluyveromyces_marxianus          0.740035\n",
      "kluyveromyces_unidentified       0.249117\n",
      "candida_metapsilosis             0.000963\n",
      "oculimacula_yallundae-ccl029     0.000963\n",
      "entoleuca_unidentified           0.000706\n",
      "wickerhamomyces_anomalus         0.000706\n",
      "yamadazyma_scolyti               0.000578\n",
      "clavispora_lusitaniae            0.000514\n",
      "debaryomyces_unidentified        0.000514\n",
      "saccharomyces_cerevisiae         0.000514\n",
      "meyerozyma_guilliermondii        0.000514\n",
      "yamadazyma_mexicana              0.000449\n",
      "candida_orthopsilosis            0.000449\n",
      "candida_albicans                 0.000321\n",
      "yarrowia_lipolytica              0.000321\n",
      "candida_unidentified             0.000321\n",
      "candida_parapsilosis             0.000257\n",
      "tuber_brumale                    0.000257\n",
      "oculimacula_yallundae-ccl031     0.000257\n",
      "pichia_kudriavzevii              0.000257\n",
      "galactomyces_geotrichum          0.000193\n",
      "rhodotorula_mucilaginosa         0.000193\n",
      "cladophialophora_unidentified    0.000193\n",
      "kodamaea_ohmeri                  0.000193\n",
      "blastobotrys_proliferans         0.000128\n",
      "cryptococcus_zero                0.000128\n",
      "cortinarius_globuliformis        0.000128\n",
      "quambalaria_cyanescens           0.000128\n",
      "asteroma_ccl068                  0.000128\n",
      "pichia_membranifaciens           0.000128\n",
      "asteroma_ccl060                  0.000064\n",
      "zymoseptoria_tritici             0.000064\n",
      "dothiorella_vidmadera            0.000064\n",
      "zygoascus_hellenicus             0.000064\n",
      "puccinia_striiformis-tritici     0.000064\n",
      "scedosporium_boydii              0.000064\n",
      "aspergillus_flavus               0.000064\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_kudriavzevii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_kudriavzevii              0.996867\n",
      "pichia_membranifaciens           0.001533\n",
      "candida_orthopsilosis            0.000333\n",
      "candida_unidentified             0.000133\n",
      "tuber_brumale                    0.000133\n",
      "yamadazyma_scolyti               0.000133\n",
      "clavispora_lusitaniae            0.000133\n",
      "yarrowia_lipolytica              0.000067\n",
      "candida_albicans                 0.000067\n",
      "candida_metapsilosis             0.000067\n",
      "candida_parapsilosis             0.000067\n",
      "galactomyces_geotrichum          0.000067\n",
      "cladophialophora_unidentified    0.000067\n",
      "kodamaea_ohmeri                  0.000067\n",
      "pyrenophora_tritici-repentis     0.000067\n",
      "wickerhamomyces_anomalus         0.000067\n",
      "yamadazyma_mexicana              0.000067\n",
      "asteroma_ccl060                  0.000067\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the query species: pichia_membranifaciens\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pichia_membranifaciens           0.994269\n",
      "pichia_kudriavzevii              0.001933\n",
      "candida_orthopsilosis            0.000400\n",
      "meyerozyma_guilliermondii        0.000400\n",
      "kodamaea_ohmeri                  0.000333\n",
      "candida_metapsilosis             0.000333\n",
      "candida_parapsilosis             0.000267\n",
      "saccharomyces_cerevisiae         0.000200\n",
      "clavispora_lusitaniae            0.000200\n",
      "kluyveromyces_unidentified       0.000200\n",
      "galactomyces_geotrichum          0.000133\n",
      "cladophialophora_unidentified    0.000133\n",
      "debaryomyces_unidentified        0.000133\n",
      "zymoseptoria_tritici             0.000133\n",
      "kluyveromyces_marxianus          0.000133\n",
      "yarrowia_lipolytica              0.000133\n",
      "quambalaria_cyanescens           0.000133\n",
      "wickerhamomyces_anomalus         0.000133\n",
      "candida_albicans                 0.000133\n",
      "oculimacula_yallundae-ccl029     0.000067\n",
      "candida_unidentified             0.000067\n",
      "rhodotorula_mucilaginosa         0.000067\n",
      "tuber_brumale                    0.000067\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running against a database that have ['Candida_orthopsilosis', 'Candida_metapsilosis', 'Aspergillus_niger'] deleted\n",
    "for species, hit_fn in sub_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pull in mapping results and analyse them at all available levels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### idea\n",
    "\n",
    "* pull in query taxid as a dictionary\n",
    "* assign taxid for each tname species from minimap2\n",
    "* generate a summary dictionary that checks concordance at each taxonmic rank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['mquality'].count().tolist(), sub_df.groupby('tname')['mquality'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "    return hit_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getquery_taxfileid(refdf_fn, species):\n",
    "    \"\"\"\n",
    "    Takes the reference dataframe filename and the species name.\n",
    "    Returns the taxfileid, which is the date/flowcellid (column 0 value) of the ref_df.\n",
    "    \"\"\"\n",
    "    ref_df = pd.read_csv(refdf_fn)\n",
    "    ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']\n",
    "    return ref_df[ref_df.name_species == species].iloc[:,0].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxid_dict(taxid_fn, taxfileid):\n",
    "    \"\"\"\n",
    "    Takes a taxonomy assignment file filename in the Qiime format and a taxonomic identifier.\n",
    "    Returns the a dictionary with the taxonomic assignment at each rank.\n",
    "    \"\"\"\n",
    "    tax_dict = {}\n",
    "    with open(taxid_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(taxfileid):\n",
    "                taxrankids = line.rstrip().split('\\t')[1].split(';')\n",
    "                for taxrank in taxrankids:\n",
    "                    tax_dict[taxrank.split('__')[0]] = taxrank.split('__')[1]\n",
    "    return tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_taxranks_results(mapping_df, tax_fn, ref_df_fn = False):\n",
    "    \"\"\"\n",
    "    This function assigns the taxonomic ranks for each hit in the mapping results dataframe.\n",
    "    It takes a mapping_df, taxonomy assignment file, and if required a reference dataframe filename.\n",
    "    Returns the mapping dataframe with assignment. \n",
    "    \"\"\"\n",
    "    for tname in mapping_df.index:\n",
    "        if ref_df_fn:\n",
    "            tmp_taxfileid = getquery_taxfileid(ref_df_fn, tname)\n",
    "        else:\n",
    "            tmp_taxfileid = tname\n",
    "        tmp_tax_dict = get_taxid_dict(tax_fn, tmp_taxfileid)\n",
    "        for key, value in tmp_tax_dict.items():\n",
    "            mapping_df.loc[tname, key] = value\n",
    "    return mapping_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarieses the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = {}\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9992679355783309,\n",
       " 'c': 0.9954079595368028,\n",
       " 'o': 0.9952083056036204,\n",
       " 'f': 0.9952083056036204,\n",
       " 'g': 0.9915479834952748,\n",
       " 's': 0.9915479834952748}"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'candida_albicans'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9995496364923117,\n",
       " 'c': 0.995560702567072,\n",
       " 'o': 0.995560702567072,\n",
       " 'f': 0.9855240301100173,\n",
       " 'g': 0.9855240301100173,\n",
       " 's': 0.49501383259345044}"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'aspergillus_niger'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9993412384716732,\n",
       " 'c': 0.9960474308300395,\n",
       " 'o': 0.9957839262187088,\n",
       " 'f': 0.9957839262187088,\n",
       " 'g': 0.9946640316205534,\n",
       " 's': 0.8432147562582345}"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test run on the qiime2 Database\n",
    "\n",
    "##### Prep on the command line\n",
    "\n",
    "cp sh_refs_qiime_ver8_dynamic_02.02.2019.fasta /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/.  \n",
    "cp sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results_v2(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    #sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['mquality'].count().tolist(), sub_df.groupby('tname')['mquality'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/media/MassStorage/tmp/TE/honours/scripts/Notebooks'"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.abspath(os.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run on test species 'penicillium_chrysogenum'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=20000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/penicillium_chrysogenum.20000.fasta!!\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 /media/MassStorage/tmp/TE/honours/analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta /media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/penicillium_chrysogenum.20000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.penicillium_chrysogenum.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.8811805461825708,\n",
       " 'c': 0.1900391415845593,\n",
       " 'o': 0.18913933504296576,\n",
       " 'f': 0.1882395285013722,\n",
       " 'g': 0.08134251136005759,\n",
       " 's': 0.0038691681288522965}"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>qlen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qstop</th>\n",
       "      <th>strand</th>\n",
       "      <th>tname</th>\n",
       "      <th>tlen</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>nmatch</th>\n",
       "      <th>alen</th>\n",
       "      <th>mquality</th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>SH1732842.08FU_FJ430779_refs_singleton</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>81</td>\n",
       "      <td>985</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1732842.08FU_FJ430779_refs_singleton</td>\n",
       "      <td>1661</td>\n",
       "      <td>500</td>\n",
       "      <td>1471</td>\n",
       "      <td>210</td>\n",
       "      <td>974</td>\n",
       "      <td>35</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Leotiomycetes</td>\n",
       "      <td>Helotiales</td>\n",
       "      <td>Helotiaceae</td>\n",
       "      <td>Acidea</td>\n",
       "      <td>Acidea_extrema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1654757.08FU_UDB014954_reps</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>1679</td>\n",
       "      <td>2520</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1654757.08FU_UDB014954_reps</td>\n",
       "      <td>1390</td>\n",
       "      <td>493</td>\n",
       "      <td>1356</td>\n",
       "      <td>60</td>\n",
       "      <td>864</td>\n",
       "      <td>13</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Orbiliomycetes</td>\n",
       "      <td>Orbiliales</td>\n",
       "      <td>unidentified</td>\n",
       "      <td>unidentified</td>\n",
       "      <td>unidentified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH2190001.08FU_AY371635_refs</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>1223</td>\n",
       "      <td>1666</td>\n",
       "      <td>+</td>\n",
       "      <td>SH2190001.08FU_AY371635_refs</td>\n",
       "      <td>516</td>\n",
       "      <td>32</td>\n",
       "      <td>508</td>\n",
       "      <td>51</td>\n",
       "      <td>476</td>\n",
       "      <td>0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_aethiopicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1560162.08FU_AY225488_reps</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>1696</td>\n",
       "      <td>2475</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1560162.08FU_AY225488_reps</td>\n",
       "      <td>1828</td>\n",
       "      <td>587</td>\n",
       "      <td>1376</td>\n",
       "      <td>45</td>\n",
       "      <td>789</td>\n",
       "      <td>0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Sordariomycetes</td>\n",
       "      <td>Microascales</td>\n",
       "      <td>Halosphaeriaceae</td>\n",
       "      <td>Lignincola</td>\n",
       "      <td>Lignincola_laevis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1900984.08FU_AF263347_refs_singleton</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>1496</td>\n",
       "      <td>1648</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1900984.08FU_AF263347_refs_singleton</td>\n",
       "      <td>507</td>\n",
       "      <td>332</td>\n",
       "      <td>500</td>\n",
       "      <td>44</td>\n",
       "      <td>168</td>\n",
       "      <td>0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_bovifimosum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH2189929.08FU_NR_103692_refs</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>1341</td>\n",
       "      <td>1736</td>\n",
       "      <td>+</td>\n",
       "      <td>SH2189929.08FU_NR_103692_refs</td>\n",
       "      <td>495</td>\n",
       "      <td>32</td>\n",
       "      <td>424</td>\n",
       "      <td>141</td>\n",
       "      <td>400</td>\n",
       "      <td>0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_griseofulvum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH2190041.08FU_KM973207_refs</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>1341</td>\n",
       "      <td>1736</td>\n",
       "      <td>+</td>\n",
       "      <td>SH2190041.08FU_KM973207_refs</td>\n",
       "      <td>495</td>\n",
       "      <td>32</td>\n",
       "      <td>424</td>\n",
       "      <td>141</td>\n",
       "      <td>400</td>\n",
       "      <td>0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_compactum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1649638.08FU_UDB016769_refs</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>2624</td>\n",
       "      <td>2874</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1649638.08FU_UDB016769_refs</td>\n",
       "      <td>1720</td>\n",
       "      <td>1429</td>\n",
       "      <td>1681</td>\n",
       "      <td>101</td>\n",
       "      <td>256</td>\n",
       "      <td>60</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Basidiomycota</td>\n",
       "      <td>Agaricomycetes</td>\n",
       "      <td>Thelephorales</td>\n",
       "      <td>Thelephoraceae</td>\n",
       "      <td>Tomentella</td>\n",
       "      <td>unidentified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1692805.08FU_JN617705_refs</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>1660</td>\n",
       "      <td>1819</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1692805.08FU_JN617705_refs</td>\n",
       "      <td>510</td>\n",
       "      <td>345</td>\n",
       "      <td>506</td>\n",
       "      <td>49</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_rolfsii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1558908.08FU_KF800655_reps</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>1124</td>\n",
       "      <td>1274</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1558908.08FU_KF800655_reps</td>\n",
       "      <td>1223</td>\n",
       "      <td>54</td>\n",
       "      <td>206</td>\n",
       "      <td>46</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporales_fam_Incertae_sedis</td>\n",
       "      <td>Pseudorobillarda</td>\n",
       "      <td>unidentified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163502 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                      qseqid  \\\n",
       "tname                                                                          \n",
       "SH1732842.08FU_FJ430779_refs_singleton  4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "SH1654757.08FU_UDB014954_reps           4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "SH2190001.08FU_AY371635_refs            4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "SH1560162.08FU_AY225488_reps            4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "SH1900984.08FU_AF263347_refs_singleton  4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "...                                                                      ...   \n",
       "SH2189929.08FU_NR_103692_refs           f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "SH2190041.08FU_KM973207_refs            f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "SH1649638.08FU_UDB016769_refs           f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "SH1692805.08FU_JN617705_refs            f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "SH1558908.08FU_KF800655_reps            f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "\n",
       "                                        qlen  qstart  qstop strand  \\\n",
       "tname                                                                \n",
       "SH1732842.08FU_FJ430779_refs_singleton  2739      81    985      +   \n",
       "SH1654757.08FU_UDB014954_reps           2739    1679   2520      +   \n",
       "SH2190001.08FU_AY371635_refs            2739    1223   1666      +   \n",
       "SH1560162.08FU_AY225488_reps            2739    1696   2475      +   \n",
       "SH1900984.08FU_AF263347_refs_singleton  2739    1496   1648      +   \n",
       "...                                      ...     ...    ...    ...   \n",
       "SH2189929.08FU_NR_103692_refs           3007    1341   1736      +   \n",
       "SH2190041.08FU_KM973207_refs            3007    1341   1736      +   \n",
       "SH1649638.08FU_UDB016769_refs           3007    2624   2874      +   \n",
       "SH1692805.08FU_JN617705_refs            3007    1660   1819      +   \n",
       "SH1558908.08FU_KF800655_reps            3007    1124   1274      +   \n",
       "\n",
       "                                                                         tname  \\\n",
       "tname                                                                            \n",
       "SH1732842.08FU_FJ430779_refs_singleton  SH1732842.08FU_FJ430779_refs_singleton   \n",
       "SH1654757.08FU_UDB014954_reps                    SH1654757.08FU_UDB014954_reps   \n",
       "SH2190001.08FU_AY371635_refs                      SH2190001.08FU_AY371635_refs   \n",
       "SH1560162.08FU_AY225488_reps                      SH1560162.08FU_AY225488_reps   \n",
       "SH1900984.08FU_AF263347_refs_singleton  SH1900984.08FU_AF263347_refs_singleton   \n",
       "...                                                                        ...   \n",
       "SH2189929.08FU_NR_103692_refs                    SH2189929.08FU_NR_103692_refs   \n",
       "SH2190041.08FU_KM973207_refs                      SH2190041.08FU_KM973207_refs   \n",
       "SH1649638.08FU_UDB016769_refs                    SH1649638.08FU_UDB016769_refs   \n",
       "SH1692805.08FU_JN617705_refs                      SH1692805.08FU_JN617705_refs   \n",
       "SH1558908.08FU_KF800655_reps                      SH1558908.08FU_KF800655_reps   \n",
       "\n",
       "                                        tlen  tstart  tend  nmatch  alen  \\\n",
       "tname                                                                      \n",
       "SH1732842.08FU_FJ430779_refs_singleton  1661     500  1471     210   974   \n",
       "SH1654757.08FU_UDB014954_reps           1390     493  1356      60   864   \n",
       "SH2190001.08FU_AY371635_refs             516      32   508      51   476   \n",
       "SH1560162.08FU_AY225488_reps            1828     587  1376      45   789   \n",
       "SH1900984.08FU_AF263347_refs_singleton   507     332   500      44   168   \n",
       "...                                      ...     ...   ...     ...   ...   \n",
       "SH2189929.08FU_NR_103692_refs            495      32   424     141   400   \n",
       "SH2190041.08FU_KM973207_refs             495      32   424     141   400   \n",
       "SH1649638.08FU_UDB016769_refs           1720    1429  1681     101   256   \n",
       "SH1692805.08FU_JN617705_refs             510     345   506      49   161   \n",
       "SH1558908.08FU_KF800655_reps            1223      54   206      46   152   \n",
       "\n",
       "                                        mquality      k              p  \\\n",
       "tname                                                                    \n",
       "SH1732842.08FU_FJ430779_refs_singleton        35  Fungi     Ascomycota   \n",
       "SH1654757.08FU_UDB014954_reps                 13  Fungi     Ascomycota   \n",
       "SH2190001.08FU_AY371635_refs                   0  Fungi     Ascomycota   \n",
       "SH1560162.08FU_AY225488_reps                   0  Fungi     Ascomycota   \n",
       "SH1900984.08FU_AF263347_refs_singleton         0  Fungi     Ascomycota   \n",
       "...                                          ...    ...            ...   \n",
       "SH2189929.08FU_NR_103692_refs                  0  Fungi     Ascomycota   \n",
       "SH2190041.08FU_KM973207_refs                   0  Fungi     Ascomycota   \n",
       "SH1649638.08FU_UDB016769_refs                 60  Fungi  Basidiomycota   \n",
       "SH1692805.08FU_JN617705_refs                   0  Fungi     Ascomycota   \n",
       "SH1558908.08FU_KF800655_reps                   0  Fungi     Ascomycota   \n",
       "\n",
       "                                                      c              o  \\\n",
       "tname                                                                    \n",
       "SH1732842.08FU_FJ430779_refs_singleton    Leotiomycetes     Helotiales   \n",
       "SH1654757.08FU_UDB014954_reps            Orbiliomycetes     Orbiliales   \n",
       "SH2190001.08FU_AY371635_refs             Eurotiomycetes     Eurotiales   \n",
       "SH1560162.08FU_AY225488_reps            Sordariomycetes   Microascales   \n",
       "SH1900984.08FU_AF263347_refs_singleton   Eurotiomycetes     Eurotiales   \n",
       "...                                                 ...            ...   \n",
       "SH2189929.08FU_NR_103692_refs            Eurotiomycetes     Eurotiales   \n",
       "SH2190041.08FU_KM973207_refs             Eurotiomycetes     Eurotiales   \n",
       "SH1649638.08FU_UDB016769_refs            Agaricomycetes  Thelephorales   \n",
       "SH1692805.08FU_JN617705_refs             Eurotiomycetes     Eurotiales   \n",
       "SH1558908.08FU_KF800655_reps            Dothideomycetes   Pleosporales   \n",
       "\n",
       "                                                                      f  \\\n",
       "tname                                                                     \n",
       "SH1732842.08FU_FJ430779_refs_singleton                      Helotiaceae   \n",
       "SH1654757.08FU_UDB014954_reps                              unidentified   \n",
       "SH2190001.08FU_AY371635_refs                             Aspergillaceae   \n",
       "SH1560162.08FU_AY225488_reps                           Halosphaeriaceae   \n",
       "SH1900984.08FU_AF263347_refs_singleton                   Aspergillaceae   \n",
       "...                                                                 ...   \n",
       "SH2189929.08FU_NR_103692_refs                            Aspergillaceae   \n",
       "SH2190041.08FU_KM973207_refs                             Aspergillaceae   \n",
       "SH1649638.08FU_UDB016769_refs                            Thelephoraceae   \n",
       "SH1692805.08FU_JN617705_refs                             Aspergillaceae   \n",
       "SH1558908.08FU_KF800655_reps            Pleosporales_fam_Incertae_sedis   \n",
       "\n",
       "                                                       g  \\\n",
       "tname                                                      \n",
       "SH1732842.08FU_FJ430779_refs_singleton            Acidea   \n",
       "SH1654757.08FU_UDB014954_reps               unidentified   \n",
       "SH2190001.08FU_AY371635_refs                 Penicillium   \n",
       "SH1560162.08FU_AY225488_reps                  Lignincola   \n",
       "SH1900984.08FU_AF263347_refs_singleton       Penicillium   \n",
       "...                                                  ...   \n",
       "SH2189929.08FU_NR_103692_refs                Penicillium   \n",
       "SH2190041.08FU_KM973207_refs                 Penicillium   \n",
       "SH1649638.08FU_UDB016769_refs                 Tomentella   \n",
       "SH1692805.08FU_JN617705_refs                 Penicillium   \n",
       "SH1558908.08FU_KF800655_reps            Pseudorobillarda   \n",
       "\n",
       "                                                               s  \n",
       "tname                                                             \n",
       "SH1732842.08FU_FJ430779_refs_singleton            Acidea_extrema  \n",
       "SH1654757.08FU_UDB014954_reps                       unidentified  \n",
       "SH2190001.08FU_AY371635_refs             Penicillium_aethiopicum  \n",
       "SH1560162.08FU_AY225488_reps                   Lignincola_laevis  \n",
       "SH1900984.08FU_AF263347_refs_singleton   Penicillium_bovifimosum  \n",
       "...                                                          ...  \n",
       "SH2189929.08FU_NR_103692_refs           Penicillium_griseofulvum  \n",
       "SH2190041.08FU_KM973207_refs               Penicillium_compactum  \n",
       "SH1649638.08FU_UDB016769_refs                       unidentified  \n",
       "SH1692805.08FU_JN617705_refs                 Penicillium_rolfsii  \n",
       "SH1558908.08FU_KF800655_reps                        unidentified  \n",
       "\n",
       "[163502 rows x 19 columns]"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9115974116524569,\n",
       " 'c': 0.4380680358650047,\n",
       " 'o': 0.432655258039657,\n",
       " 'f': 0.43007424985627085,\n",
       " 'g': 0.38932857090433143,\n",
       " 's': 0.030183117026091423}"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### These are wired results that might be linked to\n",
    "\n",
    "* database issues as you can see\n",
    "\n",
    "sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt: k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>SH1530047.08FU_HE649392_reps</td>\n",
       "      <td>52</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_chrysogenum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH2189908.08FU_AY213669_refs</td>\n",
       "      <td>34</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_chrysogenum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              count      k           p               c  \\\n",
       "tname                                                                    \n",
       "SH1530047.08FU_HE649392_reps     52  Fungi  Ascomycota  Eurotiomycetes   \n",
       "SH2189908.08FU_AY213669_refs     34  Fungi  Ascomycota  Eurotiomycetes   \n",
       "\n",
       "                                       o               f            g  \\\n",
       "tname                                                                   \n",
       "SH1530047.08FU_HE649392_reps  Eurotiales  Aspergillaceae  Penicillium   \n",
       "SH2189908.08FU_AY213669_refs  Eurotiales  Aspergillaceae  Penicillium   \n",
       "\n",
       "                                                    s  \n",
       "tname                                                  \n",
       "SH1530047.08FU_HE649392_reps  Penicillium_chrysogenum  \n",
       "SH2189908.08FU_AY213669_refs  Penicillium_chrysogenum  "
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_results[mapping_results['s'] == 'Penicillium_chrysogenum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 'Fungi',\n",
       " 'p': 'Ascomycota',\n",
       " 'c': 'Eurotiomycetes',\n",
       " 'o': 'Eurotiales',\n",
       " 'f': 'Aspergillaceae',\n",
       " 'g': 'Penicillium',\n",
       " 's': 'Penicillium_chrysogenum'}"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]\n",
    "query_tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.8811805461825708,\n",
       " 'c': 0.1900391415845593,\n",
       " 'o': 0.18913933504296576,\n",
       " 'f': 0.1882395285013722,\n",
       " 'g': 0.08134251136005759,\n",
       " 's': 0.0038691681288522965}"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=20000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/candida_albicans.20000.fasta!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 /media/MassStorage/tmp/TE/honours/analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta /media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/candida_albicans.20000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.candida_albicans.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.candida_albicans.minimap2.paf\n",
      "20180108_FAH18647/barcode03\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetales_fam_Incertae_sedis', 'g': 'Candida', 's': 'Candida_albicans'}\n"
     ]
    }
   ],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.8175055999323781,\n",
       " 'c': 0.5192088246481552,\n",
       " 'o': 0.5192088246481552,\n",
       " 'f': 0.4455855627403745,\n",
       " 'g': 0.4455855627403745,\n",
       " 's': 0.4186636236845442}"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/media/MassStorage/tmp/TE/honours/scripts/Notebooks'"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.abspath(os.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run on test species 'penicillium_chrysogenum'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=20000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode10/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/penicillium_chrysogenum.20000.fasta!!\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 /media/MassStorage/tmp/TE/honours/analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta /media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/penicillium_chrysogenum.20000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.penicillium_chrysogenum.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9551197165455781,\n",
       " 'c': 0.6736337282130203,\n",
       " 'o': 0.6466482946208081,\n",
       " 'f': 0.6438566980423034,\n",
       " 'g': 0.5488350452739702,\n",
       " 's': 0.043699223363515984}"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>qlen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qstop</th>\n",
       "      <th>strand</th>\n",
       "      <th>tname</th>\n",
       "      <th>tlen</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>nmatch</th>\n",
       "      <th>alen</th>\n",
       "      <th>mquality</th>\n",
       "      <th>cscore</th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>SH1732842.08FU_FJ430779_refs_singleton</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>81</td>\n",
       "      <td>985</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1732842.08FU_FJ430779_refs_singleton</td>\n",
       "      <td>1661</td>\n",
       "      <td>500</td>\n",
       "      <td>1471</td>\n",
       "      <td>210</td>\n",
       "      <td>974</td>\n",
       "      <td>35</td>\n",
       "      <td>1.274869</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Leotiomycetes</td>\n",
       "      <td>Helotiales</td>\n",
       "      <td>Helotiaceae</td>\n",
       "      <td>Acidea</td>\n",
       "      <td>Acidea_extrema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1654757.08FU_UDB014954_reps</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>1679</td>\n",
       "      <td>2520</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1654757.08FU_UDB014954_reps</td>\n",
       "      <td>1390</td>\n",
       "      <td>493</td>\n",
       "      <td>1356</td>\n",
       "      <td>60</td>\n",
       "      <td>864</td>\n",
       "      <td>13</td>\n",
       "      <td>1.074627</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Orbiliomycetes</td>\n",
       "      <td>Orbiliales</td>\n",
       "      <td>unidentified</td>\n",
       "      <td>unidentified</td>\n",
       "      <td>unidentified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH2190001.08FU_AY371635_refs</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>1223</td>\n",
       "      <td>1666</td>\n",
       "      <td>+</td>\n",
       "      <td>SH2190001.08FU_AY371635_refs</td>\n",
       "      <td>516</td>\n",
       "      <td>32</td>\n",
       "      <td>508</td>\n",
       "      <td>51</td>\n",
       "      <td>476</td>\n",
       "      <td>0</td>\n",
       "      <td>1.120000</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_aethiopicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1560162.08FU_AY225488_reps</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>1696</td>\n",
       "      <td>2475</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1560162.08FU_AY225488_reps</td>\n",
       "      <td>1828</td>\n",
       "      <td>587</td>\n",
       "      <td>1376</td>\n",
       "      <td>45</td>\n",
       "      <td>789</td>\n",
       "      <td>0</td>\n",
       "      <td>1.060484</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Sordariomycetes</td>\n",
       "      <td>Microascales</td>\n",
       "      <td>Halosphaeriaceae</td>\n",
       "      <td>Lignincola</td>\n",
       "      <td>Lignincola_laevis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1900984.08FU_AF263347_refs_singleton</td>\n",
       "      <td>4d0876bf-d131-46b6-8075-14de888dded8</td>\n",
       "      <td>2739</td>\n",
       "      <td>1496</td>\n",
       "      <td>1648</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1900984.08FU_AF263347_refs_singleton</td>\n",
       "      <td>507</td>\n",
       "      <td>332</td>\n",
       "      <td>500</td>\n",
       "      <td>44</td>\n",
       "      <td>168</td>\n",
       "      <td>0</td>\n",
       "      <td>1.354839</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_bovifimosum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH2189929.08FU_NR_103692_refs</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>1341</td>\n",
       "      <td>1736</td>\n",
       "      <td>+</td>\n",
       "      <td>SH2189929.08FU_NR_103692_refs</td>\n",
       "      <td>495</td>\n",
       "      <td>32</td>\n",
       "      <td>424</td>\n",
       "      <td>141</td>\n",
       "      <td>400</td>\n",
       "      <td>0</td>\n",
       "      <td>1.544402</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_griseofulvum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH2190041.08FU_KM973207_refs</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>1341</td>\n",
       "      <td>1736</td>\n",
       "      <td>+</td>\n",
       "      <td>SH2190041.08FU_KM973207_refs</td>\n",
       "      <td>495</td>\n",
       "      <td>32</td>\n",
       "      <td>424</td>\n",
       "      <td>141</td>\n",
       "      <td>400</td>\n",
       "      <td>0</td>\n",
       "      <td>1.544402</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_compactum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1649638.08FU_UDB016769_refs</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>2624</td>\n",
       "      <td>2874</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1649638.08FU_UDB016769_refs</td>\n",
       "      <td>1720</td>\n",
       "      <td>1429</td>\n",
       "      <td>1681</td>\n",
       "      <td>101</td>\n",
       "      <td>256</td>\n",
       "      <td>60</td>\n",
       "      <td>1.651613</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Basidiomycota</td>\n",
       "      <td>Agaricomycetes</td>\n",
       "      <td>Thelephorales</td>\n",
       "      <td>Thelephoraceae</td>\n",
       "      <td>Tomentella</td>\n",
       "      <td>unidentified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1692805.08FU_JN617705_refs</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>1660</td>\n",
       "      <td>1819</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1692805.08FU_JN617705_refs</td>\n",
       "      <td>510</td>\n",
       "      <td>345</td>\n",
       "      <td>506</td>\n",
       "      <td>49</td>\n",
       "      <td>161</td>\n",
       "      <td>0</td>\n",
       "      <td>1.437500</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_rolfsii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1558908.08FU_KF800655_reps</td>\n",
       "      <td>f15f2f35-a853-4012-b6cc-aac9c680cb50</td>\n",
       "      <td>3007</td>\n",
       "      <td>1124</td>\n",
       "      <td>1274</td>\n",
       "      <td>+</td>\n",
       "      <td>SH1558908.08FU_KF800655_reps</td>\n",
       "      <td>1223</td>\n",
       "      <td>54</td>\n",
       "      <td>206</td>\n",
       "      <td>46</td>\n",
       "      <td>152</td>\n",
       "      <td>0</td>\n",
       "      <td>1.433962</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Dothideomycetes</td>\n",
       "      <td>Pleosporales</td>\n",
       "      <td>Pleosporales_fam_Incertae_sedis</td>\n",
       "      <td>Pseudorobillarda</td>\n",
       "      <td>unidentified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163502 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                      qseqid  \\\n",
       "tname                                                                          \n",
       "SH1732842.08FU_FJ430779_refs_singleton  4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "SH1654757.08FU_UDB014954_reps           4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "SH2190001.08FU_AY371635_refs            4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "SH1560162.08FU_AY225488_reps            4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "SH1900984.08FU_AF263347_refs_singleton  4d0876bf-d131-46b6-8075-14de888dded8   \n",
       "...                                                                      ...   \n",
       "SH2189929.08FU_NR_103692_refs           f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "SH2190041.08FU_KM973207_refs            f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "SH1649638.08FU_UDB016769_refs           f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "SH1692805.08FU_JN617705_refs            f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "SH1558908.08FU_KF800655_reps            f15f2f35-a853-4012-b6cc-aac9c680cb50   \n",
       "\n",
       "                                        qlen  qstart  qstop strand  \\\n",
       "tname                                                                \n",
       "SH1732842.08FU_FJ430779_refs_singleton  2739      81    985      +   \n",
       "SH1654757.08FU_UDB014954_reps           2739    1679   2520      +   \n",
       "SH2190001.08FU_AY371635_refs            2739    1223   1666      +   \n",
       "SH1560162.08FU_AY225488_reps            2739    1696   2475      +   \n",
       "SH1900984.08FU_AF263347_refs_singleton  2739    1496   1648      +   \n",
       "...                                      ...     ...    ...    ...   \n",
       "SH2189929.08FU_NR_103692_refs           3007    1341   1736      +   \n",
       "SH2190041.08FU_KM973207_refs            3007    1341   1736      +   \n",
       "SH1649638.08FU_UDB016769_refs           3007    2624   2874      +   \n",
       "SH1692805.08FU_JN617705_refs            3007    1660   1819      +   \n",
       "SH1558908.08FU_KF800655_reps            3007    1124   1274      +   \n",
       "\n",
       "                                                                         tname  \\\n",
       "tname                                                                            \n",
       "SH1732842.08FU_FJ430779_refs_singleton  SH1732842.08FU_FJ430779_refs_singleton   \n",
       "SH1654757.08FU_UDB014954_reps                    SH1654757.08FU_UDB014954_reps   \n",
       "SH2190001.08FU_AY371635_refs                      SH2190001.08FU_AY371635_refs   \n",
       "SH1560162.08FU_AY225488_reps                      SH1560162.08FU_AY225488_reps   \n",
       "SH1900984.08FU_AF263347_refs_singleton  SH1900984.08FU_AF263347_refs_singleton   \n",
       "...                                                                        ...   \n",
       "SH2189929.08FU_NR_103692_refs                    SH2189929.08FU_NR_103692_refs   \n",
       "SH2190041.08FU_KM973207_refs                      SH2190041.08FU_KM973207_refs   \n",
       "SH1649638.08FU_UDB016769_refs                    SH1649638.08FU_UDB016769_refs   \n",
       "SH1692805.08FU_JN617705_refs                      SH1692805.08FU_JN617705_refs   \n",
       "SH1558908.08FU_KF800655_reps                      SH1558908.08FU_KF800655_reps   \n",
       "\n",
       "                                        tlen  tstart  tend  nmatch  alen  \\\n",
       "tname                                                                      \n",
       "SH1732842.08FU_FJ430779_refs_singleton  1661     500  1471     210   974   \n",
       "SH1654757.08FU_UDB014954_reps           1390     493  1356      60   864   \n",
       "SH2190001.08FU_AY371635_refs             516      32   508      51   476   \n",
       "SH1560162.08FU_AY225488_reps            1828     587  1376      45   789   \n",
       "SH1900984.08FU_AF263347_refs_singleton   507     332   500      44   168   \n",
       "...                                      ...     ...   ...     ...   ...   \n",
       "SH2189929.08FU_NR_103692_refs            495      32   424     141   400   \n",
       "SH2190041.08FU_KM973207_refs             495      32   424     141   400   \n",
       "SH1649638.08FU_UDB016769_refs           1720    1429  1681     101   256   \n",
       "SH1692805.08FU_JN617705_refs             510     345   506      49   161   \n",
       "SH1558908.08FU_KF800655_reps            1223      54   206      46   152   \n",
       "\n",
       "                                        mquality    cscore      k  \\\n",
       "tname                                                               \n",
       "SH1732842.08FU_FJ430779_refs_singleton        35  1.274869  Fungi   \n",
       "SH1654757.08FU_UDB014954_reps                 13  1.074627  Fungi   \n",
       "SH2190001.08FU_AY371635_refs                   0  1.120000  Fungi   \n",
       "SH1560162.08FU_AY225488_reps                   0  1.060484  Fungi   \n",
       "SH1900984.08FU_AF263347_refs_singleton         0  1.354839  Fungi   \n",
       "...                                          ...       ...    ...   \n",
       "SH2189929.08FU_NR_103692_refs                  0  1.544402  Fungi   \n",
       "SH2190041.08FU_KM973207_refs                   0  1.544402  Fungi   \n",
       "SH1649638.08FU_UDB016769_refs                 60  1.651613  Fungi   \n",
       "SH1692805.08FU_JN617705_refs                   0  1.437500  Fungi   \n",
       "SH1558908.08FU_KF800655_reps                   0  1.433962  Fungi   \n",
       "\n",
       "                                                    p                c  \\\n",
       "tname                                                                    \n",
       "SH1732842.08FU_FJ430779_refs_singleton     Ascomycota    Leotiomycetes   \n",
       "SH1654757.08FU_UDB014954_reps              Ascomycota   Orbiliomycetes   \n",
       "SH2190001.08FU_AY371635_refs               Ascomycota   Eurotiomycetes   \n",
       "SH1560162.08FU_AY225488_reps               Ascomycota  Sordariomycetes   \n",
       "SH1900984.08FU_AF263347_refs_singleton     Ascomycota   Eurotiomycetes   \n",
       "...                                               ...              ...   \n",
       "SH2189929.08FU_NR_103692_refs              Ascomycota   Eurotiomycetes   \n",
       "SH2190041.08FU_KM973207_refs               Ascomycota   Eurotiomycetes   \n",
       "SH1649638.08FU_UDB016769_refs           Basidiomycota   Agaricomycetes   \n",
       "SH1692805.08FU_JN617705_refs               Ascomycota   Eurotiomycetes   \n",
       "SH1558908.08FU_KF800655_reps               Ascomycota  Dothideomycetes   \n",
       "\n",
       "                                                    o  \\\n",
       "tname                                                   \n",
       "SH1732842.08FU_FJ430779_refs_singleton     Helotiales   \n",
       "SH1654757.08FU_UDB014954_reps              Orbiliales   \n",
       "SH2190001.08FU_AY371635_refs               Eurotiales   \n",
       "SH1560162.08FU_AY225488_reps             Microascales   \n",
       "SH1900984.08FU_AF263347_refs_singleton     Eurotiales   \n",
       "...                                               ...   \n",
       "SH2189929.08FU_NR_103692_refs              Eurotiales   \n",
       "SH2190041.08FU_KM973207_refs               Eurotiales   \n",
       "SH1649638.08FU_UDB016769_refs           Thelephorales   \n",
       "SH1692805.08FU_JN617705_refs               Eurotiales   \n",
       "SH1558908.08FU_KF800655_reps             Pleosporales   \n",
       "\n",
       "                                                                      f  \\\n",
       "tname                                                                     \n",
       "SH1732842.08FU_FJ430779_refs_singleton                      Helotiaceae   \n",
       "SH1654757.08FU_UDB014954_reps                              unidentified   \n",
       "SH2190001.08FU_AY371635_refs                             Aspergillaceae   \n",
       "SH1560162.08FU_AY225488_reps                           Halosphaeriaceae   \n",
       "SH1900984.08FU_AF263347_refs_singleton                   Aspergillaceae   \n",
       "...                                                                 ...   \n",
       "SH2189929.08FU_NR_103692_refs                            Aspergillaceae   \n",
       "SH2190041.08FU_KM973207_refs                             Aspergillaceae   \n",
       "SH1649638.08FU_UDB016769_refs                            Thelephoraceae   \n",
       "SH1692805.08FU_JN617705_refs                             Aspergillaceae   \n",
       "SH1558908.08FU_KF800655_reps            Pleosporales_fam_Incertae_sedis   \n",
       "\n",
       "                                                       g  \\\n",
       "tname                                                      \n",
       "SH1732842.08FU_FJ430779_refs_singleton            Acidea   \n",
       "SH1654757.08FU_UDB014954_reps               unidentified   \n",
       "SH2190001.08FU_AY371635_refs                 Penicillium   \n",
       "SH1560162.08FU_AY225488_reps                  Lignincola   \n",
       "SH1900984.08FU_AF263347_refs_singleton       Penicillium   \n",
       "...                                                  ...   \n",
       "SH2189929.08FU_NR_103692_refs                Penicillium   \n",
       "SH2190041.08FU_KM973207_refs                 Penicillium   \n",
       "SH1649638.08FU_UDB016769_refs                 Tomentella   \n",
       "SH1692805.08FU_JN617705_refs                 Penicillium   \n",
       "SH1558908.08FU_KF800655_reps            Pseudorobillarda   \n",
       "\n",
       "                                                               s  \n",
       "tname                                                             \n",
       "SH1732842.08FU_FJ430779_refs_singleton            Acidea_extrema  \n",
       "SH1654757.08FU_UDB014954_reps                       unidentified  \n",
       "SH2190001.08FU_AY371635_refs             Penicillium_aethiopicum  \n",
       "SH1560162.08FU_AY225488_reps                   Lignincola_laevis  \n",
       "SH1900984.08FU_AF263347_refs_singleton   Penicillium_bovifimosum  \n",
       "...                                                          ...  \n",
       "SH2189929.08FU_NR_103692_refs           Penicillium_griseofulvum  \n",
       "SH2190041.08FU_KM973207_refs               Penicillium_compactum  \n",
       "SH1649638.08FU_UDB016769_refs                       unidentified  \n",
       "SH1692805.08FU_JN617705_refs                 Penicillium_rolfsii  \n",
       "SH1558908.08FU_KF800655_reps                        unidentified  \n",
       "\n",
       "[163502 rows x 20 columns]"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9115974116524569,\n",
       " 'c': 0.4380680358650047,\n",
       " 'o': 0.432655258039657,\n",
       " 'f': 0.43007424985627085,\n",
       " 'g': 0.38932857090433143,\n",
       " 's': 0.030183117026091423}"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### These are wired results that might be linked to\n",
    "\n",
    "* database issues as you can see\n",
    "\n",
    "sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt: k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tname</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>SH2189908.08FU_AY213669_refs</td>\n",
       "      <td>844</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_chrysogenum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SH1530047.08FU_HE649392_reps</td>\n",
       "      <td>377</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>Ascomycota</td>\n",
       "      <td>Eurotiomycetes</td>\n",
       "      <td>Eurotiales</td>\n",
       "      <td>Aspergillaceae</td>\n",
       "      <td>Penicillium</td>\n",
       "      <td>Penicillium_chrysogenum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              count      k           p               c  \\\n",
       "tname                                                                    \n",
       "SH2189908.08FU_AY213669_refs    844  Fungi  Ascomycota  Eurotiomycetes   \n",
       "SH1530047.08FU_HE649392_reps    377  Fungi  Ascomycota  Eurotiomycetes   \n",
       "\n",
       "                                       o               f            g  \\\n",
       "tname                                                                   \n",
       "SH2189908.08FU_AY213669_refs  Eurotiales  Aspergillaceae  Penicillium   \n",
       "SH1530047.08FU_HE649392_reps  Eurotiales  Aspergillaceae  Penicillium   \n",
       "\n",
       "                                                    s  \n",
       "tname                                                  \n",
       "SH2189908.08FU_AY213669_refs  Penicillium_chrysogenum  \n",
       "SH1530047.08FU_HE649392_reps  Penicillium_chrysogenum  "
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_results[mapping_results['s'] == 'Penicillium_chrysogenum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 'Fungi',\n",
       " 'p': 'Ascomycota',\n",
       " 'c': 'Eurotiomycetes',\n",
       " 'o': 'Eurotiales',\n",
       " 'f': 'Aspergillaceae',\n",
       " 'g': 'Penicillium',\n",
       " 's': 'Penicillium_chrysogenum'}"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]\n",
    "query_tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9551197165455781,\n",
       " 'c': 0.6736337282130203,\n",
       " 'o': 0.6466482946208081,\n",
       " 'f': 0.6438566980423034,\n",
       " 'g': 0.5488350452739702,\n",
       " 's': 0.043699223363515984}"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=20000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/candida_albicans.20000.fasta!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 /media/MassStorage/tmp/TE/honours/analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta /media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/candida_albicans.20000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.candida_albicans.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.candida_albicans.minimap2.paf\n",
      "20180108_FAH18647/barcode03\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetales_fam_Incertae_sedis', 'g': 'Candida', 's': 'Candida_albicans'}\n"
     ]
    }
   ],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9333918831468984,\n",
       " 'c': 0.5988259221595637,\n",
       " 'o': 0.5988259221595637,\n",
       " 'f': 0.3706203198668762,\n",
       " 'g': 0.3706203198668762,\n",
       " 's': 0.30248682629194784}"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Might want to double check how your families are matched in your taxonomic input file for your known test species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_results[mapping_results['s'] == 'Candida_albicans']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "# query_tax_dict['f'] = 'Saccharomycetales_fam_Incertae_sedis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### It appears that the filtering of results by mapping quality works well for the long ITS database but not for the qiime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen',\n",
       "       'tstart', 'tend', 'nmatch', 'alen', 'mquality', 'cscore', 'k', 'p', 'c',\n",
       "       'o', 'f', 'g', 's'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_results_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(133529,)"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_results_df.tname.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(166,)"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_results_df.tname.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.index = full_results_df.tname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Asign taxonomic ranks to the full_results_df\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28992,)"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_results_df[full_results_df['g'] == 'Candida']['qseqid'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18016,)"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_results_df[full_results_df['g'] == 'Candida']['qseqid'].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.8167813733346314,\n",
       " 'c': 0.2981150162137064,\n",
       " 'o': 0.2981150162137064,\n",
       " 'f': 0.2171213743830928,\n",
       " 'g': 0.2171213743830928,\n",
       " 's': 0.17868028667929814}"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at the results unfiltered doesn't really work very well either. Might need to look into different filtering of the alignments or the qiime2 database might be just not really useful for the noise reads. Simulated reads with higher accuracy should get better here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "g\n",
       "Candida                28992\n",
       "Acidea                 16877\n",
       "Plectosphaerella       12234\n",
       "unidentified           10983\n",
       "Nakaseomyces           10297\n",
       "Oidiodendron            9564\n",
       "Tomentella              8986\n",
       "Knoxdaviesia            7567\n",
       "Lignincola              6155\n",
       "Lacrymaria              5027\n",
       "Cystoagaricus           4941\n",
       "Bannoa                  4640\n",
       "Pseudorobillarda        2835\n",
       "Deltopyxis               876\n",
       "Pestalotiopsis           712\n",
       "Leveillula               668\n",
       "Wickerhamiella           424\n",
       "Orpinomyces              384\n",
       "Sporisorium              146\n",
       "Phyllactinia             134\n",
       "Exophiala                132\n",
       "Pyrenodesmia             131\n",
       "Physcia                  126\n",
       "Physconia                123\n",
       "Rhinocladiella           111\n",
       "Pseudophaeomoniella      105\n",
       "Sclerophora              104\n",
       "Conioscypha               40\n",
       "Chromocleista             39\n",
       "Fusarium                  31\n",
       "Rhodotorula               26\n",
       "Xylaria                   19\n",
       "Lodderomyces              15\n",
       "Scheffersomyces           15\n",
       "Spathaspora               11\n",
       "Wickerhamomyces            9\n",
       "Babjeviella                7\n",
       "Obolarina                  7\n",
       "Saccharomyces              6\n",
       "Yamadazyma                 5\n",
       "Nematodospora              4\n",
       "Kluyveromyces              3\n",
       "Barnettozyma               2\n",
       "Strobilomyces              2\n",
       "Pichia                     2\n",
       "Tetrapisispora             1\n",
       "Priceomyces                1\n",
       "Cyberlindnera              1\n",
       "Galiella                   1\n",
       "Pileolaria                 1\n",
       "Histoplasma                1\n",
       "Issatchenkia               1\n",
       "Peziza                     1\n",
       "Lactarius                  1\n",
       "Saccharomycopsis           1\n",
       "Nakazawaea                 1\n",
       "Schwanniomyces             1\n",
       "Name: k, dtype: int64"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_results_df.groupby('g').count()['k'].sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on other species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":(check one reformat.sh samplereadstarget=20000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode06/length_restricted_for_use.fasta out=/media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/cortinarius_globuliformis.20000.fasta!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['cortinarius_globuliformis']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 /media/MassStorage/tmp/TE/honours/analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta /media/MassStorage/tmp/TE/honours/analysis/qiime2/subsamplereads/cortinarius_globuliformis.20000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.cortinarius_globuliformis.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.cortinarius_globuliformis.minimap2.paf\n",
      "20171103_FAH15473/barcode06\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Agaricomycetes', 'o': 'Agaricales', 'f': 'Cortinariaceae', 'g': 'Cortinarius', 's': 'Cortinarius_globuliformis'}\n"
     ]
    }
   ],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.8857976113322841,\n",
       " 'c': 0.5234237570595315,\n",
       " 'o': 0.4376446625312471,\n",
       " 'f': 0.1712804369965744,\n",
       " 'g': 0.16669752800666604,\n",
       " 's': 0.08679751874826405}"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/MassStorage/tmp/TE/honours/analysis/qiime2/sh_refs_qiime_ver8_dynamic_02_02_2019/sh_refs_qiime_ver8_dynamic_02.02.2019.cortinarius_globuliformis.minimap2.paf\n",
      "20171103_FAH15473/barcode06\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Agaricomycetes', 'o': 'Agaricales', 'f': 'Cortinariaceae', 'g': 'Cortinarius', 's': 'Cortinarius_globuliformis'}\n"
     ]
    }
   ],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.9100366018108265,\n",
       " 'c': 0.8413600462338663,\n",
       " 'o': 0.7906954344057022,\n",
       " 'f': 0.4180312078597573,\n",
       " 'g': 0.4037276054710075,\n",
       " 's': 0.13441533423232518}"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'k': 1.0,\n",
       " 'p': 0.8248654952161353,\n",
       " 'c': 0.7081179931621213,\n",
       " 'o': 0.6266929050382974,\n",
       " 'f': 0.41488431263417347,\n",
       " 'g': 0.40360711351408657,\n",
       " 's': 0.08720945641514935}"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on all species using v3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "puccinia_striiformis-tritici\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9577623888807033,\n",
      " \"c\": 0.8779038357644516,\n",
      " \"o\": 0.8779038357644516,\n",
      " \"f\": 0.8760866362162959,\n",
      " \"g\": 0.8655763469377732,\n",
      " \"s\": 0.8004027307106724\n",
      "}\n",
      "\n",
      "\n",
      "zymoseptoria_tritici\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8317647609683008,\n",
      " \"c\": 0.35651074589127685,\n",
      " \"o\": 0.18654305379969097,\n",
      " \"f\": 0.18223533267781056,\n",
      " \"g\": 0.16018167345600973,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "pyrenophora_tritici-repentis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8342309619238477,\n",
      " \"c\": 0.6039579158316634,\n",
      " \"o\": 0.5966307615230461,\n",
      " \"f\": 0.3619113226452906,\n",
      " \"g\": 0.311685871743487,\n",
      " \"s\": 0.3032940881763527\n",
      "}\n",
      "\n",
      "\n",
      "fusarium_oxysporum\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.988622754491018,\n",
      " \"c\": 0.8356661676646706,\n",
      " \"o\": 0.5900074850299402,\n",
      " \"f\": 0.5712200598802395,\n",
      " \"g\": 0.4812125748502994,\n",
      " \"s\": 0.10377994011976048\n",
      "}\n",
      "\n",
      "\n",
      "tuber_brumale\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9263061411549037,\n",
      " \"c\": 0.4925756186984418,\n",
      " \"o\": 0.4925756186984418,\n",
      " \"f\": 0.4925297891842346,\n",
      " \"g\": 0.4925297891842346,\n",
      " \"s\": 0.49243813015582033\n",
      "}\n",
      "\n",
      "\n",
      "cortinarius_globuliformis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9100366018108265,\n",
      " \"c\": 0.8413600462338663,\n",
      " \"o\": 0.7906954344057022,\n",
      " \"f\": 0.4180312078597573,\n",
      " \"g\": 0.4037276054710075,\n",
      " \"s\": 0.13441533423232518\n",
      "}\n",
      "\n",
      "\n",
      "aspergillus_niger\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9446376036373362,\n",
      " \"c\": 0.6261700989569403,\n",
      " \"o\": 0.625,\n",
      " \"f\": 0.6248997058036908,\n",
      " \"g\": 0.48849959882321475,\n",
      " \"s\": 0.15886600695373093\n",
      "}\n",
      "\n",
      "\n",
      "clavispora_lusitaniae\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9184613469438566,\n",
      " \"c\": 0.5847545541308768,\n",
      " \"o\": 0.5847545541308768,\n",
      " \"f\": 0.17710278434789825,\n",
      " \"g\": 0.17631436989086685,\n",
      " \"s\": 0.1739076310220341\n",
      "}\n",
      "\n",
      "\n",
      "kluyveromyces_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8708093701781456,\n",
      " \"c\": 0.7000514331135737,\n",
      " \"o\": 0.7000514331135737,\n",
      " \"f\": 0.6105110581194183,\n",
      " \"g\": 0.5002104081918922,\n",
      " \"s\": 0.24295132557160892\n",
      "}\n",
      "\n",
      "\n",
      "penicillium_chrysogenum\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9551197165455781,\n",
      " \"c\": 0.6736337282130203,\n",
      " \"o\": 0.6466482946208081,\n",
      " \"f\": 0.6438566980423034,\n",
      " \"g\": 0.5488350452739702,\n",
      " \"s\": 0.043699223363515984\n",
      "}\n",
      "\n",
      "\n",
      "rhodotorula_mucilaginosa\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9247644494445225,\n",
      " \"c\": 0.8070126095720246,\n",
      " \"o\": 0.8067313551774247,\n",
      " \"f\": 0.8067313551774247,\n",
      " \"g\": 0.796793699901561,\n",
      " \"s\": 0.6015562743167862\n",
      "}\n",
      "\n",
      "\n",
      "scedosporium_boydii\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9902829134061398,\n",
      " \"c\": 0.8724739874451801,\n",
      " \"o\": 0.7314472439590678,\n",
      " \"f\": 0.49552842032848915,\n",
      " \"g\": 0.3026915469945825,\n",
      " \"s\": 0.29086765844010665\n",
      "}\n",
      "\n",
      "\n",
      "blastobotrys_proliferans\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9219217814371258,\n",
      " \"c\": 0.6693020209580839,\n",
      " \"o\": 0.6693020209580839,\n",
      " \"f\": 0.5442552395209581,\n",
      " \"g\": 0.5345714820359282,\n",
      " \"s\": 0.5335890718562875\n",
      "}\n",
      "\n",
      "\n",
      "debaryomyces_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9720678352572324,\n",
      " \"c\": 0.8847560685953162,\n",
      " \"o\": 0.8847560685953162,\n",
      " \"f\": 0.022659256092347157,\n",
      " \"g\": 0.0012825994014536127,\n",
      " \"s\": 0.18393425490475512\n",
      "}\n",
      "\n",
      "\n",
      "galactomyces_geotrichum\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.940779993492307,\n",
      " \"c\": 0.7722772277227723,\n",
      " \"o\": 0.7599590945009994,\n",
      " \"f\": 0.7278389810812067,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "kodamaea_ohmeri\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8883341334613538,\n",
      " \"c\": 0.6254440710513682,\n",
      " \"o\": 0.6254440710513682,\n",
      " \"f\": 0.457081132981277,\n",
      " \"g\": 0.457081132981277,\n",
      " \"s\": 0.457081132981277\n",
      "}\n",
      "\n",
      "\n",
      "meyerozyma_guilliermondii\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9638125852328073,\n",
      " \"c\": 0.7992889148646016,\n",
      " \"o\": 0.7992889148646016,\n",
      " \"f\": 0.5187999220728619,\n",
      " \"g\": 0.5096434833430742,\n",
      " \"s\": 0.44686343269043444\n",
      "}\n",
      "\n",
      "\n",
      "wickerhamomyces_anomalus\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.967509718289581,\n",
      " \"c\": 0.8223832605461439,\n",
      " \"o\": 0.8223832605461439,\n",
      " \"f\": 0.6510534145990305,\n",
      " \"g\": 0.6503815328502184,\n",
      " \"s\": 0.6232183135768105\n",
      "}\n",
      "\n",
      "\n",
      "yamadazyma_mexicana\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9457062923799137,\n",
      " \"c\": 0.7952961259424345,\n",
      " \"o\": 0.7952961259424345,\n",
      " \"f\": 0.6078998530039357,\n",
      " \"g\": 0.5967566029683722,\n",
      " \"s\": 0.5391436293802456\n",
      "}\n",
      "\n",
      "\n",
      "yamadazyma_scolyti\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9470681306036343,\n",
      " \"c\": 0.8043342762621661,\n",
      " \"o\": 0.8043342762621661,\n",
      " \"f\": 0.4469482667689505,\n",
      " \"g\": 0.4392290358153138,\n",
      " \"s\": 0.42498921225487846\n",
      "}\n",
      "\n",
      "\n",
      "yarrowia_lipolytica\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8715733590733591,\n",
      " \"c\": 0.20608108108108109,\n",
      " \"o\": 0.20608108108108109,\n",
      " \"f\": 0.0,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "zygoascus_hellenicus\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9401688897418674,\n",
      " \"c\": 0.7915267248824489,\n",
      " \"o\": 0.7915267248824489,\n",
      " \"f\": 0.7156702811630362,\n",
      " \"g\": 0.6761347279531715,\n",
      " \"s\": 0.6508492467133672\n",
      "}\n",
      "\n",
      "\n",
      "aspergillus_flavus\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9517551755175517,\n",
      " \"c\": 0.7672367236723673,\n",
      " \"o\": 0.7666966696669667,\n",
      " \"f\": 0.7666366636663666,\n",
      " \"g\": 0.6424842484248425,\n",
      " \"s\": 0.596009600960096\n",
      "}\n",
      "\n",
      "\n",
      "cryptococcus_zero\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.885050025192543,\n",
      " \"c\": 0.7702080184265457,\n",
      " \"o\": 0.0022673288706542864,\n",
      " \"f\": 0.0001799467357662132,\n",
      " \"g\": 0.0022673288706542864,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "aspergillus_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9488148854334462,\n",
      " \"c\": 0.6485420296525198,\n",
      " \"o\": 0.647522929747855,\n",
      " \"f\": 0.6474243071764357,\n",
      " \"g\": 0.5160919162365627,\n",
      " \"s\": 0.3527729379664026\n",
      "}\n",
      "\n",
      "\n",
      "diaporthe_ccl067\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9732339336237481,\n",
      " \"c\": 0.7833116270714511,\n",
      " \"o\": 0.47774733911168993,\n",
      " \"f\": 0.005928054969236988,\n",
      " \"g\": 0.005928054969236988,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "diaporthe_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9786015538290788,\n",
      " \"c\": 0.7152053274139845,\n",
      " \"o\": 0.3137846836847947,\n",
      " \"f\": 0.2776914539400666,\n",
      " \"g\": 0.2760044395116537,\n",
      " \"s\": 0.3904550499445061\n",
      "}\n",
      "\n",
      "\n",
      "oculimacula_yallundae-ccl031\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9584735740926044,\n",
      " \"c\": 0.8043300281997635,\n",
      " \"o\": 0.6500500318384427,\n",
      " \"f\": 0.11684708450832348,\n",
      " \"g\": 0.10051851178022378,\n",
      " \"s\": 0.05130537614845811\n",
      "}\n",
      "\n",
      "\n",
      "oculimacula_yallundae-ccl029\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9535973130634612,\n",
      " \"c\": 0.8309174474102882,\n",
      " \"o\": 0.5746862294502386,\n",
      " \"f\": 0.18485946614813506,\n",
      " \"g\": 0.16377938836839315,\n",
      " \"s\": 0.09386600671734134\n",
      "}\n",
      "\n",
      "\n",
      "dothiorella_vidmadera\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9218157240657567,\n",
      " \"c\": 0.5493393799328479,\n",
      " \"o\": 0.30118170322243054,\n",
      " \"f\": 0.30070204508786463,\n",
      " \"g\": 0.2759342432302795,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "quambalaria_cyanescens\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9036699239956569,\n",
      " \"c\": 0.7672095548317047,\n",
      " \"o\": 0.766514657980456,\n",
      " \"f\": 0.7595222584147665,\n",
      " \"g\": 0.7595222584147665,\n",
      " \"s\": 0.7084907709011944\n",
      "}\n",
      "\n",
      "\n",
      "entoleuca_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9689353531043311,\n",
      " \"c\": 0.8632814166844464,\n",
      " \"o\": 0.6314913590783017,\n",
      " \"f\": 0.5993599317260507,\n",
      " \"g\": 0.03277149562620013,\n",
      " \"s\": 0.7261787924045231\n",
      "}\n",
      "\n",
      "\n",
      "asteroma_ccl060\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9771875,\n",
      " \"c\": 0.7168303571428571,\n",
      " \"o\": 0.30763392857142857,\n",
      " \"f\": 0.00026785714285714287,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "asteroma_ccl068\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9723086636900735,\n",
      " \"c\": 0.7910972804762549,\n",
      " \"o\": 0.47201551436431693,\n",
      " \"f\": 0.44576737473503814,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "saccharomyces_cerevisiae\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.957688940764517,\n",
      " \"c\": 0.7638848361054372,\n",
      " \"o\": 0.7638848361054372,\n",
      " \"f\": 0.7053302207956424,\n",
      " \"g\": 0.6633595953701001,\n",
      " \"s\": 0.5030152708880459\n",
      "}\n",
      "\n",
      "\n",
      "cladophialophora_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9241073469574361,\n",
      " \"c\": 0.4331367439308737,\n",
      " \"o\": 0.3985735838705253,\n",
      " \"f\": 0.3852237918895442,\n",
      " \"g\": 0.3569240616284918,\n",
      " \"s\": 0.7008640790015087\n",
      "}\n",
      "\n",
      "\n",
      "candida_albicans\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9333918831468984,\n",
      " \"c\": 0.5988259221595637,\n",
      " \"o\": 0.5988259221595637,\n",
      " \"f\": 0.3706203198668762,\n",
      " \"g\": 0.3706203198668762,\n",
      " \"s\": 0.30248682629194784\n",
      "}\n",
      "\n",
      "\n",
      "candida_metapsilosis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9507371119860819,\n",
      " \"c\": 0.6706803406281476,\n",
      " \"o\": 0.6706803406281476,\n",
      " \"f\": 0.6481549308671367,\n",
      " \"g\": 0.6481091475139639,\n",
      " \"s\": 0.018084424503250617\n",
      "}\n",
      "\n",
      "\n",
      "candida_orthopsilosis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.952286648551559,\n",
      " \"c\": 0.6714871275272878,\n",
      " \"o\": 0.6714871275272878,\n",
      " \"f\": 0.6451895178003961,\n",
      " \"g\": 0.6450052963662322,\n",
      " \"s\": 0.31340671487127525\n",
      "}\n",
      "\n",
      "\n",
      "candida_parapsilosis\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.950793359625791,\n",
      " \"c\": 0.6723378886544987,\n",
      " \"o\": 0.6723378886544987,\n",
      " \"f\": 0.6554159405668165,\n",
      " \"g\": 0.6553700816289095,\n",
      " \"s\": 0.19150692469962396\n",
      "}\n",
      "\n",
      "\n",
      "candida_unidentified\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9368696301830407,\n",
      " \"c\": 0.5903063130369817,\n",
      " \"o\": 0.5903063130369817,\n",
      " \"f\": 0.30033619723571164,\n",
      " \"g\": 0.30033619723571164,\n",
      " \"s\": 0.539316398954053\n",
      "}\n",
      "\n",
      "\n",
      "kluyveromyces_marxianus\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9499904834411876,\n",
      " \"c\": 0.7797392462885421,\n",
      " \"o\": 0.7797392462885421,\n",
      " \"f\": 0.694518462124096,\n",
      " \"g\": 0.578368861819566,\n",
      " \"s\": 0.5149885801294252\n",
      "}\n",
      "\n",
      "\n",
      "pichia_kudriavzevii\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9519407763105242,\n",
      " \"c\": 0.8446178471388556,\n",
      " \"o\": 0.8446178471388556,\n",
      " \"f\": 0.20768307322929172,\n",
      " \"g\": 0.20684273709483794,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "pichia_membranifaciens\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9420937443005654,\n",
      " \"c\": 0.7918110523436075,\n",
      " \"o\": 0.7918110523436075,\n",
      " \"f\": 0.7452580703994164,\n",
      " \"g\": 0.7441637789531278,\n",
      " \"s\": 0.6966076965165056\n",
      "}\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from collections import OrderedDict\n",
    "\n",
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarieses the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = OrderedDict()\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict\n",
    "\n",
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df\n",
    "    \n",
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "test_species_list = []\n",
    "for entry in ref_df.name_species.tolist():\n",
    "#     if entry[-7:] != '-ccl031' and entry[-7:] != '-ccl029':\n",
    "#         test_species_list.append(entry)\n",
    "#     else:\n",
    "#         test_species_list.append(entry[:-7])\n",
    "#         print(entry[:-7])\n",
    "    test_species_list.append(entry)\n",
    "    \n",
    "for test_species in test_species_list:\n",
    "    \n",
    "    print(test_species)\n",
    "    \n",
    "    #subsample tests species\n",
    "    fn_subsampling = {}\n",
    "    test_species = [test_species]\n",
    "    for x in test_species:\n",
    "        fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "        fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "    sub_reads_fn = {}\n",
    "    n_reads = 20000\n",
    "    for key, value in fn_subsampling.items():\n",
    "        species = key\n",
    "        in_fn = value\n",
    "        out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "        subsamplereads(in_fn, out_fn, n_reads)\n",
    "        sub_reads_fn[species] = out_fn\n",
    "        \n",
    "    ###Map the reads\n",
    "    db_fn = qiime_db_fn\n",
    "    sub_db_mapping_fn = {}\n",
    "    for species, fasta_fn in sub_reads_fn.items():\n",
    "        db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "        out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "        sub_db_mapping_fn[species] = out_fn\n",
    "        minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "        \n",
    "    ###Test out the summary results statistic for a single mapping result\n",
    "    species = test_species[0]\n",
    "    mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "    mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "    taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "    query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "    sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "            \n",
    "    print(json.dumps(sensitivity_dict, indent=1))\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test run on the qiime2 Database 99\n",
    "\n",
    "##### Prep on the command line\n",
    "\n",
    "cp sh_refs_qiime_ver8_99_02.02.2019.fasta /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/. \n",
    "cp sh_taxonomy_qiime_ver8_99_02.02.2019.txt /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_99_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_99_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###fix the query_tax_dict with the data found in the qiime database\n",
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use with Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'candida_albicans'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
