{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notebook to analyzse the efficiency of minimap mapping against a mock community"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting points from Tavish  \n",
    "reference_dataframe at '/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv'  \n",
    "custom_database at '/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta'  \n",
    "taxonomy_file at '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### workflow\n",
    "\n",
    "* Two databases\n",
    "* subsample 15000 reads per each mock community species. Save those out.\n",
    "* map reads against both databases with minimap safe out data in paf format.\n",
    "* get best hit per species (see what this means while looking at the data).\n",
    "* add the full taxonomy to each best match using the taxonomy file.\n",
    "* summarize data at different taxonomic ranks for each species.\n",
    "* pull this all together somehow (summary across all the samples? focus on species of interest e.g. deleted from analyis?)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### requirment\n",
    "\n",
    "* Bbmap (conda install bbmap https://anaconda.org/bioconda/bbmap)  \n",
    "* minimap2 (conda install minimap2 https://anaconda.org/bioconda/minimap2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fixes on the command line\n",
    "\n",
    "* fixed the taxonomy file to fit with the quiime format\n",
    "\n",
    "cat /media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file.csv | sed 's/,/\\t/' > /media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v2.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### fix the taxonomy_file_v2.csv more to reflect Qiime style\n",
    "\n",
    "* This requires to make the species name genus_species and not _species.... hope this makes sense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "old_taxonomy_file_fn = '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v2.csv'\n",
    "new_taxonomy_file_fn = '/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_v3.csv'\n",
    "with open(new_taxonomy_file_fn, 'w') as out_fh:\n",
    "    with open(old_taxonomy_file_fn, 'r') as in_fh:\n",
    "        for line in in_fh:\n",
    "            line = line.rstrip()\n",
    "            #print(line)\n",
    "            first_half = line.split('s__')[0]\n",
    "            second_half = line.split('s__')[1]\n",
    "            pattern = re.compile(r'g__\\w+;')\n",
    "            genus = re.findall(pattern, first_half)[0].replace('g__','').replace(';','')\n",
    "            new_line = F\"{first_half}s__{genus}_{second_half}\"\n",
    "            print(new_line,file=out_fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!head -2 {old_taxonomy_file_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!head -2 {new_taxonomy_file_fn}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Initial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/reference_dataframe.csv')\n",
    "max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/custom_database_labelled.fasta')\n",
    "taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_qiime.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#threads to use\n",
    "threads = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = os.path.abspath('../../analysis/Mapping_mock_gsref')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### list of species in the max database\n",
    "max_species = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ###Removed from second test databes\n",
    "species_delete = [\n",
    "# 'Candida_orthopsilosis',\n",
    "#                  'Candida_metapsilosis',\n",
    "#                  'Aspergillus_niger'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###species to be searched against both databases\n",
    "# mock_community = ['Penicillium_chrysogenum',\n",
    "#  'Aspergillus_flavus',\n",
    "#  'Aspergillus_niger',\n",
    "#  'Pichia_kudriavzevii',\n",
    "#  'Pichia_membranifaciens',\n",
    "#  'Candida_albicans',\n",
    "#  'Candida_parapsilosis',\n",
    "#  'Candida_orthopsilosis',\n",
    "#  'Candida_metapsilosis']\n",
    "\n",
    "mock_community = ['Puccinia_striiformis-tritici',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fixed_old_names = ['Kluyveromyces_lactis',\n",
    "#                    'Candida_zeylanoides',\n",
    "#                    'Cladophialophora_sp.',\n",
    "#                    'Diaporthe_sp.',\n",
    "#                    'CCL060',\n",
    "#                    'CCL068',\n",
    "#                    'CCL067',\n",
    "#                    'Aspergillus_sp.',\n",
    "#                    'Entoleuca_sp.',\n",
    "#                    'Tapesia_yallundae_CCL029',\n",
    "#                    'Tapesia_yallundae_CCL031',\n",
    "#                    'Cryptococcus_neoformans']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fixed_new_names = ['candida_unidentified',\n",
    "#                    'debaryomyces_unidentified',\n",
    "#                    'cladophialophora_unidentified',\n",
    "#                    'diaporthe_unidentified',\n",
    "#                    'asteroma_ccl060',\n",
    "#                    'asteroma_ccl068',\n",
    "#                    'diaporthe_ccl067',\n",
    "#                    'aspergillus_unidentified',\n",
    "#                    'entoleuca_unidentified',\n",
    "#                    'oculimacula_yallundae-ccl029',\n",
    "#                    'oculimacula_yallundae-ccl031',\n",
    "#                    'kluyveromyces_unidentified']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old_to_new_names = dict(zip(fixed_old_names, fixed_new_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# old_to_new_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fix databases and names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = pd.read_csv(reference_dataframe_fn)\n",
    "ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df.name_species.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_fn = os.path.join(OUT_DIR, 'gsref.db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db_list = []\n",
    "old_db_list = []\n",
    "for seq in SeqIO.parse(max_custom_database_fn, 'fasta'):\n",
    "    old_db_list.append(seq.id)\n",
    "    if seq.id.lower() in ref_df.name_species.tolist():\n",
    "        #print(seq.id)\n",
    "        seq.id = seq.name = seq.description = seq.id.lower()\n",
    "        new_db_list.append(seq)\n",
    "    else:\n",
    "        print(seq.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(new_db_list) == len(old_db_list):\n",
    "    SeqIO.write(new_db_list, new_db_fn, 'fasta')\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_db_fn = os.path.join(OUT_DIR, 'gsref.subdb.fasta')\n",
    "sub_db_list = []\n",
    "for seq in new_db_list:\n",
    "    if seq.id not in [x.lower() for x in species_delete]:\n",
    "        sub_db_list.append(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(sub_db_list) + len(species_delete) == len(new_db_list):\n",
    "    SeqIO.write(sub_db_list, sub_db_fn, 'fasta' )\n",
    "else:\n",
    "    print(\"please check!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[x.id for x in sub_db_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = [x.lower() for x in mock_community]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subsample reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 2000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MC_READ_DIR = os.path.join(OUT_DIR, 'MC_READS')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map with minimap against both databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results(fn, species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the query species: {species}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')\n",
    "    hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/custom_results/%s.json' % species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###this is running the reads against the full database\n",
    "for species, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###this is running against a database that have ['Candida_orthopsilosis', 'Candida_metapsilosis', 'Aspergillus_niger'] deleted\n",
    "for species, hit_fn in sub_db_mapping_fn.items():\n",
    "    mapping_results(hit_fn, species)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pull in mapping results and analyse them at all available levels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### idea\n",
    "\n",
    "* pull in query taxid as a dictionary\n",
    "* assign taxid for each tname species from minimap2\n",
    "* generate a summary dictionary that checks concordance at each taxonmic rank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['mquality'].count().tolist(), sub_df.groupby('tname')['mquality'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "    return hit_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getquery_taxfileid(refdf_fn, species):\n",
    "    \"\"\"\n",
    "    Takes the reference dataframe filename and the species name.\n",
    "    Returns the taxfileid, which is the date/flowcellid (column 0 value) of the ref_df.\n",
    "    \"\"\"\n",
    "    ref_df = pd.read_csv(refdf_fn)\n",
    "    ref_df['name_species'] = ref_df['genus'] +\"_\"+ ref_df['species']\n",
    "    print(ref_df[ref_df.name_species == species].iloc[:,0].values[0])\n",
    "    return ref_df[ref_df.name_species == species].iloc[:,0].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxid_dict(taxid_fn, taxfileid):\n",
    "    \"\"\"\n",
    "    Takes a taxonomy assignment file filename in the Qiime format and a taxonomic identifier.\n",
    "    Returns the a dictionary with the taxonomic assignment at each rank.\n",
    "    \"\"\"\n",
    "    tax_dict = {}\n",
    "    with open(taxid_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(taxfileid):\n",
    "                taxrankids = line.rstrip().split('\\t')[1].split(';')\n",
    "                for taxrank in taxrankids:\n",
    "                    tax_dict[taxrank.split('__')[0]] = taxrank.split('__')[1]\n",
    "    return tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_taxranks_results(mapping_df, tax_fn, ref_df_fn = False):\n",
    "    \"\"\"\n",
    "    This function assigns the taxonomic ranks for each hit in the mapping results dataframe.\n",
    "    It takes a mapping_df, taxonomy assignment file, and if required a reference dataframe filename.\n",
    "    Returns the mapping dataframe with assignment. \n",
    "    \"\"\"\n",
    "    for tname in mapping_df.index:\n",
    "        if ref_df_fn:\n",
    "            tmp_taxfileid = getquery_taxfileid(ref_df_fn, tname)\n",
    "        else:\n",
    "            tmp_taxfileid = tname\n",
    "        tmp_tax_dict = get_taxid_dict(tax_fn, tmp_taxfileid)\n",
    "        for key, value in tmp_tax_dict.items():\n",
    "            mapping_df.loc[tname, key] = value\n",
    "    return mapping_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarieses the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = {}\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'candida_albicans'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'aspergillus_niger'\n",
    "mapping_results = pull_mapping_results(sub_db_mapping_fn[species])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Assign the data taxonomics ranks for all the results\n",
    "mapping_results = assign_taxranks_results(mapping_results, taxonomy_file_fn, ref_df_fn=reference_dataframe_fn)\n",
    "\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test run on the qiime2 Database\n",
    "\n",
    "##### Prep on the command line\n",
    "\n",
    "cp sh_refs_qiime_ver8_dynamic_02.02.2019.fasta /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/.  \n",
    "cp sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results_v2(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    #sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['mquality'].count().tolist(), sub_df.groupby('tname')['mquality'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.path.abspath(os.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run on test species 'penicillium_chrysogenum'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### These are wired results that might be linked to\n",
    "\n",
    "* database issues as you can see\n",
    "\n",
    "sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt: k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_results[mapping_results['s'] == 'Penicillium_chrysogenum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]\n",
    "query_tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.path.abspath(os.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run on test species 'penicillium_chrysogenum'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### These are wired results that might be linked to\n",
    "\n",
    "* database issues as you can see\n",
    "\n",
    "sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt: k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_results[mapping_results['s'] == 'Penicillium_chrysogenum']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]\n",
    "query_tax_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Might want to double check how your families are matched in your taxonomic input file for your known test species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mapping_results[mapping_results['s'] == 'Candida_albicans']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# query_tax_dict['f'] = 'Saccharomycetales_fam_Incertae_sedis'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### It appears that the filtering of results by mapping quality works well for the long ITS database but not for the qiime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.tname.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.tname.unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.index = full_results_df.tname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Asign taxonomic ranks to the full_results_df\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df[full_results_df['g'] == 'Candida']['qseqid'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df[full_results_df['g'] == 'Candida']['qseqid'].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at the results unfiltered doesn't really work very well either. Might need to look into different filtering of the alignments or the qiime2 database might be just not really useful for the noise reads. Simulated reads with higher accuracy should get better here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results_df.groupby('g').count()['k'].sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###There must be more or equal number of mapping results compared to number of mapped reads\n",
    "mapping_results['count'].sum() >= full_results_df.qseqid.unique().shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on other species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['cortinarius_globuliformis']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Test out the summary results statistic for a single mapping result\n",
    "species = test_species[0]\n",
    "print(sub_db_mapping_fn[species])\n",
    "mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "print(taxfileid)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "print(query_tax_dict)\n",
    "\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Testing on all species using v3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from collections import OrderedDict\n",
    "\n",
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarises the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = OrderedDict()\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        tmps_df = pd.DataFrame(data=None)\n",
    "        if tax_rank == 's':\n",
    "            for index, row in mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]].iterrows():\n",
    "                if row['s'] == 'unidentified' and row['g'] != query_tax_dict['g']:\n",
    "                    mapping_df.drop(index, axis=0, inplace=True)\n",
    "                else:\n",
    "                    continue\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        else:\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict\n",
    "\n",
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df\n",
    "    \n",
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "test_species_list = []\n",
    "for entry in ref_df.name_species.tolist():\n",
    "#     if entry[-7:] != '-ccl031' and entry[-7:] != '-ccl029':\n",
    "#         test_species_list.append(entry)\n",
    "#     else:\n",
    "#         test_species_list.append(entry[:-7])\n",
    "#         print(entry[:-7])\n",
    "    test_species_list.append(entry)\n",
    "    \n",
    "for test_species in test_species_list:\n",
    "    \n",
    "    print(test_species)\n",
    "    \n",
    "    #subsample tests species\n",
    "    fn_subsampling = {}\n",
    "    test_species = [test_species]\n",
    "    for x in test_species:\n",
    "        fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "        fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "    sub_reads_fn = {}\n",
    "    n_reads = 20000\n",
    "    for key, value in fn_subsampling.items():\n",
    "        species = key\n",
    "        in_fn = value\n",
    "        out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "        subsamplereads(in_fn, out_fn, n_reads)\n",
    "        sub_reads_fn[species] = out_fn\n",
    "        \n",
    "    ###Map the reads\n",
    "    db_fn = qiime_db_fn\n",
    "    sub_db_mapping_fn = {}\n",
    "    for species, fasta_fn in sub_reads_fn.items():\n",
    "        db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "        out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "        sub_db_mapping_fn[species] = out_fn\n",
    "        minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "        \n",
    "    ###Test out the summary results statistic for a single mapping result\n",
    "    species = test_species[0]\n",
    "    mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "    mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "    taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "    \n",
    "    query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "\n",
    "    print(type(query_tax_dict))\n",
    "    print(query_tax_dict)\n",
    "    \n",
    "    sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "        \n",
    "    print(json.dumps(sensitivity_dict, indent=1))\n",
    "#     print('\\n')\n",
    "#     with open('/media/MassStorage/tmp/TE/honours/analysis/Mapping/qiime_results/%s.json' % species, 'w+') as fp:\n",
    "#         json.dump(sensitivity_dict, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test run on the qiime2 Database 99\n",
    "\n",
    "##### Prep on the command line\n",
    "\n",
    "cp sh_refs_qiime_ver8_99_02.02.2019.fasta /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/. \n",
    "cp sh_taxonomy_qiime_ver8_99_02.02.2019.txt /media/WorkingStorage/ben.working/students/tavish/analysis/qiime2/db/."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_99_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_99_02.02.2019.txt')\n",
    "threads = 10\n",
    "QIIME_DIR = os.path.abspath('../../analysis/qiime2/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##mapping folder\n",
    "mapping_dir = os.path.join(QIIME_DIR, os.path.basename(qiime_db_fn).replace('.fasta', '').replace('.','_'))\n",
    "print(mapping_dir)\n",
    "if not os.path.exists(mapping_dir):\n",
    "    os.mkdir(mapping_dir)\n",
    "subsampling_dir = os.path.join(QIIME_DIR, 'subsamplereads')\n",
    "if not os.path.exists(subsampling_dir):\n",
    "    os.mkdir(subsampling_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['penicillium_chrysogenum']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'penicillium_chrysogenum'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###fix the query_tax_dict with the data found in the qiime database\n",
    "query_tax_dict = {}\n",
    "taxrank = 'k__Fungi;p__Ascomycota;c__Eurotiomycetes;o__Eurotiales;f__Aspergillaceae;g__Penicillium;s__Penicillium_chrysogenum'\n",
    "for rank_id in taxrank.split(';'):\n",
    "    query_tax_dict[rank_id.split('__')[0]] = rank_id.split('__')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_accuracy_dict(mapping_results, query_tax_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use with Candida albicans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#subsample tests species\n",
    "fn_subsampling = {}\n",
    "test_species = ['candida_albicans']\n",
    "for x in test_species:\n",
    "    fn_subsampling[x] = (ref_df[(ref_df['species'] == x.split('_')[1]) & (ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "sub_reads_fn = {}\n",
    "n_reads = 20000\n",
    "for key, value in fn_subsampling.items():\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn\n",
    "\n",
    "###Map the reads\n",
    "db_fn = qiime_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "\n",
    "###Test out the summary results statistic for a single mapping result\n",
    "species = 'candida_albicans'\n",
    "mapping_results , full_results_df = pull_mapping_results_v2(sub_db_mapping_fn[species])\n",
    "mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "taxfileid = getquery_taxfileid(reference_dataframe_fn, species)\n",
    "query_tax_dict = get_taxid_dict(taxonomy_file_fn, taxfileid)\n",
    "###fix family level for 'penicillium_chrysogenum'\n",
    "sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "\n",
    "\n",
    "full_results_df.index = full_results_df.tname\n",
    "###Also look at the full results dataframe to explore results a bit more\n",
    "for tname in full_results_df.tname.unique():\n",
    "\n",
    "    tmp_tax_dict = get_taxid_dict(qiime_tax_fn, tname)\n",
    "    for key, value in tmp_tax_dict.items():\n",
    "        full_results_df.loc[tname, key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###looking at unfiltered results\n",
    "###look at the results unfiltered\n",
    "full_results_df['count'] = 1\n",
    "\n",
    "get_accuracy_dict(full_results_df, query_tax_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply to wheat species for qiime and consensus respectively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>species</th>\n",
       "      <th>genus</th>\n",
       "      <th>family</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>phylum</th>\n",
       "      <th>kingdom</th>\n",
       "      <th># raw reads</th>\n",
       "      <th># reads after homology filtering</th>\n",
       "      <th># reads after length filtering</th>\n",
       "      <th># for use</th>\n",
       "      <th>path to raw reads</th>\n",
       "      <th>path to homology filtering</th>\n",
       "      <th>path to length filtering</th>\n",
       "      <th>path for use</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>barcode01</td>\n",
       "      <td>puccinia_striiformis</td>\n",
       "      <td>puccinia</td>\n",
       "      <td>pucciniaceae</td>\n",
       "      <td>pucciniales</td>\n",
       "      <td>pucciniomycetes</td>\n",
       "      <td>basidiomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>115052</td>\n",
       "      <td>23477</td>\n",
       "      <td>21590</td>\n",
       "      <td>21590.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode01/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode01/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode01/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode01/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>barcode02</td>\n",
       "      <td>zymoseptoria_tritici</td>\n",
       "      <td>zymoseptoria</td>\n",
       "      <td>mycosphaerellaceae</td>\n",
       "      <td>capnodiales</td>\n",
       "      <td>dothideomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>151005</td>\n",
       "      <td>20416</td>\n",
       "      <td>18100</td>\n",
       "      <td>18100.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode02/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode02/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode02/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode02/lengt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>barcode05</td>\n",
       "      <td>pyrenophora_tritici-repentis</td>\n",
       "      <td>pyrenophora</td>\n",
       "      <td>pleosporaceae</td>\n",
       "      <td>pleosporales</td>\n",
       "      <td>dothideomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>134094</td>\n",
       "      <td>24785</td>\n",
       "      <td>22433</td>\n",
       "      <td>22433.0</td>\n",
       "      <td>analysis/Concatenated/wheat/barcode05/merged.f...</td>\n",
       "      <td>analysis/Python_Processing/wheat/barcode05/com...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode05/lengt...</td>\n",
       "      <td>analysis/Length_Filtered/wheat/barcode05/lengt...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0                       species         genus              family  \\\n",
       "0  barcode01          puccinia_striiformis      puccinia        pucciniaceae   \n",
       "1  barcode02          zymoseptoria_tritici  zymoseptoria  mycosphaerellaceae   \n",
       "2  barcode05  pyrenophora_tritici-repentis   pyrenophora       pleosporaceae   \n",
       "\n",
       "          order            class         phylum kingdom  # raw reads  \\\n",
       "0   pucciniales  pucciniomycetes  basidiomycota   fungi       115052   \n",
       "1   capnodiales  dothideomycetes     ascomycota   fungi       151005   \n",
       "2  pleosporales  dothideomycetes     ascomycota   fungi       134094   \n",
       "\n",
       "   # reads after homology filtering  # reads after length filtering  \\\n",
       "0                             23477                           21590   \n",
       "1                             20416                           18100   \n",
       "2                             24785                           22433   \n",
       "\n",
       "   # for use                                  path to raw reads  \\\n",
       "0    21590.0  analysis/Concatenated/wheat/barcode01/merged.f...   \n",
       "1    18100.0  analysis/Concatenated/wheat/barcode02/merged.f...   \n",
       "2    22433.0  analysis/Concatenated/wheat/barcode05/merged.f...   \n",
       "\n",
       "                          path to homology filtering  \\\n",
       "0  analysis/Python_Processing/wheat/barcode01/com...   \n",
       "1  analysis/Python_Processing/wheat/barcode02/com...   \n",
       "2  analysis/Python_Processing/wheat/barcode05/com...   \n",
       "\n",
       "                            path to length filtering  \\\n",
       "0  analysis/Length_Filtered/wheat/barcode01/lengt...   \n",
       "1  analysis/Length_Filtered/wheat/barcode02/lengt...   \n",
       "2  analysis/Length_Filtered/wheat/barcode05/lengt...   \n",
       "\n",
       "                                        path for use  \n",
       "0  analysis/Length_Filtered/wheat/barcode01/lengt...  \n",
       "1  analysis/Length_Filtered/wheat/barcode02/lengt...  \n",
       "2  analysis/Length_Filtered/wheat/barcode05/lengt...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "\n",
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')\n",
    "subsampling_dir = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/subsample_reads')\n",
    "mapping_dir = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/qiime_results')\n",
    "\n",
    "wheat_reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/wheat_reference_dataframe.csv')\n",
    "wheat_max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/wheat_database_labelled.fasta')\n",
    "wheat_taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/wheat_taxonomy_file_qiime.csv')\n",
    "\n",
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "\n",
    "wheat_ref_df = pd.read_csv(wheat_reference_dataframe_fn)\n",
    "wheat_ref_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Qiime Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "puccinia_striiformis\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Pucciniomycetes', 'o': 'Pucciniales', 'f': 'Pucciniaceae', 'g': 'Puccinia', 's': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.4479638009049774,\n",
      " \"c\": 0.0,\n",
      " \"o\": 0.0,\n",
      " \"f\": 0.0,\n",
      " \"g\": 0.0,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "zymoseptoria_tritici\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Dothideomycetes', 'o': 'Capnodiales', 'f': 'Mycosphaerellaceae', 'g': 'Zymoseptoria', 's': 'Zymoseptoria_tritici'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.6832579185520362,\n",
      " \"c\": 0.20361990950226244,\n",
      " \"o\": 0.09049773755656108,\n",
      " \"f\": 0.08597285067873303,\n",
      " \"g\": 0.08597285067873303,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "pyrenophora_tritici-repentis\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Dothideomycetes', 'o': 'Pleosporales', 'f': 'Pleosporaceae', 'g': 'Pyrenophora', 's': 'Pyrenophora_tritici-repentis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.6635514018691588,\n",
      " \"c\": 0.2570093457943925,\n",
      " \"o\": 0.2570093457943925,\n",
      " \"f\": 0.10747663551401869,\n",
      " \"g\": 0.08411214953271028,\n",
      " \"s\": 0.08411214953271028\n",
      "}\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from collections import OrderedDict\n",
    "\n",
    "def assign_taxranks_results(mapping_df, tax_fn, ref_df_fn = False):\n",
    "    \"\"\"\n",
    "    This function assigns the taxonomic ranks for each hit in the mapping results dataframe.\n",
    "    It takes a mapping_df, taxonomy assignment file, and if required a reference dataframe filename.\n",
    "    Returns the mapping dataframe with assignment. \n",
    "    \"\"\"\n",
    "    for tname in mapping_df.index:\n",
    "        if ref_df_fn:\n",
    "            tmp_taxfileid = getquery_taxfileid(ref_df_fn, tname)\n",
    "        else:\n",
    "            tmp_taxfileid = tname\n",
    "        tmp_tax_dict = get_taxid_dict(tax_fn, tmp_taxfileid)\n",
    "        for key, value in tmp_tax_dict.items():\n",
    "            mapping_df.loc[tname, key] = value\n",
    "    return mapping_df\n",
    "\n",
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarises the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = OrderedDict()\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        tmps_df = pd.DataFrame(data=None)\n",
    "        if tax_rank == 's':\n",
    "            for index, row in mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]].iterrows():\n",
    "                if row['s'] == 'unidentified' and row['g'] != query_tax_dict['g']:\n",
    "                    mapping_df.drop(index, axis=0, inplace=True)\n",
    "                else:\n",
    "                    continue\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        else:\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict\n",
    "\n",
    "def getquery_taxfileid(refdf_fn, species):\n",
    "    \"\"\"\n",
    "    Takes the reference dataframe filename and the species name.\n",
    "    Returns the taxfileid, which is the date/flowcellid (column 0 value) of the ref_df.\n",
    "    \"\"\"\n",
    "    ref_df = pd.read_csv(refdf_fn)\n",
    "    return ref_df[ref_df.species == species].iloc[:,0].values[0]\n",
    "\n",
    "def get_taxid_dict(taxid_fn, taxfileid):\n",
    "    \"\"\"\n",
    "    Takes a taxonomy assignment file filename in the Qiime format and a taxonomic identifier.\n",
    "    Returns the a dictionary with the taxonomic assignment at each rank.\n",
    "    \"\"\"\n",
    "    tax_dict = {}\n",
    "    with open(taxid_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(taxfileid):\n",
    "                taxrankids = line.rstrip().split('\\t')[1].split(';')\n",
    "                for taxrank in taxrankids:\n",
    "                    tax_dict[taxrank.split('__')[0]] = taxrank.split('__')[1]\n",
    "    return tax_dict\n",
    "\n",
    "def mapping_results(fn, species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the query species: {species}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')\n",
    "    hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/custom_results/%s.json' % species)\n",
    "\n",
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df\n",
    "    \n",
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "test_species_list = []\n",
    "for entry in wheat_ref_df.species.tolist():\n",
    "#     if entry[-7:] != '-ccl031' and entry[-7:] != '-ccl029':\n",
    "#         test_species_list.append(entry)\n",
    "#     else:\n",
    "#         test_species_list.append(entry[:-7])\n",
    "#         print(entry[:-7])\n",
    "    test_species_list.append(entry)\n",
    "    \n",
    "for test_species in test_species_list:\n",
    "    \n",
    "    print(test_species)\n",
    "    \n",
    "    #subsample tests species\n",
    "    fn_subsampling = {}\n",
    "    test_species = [test_species]\n",
    "    for x in test_species:\n",
    "#         print((wheat_ref_df['species'] == x))\n",
    "        fn_subsampling[x] = (wheat_ref_df[(wheat_ref_df['species'] == x) & (wheat_ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "        fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "    sub_reads_fn = {}\n",
    "    n_reads = 200\n",
    "    for key, value in fn_subsampling.items():\n",
    "        species = key\n",
    "        in_fn = value\n",
    "        out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "        subsamplereads(in_fn, out_fn, n_reads)\n",
    "        sub_reads_fn[species] = out_fn\n",
    "        \n",
    "    ###Map the reads\n",
    "    db_fn = qiime_db_fn\n",
    "    sub_db_mapping_fn = {}\n",
    "    for species, fasta_fn in sub_reads_fn.items():\n",
    "        db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "        out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "        sub_db_mapping_fn[species] = out_fn\n",
    "        minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "        \n",
    "    ###Test out the summary results statistic for a single mapping result\n",
    "    species = test_species[0]\n",
    "    mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "    mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "    taxfileid = getquery_taxfileid(wheat_reference_dataframe_fn, species)\n",
    "    \n",
    "    query_tax_dict = get_taxid_dict(wheat_taxonomy_file_fn, taxfileid)\n",
    "    print(query_tax_dict)\n",
    "    \n",
    "    sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "        \n",
    "    print(json.dumps(sensitivity_dict, indent=1))\n",
    "    print('\\n')\n",
    "    with open('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/qiime_results/%s.json' % species, 'w+') as fp:\n",
    "        json.dump(sensitivity_dict, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Custom Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### list of species in the max database\n",
    "max_species = ['Puccinia_striiformis',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']\n",
    "references = {}\n",
    "for species in max_species:\n",
    "    references[species] = species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_wheat.csv')\n",
    "species_file = {}\n",
    "genus_file = {}\n",
    "family_file = {}\n",
    "order_file = {}\n",
    "\n",
    "with open(taxonomy_file_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            species_file[references[line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]].lower()] = line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]\n",
    "            genus_file[references[line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]].lower()] = line.split('\\t')[1].split(';')[5].split('__')[1].split('\\n')[0]\n",
    "            family_file[references[line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]].lower()] = line.split('\\t')[1].split(';')[4].split('__')[1].split('\\n')[0]\n",
    "            order_file[references[line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]].lower()] = line.split('\\t')[1].split(';')[3].split('__')[1].split('\\n')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = ['barcode01','barcode02',\n",
    "                 'barcode05']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 2000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = os.path.abspath('../../analysis/Mapping/wheat')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)\n",
    "MC_READ_DIR = os.path.join(OUT_DIR, 'MC_READS')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)\n",
    "sub_db_fn = os.path.join('../../analysis/Mapping/gsref.subdb.fasta')\n",
    "new_db_fn = os.path.join('../../analysis/Mapping/gsref.db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'species', 'genus', 'family', 'order', 'class', 'phylum',\n",
       "       'kingdom', '# raw reads', '# reads after homology filtering',\n",
       "       '# reads after length filtering', '# for use', 'path to raw reads',\n",
       "       'path to homology filtering', 'path to length filtering',\n",
       "       'path for use'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wheat_ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode01\n",
      "barcode02\n",
      "barcode05\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'barcode01': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta',\n",
       " 'barcode02': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta',\n",
       " 'barcode05': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    print(x)\n",
    "    fn_subsampling[x] = wheat_ref_df[wheat_ref_df['Unnamed: 0'] == x]['path for use'].tolist()[0]\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode01\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta\n",
      ":(check one reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode01/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.2000.fasta!!\n",
      "\n",
      "barcode02\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta\n",
      ":(check one reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode02/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.2000.fasta!!\n",
      "\n",
      "barcode05\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta\n",
      ":(check one reformat.sh samplereadstarget=2000 in=/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/wheat/barcode05/length_restricted_reads.fasta out=/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.2000.fasta!!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    print(key)\n",
    "    print(value)\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "    subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map with minimap against both databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../analysis/Mapping/gsref.subdb.fasta\n",
      "../../analysis/Mapping/gsref.db.fasta\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'../../analysis/Mapping/gsref.subdb.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb',\n",
       " '../../analysis/Mapping/gsref.db.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    print(x)\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode01.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode02.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_subdb/gsref.subdb.barcode05.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode01.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode01.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode02.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode02.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/MC_READS/barcode05.2000.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/gsref_db/gsref.db.barcode05.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results_v2(fn, species,expected_species, tax_rank, tax_rank_file):\n",
    "    import numpy as np\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    if tax_rank == 's' or tax_rank == 'species':\n",
    "        new_new_hit_series = hit_series\n",
    "        new_new_hit_series.sort_values(ascending=False, inplace=True)\n",
    "    else:\n",
    "        new_hit_series = pd.Series()\n",
    "        new_new_hit_series = pd.Series()\n",
    "        for index in hit_series.index.unique():\n",
    "            new_hit_series.at[index] = data=np.sum(hit_series[index])\n",
    "        new_hit_series = new_hit_series.rename(tax_rank_file, axis='index')\n",
    "        for index in new_hit_series.index.unique():\n",
    "            new_new_hit_series.at[index] = data=np.sum(new_hit_series[index])\n",
    "        new_new_hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the sample type expected: {tax_rank_file[expected_species[species]]}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(new_new_hit_series,'\\n')\n",
    "    new_new_hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/wheat/custom_results/%s_%s.json' % (tax_rank,expected_species[species]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "expected_rank_s = {}\n",
    "expected_rank_g = {}\n",
    "expected_rank_f = {}\n",
    "expected_rank_o = {}\n",
    "\n",
    "for column, row in wheat_ref_df.iterrows():\n",
    "    expected_rank_s[row['Unnamed: 0']] = row['species']\n",
    "    expected_rank_g[row['Unnamed: 0']] = row['genus']\n",
    "    expected_rank_f[row['Unnamed: 0']] = row['family']\n",
    "    expected_rank_o[row['Unnamed: 0']] = row['order']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia_striiformis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "cryptococcus_zero                0.641251\n",
      "blastobotrys_proliferans         0.160313\n",
      "saccharomyces_cerevisiae         0.073314\n",
      "cortinarius_globuliformis        0.040567\n",
      "rhodotorula_mucilaginosa         0.008309\n",
      "aspergillus_flavus               0.007820\n",
      "meyerozyma_guilliermondii        0.006354\n",
      "puccinia_striiformis             0.005376\n",
      "aspergillus_unidentified         0.003910\n",
      "candida_albicans                 0.003910\n",
      "candida_unidentified             0.003910\n",
      "kluyveromyces_unidentified       0.003910\n",
      "zymoseptoria_tritici             0.003910\n",
      "wickerhamomyces_anomalus         0.003910\n",
      "entoleuca_unidentified           0.003421\n",
      "candida_metapsilosis             0.003421\n",
      "quambalaria_cyanescens           0.003421\n",
      "debaryomyces_unidentified        0.002933\n",
      "pyrenophora_tritici-repentis     0.002933\n",
      "dothiorella_vidmadera            0.002444\n",
      "oculimacula_yallundae-ccl031     0.002444\n",
      "kluyveromyces_marxianus          0.001955\n",
      "penicillium_chrysogenum          0.001466\n",
      "zygoascus_hellenicus             0.000978\n",
      "pichia_kudriavzevii              0.000978\n",
      "oculimacula_yallundae-ccl029     0.000978\n",
      "cladophialophora_unidentified    0.000978\n",
      "tuber_brumale                    0.000978\n",
      "asteroma_ccl068                  0.000489\n",
      "aspergillus_niger                0.000489\n",
      "yarrowia_lipolytica              0.000489\n",
      "clavispora_lusitaniae            0.000489\n",
      "yamadazyma_scolyti               0.000489\n",
      "candida_orthopsilosis            0.000489\n",
      "candida_parapsilosis             0.000489\n",
      "fusarium_oxysporum               0.000489\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Zymoseptoria_tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zymoseptoria_tritici             0.517073\n",
      "debaryomyces_unidentified        0.122927\n",
      "cryptococcus_zero                0.108780\n",
      "cortinarius_globuliformis        0.060976\n",
      "pyrenophora_tritici-repentis     0.042439\n",
      "cladophialophora_unidentified    0.029756\n",
      "blastobotrys_proliferans         0.025366\n",
      "aspergillus_unidentified         0.024878\n",
      "quambalaria_cyanescens           0.007805\n",
      "aspergillus_flavus               0.006341\n",
      "saccharomyces_cerevisiae         0.006341\n",
      "candida_unidentified             0.005854\n",
      "meyerozyma_guilliermondii        0.004390\n",
      "kluyveromyces_unidentified       0.004390\n",
      "rhodotorula_mucilaginosa         0.004390\n",
      "candida_albicans                 0.003902\n",
      "dothiorella_vidmadera            0.003902\n",
      "candida_orthopsilosis            0.002439\n",
      "penicillium_chrysogenum          0.002439\n",
      "candida_metapsilosis             0.002439\n",
      "candida_parapsilosis             0.001951\n",
      "entoleuca_unidentified           0.001951\n",
      "aspergillus_niger                0.001463\n",
      "wickerhamomyces_anomalus         0.001463\n",
      "zygoascus_hellenicus             0.000976\n",
      "kluyveromyces_marxianus          0.000976\n",
      "tuber_brumale                    0.000976\n",
      "yamadazyma_mexicana              0.000488\n",
      "pichia_membranifaciens           0.000488\n",
      "clavispora_lusitaniae            0.000488\n",
      "diaporthe_unidentified           0.000488\n",
      "oculimacula_yallundae-ccl029     0.000488\n",
      "oculimacula_yallundae-ccl031     0.000488\n",
      "scedosporium_boydii              0.000488\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pyrenophora_tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pyrenophora_tritici-repentis     0.365292\n",
      "cryptococcus_zero                0.191755\n",
      "meyerozyma_guilliermondii        0.112656\n",
      "cortinarius_globuliformis        0.080537\n",
      "candida_orthopsilosis            0.044104\n",
      "blastobotrys_proliferans         0.039310\n",
      "oculimacula_yallundae-ccl031     0.030201\n",
      "candida_unidentified             0.013423\n",
      "tuber_brumale                    0.012464\n",
      "quambalaria_cyanescens           0.011505\n",
      "saccharomyces_cerevisiae         0.011505\n",
      "candida_albicans                 0.009108\n",
      "aspergillus_flavus               0.007191\n",
      "dothiorella_vidmadera            0.006711\n",
      "rhodotorula_mucilaginosa         0.006232\n",
      "kluyveromyces_unidentified       0.006232\n",
      "wickerhamomyces_anomalus         0.005753\n",
      "entoleuca_unidentified           0.004794\n",
      "candida_metapsilosis             0.004794\n",
      "zymoseptoria_tritici             0.004794\n",
      "yamadazyma_scolyti               0.003835\n",
      "aspergillus_unidentified         0.003835\n",
      "kluyveromyces_marxianus          0.003356\n",
      "cladophialophora_unidentified    0.003356\n",
      "debaryomyces_unidentified        0.002876\n",
      "oculimacula_yallundae-ccl029     0.002876\n",
      "aspergillus_niger                0.002876\n",
      "candida_parapsilosis             0.001918\n",
      "galactomyces_geotrichum          0.001918\n",
      "yamadazyma_mexicana              0.001438\n",
      "pichia_kudriavzevii              0.000959\n",
      "zygoascus_hellenicus             0.000479\n",
      "clavispora_lusitaniae            0.000479\n",
      "penicillium_chrysogenum          0.000479\n",
      "yarrowia_lipolytica              0.000479\n",
      "asteroma_ccl060                  0.000479\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 's', species_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia\n",
      "\n",
      "These are the results:\n",
      "Cryptococcus        0.641251\n",
      "Blastobotrys        0.160313\n",
      "Saccharomyces       0.073314\n",
      "Cortinarius         0.040567\n",
      "Aspergillus         0.012219\n",
      "Candida             0.012219\n",
      "Rhodotorula         0.008309\n",
      "Meyerozyma          0.006354\n",
      "Kluyveromyces       0.005865\n",
      "Puccinia            0.005376\n",
      "Zymoseptoria        0.003910\n",
      "Wickerhamomyces     0.003910\n",
      "Entoleuca           0.003421\n",
      "Quambalaria         0.003421\n",
      "Oculimacula         0.003421\n",
      "Pyrenophora         0.002933\n",
      "Debaryomyces        0.002933\n",
      "Dothiorella         0.002444\n",
      "Penicillium         0.001466\n",
      "Zygoascus           0.000978\n",
      "Tuber               0.000978\n",
      "Pichia              0.000978\n",
      "Cladophialophora    0.000978\n",
      "Fusarium            0.000489\n",
      "Clavispora          0.000489\n",
      "Yamadazyma          0.000489\n",
      "Yarrowia            0.000489\n",
      "Asteroma            0.000489\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Zymoseptoria\n",
      "\n",
      "These are the results:\n",
      "Zymoseptoria        0.517073\n",
      "Debaryomyces        0.122927\n",
      "Cryptococcus        0.108780\n",
      "Cortinarius         0.060976\n",
      "Pyrenophora         0.042439\n",
      "Aspergillus         0.032683\n",
      "Cladophialophora    0.029756\n",
      "Blastobotrys        0.025366\n",
      "Candida             0.016585\n",
      "Quambalaria         0.007805\n",
      "Saccharomyces       0.006341\n",
      "Kluyveromyces       0.005366\n",
      "Rhodotorula         0.004390\n",
      "Meyerozyma          0.004390\n",
      "Dothiorella         0.003902\n",
      "Penicillium         0.002439\n",
      "Entoleuca           0.001951\n",
      "Wickerhamomyces     0.001463\n",
      "Zygoascus           0.000976\n",
      "Oculimacula         0.000976\n",
      "Tuber               0.000976\n",
      "Diaporthe           0.000488\n",
      "Yamadazyma          0.000488\n",
      "Clavispora          0.000488\n",
      "Pichia              0.000488\n",
      "Scedosporium        0.000488\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pyrenophora\n",
      "\n",
      "These are the results:\n",
      "Pyrenophora         0.365292\n",
      "Cryptococcus        0.191755\n",
      "Meyerozyma          0.112656\n",
      "Cortinarius         0.080537\n",
      "Candida             0.073346\n",
      "Blastobotrys        0.039310\n",
      "Oculimacula         0.033078\n",
      "Aspergillus         0.013902\n",
      "Tuber               0.012464\n",
      "Quambalaria         0.011505\n",
      "Saccharomyces       0.011505\n",
      "Kluyveromyces       0.009588\n",
      "Dothiorella         0.006711\n",
      "Rhodotorula         0.006232\n",
      "Wickerhamomyces     0.005753\n",
      "Yamadazyma          0.005273\n",
      "Entoleuca           0.004794\n",
      "Zymoseptoria        0.004794\n",
      "Cladophialophora    0.003356\n",
      "Debaryomyces        0.002876\n",
      "Galactomyces        0.001918\n",
      "Pichia              0.000959\n",
      "Zygoascus           0.000479\n",
      "Penicillium         0.000479\n",
      "Clavispora          0.000479\n",
      "Yarrowia            0.000479\n",
      "Asteroma            0.000479\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 'g', genus_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pucciniaceae\n",
      "\n",
      "These are the results:\n",
      "Tremellaceae           0.641251\n",
      "Trichomonascaceae      0.161290\n",
      "Saccharomycetaceae     0.091398\n",
      "Cortinariaceae         0.040567\n",
      "Aspergillaceae         0.013685\n",
      "Debaryomycetaceae      0.009775\n",
      "Sporidiobolaceae       0.008309\n",
      "Pucciniaceae           0.005376\n",
      "Mycosphaerellaceae     0.003910\n",
      "Phaffomycetaceae       0.003910\n",
      "Dermateaceae           0.003421\n",
      "Quambalariaceae        0.003421\n",
      "Xylariaceae            0.003421\n",
      "Pleosporaceae          0.002933\n",
      "Botryosphaeriaceae     0.002444\n",
      "Pichiaceae             0.000978\n",
      "Herpotrichiellaceae    0.000978\n",
      "Tuberaceae             0.000978\n",
      "Nectriaceae            0.000489\n",
      "Dipodascaceae          0.000489\n",
      "Metschnikowiaceae      0.000489\n",
      "Gnomoniaceae           0.000489\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Mycosphaerellaceae\n",
      "\n",
      "These are the results:\n",
      "Mycosphaerellaceae     0.517073\n",
      "Debaryomycetaceae      0.127805\n",
      "Tremellaceae           0.108780\n",
      "Cortinariaceae         0.060976\n",
      "Pleosporaceae          0.042439\n",
      "Aspergillaceae         0.035122\n",
      "Herpotrichiellaceae    0.029756\n",
      "Saccharomycetaceae     0.028293\n",
      "Trichomonascaceae      0.026341\n",
      "Quambalariaceae        0.007805\n",
      "Sporidiobolaceae       0.004390\n",
      "Botryosphaeriaceae     0.003902\n",
      "Xylariaceae            0.001951\n",
      "Phaffomycetaceae       0.001463\n",
      "Dermateaceae           0.000976\n",
      "Tuberaceae             0.000976\n",
      "Pichiaceae             0.000488\n",
      "Diaporthaceae          0.000488\n",
      "Microascaceae          0.000488\n",
      "Metschnikowiaceae      0.000488\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pleosporaceae\n",
      "\n",
      "These are the results:\n",
      "Pleosporaceae          0.365292\n",
      "Tremellaceae           0.191755\n",
      "Debaryomycetaceae      0.120805\n",
      "Saccharomycetaceae     0.094439\n",
      "Cortinariaceae         0.080537\n",
      "Trichomonascaceae      0.039789\n",
      "Dermateaceae           0.033078\n",
      "Aspergillaceae         0.014382\n",
      "Tuberaceae             0.012464\n",
      "Quambalariaceae        0.011505\n",
      "Botryosphaeriaceae     0.006711\n",
      "Sporidiobolaceae       0.006232\n",
      "Phaffomycetaceae       0.005753\n",
      "Xylariaceae            0.004794\n",
      "Mycosphaerellaceae     0.004794\n",
      "Herpotrichiellaceae    0.003356\n",
      "Dipodascaceae          0.002397\n",
      "Pichiaceae             0.000959\n",
      "Metschnikowiaceae      0.000479\n",
      "Gnomoniaceae           0.000479\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 'f', family_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pucciniales\n",
      "\n",
      "These are the results:\n",
      "Tremellales          0.641251\n",
      "Saccharomycetales    0.268328\n",
      "Agaricales           0.040567\n",
      "Eurotiales           0.013685\n",
      "Sporidiobolales      0.008309\n",
      "Pucciniales          0.005376\n",
      "Capnodiales          0.003910\n",
      "Microstromatales     0.003421\n",
      "Helotiales           0.003421\n",
      "Xylariales           0.003421\n",
      "Pleosporales         0.002933\n",
      "Botryosphaeriales    0.002444\n",
      "Pezizales            0.000978\n",
      "Chaetothyriales      0.000978\n",
      "Hypocreales          0.000489\n",
      "Diaporthales         0.000489\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Capnodiales\n",
      "\n",
      "These are the results:\n",
      "Capnodiales          0.517073\n",
      "Saccharomycetales    0.184878\n",
      "Tremellales          0.108780\n",
      "Agaricales           0.060976\n",
      "Pleosporales         0.042439\n",
      "Eurotiales           0.035122\n",
      "Chaetothyriales      0.029756\n",
      "Microstromatales     0.007805\n",
      "Sporidiobolales      0.004390\n",
      "Botryosphaeriales    0.003902\n",
      "Xylariales           0.001951\n",
      "Pezizales            0.000976\n",
      "Helotiales           0.000976\n",
      "Microascales         0.000488\n",
      "Diaporthales         0.000488\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pleosporales\n",
      "\n",
      "These are the results:\n",
      "Pleosporales         0.365292\n",
      "Saccharomycetales    0.264621\n",
      "Tremellales          0.191755\n",
      "Agaricales           0.080537\n",
      "Helotiales           0.033078\n",
      "Eurotiales           0.014382\n",
      "Pezizales            0.012464\n",
      "Microstromatales     0.011505\n",
      "Botryosphaeriales    0.006711\n",
      "Sporidiobolales      0.006232\n",
      "Capnodiales          0.004794\n",
      "Xylariales           0.004794\n",
      "Chaetothyriales      0.003356\n",
      "Diaporthales         0.000479\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 'o', order_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "y-axis = proportion of reads assigned to taxonomic rank (at 4 taxonomic ranks - s,g,f,o) <br>\n",
    "x-axis = taxonomic ranks for an expected species for each of custom/qiime/ML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply to mock_database for qiime and consensus respectively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>species</th>\n",
       "      <th>genus</th>\n",
       "      <th>family</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>phylum</th>\n",
       "      <th>kingdom</th>\n",
       "      <th># raw reads</th>\n",
       "      <th># reads after homology filtering</th>\n",
       "      <th># reads after length filtering</th>\n",
       "      <th># for use</th>\n",
       "      <th>path to raw reads</th>\n",
       "      <th>path to homology filtering</th>\n",
       "      <th>path to length filtering</th>\n",
       "      <th>path for use</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>20171103_FAH15473/barcode01</td>\n",
       "      <td>puccinia_striiformis</td>\n",
       "      <td>puccinia</td>\n",
       "      <td>pucciniaceae</td>\n",
       "      <td>pucciniales</td>\n",
       "      <td>pucciniomycetes</td>\n",
       "      <td>basidiomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>272465</td>\n",
       "      <td>122080</td>\n",
       "      <td>113337</td>\n",
       "      <td>113337.0</td>\n",
       "      <td>analysis/Concatenated/20171103_FAH15473/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20171103_FAH15473/b...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>20171103_FAH15473/barcode02</td>\n",
       "      <td>zymoseptoria_tritici</td>\n",
       "      <td>zymoseptoria</td>\n",
       "      <td>mycosphaerellaceae</td>\n",
       "      <td>capnodiales</td>\n",
       "      <td>dothideomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>413127</td>\n",
       "      <td>143363</td>\n",
       "      <td>133089</td>\n",
       "      <td>133089.0</td>\n",
       "      <td>analysis/Concatenated/20171103_FAH15473/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20171103_FAH15473/b...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>20171103_FAH15473/barcode03</td>\n",
       "      <td>pyrenophora_tritici-repentis</td>\n",
       "      <td>pyrenophora</td>\n",
       "      <td>pleosporaceae</td>\n",
       "      <td>pleosporales</td>\n",
       "      <td>dothideomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>260896</td>\n",
       "      <td>97584</td>\n",
       "      <td>90015</td>\n",
       "      <td>90015.0</td>\n",
       "      <td>analysis/Concatenated/20171103_FAH15473/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20171103_FAH15473/b...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>20171103_FAH15473/barcode07</td>\n",
       "      <td>aspergillus_niger</td>\n",
       "      <td>aspergillus</td>\n",
       "      <td>aspergillaceae</td>\n",
       "      <td>eurotiales</td>\n",
       "      <td>eurotiomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>171615</td>\n",
       "      <td>65065</td>\n",
       "      <td>59406</td>\n",
       "      <td>59406.0</td>\n",
       "      <td>analysis/Concatenated/20171103_FAH15473/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20171103_FAH15473/b...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>20171103_FAH15473/barcode11</td>\n",
       "      <td>rhodotorula_mucilaginosa</td>\n",
       "      <td>rhodotorula</td>\n",
       "      <td>sporidiobolaceae</td>\n",
       "      <td>sporidiobolales</td>\n",
       "      <td>microbotryomycetes</td>\n",
       "      <td>basidiomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>318405</td>\n",
       "      <td>127801</td>\n",
       "      <td>117801</td>\n",
       "      <td>117801.0</td>\n",
       "      <td>analysis/Concatenated/20171103_FAH15473/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20171103_FAH15473/b...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>20171103_FAH15473/barcode12</td>\n",
       "      <td>scedosporium_boydii</td>\n",
       "      <td>scedosporium</td>\n",
       "      <td>microascaceae</td>\n",
       "      <td>microascales</td>\n",
       "      <td>sordariomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>331947</td>\n",
       "      <td>102481</td>\n",
       "      <td>93723</td>\n",
       "      <td>93723.0</td>\n",
       "      <td>analysis/Concatenated/20171103_FAH15473/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20171103_FAH15473/b...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20171103_FAH15473/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>20171207_FAH18654/barcode12</td>\n",
       "      <td>aspergillus_flavus</td>\n",
       "      <td>aspergillus</td>\n",
       "      <td>aspergillaceae</td>\n",
       "      <td>eurotiales</td>\n",
       "      <td>eurotiomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>125014</td>\n",
       "      <td>54061</td>\n",
       "      <td>51340</td>\n",
       "      <td>51340.0</td>\n",
       "      <td>analysis/Concatenated/20171207_FAH18654/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20171207_FAH18654/b...</td>\n",
       "      <td>analysis/Length_Filtered/20171207_FAH18654/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20171207_FAH18654/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>20180108_FAH18647/barcode01</td>\n",
       "      <td>saccharomyces_cerevisiae</td>\n",
       "      <td>saccharomyces</td>\n",
       "      <td>saccharomycetaceae</td>\n",
       "      <td>saccharomycetales</td>\n",
       "      <td>saccharomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>96837</td>\n",
       "      <td>33025</td>\n",
       "      <td>30260</td>\n",
       "      <td>30260.0</td>\n",
       "      <td>analysis/Concatenated/20180108_FAH18647/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20180108_FAH18647/b...</td>\n",
       "      <td>analysis/Length_Filtered/20180108_FAH18647/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20180108_FAH18647/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>20180108_FAH18647/barcode03</td>\n",
       "      <td>candida_albicans</td>\n",
       "      <td>candida</td>\n",
       "      <td>saccharomycetaceae</td>\n",
       "      <td>saccharomycetales</td>\n",
       "      <td>saccharomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>91031</td>\n",
       "      <td>26114</td>\n",
       "      <td>25597</td>\n",
       "      <td>25597.0</td>\n",
       "      <td>analysis/Concatenated/20180108_FAH18647/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20180108_FAH18647/b...</td>\n",
       "      <td>analysis/Length_Filtered/20180108_FAH18647/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20180108_FAH18647/bar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>20180108_FAH18647/barcode05</td>\n",
       "      <td>candida_orthopsilosis</td>\n",
       "      <td>candida</td>\n",
       "      <td>saccharomycetaceae</td>\n",
       "      <td>saccharomycetales</td>\n",
       "      <td>saccharomycetes</td>\n",
       "      <td>ascomycota</td>\n",
       "      <td>fungi</td>\n",
       "      <td>104214</td>\n",
       "      <td>32490</td>\n",
       "      <td>29765</td>\n",
       "      <td>29765.0</td>\n",
       "      <td>analysis/Concatenated/20180108_FAH18647/barcod...</td>\n",
       "      <td>analysis/Python_Processing/20180108_FAH18647/b...</td>\n",
       "      <td>analysis/Length_Filtered/20180108_FAH18647/bar...</td>\n",
       "      <td>analysis/Length_Filtered/20180108_FAH18647/bar...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0                       species          genus  \\\n",
       "0  20171103_FAH15473/barcode01          puccinia_striiformis       puccinia   \n",
       "1  20171103_FAH15473/barcode02          zymoseptoria_tritici   zymoseptoria   \n",
       "2  20171103_FAH15473/barcode03  pyrenophora_tritici-repentis    pyrenophora   \n",
       "3  20171103_FAH15473/barcode07             aspergillus_niger    aspergillus   \n",
       "4  20171103_FAH15473/barcode11      rhodotorula_mucilaginosa    rhodotorula   \n",
       "5  20171103_FAH15473/barcode12           scedosporium_boydii   scedosporium   \n",
       "6  20171207_FAH18654/barcode12            aspergillus_flavus    aspergillus   \n",
       "7  20180108_FAH18647/barcode01      saccharomyces_cerevisiae  saccharomyces   \n",
       "8  20180108_FAH18647/barcode03              candida_albicans        candida   \n",
       "9  20180108_FAH18647/barcode05         candida_orthopsilosis        candida   \n",
       "\n",
       "               family              order               class         phylum  \\\n",
       "0        pucciniaceae        pucciniales     pucciniomycetes  basidiomycota   \n",
       "1  mycosphaerellaceae        capnodiales     dothideomycetes     ascomycota   \n",
       "2       pleosporaceae       pleosporales     dothideomycetes     ascomycota   \n",
       "3      aspergillaceae         eurotiales      eurotiomycetes     ascomycota   \n",
       "4    sporidiobolaceae    sporidiobolales  microbotryomycetes  basidiomycota   \n",
       "5       microascaceae       microascales     sordariomycetes     ascomycota   \n",
       "6      aspergillaceae         eurotiales      eurotiomycetes     ascomycota   \n",
       "7  saccharomycetaceae  saccharomycetales     saccharomycetes     ascomycota   \n",
       "8  saccharomycetaceae  saccharomycetales     saccharomycetes     ascomycota   \n",
       "9  saccharomycetaceae  saccharomycetales     saccharomycetes     ascomycota   \n",
       "\n",
       "  kingdom  # raw reads  # reads after homology filtering  \\\n",
       "0   fungi       272465                            122080   \n",
       "1   fungi       413127                            143363   \n",
       "2   fungi       260896                             97584   \n",
       "3   fungi       171615                             65065   \n",
       "4   fungi       318405                            127801   \n",
       "5   fungi       331947                            102481   \n",
       "6   fungi       125014                             54061   \n",
       "7   fungi        96837                             33025   \n",
       "8   fungi        91031                             26114   \n",
       "9   fungi       104214                             32490   \n",
       "\n",
       "   # reads after length filtering  # for use  \\\n",
       "0                          113337   113337.0   \n",
       "1                          133089   133089.0   \n",
       "2                           90015    90015.0   \n",
       "3                           59406    59406.0   \n",
       "4                          117801   117801.0   \n",
       "5                           93723    93723.0   \n",
       "6                           51340    51340.0   \n",
       "7                           30260    30260.0   \n",
       "8                           25597    25597.0   \n",
       "9                           29765    29765.0   \n",
       "\n",
       "                                   path to raw reads  \\\n",
       "0  analysis/Concatenated/20171103_FAH15473/barcod...   \n",
       "1  analysis/Concatenated/20171103_FAH15473/barcod...   \n",
       "2  analysis/Concatenated/20171103_FAH15473/barcod...   \n",
       "3  analysis/Concatenated/20171103_FAH15473/barcod...   \n",
       "4  analysis/Concatenated/20171103_FAH15473/barcod...   \n",
       "5  analysis/Concatenated/20171103_FAH15473/barcod...   \n",
       "6  analysis/Concatenated/20171207_FAH18654/barcod...   \n",
       "7  analysis/Concatenated/20180108_FAH18647/barcod...   \n",
       "8  analysis/Concatenated/20180108_FAH18647/barcod...   \n",
       "9  analysis/Concatenated/20180108_FAH18647/barcod...   \n",
       "\n",
       "                          path to homology filtering  \\\n",
       "0  analysis/Python_Processing/20171103_FAH15473/b...   \n",
       "1  analysis/Python_Processing/20171103_FAH15473/b...   \n",
       "2  analysis/Python_Processing/20171103_FAH15473/b...   \n",
       "3  analysis/Python_Processing/20171103_FAH15473/b...   \n",
       "4  analysis/Python_Processing/20171103_FAH15473/b...   \n",
       "5  analysis/Python_Processing/20171103_FAH15473/b...   \n",
       "6  analysis/Python_Processing/20171207_FAH18654/b...   \n",
       "7  analysis/Python_Processing/20180108_FAH18647/b...   \n",
       "8  analysis/Python_Processing/20180108_FAH18647/b...   \n",
       "9  analysis/Python_Processing/20180108_FAH18647/b...   \n",
       "\n",
       "                            path to length filtering  \\\n",
       "0  analysis/Length_Filtered/20171103_FAH15473/bar...   \n",
       "1  analysis/Length_Filtered/20171103_FAH15473/bar...   \n",
       "2  analysis/Length_Filtered/20171103_FAH15473/bar...   \n",
       "3  analysis/Length_Filtered/20171103_FAH15473/bar...   \n",
       "4  analysis/Length_Filtered/20171103_FAH15473/bar...   \n",
       "5  analysis/Length_Filtered/20171103_FAH15473/bar...   \n",
       "6  analysis/Length_Filtered/20171207_FAH18654/bar...   \n",
       "7  analysis/Length_Filtered/20180108_FAH18647/bar...   \n",
       "8  analysis/Length_Filtered/20180108_FAH18647/bar...   \n",
       "9  analysis/Length_Filtered/20180108_FAH18647/bar...   \n",
       "\n",
       "                                        path for use  \n",
       "0  analysis/Length_Filtered/20171103_FAH15473/bar...  \n",
       "1  analysis/Length_Filtered/20171103_FAH15473/bar...  \n",
       "2  analysis/Length_Filtered/20171103_FAH15473/bar...  \n",
       "3  analysis/Length_Filtered/20171103_FAH15473/bar...  \n",
       "4  analysis/Length_Filtered/20171103_FAH15473/bar...  \n",
       "5  analysis/Length_Filtered/20171103_FAH15473/bar...  \n",
       "6  analysis/Length_Filtered/20171207_FAH18654/bar...  \n",
       "7  analysis/Length_Filtered/20180108_FAH18647/bar...  \n",
       "8  analysis/Length_Filtered/20180108_FAH18647/bar...  \n",
       "9  analysis/Length_Filtered/20180108_FAH18647/bar...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "\n",
    "INPUT_BASEDIR = os.path.abspath('/media/MassStorage/tmp/TE/honours')\n",
    "subsampling_dir = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads')\n",
    "mapping_dir = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/qiime_results')\n",
    "\n",
    "mock_reference_dataframe_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/mock_reference_dataframe.csv')\n",
    "mock_max_custom_database_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/database/mock_database_labelled.fasta')\n",
    "mock_taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/mock_taxonomy_file_qiime.csv')\n",
    "\n",
    "qiime_db_fn = os.path.abspath('../../analysis/qiime2/db/sh_refs_qiime_ver8_dynamic_02.02.2019.fasta')\n",
    "qiime_tax_fn = os.path.abspath('../../analysis/qiime2/db/sh_taxonomy_qiime_ver8_dynamic_02.02.2019.txt')\n",
    "threads = 10\n",
    "\n",
    "mock_ref_df = pd.read_csv(mock_reference_dataframe_fn)\n",
    "mock_ref_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Qiime Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "puccinia_striiformis\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Pucciniomycetes', 'o': 'Pucciniales', 'f': 'Pucciniaceae', 'g': 'Puccinia', 's': 'Puccinia_striiformis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9585492227979274,\n",
      " \"c\": 0.8497409326424871,\n",
      " \"o\": 0.8497409326424871,\n",
      " \"f\": 0.8497409326424871,\n",
      " \"g\": 0.8497409326424871,\n",
      " \"s\": 0.8031088082901554\n",
      "}\n",
      "\n",
      "\n",
      "zymoseptoria_tritici\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Dothideomycetes', 'o': 'Capnodiales', 'f': 'Mycosphaerellaceae', 'g': 'Zymoseptoria', 's': 'Zymoseptoria_tritici'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8894009216589862,\n",
      " \"c\": 0.4055299539170507,\n",
      " \"o\": 0.2073732718894009,\n",
      " \"f\": 0.1935483870967742,\n",
      " \"g\": 0.17972350230414746,\n",
      " \"s\": 0.0\n",
      "}\n",
      "\n",
      "\n",
      "pyrenophora_tritici-repentis\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Dothideomycetes', 'o': 'Pleosporales', 'f': 'Pleosporaceae', 'g': 'Pyrenophora', 's': 'Pyrenophora_tritici-repentis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.8773584905660378,\n",
      " \"c\": 0.6462264150943396,\n",
      " \"o\": 0.6367924528301887,\n",
      " \"f\": 0.32547169811320753,\n",
      " \"g\": 0.2688679245283019,\n",
      " \"s\": 0.25943396226415094\n",
      "}\n",
      "\n",
      "\n",
      "aspergillus_niger\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Eurotiomycetes', 'o': 'Eurotiales', 'f': 'Aspergillaceae', 'g': 'Aspergillus', 's': 'Aspergillus_niger'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.937037037037037,\n",
      " \"c\": 0.5851851851851851,\n",
      " \"o\": 0.5814814814814815,\n",
      " \"f\": 0.5814814814814815,\n",
      " \"g\": 0.44074074074074077,\n",
      " \"s\": 0.16296296296296298\n",
      "}\n",
      "\n",
      "\n",
      "rhodotorula_mucilaginosa\n",
      "{'k': 'Fungi', 'p': 'Basidiomycota', 'c': 'Microbotryomycetes', 'o': 'Sporidiobolales', 'f': 'Sporidiobolaceae', 'g': 'Rhodotorula', 's': 'Rhodotorula_mucilaginosa'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9389671361502347,\n",
      " \"c\": 0.8356807511737089,\n",
      " \"o\": 0.8356807511737089,\n",
      " \"f\": 0.8356807511737089,\n",
      " \"g\": 0.8028169014084507,\n",
      " \"s\": 0.6009389671361502\n",
      "}\n",
      "\n",
      "\n",
      "scedosporium_boydii\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Sordariomycetes', 'o': 'Microascales', 'f': 'Microascaceae', 'g': 'Scedosporium', 's': 'Scedosporium_boydii'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9919354838709677,\n",
      " \"c\": 0.8709677419354839,\n",
      " \"o\": 0.7338709677419355,\n",
      " \"f\": 0.5161290322580645,\n",
      " \"g\": 0.3185483870967742,\n",
      " \"s\": 0.3024193548387097\n",
      "}\n",
      "\n",
      "\n",
      "aspergillus_flavus\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Eurotiomycetes', 'o': 'Eurotiales', 'f': 'Aspergillaceae', 'g': 'Aspergillus', 's': 'Aspergillus_flavus'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.956386292834891,\n",
      " \"c\": 0.7476635514018691,\n",
      " \"o\": 0.7476635514018691,\n",
      " \"f\": 0.7476635514018691,\n",
      " \"g\": 0.6199376947040498,\n",
      " \"s\": 0.5638629283489096\n",
      "}\n",
      "\n",
      "\n",
      "saccharomyces_cerevisiae\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetaceae', 'g': 'Saccharomyces', 's': 'Saccharomyces_cerevisiae'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9455445544554455,\n",
      " \"c\": 0.7524752475247525,\n",
      " \"o\": 0.7524752475247525,\n",
      " \"f\": 0.693069306930693,\n",
      " \"g\": 0.6782178217821783,\n",
      " \"s\": 0.5247524752475248\n",
      "}\n",
      "\n",
      "\n",
      "candida_albicans\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetaceae', 'g': 'Candida', 's': 'Candida_albicans'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9380952380952381,\n",
      " \"c\": 0.6142857142857143,\n",
      " \"o\": 0.6142857142857143,\n",
      " \"f\": 0.19523809523809524,\n",
      " \"g\": 0.41904761904761906,\n",
      " \"s\": 0.36666666666666664\n",
      "}\n",
      "\n",
      "\n",
      "candida_orthopsilosis\n",
      "{'k': 'Fungi', 'p': 'Ascomycota', 'c': 'Saccharomycetes', 'o': 'Saccharomycetales', 'f': 'Saccharomycetaceae', 'g': 'Candida', 's': 'Candida_orthopsilosis'}\n",
      "{\n",
      " \"k\": 1.0,\n",
      " \"p\": 0.9773755656108597,\n",
      " \"c\": 0.6923076923076923,\n",
      " \"o\": 0.6923076923076923,\n",
      " \"f\": 0.00904977375565611,\n",
      " \"g\": 0.6470588235294118,\n",
      " \"s\": 0.2986425339366516\n",
      "}\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from collections import OrderedDict\n",
    "\n",
    "def assign_taxranks_results(mapping_df, tax_fn, ref_df_fn = False):\n",
    "    \"\"\"\n",
    "    This function assigns the taxonomic ranks for each hit in the mapping results dataframe.\n",
    "    It takes a mapping_df, taxonomy assignment file, and if required a reference dataframe filename.\n",
    "    Returns the mapping dataframe with assignment. \n",
    "    \"\"\"\n",
    "    for tname in mapping_df.index:\n",
    "        if ref_df_fn:\n",
    "            tmp_taxfileid = getquery_taxfileid(ref_df_fn, tname)\n",
    "        else:\n",
    "            tmp_taxfileid = tname\n",
    "        tmp_tax_dict = get_taxid_dict(tax_fn, tmp_taxfileid)\n",
    "        for key, value in tmp_tax_dict.items():\n",
    "            mapping_df.loc[tname, key] = value\n",
    "    return mapping_df\n",
    "\n",
    "def get_accuracy_dict(mapping_df, query_tax_dict):\n",
    "    \"\"\"\n",
    "    Summarises the mapping accuracy of the mapping results at all taxonomic ranks.\n",
    "    Takes the mapping_df with taxnomonic assignments and a taxnomic dictionary of the known query.\n",
    "    Returns an accuracy dictionary for each taxnomic rank ['k', 'p', 'c', 'o', 'f', 'g', 's']. \n",
    "    Right now this function takes a qiime tax \n",
    "    \"\"\"\n",
    "    accuracy_dict = OrderedDict()\n",
    "    total_count = mapping_df['count'].sum()\n",
    "    for tax_rank in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        tmps_df = pd.DataFrame(data=None)\n",
    "        if tax_rank == 's':\n",
    "            for index, row in mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]].iterrows():\n",
    "                if row['s'] == 'unidentified' and row['g'] != query_tax_dict['g']:\n",
    "                    mapping_df.drop(index, axis=0, inplace=True)\n",
    "                else:\n",
    "                    continue\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        else:\n",
    "            hit_count = mapping_df[mapping_df[tax_rank] == query_tax_dict[tax_rank]]['count'].sum()\n",
    "        accuracy_dict[tax_rank] = hit_count/total_count\n",
    "    return accuracy_dict\n",
    "\n",
    "def getquery_taxfileid(refdf_fn, species):\n",
    "    \"\"\"\n",
    "    Takes the reference dataframe filename and the species name.\n",
    "    Returns the taxfileid, which is the date/flowcellid (column 0 value) of the ref_df.\n",
    "    \"\"\"\n",
    "    ref_df = pd.read_csv(refdf_fn)\n",
    "    return ref_df[ref_df.species == species].iloc[:,0].values[0]\n",
    "\n",
    "def get_taxid_dict(taxid_fn, taxfileid):\n",
    "    \"\"\"\n",
    "    Takes a taxonomy assignment file filename in the Qiime format and a taxonomic identifier.\n",
    "    Returns the a dictionary with the taxonomic assignment at each rank.\n",
    "    \"\"\"\n",
    "    tax_dict = {}\n",
    "    with open(taxid_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(taxfileid):\n",
    "                taxrankids = line.rstrip().split('\\t')[1].split(';')\n",
    "                for taxrank in taxrankids:\n",
    "                    tax_dict[taxrank.split('__')[0]] = taxrank.split('__')[1]\n",
    "    return tax_dict\n",
    "\n",
    "def mapping_results(fn, species):\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the query species: {species}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(hit_series,'\\n')\n",
    "    hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/custom_results/%s.json' % species)\n",
    "\n",
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "def pull_mapping_results_v3(fn):\n",
    "    \"\"\"\n",
    "    Takes a minimap2 paf and reads it in with the first 12 columns. Ignores the rest.\n",
    "    Filters for each read the best hit on mquality first taking the highest value.\n",
    "    Filters for each read by the number of nmatches in the second step.\n",
    "    Returns a dataframe that has the tnames as index and the counts of hits as column 'count'.\n",
    "    The dataframe has also the taxrank columns ['k', 'p', 'c', 'o', 'f', 'g', 's'] that are all False to start with.\n",
    "    \"\"\"\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    tmp_df['cscore'] = tmp_df['alen']/(tmp_df['alen']-tmp_df['nmatch'])\n",
    "    sub_df = tmp_df[tmp_df['cscore'] == tmp_df.groupby('qseqid')['cscore'].transform(max)].reset_index(drop=True)\n",
    "#     sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_df = pd.DataFrame(sub_df.groupby('tname')['cscore'].count().tolist(), sub_df.groupby('tname')['cscore'].count().index, columns=['count'])\n",
    "    hit_df.sort_values(by='count', ascending=False, inplace=True)\n",
    "    for key in ['k', 'p', 'c', 'o', 'f', 'g', 's']:\n",
    "        hit_df[key] = False\n",
    "        tmp_df[key] = False\n",
    "    return hit_df, tmp_df\n",
    "    \n",
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "\n",
    "test_species_list = []\n",
    "for entry in mock_ref_df.species.tolist():\n",
    "#     if entry[-7:] != '-ccl031' and entry[-7:] != '-ccl029':\n",
    "#         test_species_list.append(entry)\n",
    "#     else:\n",
    "#         test_species_list.append(entry[:-7])\n",
    "#         print(entry[:-7])\n",
    "    test_species_list.append(entry)\n",
    "    \n",
    "for test_species in test_species_list:\n",
    "    \n",
    "    print(test_species)\n",
    "    \n",
    "    #subsample tests species\n",
    "    fn_subsampling = {}\n",
    "    test_species = [test_species]\n",
    "    for x in test_species:\n",
    "#         print((mock_ref_df['species'] == x))\n",
    "        fn_subsampling[x] = (mock_ref_df[(mock_ref_df['species'] == x) & (mock_ref_df['genus'] == x.split('_')[0])]['path for use'].tolist()[0])\n",
    "        fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "\n",
    "    sub_reads_fn = {}\n",
    "    n_reads = 200\n",
    "    for key, value in fn_subsampling.items():\n",
    "        species = key\n",
    "        in_fn = value\n",
    "        out_fn = os.path.join(subsampling_dir, F'{species}.{n_reads}.fasta')\n",
    "        subsamplereads(in_fn, out_fn, n_reads)\n",
    "        sub_reads_fn[species] = out_fn\n",
    "        \n",
    "    ###Map the reads\n",
    "    db_fn = qiime_db_fn\n",
    "    sub_db_mapping_fn = {}\n",
    "    for species, fasta_fn in sub_reads_fn.items():\n",
    "        db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "        out_fn = os.path.join(mapping_dir, F\"{db_name}.{species}.minimap2.paf\")\n",
    "        sub_db_mapping_fn[species] = out_fn\n",
    "        minimapmapping(fasta_fn, db_fn, out_fn, threads)\n",
    "        \n",
    "    ###Test out the summary results statistic for a single mapping result\n",
    "    species = test_species[0]\n",
    "    mapping_results , full_results_df = pull_mapping_results_v3(sub_db_mapping_fn[species])\n",
    "    mapping_results = assign_taxranks_results(mapping_results, qiime_tax_fn)\n",
    "    taxfileid = getquery_taxfileid(mock_reference_dataframe_fn, species)\n",
    "    \n",
    "    query_tax_dict = get_taxid_dict(mock_taxonomy_file_fn, taxfileid)\n",
    "    print(query_tax_dict)\n",
    "    \n",
    "    sensitivity_dict = get_accuracy_dict(mapping_results, query_tax_dict)\n",
    "        \n",
    "    print(json.dumps(sensitivity_dict, indent=1))\n",
    "    print('\\n')\n",
    "    with open('/media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/qiime_results/%s.json' % species, 'w+') as fp:\n",
    "        json.dump(sensitivity_dict, fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Custom Database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### list of species in the max database\n",
    "max_species = ['Puccinia_striiformis',\n",
    "             'Zymoseptoria_tritici',\n",
    "             'Pyrenophora_tritici-repentis',\n",
    "             'Fusarium_oxysporum',\n",
    "             'Tuber_brumale',\n",
    "             'Cortinarius_globuliformis',\n",
    "             'Aspergillus_niger',\n",
    "             'Clavispora_lusitaniae',\n",
    "             'Kluyveromyces_unidentified',\n",
    "             'Penicillium_chrysogenum',\n",
    "             'Rhodotorula_mucilaginosa',\n",
    "             'Scedosporium_boydii',\n",
    "             'Blastobotrys_proliferans',\n",
    "             'Debaryomyces_unidentified',\n",
    "             'Galactomyces_geotrichum',\n",
    "             'Kodamaea_ohmeri',\n",
    "             'Meyerozyma_guilliermondii',\n",
    "             'Wickerhamomyces_anomalus',\n",
    "             'Yamadazyma_mexicana',\n",
    "             'Yamadazyma_scolyti',\n",
    "             'Yarrowia_lipolytica',\n",
    "             'Zygoascus_hellenicus',\n",
    "             'Aspergillus_flavus',\n",
    "             'Cryptococcus_zero',\n",
    "             'Aspergillus_unidentified',\n",
    "             'Diaporthe_CCL067',\n",
    "             'Diaporthe_unidentified',\n",
    "             'Oculimacula_yallundae-CCL031',\n",
    "             'Oculimacula_yallundae-CCL029',\n",
    "             'Dothiorella_vidmadera',\n",
    "             'Quambalaria_cyanescens',\n",
    "             'Entoleuca_unidentified',\n",
    "             'Asteroma_CCL060',\n",
    "             'Asteroma_CCL068',\n",
    "             'Saccharomyces_cerevisiae',\n",
    "             'Cladophialophora_unidentified',\n",
    "             'Candida_albicans',\n",
    "             'Candida_metapsilosis',\n",
    "             'Candida_orthopsilosis',\n",
    "             'Candida_parapsilosis',\n",
    "             'Candida_unidentified',\n",
    "             'Kluyveromyces_marxianus',\n",
    "             'Pichia_kudriavzevii',\n",
    "             'Pichia_membranifaciens']\n",
    "references = {}\n",
    "for species in max_species:\n",
    "    references[species] = species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy_file_fn = os.path.abspath('/media/MassStorage/tmp/TE/honours/analysis/Stats/taxonomy_file_mock.csv')\n",
    "species_file = {}\n",
    "genus_file = {}\n",
    "family_file = {}\n",
    "order_file = {}\n",
    "\n",
    "with open(taxonomy_file_fn, 'r') as fh:\n",
    "        for line in fh:\n",
    "            species_file[references[line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]].lower()] = line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]\n",
    "            genus_file[references[line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]].lower()] = line.split('\\t')[1].split(';')[5].split('__')[1].split('\\n')[0]\n",
    "            family_file[references[line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]].lower()] = line.split('\\t')[1].split(';')[4].split('__')[1].split('\\n')[0]\n",
    "            order_file[references[line.split('\\t')[1].split(';')[6].split('__')[1].split('\\n')[0]].lower()] = line.split('\\t')[1].split(';')[3].split('__')[1].split('\\n')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mock_community = ['aspergillus_flavus',\n",
    "                  'aspergillus_niger',\n",
    "                  'candida_albicans',\n",
    "                  'candida_orthopsilosis',\n",
    "                  'puccinia_striiformis',\n",
    "                  'pyrenophora_tritici-repentis',\n",
    "                  'rhodotorula_mucilaginosa',\n",
    "                  'saccharomyces_cerevisiae',\n",
    "                  'scedosporium_boydii',\n",
    "                  'zymoseptoria_tritici']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsamplereads(in_fn, out_fn, n_reads):\n",
    "    command = F'reformat.sh samplereadstarget={n_reads} in={in_fn} out={out_fn}'\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_reads = 200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUT_DIR = os.path.abspath('../../analysis/Mapping/mock')\n",
    "if not os.path.exists(OUT_DIR):\n",
    "    os.mkdir(OUT_DIR)\n",
    "MC_READ_DIR = os.path.join(OUT_DIR, 'subsample_reads')\n",
    "if not os.path.exists(MC_READ_DIR):\n",
    "    os.mkdir(MC_READ_DIR)\n",
    "sub_db_fn = os.path.join('../../analysis/Mapping/gsref.subdb.fasta')\n",
    "new_db_fn = os.path.join('../../analysis/Mapping/gsref.db.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0', 'species', 'genus', 'family', 'order', 'class', 'phylum',\n",
       "       'kingdom', '# raw reads', '# reads after homology filtering',\n",
       "       '# reads after length filtering', '# for use', 'path to raw reads',\n",
       "       'path to homology filtering', 'path to length filtering',\n",
       "       'path for use'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mock_ref_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aspergillus_flavus\n",
      "aspergillus_niger\n",
      "candida_albicans\n",
      "candida_orthopsilosis\n",
      "puccinia_striiformis\n",
      "pyrenophora_tritici-repentis\n",
      "rhodotorula_mucilaginosa\n",
      "saccharomyces_cerevisiae\n",
      "scedosporium_boydii\n",
      "zymoseptoria_tritici\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'aspergillus_flavus': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta',\n",
       " 'aspergillus_niger': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta',\n",
       " 'candida_albicans': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta',\n",
       " 'candida_orthopsilosis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta',\n",
       " 'puccinia_striiformis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode01/length_restricted_for_use.fasta',\n",
       " 'pyrenophora_tritici-repentis': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode03/length_restricted_for_use.fasta',\n",
       " 'rhodotorula_mucilaginosa': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode11/length_restricted_for_use.fasta',\n",
       " 'saccharomyces_cerevisiae': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode01/length_restricted_for_use.fasta',\n",
       " 'scedosporium_boydii': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode12/length_restricted_for_use.fasta',\n",
       " 'zymoseptoria_tritici': '/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_for_use.fasta'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn_subsampling = {}\n",
    "for x in mock_community:\n",
    "    print(x)\n",
    "    fn_subsampling[x] = mock_ref_df[mock_ref_df['species'] == x]['path for use'].tolist()[0]\n",
    "    fn_subsampling[x] = os.path.join(INPUT_BASEDIR, fn_subsampling[x])\n",
    "fn_subsampling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aspergillus_flavus\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171207_FAH18654/barcode12/length_restricted_for_use.fasta\n",
      "aspergillus_niger\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode07/length_restricted_for_use.fasta\n",
      "candida_albicans\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode03/length_restricted_for_use.fasta\n",
      "candida_orthopsilosis\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode05/length_restricted_for_use.fasta\n",
      "puccinia_striiformis\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode01/length_restricted_for_use.fasta\n",
      "pyrenophora_tritici-repentis\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode03/length_restricted_for_use.fasta\n",
      "rhodotorula_mucilaginosa\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode11/length_restricted_for_use.fasta\n",
      "saccharomyces_cerevisiae\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20180108_FAH18647/barcode01/length_restricted_for_use.fasta\n",
      "scedosporium_boydii\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode12/length_restricted_for_use.fasta\n",
      "zymoseptoria_tritici\n",
      "/media/MassStorage/tmp/TE/honours/analysis/Length_Filtered/20171103_FAH15473/barcode02/length_restricted_for_use.fasta\n"
     ]
    }
   ],
   "source": [
    "sub_reads_fn = {}\n",
    "for key, value in fn_subsampling.items():\n",
    "    print(key)\n",
    "    print(value)\n",
    "    species = key\n",
    "    in_fn = value\n",
    "    out_fn = os.path.join(MC_READ_DIR, F'{species}.{n_reads}.fasta')\n",
    "#     subsamplereads(in_fn, out_fn, n_reads)\n",
    "    sub_reads_fn[species] = out_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Map with minimap against both databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minimapmapping(fasta_fn, ref_fn, out_fn, threads):\n",
    "    command = F\"minimap2 -x map-ont -t {threads} {ref_fn} {fasta_fn} -o {out_fn}\"\n",
    "    out = subprocess.getstatusoutput(command)\n",
    "    if out[0] == 0:\n",
    "        print(F\":)Completed {command}\\n\")\n",
    "    else:\n",
    "        print(F\":(check one {command}!!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../analysis/Mapping/gsref.subdb.fasta\n",
      "../../analysis/Mapping/gsref.db.fasta\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'../../analysis/Mapping/gsref.subdb.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb',\n",
       " '../../analysis/Mapping/gsref.db.fasta': '/media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db'}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbases_fn = {}\n",
    "for x in [sub_db_fn, new_db_fn]:\n",
    "    print(x)\n",
    "    dbases_fn[x] = os.path.join(OUT_DIR, os.path.basename(x).replace('.fasta', '').replace('.','_'))\n",
    "    if not os.path.exists(dbases_fn[x]):\n",
    "        os.mkdir(dbases_fn[x])\n",
    "dbases_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/aspergillus_flavus.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.aspergillus_flavus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/aspergillus_niger.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.aspergillus_niger.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/candida_albicans.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.candida_albicans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/candida_orthopsilosis.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.candida_orthopsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/puccinia_striiformis.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.puccinia_striiformis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/pyrenophora_tritici-repentis.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.pyrenophora_tritici-repentis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/rhodotorula_mucilaginosa.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.rhodotorula_mucilaginosa.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/saccharomyces_cerevisiae.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.saccharomyces_cerevisiae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/scedosporium_boydii.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.scedosporium_boydii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.subdb.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/zymoseptoria_tritici.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_subdb/gsref.subdb.zymoseptoria_tritici.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = sub_db_fn\n",
    "sub_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    sub_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/aspergillus_flavus.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.aspergillus_flavus.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/aspergillus_niger.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.aspergillus_niger.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/candida_albicans.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.candida_albicans.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/candida_orthopsilosis.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.candida_orthopsilosis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/puccinia_striiformis.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.puccinia_striiformis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/pyrenophora_tritici-repentis.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.pyrenophora_tritici-repentis.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/rhodotorula_mucilaginosa.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.rhodotorula_mucilaginosa.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/saccharomyces_cerevisiae.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.saccharomyces_cerevisiae.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/scedosporium_boydii.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.scedosporium_boydii.minimap2.paf\n",
      "\n",
      ":)Completed minimap2 -x map-ont -t 10 ../../analysis/Mapping/gsref.db.fasta /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/subsample_reads/zymoseptoria_tritici.200.fasta -o /media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/gsref_db/gsref.db.zymoseptoria_tritici.minimap2.paf\n",
      "\n"
     ]
    }
   ],
   "source": [
    "db_fn = new_db_fn\n",
    "new_db_mapping_fn = {}\n",
    "for species, fasta_fn in sub_reads_fn.items():\n",
    "    tmp_out = dbases_fn[db_fn]\n",
    "    db_name = os.path.basename(db_fn).replace('.fasta', '')\n",
    "    out_fn = os.path.join(tmp_out, F\"{db_name}.{species}.minimap2.paf\")\n",
    "    new_db_mapping_fn[species] = out_fn\n",
    "    minimapmapping(fasta_fn, db_fn, out_fn, threads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at mapping results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mapping_results_v2(fn, species,expected_species, tax_rank, tax_rank_file):\n",
    "    import numpy as np\n",
    "    min_header = ['qseqid', 'qlen', 'qstart', 'qstop', 'strand', 'tname', 'tlen', 'tstart', 'tend', 'nmatch', 'alen', 'mquality']\n",
    "    tmp_df = pd.read_csv(fn, sep='\\t', header = None, usecols=[x for x in range(0,12)], names=min_header)\n",
    "    sub_df = tmp_df[tmp_df['mquality'] == tmp_df.groupby('qseqid')['mquality'].transform(max)].reset_index(drop=True)\n",
    "    sub_df = sub_df[sub_df['nmatch'] == sub_df.groupby('qseqid')['nmatch'].transform(max)].reset_index(drop=True)\n",
    "    hit_series = pd.Series(sub_df.groupby('tname')['mquality'].count().tolist()/sub_df.groupby('tname')['mquality'].count().sum(),\n",
    "                      sub_df.groupby('tname')['mquality'].count().index)\n",
    "    if tax_rank == 's' or tax_rank == 'species':\n",
    "        new_new_hit_series = hit_series\n",
    "        new_new_hit_series.sort_values(ascending=False, inplace=True)\n",
    "    else:\n",
    "        new_hit_series = pd.Series()\n",
    "        new_new_hit_series = pd.Series()\n",
    "        for index in hit_series.index.unique():\n",
    "            new_hit_series.at[index] = data=np.sum(hit_series[index])\n",
    "        new_hit_series = new_hit_series.rename(tax_rank_file, axis='index')\n",
    "        for index in new_hit_series.index.unique():\n",
    "            new_new_hit_series.at[index] = data=np.sum(new_hit_series[index])\n",
    "        new_new_hit_series.sort_values(ascending=False, inplace=True)\n",
    "    print(sub_df.qseqid.unique().shape == tmp_df.qseqid.unique().shape)\n",
    "    print('##########\\n')\n",
    "    print(F\"This was the sample type expected: {tax_rank_file[expected_species[species]]}\\n\")\n",
    "    print(F\"These are the results:\")\n",
    "    print(new_new_hit_series,'\\n')\n",
    "    new_new_hit_series.to_json('/media/MassStorage/tmp/TE/honours/analysis/Mapping/mock/custom_results/%s_%s.json' % (tax_rank,expected_species[species]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "puccinia_striiformis\n",
      "zymoseptoria_tritici\n",
      "pyrenophora_tritici-repentis\n",
      "aspergillus_niger\n",
      "rhodotorula_mucilaginosa\n",
      "scedosporium_boydii\n",
      "aspergillus_flavus\n",
      "saccharomyces_cerevisiae\n",
      "candida_albicans\n",
      "candida_orthopsilosis\n"
     ]
    }
   ],
   "source": [
    "expected_rank_s = {}\n",
    "expected_rank_g = {}\n",
    "expected_rank_f = {}\n",
    "expected_rank_o = {}\n",
    "\n",
    "for column, row in mock_ref_df.iterrows():\n",
    "    print(row['species'])\n",
    "    expected_rank_s[row['species']] = row['species']\n",
    "    expected_rank_g[row['species']] = row['genus']\n",
    "    expected_rank_f[row['species']] = row['family']\n",
    "    expected_rank_o[row['species']] = row['order']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Aspergillus_flavus\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_flavus    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Aspergillus_niger\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "aspergillus_niger           0.831683\n",
      "aspergillus_unidentified    0.153465\n",
      "aspergillus_flavus          0.009901\n",
      "scedosporium_boydii         0.004950\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Candida_albicans\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_albicans              0.524752\n",
      "candida_unidentified          0.316832\n",
      "candida_orthopsilosis         0.094059\n",
      "candida_metapsilosis          0.044554\n",
      "candida_parapsilosis          0.014851\n",
      "kluyveromyces_unidentified    0.004950\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Candida_orthopsilosis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "candida_orthopsilosis        0.719626\n",
      "candida_metapsilosis         0.228972\n",
      "candida_parapsilosis         0.037383\n",
      "meyerozyma_guilliermondii    0.014019\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia_striiformis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "puccinia_striiformis    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pyrenophora_tritici-repentis\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "pyrenophora_tritici-repentis    0.970\n",
      "zymoseptoria_tritici            0.025\n",
      "debaryomyces_unidentified       0.005\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Rhodotorula_mucilaginosa\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "rhodotorula_mucilaginosa    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Saccharomyces_cerevisiae\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "saccharomyces_cerevisiae    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Scedosporium_boydii\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "scedosporium_boydii    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Zymoseptoria_tritici\n",
      "\n",
      "These are the results:\n",
      "tname\n",
      "zymoseptoria_tritici       0.995\n",
      "penicillium_chrysogenum    0.005\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 's', species_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Aspergillus\n",
      "\n",
      "These are the results:\n",
      "Aspergillus    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Aspergillus\n",
      "\n",
      "These are the results:\n",
      "Aspergillus     0.99505\n",
      "Scedosporium    0.00495\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Candida\n",
      "\n",
      "These are the results:\n",
      "Candida          0.99505\n",
      "Kluyveromyces    0.00495\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Candida\n",
      "\n",
      "These are the results:\n",
      "Candida       0.985981\n",
      "Meyerozyma    0.014019\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Puccinia\n",
      "\n",
      "These are the results:\n",
      "Puccinia    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pyrenophora\n",
      "\n",
      "These are the results:\n",
      "Pyrenophora     0.970\n",
      "Zymoseptoria    0.025\n",
      "Debaryomyces    0.005\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Rhodotorula\n",
      "\n",
      "These are the results:\n",
      "Rhodotorula    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Saccharomyces\n",
      "\n",
      "These are the results:\n",
      "Saccharomyces    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Scedosporium\n",
      "\n",
      "These are the results:\n",
      "Scedosporium    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Zymoseptoria\n",
      "\n",
      "These are the results:\n",
      "Zymoseptoria    0.995\n",
      "Penicillium     0.005\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 'g', genus_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Aspergillaceae\n",
      "\n",
      "These are the results:\n",
      "Aspergillaceae    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Aspergillaceae\n",
      "\n",
      "These are the results:\n",
      "Aspergillaceae    0.99505\n",
      "Microascaceae     0.00495\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Saccharomycetaceae\n",
      "\n",
      "These are the results:\n",
      "Saccharomycetaceae    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Saccharomycetaceae\n",
      "\n",
      "These are the results:\n",
      "Saccharomycetaceae    0.985981\n",
      "Debaryomycetaceae     0.014019\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pucciniaceae\n",
      "\n",
      "These are the results:\n",
      "Pucciniaceae    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pleosporaceae\n",
      "\n",
      "These are the results:\n",
      "Pleosporaceae         0.970\n",
      "Mycosphaerellaceae    0.025\n",
      "Debaryomycetaceae     0.005\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Sporidiobolaceae\n",
      "\n",
      "These are the results:\n",
      "Sporidiobolaceae    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Saccharomycetaceae\n",
      "\n",
      "These are the results:\n",
      "Saccharomycetaceae    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Microascaceae\n",
      "\n",
      "These are the results:\n",
      "Microascaceae    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Mycosphaerellaceae\n",
      "\n",
      "These are the results:\n",
      "Mycosphaerellaceae    0.995\n",
      "Aspergillaceae        0.005\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 'f', family_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Eurotiales\n",
      "\n",
      "These are the results:\n",
      "Eurotiales    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Eurotiales\n",
      "\n",
      "These are the results:\n",
      "Eurotiales      0.99505\n",
      "Microascales    0.00495\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Saccharomycetales\n",
      "\n",
      "These are the results:\n",
      "Saccharomycetales    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Saccharomycetales\n",
      "\n",
      "These are the results:\n",
      "Saccharomycetales    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pucciniales\n",
      "\n",
      "These are the results:\n",
      "Pucciniales    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Pleosporales\n",
      "\n",
      "These are the results:\n",
      "Pleosporales         0.970\n",
      "Capnodiales          0.025\n",
      "Saccharomycetales    0.005\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Sporidiobolales\n",
      "\n",
      "These are the results:\n",
      "Sporidiobolales    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Saccharomycetales\n",
      "\n",
      "These are the results:\n",
      "Saccharomycetales    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Microascales\n",
      "\n",
      "These are the results:\n",
      "Microascales    1.0\n",
      "dtype: float64 \n",
      "\n",
      "True\n",
      "##########\n",
      "\n",
      "This was the sample type expected: Capnodiales\n",
      "\n",
      "These are the results:\n",
      "Capnodiales    0.995\n",
      "Eurotiales     0.005\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "###this is running the reads against the full database\n",
    "for barcode, hit_fn in new_db_mapping_fn.items():\n",
    "    mapping_results_v2(hit_fn, barcode, expected_rank_s, 'o', order_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
